{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "2b0efc03-e462-485e-97b5-a8139fcdbc3b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "The legacy packages maptools, rgdal, and rgeos, underpinning this package\n",
      "will retire shortly. Please refer to R-spatial evolution reports on\n",
      "https://r-spatial.org/r/2023/05/15/evolution4.html for details.\n",
      "This package is now running under evolution status 0 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(infercnv)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "6a8e7f58-3aad-459b-b9e8-44ef6581bc2d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(rtracklayer)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "100d4f1c-965a-4f83-ade0-dd66da6cd150",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "── \u001b[1mAttaching core tidyverse packages\u001b[22m ──────────────────────── tidyverse 2.0.0 ──\n",
      "\u001b[32m✔\u001b[39m \u001b[34mdplyr    \u001b[39m 1.1.2     \u001b[32m✔\u001b[39m \u001b[34mreadr    \u001b[39m 2.1.4\n",
      "\u001b[32m✔\u001b[39m \u001b[34mforcats  \u001b[39m 1.0.0     \u001b[32m✔\u001b[39m \u001b[34mstringr  \u001b[39m 1.5.0\n",
      "\u001b[32m✔\u001b[39m \u001b[34mggplot2  \u001b[39m 3.4.2     \u001b[32m✔\u001b[39m \u001b[34mtibble   \u001b[39m 3.2.1\n",
      "\u001b[32m✔\u001b[39m \u001b[34mlubridate\u001b[39m 1.9.2     \u001b[32m✔\u001b[39m \u001b[34mtidyr    \u001b[39m 1.3.0\n",
      "\u001b[32m✔\u001b[39m \u001b[34mpurrr    \u001b[39m 1.0.1     \n",
      "── \u001b[1mConflicts\u001b[22m ────────────────────────────────────────── tidyverse_conflicts() ──\n",
      "\u001b[31m✖\u001b[39m \u001b[34mlubridate\u001b[39m::\u001b[32m%within%()\u001b[39m masks \u001b[34mIRanges\u001b[39m::%within%()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mcollapse()\u001b[39m     masks \u001b[34mIRanges\u001b[39m::collapse()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mcombine()\u001b[39m      masks \u001b[34mBiocGenerics\u001b[39m::combine()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mdesc()\u001b[39m         masks \u001b[34mIRanges\u001b[39m::desc()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mtidyr\u001b[39m::\u001b[32mexpand()\u001b[39m       masks \u001b[34mS4Vectors\u001b[39m::expand()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfilter()\u001b[39m       masks \u001b[34mstats\u001b[39m::filter()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfirst()\u001b[39m        masks \u001b[34mS4Vectors\u001b[39m::first()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlag()\u001b[39m          masks \u001b[34mstats\u001b[39m::lag()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mggplot2\u001b[39m::\u001b[32mPosition()\u001b[39m   masks \u001b[34mBiocGenerics\u001b[39m::Position(), \u001b[34mbase\u001b[39m::Position()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mpurrr\u001b[39m::\u001b[32mreduce()\u001b[39m       masks \u001b[34mGenomicRanges\u001b[39m::reduce(), \u001b[34mIRanges\u001b[39m::reduce()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mrename()\u001b[39m       masks \u001b[34mS4Vectors\u001b[39m::rename()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mlubridate\u001b[39m::\u001b[32msecond()\u001b[39m   masks \u001b[34mS4Vectors\u001b[39m::second()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mlubridate\u001b[39m::\u001b[32msecond<-()\u001b[39m masks \u001b[34mS4Vectors\u001b[39m::second<-()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mslice()\u001b[39m        masks \u001b[34mIRanges\u001b[39m::slice()\n",
      "\u001b[36mℹ\u001b[39m Use the conflicted package (\u001b[3m\u001b[34m<http://conflicted.r-lib.org/>\u001b[39m\u001b[23m) to force all conflicts to become errors\n"
     ]
    }
   ],
   "source": [
    "library(tidyverse)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "771f7713-fd5a-414e-87b2-636579a5b88b",
   "metadata": {},
   "source": [
    "# Prepare files for infercnv"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1d90a1cd-c397-440b-bf06-217f9bb74fe3",
   "metadata": {
    "tags": []
   },
   "source": [
    "### prepare gene position file (one time only, then read from disk)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fa6b2fb8-bda8-44f6-b5e7-f0bace46d469",
   "metadata": {},
   "source": [
    "#### I downloaded the gtf file from https://support.10xgenomics.com/single-cell-gene-expression/software/release-notes/build#grch38_1.2.0 and filtered as per the instructions there"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b03348c7-9e67-4049-be2a-961dda2f6335",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf <- rtracklayer::import('data//Homo_sapiens.GRCh38.84.filtered.gtf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "4c438f4a-ae0e-4f58-aeda-2c28f1593866",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf_df=as.data.frame(gtf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "3f4edfef-d862-48ba-a58a-798df69b5de1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 31</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>seqnames</th><th scope=col>start</th><th scope=col>end</th><th scope=col>width</th><th scope=col>strand</th><th scope=col>source</th><th scope=col>type</th><th scope=col>score</th><th scope=col>phase</th><th scope=col>gene_id</th><th scope=col>⋯</th><th scope=col>havana_transcript</th><th scope=col>havana_transcript_version</th><th scope=col>tag</th><th scope=col>transcript_support_level</th><th scope=col>exon_number</th><th scope=col>exon_id</th><th scope=col>exon_version</th><th scope=col>ccds_id</th><th scope=col>protein_id</th><th scope=col>protein_version</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>⋯</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>1</td><td>29554</td><td>31109</td><td>1556</td><td>+</td><td>havana</td><td>gene      </td><td>NA</td><td>NA</td><td>ENSG00000243485</td><td>⋯</td><td>NA                </td><td>NA</td><td>NA   </td><td>NA</td><td>NA</td><td>NA             </td><td>NA</td><td>NA</td><td>NA</td><td>NA</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>1</td><td>29554</td><td>31097</td><td>1544</td><td>+</td><td>havana</td><td>transcript</td><td>NA</td><td>NA</td><td>ENSG00000243485</td><td>⋯</td><td>OTTHUMT00000002840</td><td>1 </td><td>basic</td><td>5 </td><td>NA</td><td>NA             </td><td>NA</td><td>NA</td><td>NA</td><td>NA</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>1</td><td>29554</td><td>30039</td><td> 486</td><td>+</td><td>havana</td><td>exon      </td><td>NA</td><td>NA</td><td>ENSG00000243485</td><td>⋯</td><td>OTTHUMT00000002840</td><td>1 </td><td>basic</td><td>5 </td><td>1 </td><td>ENSE00001947070</td><td>1 </td><td>NA</td><td>NA</td><td>NA</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>1</td><td>30564</td><td>30667</td><td> 104</td><td>+</td><td>havana</td><td>exon      </td><td>NA</td><td>NA</td><td>ENSG00000243485</td><td>⋯</td><td>OTTHUMT00000002840</td><td>1 </td><td>basic</td><td>5 </td><td>2 </td><td>ENSE00001922571</td><td>1 </td><td>NA</td><td>NA</td><td>NA</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>1</td><td>30976</td><td>31097</td><td> 122</td><td>+</td><td>havana</td><td>exon      </td><td>NA</td><td>NA</td><td>ENSG00000243485</td><td>⋯</td><td>OTTHUMT00000002840</td><td>1 </td><td>basic</td><td>5 </td><td>3 </td><td>ENSE00001827679</td><td>1 </td><td>NA</td><td>NA</td><td>NA</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>1</td><td>30267</td><td>31109</td><td> 843</td><td>+</td><td>havana</td><td>transcript</td><td>NA</td><td>NA</td><td>ENSG00000243485</td><td>⋯</td><td>OTTHUMT00000002841</td><td>2 </td><td>basic</td><td>5 </td><td>NA</td><td>NA             </td><td>NA</td><td>NA</td><td>NA</td><td>NA</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 31\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "  & seqnames & start & end & width & strand & source & type & score & phase & gene\\_id & ⋯ & havana\\_transcript & havana\\_transcript\\_version & tag & transcript\\_support\\_level & exon\\_number & exon\\_id & exon\\_version & ccds\\_id & protein\\_id & protein\\_version\\\\\n",
       "  & <fct> & <int> & <int> & <int> & <fct> & <fct> & <fct> & <dbl> & <int> & <chr> & ⋯ & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & 1 & 29554 & 31109 & 1556 & + & havana & gene       & NA & NA & ENSG00000243485 & ⋯ & NA                 & NA & NA    & NA & NA & NA              & NA & NA & NA & NA\\\\\n",
       "\t2 & 1 & 29554 & 31097 & 1544 & + & havana & transcript & NA & NA & ENSG00000243485 & ⋯ & OTTHUMT00000002840 & 1  & basic & 5  & NA & NA              & NA & NA & NA & NA\\\\\n",
       "\t3 & 1 & 29554 & 30039 &  486 & + & havana & exon       & NA & NA & ENSG00000243485 & ⋯ & OTTHUMT00000002840 & 1  & basic & 5  & 1  & ENSE00001947070 & 1  & NA & NA & NA\\\\\n",
       "\t4 & 1 & 30564 & 30667 &  104 & + & havana & exon       & NA & NA & ENSG00000243485 & ⋯ & OTTHUMT00000002840 & 1  & basic & 5  & 2  & ENSE00001922571 & 1  & NA & NA & NA\\\\\n",
       "\t5 & 1 & 30976 & 31097 &  122 & + & havana & exon       & NA & NA & ENSG00000243485 & ⋯ & OTTHUMT00000002840 & 1  & basic & 5  & 3  & ENSE00001827679 & 1  & NA & NA & NA\\\\\n",
       "\t6 & 1 & 30267 & 31109 &  843 & + & havana & transcript & NA & NA & ENSG00000243485 & ⋯ & OTTHUMT00000002841 & 2  & basic & 5  & NA & NA              & NA & NA & NA & NA\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 31\n",
       "\n",
       "| <!--/--> | seqnames &lt;fct&gt; | start &lt;int&gt; | end &lt;int&gt; | width &lt;int&gt; | strand &lt;fct&gt; | source &lt;fct&gt; | type &lt;fct&gt; | score &lt;dbl&gt; | phase &lt;int&gt; | gene_id &lt;chr&gt; | ⋯ ⋯ | havana_transcript &lt;chr&gt; | havana_transcript_version &lt;chr&gt; | tag &lt;chr&gt; | transcript_support_level &lt;chr&gt; | exon_number &lt;chr&gt; | exon_id &lt;chr&gt; | exon_version &lt;chr&gt; | ccds_id &lt;chr&gt; | protein_id &lt;chr&gt; | protein_version &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | 1 | 29554 | 31109 | 1556 | + | havana | gene       | NA | NA | ENSG00000243485 | ⋯ | NA                 | NA | NA    | NA | NA | NA              | NA | NA | NA | NA |\n",
       "| 2 | 1 | 29554 | 31097 | 1544 | + | havana | transcript | NA | NA | ENSG00000243485 | ⋯ | OTTHUMT00000002840 | 1  | basic | 5  | NA | NA              | NA | NA | NA | NA |\n",
       "| 3 | 1 | 29554 | 30039 |  486 | + | havana | exon       | NA | NA | ENSG00000243485 | ⋯ | OTTHUMT00000002840 | 1  | basic | 5  | 1  | ENSE00001947070 | 1  | NA | NA | NA |\n",
       "| 4 | 1 | 30564 | 30667 |  104 | + | havana | exon       | NA | NA | ENSG00000243485 | ⋯ | OTTHUMT00000002840 | 1  | basic | 5  | 2  | ENSE00001922571 | 1  | NA | NA | NA |\n",
       "| 5 | 1 | 30976 | 31097 |  122 | + | havana | exon       | NA | NA | ENSG00000243485 | ⋯ | OTTHUMT00000002840 | 1  | basic | 5  | 3  | ENSE00001827679 | 1  | NA | NA | NA |\n",
       "| 6 | 1 | 30267 | 31109 |  843 | + | havana | transcript | NA | NA | ENSG00000243485 | ⋯ | OTTHUMT00000002841 | 2  | basic | 5  | NA | NA              | NA | NA | NA | NA |\n",
       "\n"
      ],
      "text/plain": [
       "  seqnames start end   width strand source type       score phase\n",
       "1 1        29554 31109 1556  +      havana gene       NA    NA   \n",
       "2 1        29554 31097 1544  +      havana transcript NA    NA   \n",
       "3 1        29554 30039  486  +      havana exon       NA    NA   \n",
       "4 1        30564 30667  104  +      havana exon       NA    NA   \n",
       "5 1        30976 31097  122  +      havana exon       NA    NA   \n",
       "6 1        30267 31109  843  +      havana transcript NA    NA   \n",
       "  gene_id         ⋯ havana_transcript  havana_transcript_version tag  \n",
       "1 ENSG00000243485 ⋯ NA                 NA                        NA   \n",
       "2 ENSG00000243485 ⋯ OTTHUMT00000002840 1                         basic\n",
       "3 ENSG00000243485 ⋯ OTTHUMT00000002840 1                         basic\n",
       "4 ENSG00000243485 ⋯ OTTHUMT00000002840 1                         basic\n",
       "5 ENSG00000243485 ⋯ OTTHUMT00000002840 1                         basic\n",
       "6 ENSG00000243485 ⋯ OTTHUMT00000002841 2                         basic\n",
       "  transcript_support_level exon_number exon_id         exon_version ccds_id\n",
       "1 NA                       NA          NA              NA           NA     \n",
       "2 5                        NA          NA              NA           NA     \n",
       "3 5                        1           ENSE00001947070 1            NA     \n",
       "4 5                        2           ENSE00001922571 1            NA     \n",
       "5 5                        3           ENSE00001827679 1            NA     \n",
       "6 5                        NA          NA              NA           NA     \n",
       "  protein_id protein_version\n",
       "1 NA         NA             \n",
       "2 NA         NA             \n",
       "3 NA         NA             \n",
       "4 NA         NA             \n",
       "5 NA         NA             \n",
       "6 NA         NA             "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(gtf_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6055a26b-0129-4108-a482-84bd67148321",
   "metadata": {},
   "source": [
    "#### drop duplicate rows -- they're really similar so can just randomly pick one"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "4c5ed90e-c091-4e34-8d85-960871bd05a9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "33660"
      ],
      "text/latex": [
       "33660"
      ],
      "text/markdown": [
       "33660"
      ],
      "text/plain": [
       "[1] 33660"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "length(unique((gtf_df %>% filter(type==\"gene\") %>% select(gene_name, seqnames, start, end))$gene_name))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "49d0c405-07ea-4366-8850-f5a0604b4bf3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "33694"
      ],
      "text/latex": [
       "33694"
      ],
      "text/markdown": [
       "33694"
      ],
      "text/plain": [
       "[1] 33694"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(gtf_df %>% filter(type==\"gene\") %>% select(start, end, source, gene_id, gene_name))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "f6447ceb-28f3-4bef-9793-4da3ae055560",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "33660"
      ],
      "text/latex": [
       "33660"
      ],
      "text/markdown": [
       "33660"
      ],
      "text/plain": [
       "[1] 33660"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "gtf_df = subset(gtf_df %>% filter(type==\"gene\") %>% select(gene_name, seqnames, start, end),!duplicated((gtf_df %>% filter(type==\"gene\"))$gene_name))\n",
    "nrow(gtf_df)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6fc3c328-060e-444b-94a8-5f29adcbab28",
   "metadata": {},
   "source": [
    "#### filter to genes on the 23 chromosomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "72edd295-3dfb-44a9-8ae4-e86b18873552",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf_df = gtf_df %>% filter(seqnames %in% c(c(1:22), \"X\", \"Y\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "4c9b6e8a-0d95-4b9e-a1bf-133d3a7726a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "write.table(gtf_df, file=\"data/infercnv_input_files/cd138_mm/gene_pos.txt\", col.names=F, row.names=F, quote=F, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cb3a348b-d368-4f0c-ae07-b527f552ceb4",
   "metadata": {},
   "source": [
    "#### prepared the other 2 input files in python, writing from anndata object to disk"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a646c6b8-ee4d-4963-afb9-061ec7472e3e",
   "metadata": {},
   "source": [
    "# Run infercnv"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "15f2048a-5ef9-49c1-ba0a-42971470c7c8",
   "metadata": {},
   "source": [
    "#### start by running with analysis_mode=\"subclusters\" through step 15 and inspect outputs -- it finds too many subclusters, going to run with analysis_mode=\"samples\" (below)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "81b7bdd4-960a-44f8-ba95-54f36b6025e5",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2023-08-08 14:25:32] Parsing matrix: data/infercnv_input_files/crc/GSE132465_GEO_processed_CRC_10X_raw_UMI_count_matrix.txt\n",
      "INFO [2023-08-08 14:39:02] Parsing gene order file: data/infercnv_input_files/cd138_mm/gene_pos.txt\n",
      "INFO [2023-08-08 14:39:02] Parsing cell annotations file: data/infercnv_input_files/crc/cell_annots.txt\n",
      "INFO [2023-08-08 14:39:02] ::order_reduce:Start.\n",
      "INFO [2023-08-08 14:39:12] .order_reduce(): expr and order match.\n",
      "INFO [2023-08-08 14:39:26] ::process_data:order_reduce:Reduction from positional data, new dimensions (r,c) = 33694,63689 Total=622669375 Min=0 Max=26489.\n",
      "INFO [2023-08-08 14:39:35] num genes removed taking into account provided gene ordering list: 76 = 0.225559446785778% removed.\n",
      "INFO [2023-08-08 14:39:37] -filtering out cells < 100 or > Inf, removing 0 % of cells\n",
      "WARN [2023-08-08 14:39:57] Please use \"options(scipen = 100)\" before running infercnv if you are using the analysis_mode=\"subclusters\" option or you may encounter an error while the hclust is being generated.\n",
      "INFO [2023-08-08 14:40:48] validating infercnv_obj\n"
     ]
    }
   ],
   "source": [
    "infercnv_obj = CreateInfercnvObject(raw_counts_matrix=\"data/infercnv_input_files/crc/GSE132465_GEO_processed_CRC_10X_raw_UMI_count_matrix.txt\",\n",
    "                                    annotations_file=\"data/infercnv_input_files/crc/cell_annots.txt\",\n",
    "                                    delim=\"\\t\",\n",
    "                                    gene_order_file=\"data/infercnv_input_files/cd138_mm/gene_pos.txt\",\n",
    "                                    ref_group_names=c('T cells','B cells','Stromal cells','Epithelial cells','Myeloids','Mast cells'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "bf354959-bd3b-45f1-ad88-fe78e4d7e252",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(scipen = 100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "f0f50d10-f2c2-4b61-83c3-1036d95eb546",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2023-08-08 14:45:18] ::process_data:Start\n",
      "INFO [2023-08-08 14:45:18] Checking for saved results.\n",
      "INFO [2023-08-08 14:45:18] Trying to reload from step 1\n",
      "INFO [2023-08-08 14:46:21] Using backup from step 1\n",
      "INFO [2023-08-08 14:46:21] \n",
      "\n",
      "\tSTEP 1: incoming data\n",
      "\n",
      "INFO [2023-08-08 14:46:21] \n",
      "\n",
      "\tSTEP 02: Removing lowly expressed genes\n",
      "\n",
      "INFO [2023-08-08 14:46:21] ::above_min_mean_expr_cutoff:Start\n",
      "INFO [2023-08-08 14:46:26] Removing 26541 genes from matrix as below mean expr threshold: 0.1\n",
      "INFO [2023-08-08 14:46:29] validating infercnv_obj\n",
      "INFO [2023-08-08 14:46:29] There are 7077 genes and 63689 cells remaining in the expr matrix.\n",
      "INFO [2023-08-08 14:46:49] no genes removed due to min cells/gene filter\n",
      "INFO [2023-08-08 14:48:52] \n",
      "\n",
      "\tSTEP 03: normalization by sequencing depth\n",
      "\n",
      "INFO [2023-08-08 14:48:52] normalizing counts matrix by depth\n",
      "INFO [2023-08-08 14:49:03] Computed total sum normalization factor as median libsize: 3896.000000\n",
      "INFO [2023-08-08 14:49:05] Adding h-spike\n",
      "INFO [2023-08-08 14:49:05] -hspike modeling of T cells\n",
      "INFO [2023-08-08 14:51:40] -hspike modeling of B cells\n",
      "INFO [2023-08-08 14:54:19] -hspike modeling of Stromal cells\n",
      "INFO [2023-08-08 14:56:56] -hspike modeling of Epithelial cells\n",
      "INFO [2023-08-08 14:59:33] -hspike modeling of Myeloids\n",
      "INFO [2023-08-08 15:02:10] -hspike modeling of Mast cells\n",
      "INFO [2023-08-08 15:04:47] validating infercnv_obj\n",
      "INFO [2023-08-08 15:04:47] normalizing counts matrix by depth\n",
      "INFO [2023-08-08 15:04:47] Using specified normalization factor: 3896.000000\n",
      "INFO [2023-08-08 15:07:03] \n",
      "\n",
      "\tSTEP 04: log transformation of data\n",
      "\n",
      "INFO [2023-08-08 15:07:03] transforming log2xplus1()\n",
      "INFO [2023-08-08 15:07:18] -mirroring for hspike\n",
      "INFO [2023-08-08 15:07:18] transforming log2xplus1()\n",
      "INFO [2023-08-08 15:09:33] \n",
      "\n",
      "\tSTEP 08: removing average of reference data (before smoothing)\n",
      "\n",
      "INFO [2023-08-08 15:09:33] ::subtract_ref_expr_from_obs:Start inv_log=FALSE, use_bounds=TRUE\n",
      "INFO [2023-08-08 15:09:33] subtracting mean(normal) per gene per cell across all data\n",
      "INFO [2023-08-08 15:09:54] -subtracting expr per gene, use_bounds=TRUE\n",
      "INFO [2023-08-08 15:10:31] -mirroring for hspike\n",
      "INFO [2023-08-08 15:10:31] ::subtract_ref_expr_from_obs:Start inv_log=FALSE, use_bounds=TRUE\n",
      "INFO [2023-08-08 15:10:31] subtracting mean(normal) per gene per cell across all data\n",
      "INFO [2023-08-08 15:10:35] -subtracting expr per gene, use_bounds=TRUE\n",
      "INFO [2023-08-08 15:14:30] \n",
      "\n",
      "\tSTEP 09: apply max centered expression threshold: 3\n",
      "\n",
      "INFO [2023-08-08 15:14:30] ::process_data:setting max centered expr, threshold set to: +/-:  3\n",
      "INFO [2023-08-08 15:14:45] -mirroring for hspike\n",
      "INFO [2023-08-08 15:14:45] ::process_data:setting max centered expr, threshold set to: +/-:  3\n",
      "INFO [2023-08-08 15:18:31] \n",
      "\n",
      "\tSTEP 10: Smoothing data per cell by chromosome\n",
      "\n",
      "INFO [2023-08-08 15:18:31] smooth_by_chromosome: chr: 1\n",
      "INFO [2023-08-08 15:19:54] smooth_by_chromosome: chr: 2\n",
      "INFO [2023-08-08 15:21:10] smooth_by_chromosome: chr: 3\n",
      "INFO [2023-08-08 15:22:22] smooth_by_chromosome: chr: 4\n",
      "INFO [2023-08-08 15:23:29] smooth_by_chromosome: chr: 5\n",
      "INFO [2023-08-08 15:24:35] smooth_by_chromosome: chr: 6\n",
      "INFO [2023-08-08 15:25:48] smooth_by_chromosome: chr: 7\n",
      "INFO [2023-08-08 15:26:57] smooth_by_chromosome: chr: X\n",
      "INFO [2023-08-08 15:28:03] smooth_by_chromosome: chr: 8\n",
      "INFO [2023-08-08 15:29:07] smooth_by_chromosome: chr: 9\n",
      "INFO [2023-08-08 15:30:16] smooth_by_chromosome: chr: 11\n",
      "INFO [2023-08-08 15:31:28] smooth_by_chromosome: chr: 10\n",
      "INFO [2023-08-08 15:32:34] smooth_by_chromosome: chr: 12\n",
      "INFO [2023-08-08 15:33:43] smooth_by_chromosome: chr: 13\n",
      "INFO [2023-08-08 15:34:41] smooth_by_chromosome: chr: 14\n",
      "INFO [2023-08-08 15:35:43] smooth_by_chromosome: chr: 15\n",
      "INFO [2023-08-08 15:36:45] smooth_by_chromosome: chr: 16\n",
      "INFO [2023-08-08 15:37:51] smooth_by_chromosome: chr: 17\n",
      "INFO [2023-08-08 15:39:03] smooth_by_chromosome: chr: 18\n",
      "INFO [2023-08-08 15:39:48] smooth_by_chromosome: chr: 20\n",
      "INFO [2023-08-08 15:40:51] smooth_by_chromosome: chr: 19\n",
      "INFO [2023-08-08 15:42:08] smooth_by_chromosome: chr: Y\n",
      "INFO [2023-08-08 15:42:13] smooth_by_chromosome: chr: 22\n",
      "INFO [2023-08-08 15:43:15] smooth_by_chromosome: chr: 21\n",
      "INFO [2023-08-08 15:43:54] -mirroring for hspike\n",
      "INFO [2023-08-08 15:43:54] smooth_by_chromosome: chr: chrA\n",
      "INFO [2023-08-08 15:43:55] smooth_by_chromosome: chr: chr_0\n",
      "INFO [2023-08-08 15:43:57] smooth_by_chromosome: chr: chr_B\n",
      "INFO [2023-08-08 15:43:58] smooth_by_chromosome: chr: chr_0pt5\n",
      "INFO [2023-08-08 15:43:59] smooth_by_chromosome: chr: chr_C\n",
      "INFO [2023-08-08 15:44:00] smooth_by_chromosome: chr: chr_1pt5\n",
      "INFO [2023-08-08 15:44:02] smooth_by_chromosome: chr: chr_D\n",
      "INFO [2023-08-08 15:44:03] smooth_by_chromosome: chr: chr_2pt0\n",
      "INFO [2023-08-08 15:44:04] smooth_by_chromosome: chr: chr_E\n",
      "INFO [2023-08-08 15:44:06] smooth_by_chromosome: chr: chr_3pt0\n",
      "INFO [2023-08-08 15:44:07] smooth_by_chromosome: chr: chr_F\n",
      "INFO [2023-08-08 15:48:06] \n",
      "\n",
      "\tSTEP 11: re-centering data across chromosome after smoothing\n",
      "\n",
      "INFO [2023-08-08 15:48:06] ::center_smooth across chromosomes per cell\n",
      "INFO [2023-08-08 15:49:19] -mirroring for hspike\n",
      "INFO [2023-08-08 15:49:19] ::center_smooth across chromosomes per cell\n",
      "INFO [2023-08-08 15:53:31] \n",
      "\n",
      "\tSTEP 12: removing average of reference data (after smoothing)\n",
      "\n",
      "INFO [2023-08-08 15:53:31] ::subtract_ref_expr_from_obs:Start inv_log=FALSE, use_bounds=TRUE\n",
      "INFO [2023-08-08 15:53:31] subtracting mean(normal) per gene per cell across all data\n",
      "INFO [2023-08-08 15:53:51] -subtracting expr per gene, use_bounds=TRUE\n",
      "INFO [2023-08-08 15:54:28] -mirroring for hspike\n",
      "INFO [2023-08-08 15:54:28] ::subtract_ref_expr_from_obs:Start inv_log=FALSE, use_bounds=TRUE\n",
      "INFO [2023-08-08 15:54:28] subtracting mean(normal) per gene per cell across all data\n",
      "INFO [2023-08-08 15:54:32] -subtracting expr per gene, use_bounds=TRUE\n",
      "INFO [2023-08-08 15:57:48] \n",
      "\n",
      "\tSTEP 14: invert log2(FC) to FC\n",
      "\n",
      "INFO [2023-08-08 15:57:48] invert_log2(), computing 2^x\n",
      "INFO [2023-08-08 15:57:59] -mirroring for hspike\n",
      "INFO [2023-08-08 15:57:59] invert_log2(), computing 2^x\n",
      "INFO [2023-08-08 16:02:08] \n",
      "\n",
      "\tSTEP 15: computing tumor subclusters via leiden\n",
      "\n",
      "INFO [2023-08-08 16:02:08] define_signif_tumor_subclusters(p_val=0.1\n",
      "INFO [2023-08-08 16:02:15] define_signif_tumor_subclusters(), tumor: malignant_SMC01\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  GPN2, SFN, ARID1A, RPS6KA1, NUDC, CD52, SLC9A1, SH3BGRL3, TMEM57, TMEM50A \n",
      "\t   LDLRAP1, RSRP1, MTFR1L, SEPN1, ZNF593, HMGN2, STMN1, TMEM222, SYF2, SYTL1 \n",
      "\t   RUNX3, WASF2, CLIC4, IFI6, SRRM1, STX12, RCAN3, SRSF10, PPP1R8, PNRC2 \n",
      "Negative:  PARD6B, PTPN1, CEBPB, TMEM189, SNAI1, RNF114, SPATA2, B4GALT5, ZFAS1, YTHDF1 \n",
      "\t   ZNFX1, DDX27, ARFGAP1, STAU1, PPDPF, CSE1L, PTK6, ARFGEF2, HELZ2, SULF2 \n",
      "\t   KIF3B, STMN3, NCOA3, POFUT1, ARFRP1, ZMYND8, PLAGL2, SLC2A4RG, TM9SF4, TP53RK \n",
      "PC_ 2 \n",
      "Positive:  SPP1, PPM1K, SPARCL1, PYURF, NUDT9, FAM13A, HSD17B11, GPRIN3, CDS1, PDLIM5 \n",
      "\t   MRPS18C, COQ2, RAP1GDS1, PLAC8, TSPAN5, COPS4, EIF4E, THAP9-AS1, METAP1, SEC31A \n",
      "\t   ADH5, ENOPH1, ADH1B, HNRNPDL, ADH1C, HNRNPD, DAPP1, RASGEF1B, ANTXR2, LAMTOR3 \n",
      "Negative:  OVOL1, RNASEH2C, CFL1, MUS81, KAT5, RELA, MAP3K11, SIPA1, EHBP1L1, FAM89B \n",
      "\t   EIF1AD, SSSCA1, BANF1, LTBP3, MRPL49, CAPN1, SYVN1, SF3B2, FAU, CDC42EP2 \n",
      "\t   SCYL1, TM7SF2, DPF2, RAB1B, FRMD8, VPS51, MALAT1, YIF1A, NEAT1, ZFPL1 \n",
      "PC_ 3 \n",
      "Positive:  ETS2, PSMG1, DYRK1A, BRWD1, DSCR3, TTC3, HMGN1, PIGP, CBR1, WRB \n",
      "\t   RUNX1, BACE2, RCAN1, FAM3B, SLC5A3, MX2, MRPS6, MX1, ATP5O, TMPRSS2 \n",
      "\t   CRYZL1, ABCG1, SON, TFF3, GART, TFF2, TFF1, TMEM50B, SLC37A1, NDUFV3 \n",
      "Negative:  MYO19, GGNBP2, DHRS11, ZNHIT3, AATF, CCL4L2, CCL3L3, DUSP14, CCL4, CCL18 \n",
      "\t   CCL3, CCL15, SYNRG, CCL5, TAF15, AP2B1, DDX52, RP11-1094M14.11, MRPL45, NLE1 \n",
      "\t   MLLT6, RFFL, CISD3, ZNF830, PSMB3, CCL13, CWC25, CCL8, RPL23, LASP1 \n",
      "PC_ 4 \n",
      "Positive:  ODC1, NOL10, HPCAL1, PDIA6, RRM2, IAH1, CPSF3, PQLC3, KLF11, ADAM17 \n",
      "\t   YWHAQ, ITGB1BP1, ROCK2, ASAP2, LPIN1, MBOAT2, FAM84A, KIDINS220, DDX1, ID2 \n",
      "\t   FAM49A, AC092580.4, VSNL1, RPS7, SMC6, RDH14, RNASEH1, TTC32, ADI1, LAPTM4A \n",
      "Negative:  S100A9, S100A8, S100A11, S100A6, S100A10, S100A4, THEM4, S100A2, MRPL9, S100A16 \n",
      "\t   SNX27, S100A14, TUFT1, S100A13, CGN, POGZ, CHTOP, PSMB4, SNAPIN, ILF2 \n",
      "\t   GATAD2B, SLC39A1, JTB, RAB13, RPS27, NEAT1, MALAT1, FRMD8, SCYL1, LTBP3 \n",
      "PC_ 5 \n",
      "Positive:  C21orf2, UBE2G2, PFKL, SUMO3, PTTG1IP, ITGB2, AGPAT3, FAM207A, RRP1, COL18A1 \n",
      "\t   CSTB, PDXK, COL6A1, RRP1B, COL6A2, U2AF1, SPATC1L, PKNOX1, LSS, NDUFV3 \n",
      "\t   YBEY, SLC37A1, PRMT2, TFF1, TFF2, TFF3, ABCG1, TMPRSS2, MX1, MX2 \n",
      "Negative:  HLA-B, HLA-C, MICA, DDR1, IER3, FLOT1, NFKBIL1, TUBB, PPP1R18, TNF \n",
      "\t   C6orf136, LTB, MRPS18B, PPP1R10, ABCF1, LST1, GNL1, HLA-E, AIF1, RPP21 \n",
      "\t   HCG18, PRRC2A, TRIM26, BAG6, TRIM15, C6orf47, TRIM31, PPP1R11, GPANK1, ZNRD1 \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2023-08-08 16:02:25] define_signif_tumor_subclusters(), tumor: malignant_SMC02\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  ADAMDEC1, ADAM28, PPP2R2A, SLC25A37, R3HCC1, BNIP3L, TNFRSF10B, DPYSL2, EGR3, PTK2B \n",
      "\t   BIN3, PDLIM2, EPHX2, SORBS3, CLU, PPP3CC, CCDC25, SLC39A14, REEP4, DUSP4 \n",
      "\t   DOK2, SARAF, ATP6V1B2, RP11-489E7.4, INTS10, LEPROTL1, RP11-51J9.5, ASAH1, DCTN6, RBPMS \n",
      "Negative:  KIF3B, NOL4L, ASXL1, COMMD7, CBFA2T2, CDK5RAP1, POFUT1, MAPRE1, NECAB3, CHMP4B \n",
      "\t   RALY, PLAGL2, EIF2S2, TM9SF4, AHCY, PDRG1, DSTN, SNRPB2, KIF16B, TPX2 \n",
      "\t   ITCH, RRBP1, NDUFAF5, ESF1, JAG1, SNX5, BCL2L1, MKKS, DYNLRB1, PLCB4 \n",
      "PC_ 2 \n",
      "Positive:  APOBEC3G, RPL3, APOBEC3C, CBX6, MIEF1, SUN2, ATF4, GTPBP1, JOSD1, RPS19BP1 \n",
      "\t   SNX6, BAZ1A, EAPP, SRP54, SPTSSA, FAM177A1, TOMM22, EGLN3, PPP2R3C, ARHGAP5 \n",
      "\t   HECTD1, PSMA6, STRN3, SCFD1, STXBP6, GZMB, NFKBIA, CBY1, RALGAPA1, GZMH \n",
      "Negative:  EREG, MTHFD2L, AREG, CXCL2, PARM1, CXCL3, RCHY1, G3BP2, USO1, CXCL5 \n",
      "\t   NAAA, PPBP, SDAD1, CXCL1, CXCL8, CXCL9, CXCL10, RASSF6, NUP54, ANKRD17 \n",
      "\t   SCARB2, DCK, SHROOM3, GRSF1, SEPT11, UTP3, CCNI, CCNG2, JCHAIN, CXCL13 \n",
      "PC_ 3 \n",
      "Positive:  OSTC, CCDC109B, RPL34, CASP6, PLA2G12A, GAR1, TIFA, HADH, LARP7, PML \n",
      "\t   ISLR, LOXL1, UBL7, CAMK2D, CD276, CLK3, CSK, NPTN, SCAMP2, SGMS2 \n",
      "\t   SNHG8, ADPGK, COX5A, ARIH1, RPP25, SEC24D, HEXA, COMMD4, PKM, PAPSS1 \n",
      "Negative:  NEAT1, FRMD8, MALAT1, EHD1, ATG2A, DPF2, SF1, ARL2, CDC42EP2, RPS6KA4 \n",
      "\t   SAC3D1, SCYL1, CAPN1, ZFPL1, CCDC88B, MRPL49, VPS51, AP003774.1, LTBP3, SYVN1 \n",
      "\t   TM7SF2, PRDX5, SSSCA1, FAU, TRMT112, ESRRA, FAM89B, GPR137, BAD, EHBP1L1 \n",
      "PC_ 4 \n",
      "Positive:  DUSP23, ACKR1, TAGLN2, IFI16, ETV3, SH2D2A, LINC01133, PRCC, HDGF, IGSF8 \n",
      "\t   MRPL24, PEA15, DCAF8, COPA, NCSTN, SLC1A5, AP2S1, STRN4, PRKD2, ARHGAP35 \n",
      "\t   CD48, ILVBL, NDUFB7, CALM3, BRD4, TMEM160, AKAP8, TECR, AKAP8L, DNAJB1 \n",
      "Negative:  SLC5A6, PREB, KHK, EMILIN1, OST4, TMEM214, SLC35F6, HADHB, HADHA, RAB10 \n",
      "\t   ASXL2, PTRHD1, NCOA1, HDAC3, PCDH1, GNPDA1, DIAPH1, ITSN2, TAF7, NDFIP1 \n",
      "\t   TP53I3, ZMAT2, ARHGAP26, SF3B6, HARS, ETF1, NR3C1, HSPA9, PAIP2, UBXN2A \n",
      "PC_ 5 \n",
      "Positive:  NKG7, ETFB, KLK10, KLK6, KLK1, SCFD1, STXBP6, STRN3, GZMB, HECTD1 \n",
      "\t   C19orf48, ARHGAP5, PPP2R1A, CLEC11A, EGLN3, GZMH, JOSD2, SPTSSA, EAPP, EMC10 \n",
      "\t   SDR39U1, NR1H2, SNX6, MYH14, DHRS1, BAZ1A, VRK3, ATF5, SRP54, RABGGTA \n",
      "Negative:  SMAD3, AAGAB, C15orf61, RPL4, SNAPC5, PIAS1, MAP2K1, TIPIN, CLN6, RAB11A \n",
      "\t   ANP32A, DENND4A, VWA9, GLCE, HACD3, DPP8, RPLP1, CLPX, TLE3, PDCD7 \n",
      "\t   UACA, MTFMT, PKM, SPG21, OAZ2, HEXA, KIAA0101, ARIH1, PPIB, ADPGK \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2023-08-08 16:02:31] define_signif_tumor_subclusters(), tumor: malignant_SMC03\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  DRAM2, LRIF1, EPS8L3, CD53, DENND2D, AHCYL1, GSTM3, LAMTOR5, WDR77, ATP5F1 \n",
      "\t   GSTM4, C1orf162, RAP1A, GNAI3, CTTNBP2NL, PSMA5, CAPZA1, RHOC, SORT1, SARS \n",
      "\t   KIAA1324, TMEM167B, ABT1, RP1-313I6.12, BTN2A1, ZNF165, ZSCAN16-AS1, BTN3A2, TRIM27, TAF13 \n",
      "Negative:  NDUFV1, GSTP1, NUDT8, CDK2AP2, AIP, TMEM134, TFF1, TFF2, TFF3, ABCG1 \n",
      "\t   AAAS, SP1, TMPRSS2, C12orf10, PRR13, PCBP2, PFDN5, MX1, TARBP2, ITGB7 \n",
      "\t   IGFBP6, ATF7, MX2, EIF4B, ATP5G2, FAM3B, KRT18, CALCOCO1, KRT8, BACE2 \n",
      "PC_ 2 \n",
      "Positive:  KCNN4, PLAUR, PVR, BCL3, CBLC, BCAM, ZNF428, PVRL2, ETHE1, TOMM40 \n",
      "\t   CEACAM1, APOE, APOC1, PAFAH1B3, CLPTM1, ERF, RELB, GSK3A, CLASRP, DEDD2 \n",
      "\t   GEMIN7, RABAC1, PPP1R37, ARHGEF1, TRAPPC6A, CD3EAP, CD79A, ERCC1, RPS19, FOSB \n",
      "Negative:  ITGAV, ZC3H15, TFPI, NCKAP1, COL3A1, FRZB, DNAJC10, COL5A2, SSFA2, UBE2E3 \n",
      "\t   SLC40A1, CWC22, ASNSD1, SESTD1, PLEKHA3, ORMDL1, PRKRA, HIBCH, AGPS, INPP1 \n",
      "\t   NFE2L2, MFSD6, HNRNPA3, NAB1, MTX2, GLS, STAT1, STAT4, MYO1B, NABP1 \n",
      "PC_ 3 \n",
      "Positive:  NUP54, SCARB2, CXCL10, SHROOM3, CXCL9, SEPT11, SDAD1, NAAA, CCNI, USO1 \n",
      "\t   G3BP2, CCNG2, RCHY1, CXCL13, PARM1, CNOT6L, MRPL1, AREG, ANXA3, EREG \n",
      "\t   ANTXR2, MTHFD2L, CXCL2, RASGEF1B, CXCL3, CXCL5, HNRNPD, PPBP, HNRNPDL, CXCL1 \n",
      "Negative:  FAU, TM7SF2, VPS51, SYVN1, ZFPL1, MRPL49, SAC3D1, CAPN1, RPL3, APOBEC3G \n",
      "\t   ARL2, APOBEC3C, CDC42EP2, MIEF1, CBX6, ATG2A, ATF4, DPF2, RPS19BP1, SUN2 \n",
      "\t   FRMD8, EHD1, GTPBP1, NEAT1, JOSD1, SF1, MALAT1, TOMM22, SCYL1, CBY1 \n",
      "PC_ 4 \n",
      "Positive:  DUOX2, DUOXA2, SORD, C15orf48, TRIM69, BLOC1S6, B2M, SQRDL, EIF3J, DUT \n",
      "\t   CASC4, FBN1, MFAP1, EID1, SERF2, COPS2, PDIA3, CKMT1A, GALK2, CKMT1B \n",
      "\t   DPF2, CDC42EP2, FRMD8, CAPN1, NEAT1, MALAT1, MRPL49, SCYL1, CCNDBP1, FGF7 \n",
      "Negative:  RPL18, SULT2B1, LMTK3, FUT2, CYTH2, BCAT2, GRWD1, PPP1R15A, KDELR1, NUCB1 \n",
      "\t   EMP3, BAX, FTL, CARD8, RUVBL2, SEPW1, GLTSCR2, SNRNP70, TRPM4, NAPA \n",
      "\t   CD37, C5AR1, PIH1D1, INAFM1, BBC3, SAE1, TMEM160, ARHGAP35, AP2S1, SLC1A5 \n",
      "PC_ 5 \n",
      "Positive:  RP11-12G12.7, PCCA, TPP2, TM9SF2, GPR183, TEX30, UBAC2, EFNB2, STK24, ARGLU1 \n",
      "\t   FARP1, ABHD13, TNFSF13B, IPO5, RP11-54H7.4, IRS2, RAP2A, COL4A1, MBNL2, COL4A2 \n",
      "\t   DNAJC3, RAB20, CARKD, SPRY2, CARS2, ING1, NDFIP2, ANKRD10, RBM26, ARHGEF7 \n",
      "Negative:  SCNM1, VPS72, PSMD4, CDC42SE1, PI4KB, C1orf56, CERS2, CTSK, CTSS, GOLPH3L \n",
      "\t   CGN, RPS27, TUFT1, ENSA, RAB13, SLC39A1, JTB, TPM3, GATAD2B, ILF2 \n",
      "\t   SNAPIN, C1orf43, SNX27, CHTOP, UBAP2L, MCL1, S100A13, MRPL9, HAX1, UBE2Q1 \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2023-08-08 16:02:39] define_signif_tumor_subclusters(), tumor: malignant_SMC04\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  SAMD4B, PAF1, GMFG, PAK4, MED29, MRPS12, ZFP36, RPS16, NFKBIB, SIRT2 \n",
      "\t   SUPT5H, HNRNPL, TIMM50, ECH1, LGALS4, FBL, ACTN4, EIF3K, FCGBP, FAM98C \n",
      "\t   PSMD8, PSMC4, KCNK6, YIF1B, AKT2, C19orf33, PPP1R14A, PLD3, SPINT2, ZNF146 \n",
      "Negative:  SQRDL, DUT, BLOC1S6, FBN1, C15orf48, EID1, DUOXA2, COPS2, GALK2, DUOX2 \n",
      "\t   FGF7, SORD, DTWD1, TRIM69, GABPB1, B2M, CLDN23, DEFA5, GABPB1-AS1, DEFA6 \n",
      "\t   MFHAS1, AGPAT5, MSRA, CTD-2336O2.1, FDFT1, CLN8, VPS37A, MTUS1, CNOT7, CTSB \n",
      "PC_ 2 \n",
      "Positive:  IL6, TOMM7, CDCA7L, KLHL7, NUPL2, GPNMB, ITGB8, MALSU1, TRA2A, CTA-293F17.1 \n",
      "\t   CYCS, NFE2L3, HNRNPA2B1, CBX3, MACC1, SNX10, SKAP2, TWISTNB, HOTAIRM1, HOXA9 \n",
      "\t   SNX13, HOXA10, AHR, HIBADH, TAX1BP1, AGR3, CREB5, AGR2, CPVL, TSPAN13 \n",
      "Negative:  EIF1AD, BANF1, SF3B2, MUS81, RAB1B, YIF1A, CFL1, CD248, OVOL1, BRMS1 \n",
      "\t   MRPL11, RNASEH2C, DPP3, KAT5, ZDHHC24, CTSF, RELA, CCS, MAP3K11, SIPA1 \n",
      "\t   EHBP1L1, FAM89B, SSSCA1, CHMP4A, IRF9, PSME2, NEDD8, GMPR2, EMC9, LTBP3 \n",
      "PC_ 3 \n",
      "Positive:  STK17B, RP11-347P5.1, PGAP1, ANKRD44, SF3B1, COQ10B, HSPD1, HSPE1, MOB4, SATB2 \n",
      "\t   C2orf69, C2orf47, SPATS2L, BZW1, CLK1, PPIL3, FAM126B, KIAA1324, TMEM167B, TAF13 \n",
      "\t   SARS, GPSM2, STXBP3, SORT1, PSMA5, GNAI3, GSTM4, GSTM3, EPS8L3, AHCYL1 \n",
      "Negative:  PVRL2, TOMM40, BCAM, APOE, APOC1, CBLC, CLPTM1, RELB, CLASRP, BCL3 \n",
      "\t   GEMIN7, PPP1R37, PVR, TRAPPC6A, CD3EAP, ERCC1, KCNN4, FOSB, VASP, PLAUR \n",
      "\t   OPA3, ZNF428, EML2, SNRPD2, ETHE1, SYMPK, CEACAM1, FOXA3, IGFL2, PAFAH1B3 \n",
      "PC_ 4 \n",
      "Positive:  TK1, SYNGR2, TMC8, ALYREF, TMC6, ANAPC11, PCYT2, SEPT9, SIRT7, SEC14L1 \n",
      "\t   MAFG, MFSD11, PYCR1, SRSF2, METTL23, ASPSCR1, JMJD6, CUL4A, PCID2, ATP11A \n",
      "\t   LAMP1, ARHGEF7, ANKRD10, CARS2, ING1, GRTP1, CARKD, RP11-54H7.4, RAB20, IRS2 \n",
      "Negative:  GSTK1, TRBC2, AC093673.5, PRSS2, ZYX, TRBC1, TCAF1, ARHGEF35, SSBP1, ARHGEF5 \n",
      "\t   MRPS33, CUL1, NDUFB2, EZH2, PDIA4, ADCK2, ATP6V0E2, MKRN1, RARRES2, REPIN1 \n",
      "\t   SLC37A3, GIMAP7, KDM7A, GIMAP4, PARP12, FXYD3, GRAMD1A, FXYD1, TMEM176B, UBA2 \n",
      "PC_ 5 \n",
      "Positive:  EIF3A, FAM45A, SFXN4, PRDX3, RGS10, TIAL1, BAG3, MCMBP, SEC23IP, ACSL5 \n",
      "\t   WDR11, NSMCE4A, SHOC2, TACC2, BBIP1, PLEKHA1, PDCD4, ARL2, SAC3D1, ATG2A \n",
      "\t   EHD1, ZFPL1, SF1, HTRA1, SMC3, VPS51, RPS6KA4, TM7SF2, CCDC88B, FAU \n",
      "Negative:  SNW1, SLIRP, SEL1L, SPTLC2, GPR65, ZC3H14, AHSA1, FOXN3, PSMC1, GSTZ1 \n",
      "\t   CALM1, GPATCH2L, IFT43, CCDC88C, C14orf1, BATF, PPP4R3A, FOS, TMED10, TC2N \n",
      "\t   ACYP1, TRIP11, EIF2B2, ATXN3, PGF, NDUFB1, DLST, CPSF2, YLPM1, ISCA2 \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2023-08-08 16:02:46] define_signif_tumor_subclusters(), tumor: malignant_SMC05\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  RIPK2, NBN, CPNE3, DECR1, RMDN1, OTUD6B-AS1, WWP1, CA2, FAM92A1, C8orf59 \n",
      "\t   LRRCC1, PDP1, CHMP4C, CDH17, GEM, ZFAND1, KIAA1429, ESRP1, FABP5, NDUFAF6 \n",
      "\t   PLEKHF2, PAG1, SDC2, PTDSS1, CPQ, LAPTM4B, MATN2, UQCRB, MTDH, RPL30 \n",
      "Negative:  EIF1AD, MUS81, CFL1, BANF1, OVOL1, SF3B2, RNASEH2C, KAT5, RELA, RAB1B \n",
      "\t   MAP3K11, DHRS1, RABGGTA, SIPA1, YIF1A, TINF2, GMPR2, NEDD8, EHBP1L1, CHMP4A \n",
      "\t   CD248, IRF9, FAM89B, PSME2, EMC9, BRMS1, SSSCA1, PSME1, LTBP3, DCAF11 \n",
      "PC_ 2 \n",
      "Positive:  SIVA1, AKT1, INF2, C14orf2, CEP170B, ZFYVE21, APOPT1, KLC1, BAG5, TRMT61A \n",
      "\t   CKB, MARK3, EIF5, TNFAIP2, RGS5, PBX1, MGST3, HSD17B7, CDC42BPB, UAP1 \n",
      "\t   ATF6, UHMK1, DUSP12, AMN, FCGR2B, FCGR3A, HSPA6, SFT2D2, TRAF3, FCGR2A \n",
      "Negative:  CD63, BLOC1S1, ORMDL2, PYM1, ITGA5, RAB5B, HNRNPA1, COPZ1, RPS26, CBX5 \n",
      "\t   ERBB3, SMUG1, PA2G4, CALCOCO1, RPL41, ATP5G2, ESYT1, ATF7, MYL6B, TARBP2 \n",
      "\t   MYL6, PCBP2, SMARCC2, PRR13, RNF41, SP1, NABP2, AAAS, SLC39A5, CS \n",
      "PC_ 3 \n",
      "Positive:  SLC35E1, ATP5D, MIDN, C19orf24, STK11, CIRBP, CHERP, KLF2, GPX4, POLR2E \n",
      "\t   HMHA1, AP1M1, CNN2, TMEM259, FAM32A, WDR18, HSH2D, ARID3A, AKAP17A, RAB8A \n",
      "\t   DHRSX, ASMTL, ZBED1, SLC25A6, MXRA5, ARSD, R3HDM4, ARSE, CD99, PRKX \n",
      "Negative:  CBX3, SNX10, HNRNPA2B1, NFE2L3, CYCS, SKAP2, TRA2A, HOTAIRM1, HOXA9, MALSU1 \n",
      "\t   HOXA10, GPNMB, HIBADH, TAX1BP1, NUPL2, KLHL7, CREB5, TOMM7, CPVL, IL6 \n",
      "\t   CDCA7L, PRR15, ITGB8, GGCT, CTA-293F17.1, GARS, MACC1, AQP1, LSM5, TWISTNB \n",
      "PC_ 4 \n",
      "Positive:  AHSA1, GSTZ1, GPATCH2L, SPTLC2, IFT43, SLIRP, C14orf1, SNW1, BATF, SEL1L \n",
      "\t   FOS, GPR65, TMED10, ACYP1, ZC3H14, DLST, EIF2B2, PGF, FOXN3, PSMC1 \n",
      "\t   CALM1, CCDC88C, PPP4R3A, TC2N, TRIP11, ATXN3, NDUFB1, CPSF2, LGMN, GABPB1 \n",
      "Negative:  GPATCH4, ISG20L2, MRPL24, APOA1BP, HDGF, MEF2D, PRCC, CCT3, GLMP, SH2D2A \n",
      "\t   SMG5, ETV3, PMF1, IFI16, ACKR1, SLC25A44, DUSP23, SEMA4A, TAGLN2, LINC01133 \n",
      "\t   LMNA, IGSF8, RAB25, PEA15, LAMTOR2, DCAF8, SSR2, COPA, ARHGEF2, NCSTN \n",
      "PC_ 5 \n",
      "Positive:  PSMC1, CALM1, PPP4R3A, CCDC88C, TRIP11, TC2N, FOXN3, ATXN3, ZC3H14, NDUFB1 \n",
      "\t   GPR65, CPSF2, LGMN, SEL1L, ITPK1, MOAP1, TMEM251, C14orf142, SNW1, BTBD7 \n",
      "\t   DDX24, SLIRP, IFI27L1, SPTLC2, AHSA1, GSTZ1, GPATCH2L, C3orf52, ATG3, CD96 \n",
      "Negative:  CTSA, ACOT8, TNNC2, UBE2C, PLTP, PCIF1, DNTTIP1, WFDC2, PIGT, MMP9 \n",
      "\t   SYS1, CD40, SDC4, SLC35C2, SLPI, PI3, TP53RK, ZMYND8, STK4, NCOA3 \n",
      "\t   TOMM34, SULF2, YWHAB, ARFGEF2, CSE1L, ADA, STAU1, PKIG, DDX27, SERINC3 \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2023-08-08 16:02:49] define_signif_tumor_subclusters(), tumor: malignant_SMC06\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  RNF2, FAM129A, EDEM3, C1orf21, TSEN15, ARPC5, SMG7, LAMC2, ZNF281, C1orf106 \n",
      "\t   TMEM9, DTWD1, FGF7, GABPB1, GALK2, COPS2, GABPB1-AS1, PHLDA3, EID1, FBN1 \n",
      "\t   USP8, DUT, SQRDL, SPPL2A, BLOC1S6, C15orf48, FAM107B, CSRP1, LYSMD2, DUOXA2 \n",
      "Negative:  RTKN, WDR54, MOGS, DCTN1, AUP1, MTHFD2, MOB1A, HTRA2, BOLA3, HK2 \n",
      "\t   DGUOK, STAMBP, POLE4, DUSP11, MRPL19, TPRKB, GCFC2, CCT7, REG1B, PRADC1 \n",
      "\t   REG1A, SPR, REG3A, ZNF638, MPHOSPH10, SUCLG1, MCEE, NAGK, TRABD2A, TEX261 \n",
      "PC_ 2 \n",
      "Positive:  MUS81, CFL1, OVOL1, RNASEH2C, KAT5, RELA, MAP3K11, SIPA1, EIF1AD, EHBP1L1 \n",
      "\t   FAM89B, BANF1, SSSCA1, SF3B2, LTBP3, RAB1B, SCYL1, MALAT1, YIF1A, NEAT1 \n",
      "\t   FRMD8, DPF2, CD248, CDC42EP2, CAPN1, MRPL49, SYVN1, BRMS1, FAU, TM7SF2 \n",
      "Negative:  S100A11, S100A10, S100A9, THEM4, S100A8, MRPL9, S100A6, SNX27, S100A4, TUFT1 \n",
      "\t   S100A2, CGN, S100A16, SPP1, POGZ, S100A14, PPM1K, SPARCL1, PYURF, PSMB4 \n",
      "\t   S100A13, NUDT9, FAM13A, GPRIN3, HSD17B11, SELENBP1, CHTOP, PDLIM5, CDS1, RAP1GDS1 \n",
      "PC_ 3 \n",
      "Positive:  MRPL45, MLLT6, DDX52, CISD3, SYNRG, DUSP14, PSMB3, AATF, CWC25, DHRS11 \n",
      "\t   GGNBP2, RPL23, MYO19, LASP1, ZNHIT3, CCL4L2, RPL19, CCL3L3, CCL4, MED1 \n",
      "\t   CCL3, CCL18, CDK12, CCL15, CCL5, TAF15, PPP1R1B, AP2B1, RP11-1094M14.11, STARD3 \n",
      "Negative:  NAB1, GLS, MFSD6, STAT1, STAT4, INPP1, MYO1B, NABP1, HIBCH, SLC39A10 \n",
      "\t   STK17B, RP11-347P5.1, ORMDL1, PGAP1, ASNSD1, ANKRD44, SF3B1, SLC40A1, COQ10B, COL5A2 \n",
      "\t   HSPD1, HSPE1, COL3A1, MOB4, SATB2, TFPI, C2orf69, C2orf47, ITGAV, SPATS2L \n",
      "PC_ 4 \n",
      "Positive:  PIH1D1, CD37, TRPM4, SNRNP70, RUVBL2, FTL, BAX, NUCB1, PPP1R15A, BCAT2 \n",
      "\t   FUT2, RPL18, SULT2B1, LMTK3, CYTH2, GRWD1, KDELR1, EMP3, CARD8, AKT1S1 \n",
      "\t   SEPW1, TBC1D17, GLTSCR2, IL4I1, NUP62, NAPA, ATF5, C5AR1, VRK3, INAFM1 \n",
      "Negative:  AP1B1, EWSR1, XBP1, HSCB, PITPNB, TPST2, HPS4, KIAA1671, SNRPD3, GUCD1 \n",
      "\t   CABIN1, DDT, MIF, DERL3, SMARCB1, MMP11, CHCHD10, VPREB3, IGLC7, IGLC6 \n",
      "\t   IGLC3, IGLC2, IGLL5, IGLV3-1, IGLV6-57, MAPK1, SDF2L1, YDJC, UBE2L3, THAP7 \n",
      "PC_ 5 \n",
      "Positive:  PMVK, PBXIP1, ADAR, SHC1, UBE2Q1, HAX1, CKS1B, UBAP2L, FLAD1, C1orf43 \n",
      "\t   ZBTB7B, TPM3, ADAM15, RPS27, EFNA4, RAB13, EFNA3, JTB, EFNA1, SLC39A1 \n",
      "\t   SLC50A1, GATAD2B, DPM3, ILF2, KRTCAP2, SNAPIN, MUC1, CHTOP, MTX1, S100A13 \n",
      "Negative:  RFFL, ZNF830, NLE1, CCL13, CCL8, CCL11, RP11-1094M14.11, CCL2, AP2B1, TMEM98 \n",
      "\t   TAF15, MYO1D, CCL5, PSMD11, CCL15, ZNF207, CCL18, RHOT1, CCL3, CCL4 \n",
      "\t   SUZ12, CCL3L3, CCL4L2, UTP6, ZNHIT3, MYO19, COPRS, GGNBP2, DHRS11, AATF \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2023-08-08 16:02:54] define_signif_tumor_subclusters(), tumor: malignant_SMC07\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  POFUT1, PLAGL2, KIF3B, TM9SF4, ASXL1, NOL4L, PDRG1, COMMD7, MAPRE1, CDK5RAP1 \n",
      "\t   CBFA2T2, TPX2, NECAB3, CHMP4B, BCL2L1, RALY, ID1, EIF2S2, MGME1, SNX5 \n",
      "\t   PET117, RRBP1, SEC23B, LINC00493, DTD1, AHCY, DSTN, HM13, RP1-122P22.4, RIN2 \n",
      "Negative:  HMGCS2, REG4, SEC22B, PHGDH, FAM46C, RP11-14N7.2, MAN1A2, CH17-373J23.1, CD2, RNF115 \n",
      "\t   CD58, RBM8A, ATP1A1, SIKE1, POLR3GL, CSDE1, TXNIP, NRAS, BCAS2, CHD1L \n",
      "\t   TRIM33, PDE4DIP, OLFML3, HIPK1, CH17-189H20.1, BCL2L15, LINC00869, PTPN22, HIST2H2BE, RSBN1 \n",
      "PC_ 2 \n",
      "Positive:  CHCHD10, MMP11, VPREB3, IGLC7, SMARCB1, DERL3, IGLC6, IGLC3, MIF, IGLC2 \n",
      "\t   DDT, IGLL5, CABIN1, IGLV3-1, GUCD1, IGLV6-57, MAPK1, SNRPD3, SDF2L1, KIAA1671 \n",
      "\t   YDJC, HPS4, UBE2L3, TPST2, THAP7, PITPNB, CRKL, SNAP29, HSCB, MED15 \n",
      "Negative:  CD52, HMGN2, SH3BGRL3, RPS6KA1, ZNF593, ARID1A, SFN, GPN2, NUDC, STMN1 \n",
      "\t   SLC9A1, MTFR1L, TMEM222, SEPN1, SYTL1, WASF2, LDLRAP1, IFI6, STX12, TMEM57 \n",
      "\t   RPL41, PPP1R8, TMEM50A, THEMIS2, ESYT1, PA2G4, RPL19, MED1, RPA2, CDK12 \n",
      "PC_ 3 \n",
      "Positive:  ZNF791, ACP5, WDR83OS, DHPS, ELOF1, C19orf43, ECSIT, ASNA1, HOOK2, PRKCSH \n",
      "\t   JUNB, PRDX2, TMEM205, RNASEH2A, DNASE2, GCDH, LDLR, FARSA, CALR, SMARCA4 \n",
      "\t   RAD23A, GADD45GIP1, C19orf52, NFIX, TRMT1, YIPF2, NACC1, CARM1, STX10, IER2 \n",
      "Negative:  ADH5, METAP1, ADH1B, EIF4E, TSPAN5, ADH1C, RAP1GDS1, DAPP1, PDLIM5, LAMTOR3 \n",
      "\t   GPRIN3, FAM13A, DNAJB14, PYURF, SPP1, PPM1K, SPARCL1, H2AFZ, NUDT9, PPP3CA \n",
      "\t   FLJ20021, SLC39A8, HSD17B11, NFKB1, MANBA, CDS1, UBE2D3, CISD2, MRPS18C, BDH2 \n",
      "PC_ 4 \n",
      "Positive:  S100A6, S100A8, S100A9, S100A4, S100A11, S100A2, S100A16, S100A10, S100A14, THEM4 \n",
      "\t   S100A13, MRPL9, SNX27, CHTOP, SNAPIN, TUFT1, ILF2, CGN, GATAD2B, POGZ \n",
      "\t   SLC39A1, PSMB4, JTB, SELENBP1, RAB13, RPS27, PI4KB, TPM3, PSMD4, C1orf43 \n",
      "Negative:  FAU, SYVN1, MRPL49, CAPN1, TM7SF2, CDC42EP2, DPF2, VPS51, FRMD8, ZFPL1 \n",
      "\t   ORMDL3, NEAT1, MIEN1, ERBB2, STARD3, PSMD3, PPP1R1B, MED24, SAC3D1, THRA \n",
      "\t   MSL1, CDK12, MALAT1, CASC3, MED1, ARL2, WIPF2, RPL19, GZMK, NDUFS4 \n",
      "PC_ 5 \n",
      "Positive:  GZMH, GZMB, SDR39U1, STXBP6, DHRS1, SCFD1, RABGGTA, STRN3, TINF2, HECTD1 \n",
      "\t   GMPR2, ARHGAP5, NEDD8, EGLN3, CHMP4A, SPTSSA, IRF9, EAPP, PSME2, SNX6 \n",
      "\t   EMC9, BAZ1A, PSME1, SRP54, DCAF11, PCK2, FAM177A1, DHRS4L2, PPP2R3C, DHRS4 \n",
      "Negative:  EMP3, CARD8, KDELR1, SEPW1, GLTSCR2, GRWD1, CYTH2, NAPA, LMTK3, C5AR1 \n",
      "\t   SULT2B1, INAFM1, RPL18, BBC3, FUT2, SAE1, BCAT2, TMEM160, PPP1R15A, ARHGAP35 \n",
      "\t   NUCB1, AP2S1, BAX, SLC1A5, FTL, STRN4, RUVBL2, PRKD2, SNRNP70, CALM3 \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2023-08-08 16:03:06] define_signif_tumor_subclusters(), tumor: malignant_SMC08\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  RRBP1, SNX5, MGME1, DSTN, PET117, SEC23B, SNRPB2, LINC00493, KIF16B, DTD1 \n",
      "\t   NDUFAF5, RP1-122P22.4, ESF1, RIN2, JAG1, NAA20, MKKS, CRNKL1, PLCB4, RALGAPA2 \n",
      "\t   TMX4, BMP2, XRN2, FERMT1, FOXA2, CRLS1, THBD, TRMT6, CD93, NXT1 \n",
      "Negative:  MALAT1, SCYL1, LTBP3, NEAT1, SSSCA1, FRMD8, FAM89B, DPF2, EHBP1L1, CDC42EP2 \n",
      "\t   SIPA1, MAP3K11, CAPN1, RELA, MRPL49, SYVN1, KAT5, FAU, RNASEH2C, TM7SF2 \n",
      "\t   VPS51, OVOL1, ZFPL1, CFL1, SAC3D1, ARL2, MUS81, ATG2A, EHD1, SF1 \n",
      "PC_ 2 \n",
      "Positive:  SC5D, ARHGEF12, OAF, TRIM29, SORL1, HSPA8, THY1, TBRG1, MCAM, SIAE \n",
      "\t   VSIG2, DPAGT1, ESAM, H2AFX, TMEM218, EI24, HMBS, STT3A, HYOU1, RPUSD4 \n",
      "\t   SRPRA, SLC37A4, DCPS, TRAPPC4, ST3GAL4, RPS25, ETS1, DDX6, ARHGAP32, ARCN1 \n",
      "Negative:  SCARB2, NUP54, SHROOM3, CXCL10, SEPT11, CXCL9, CCNI, CCNG2, SDAD1, CXCL13 \n",
      "\t   CNOT6L, NAAA, MRPL1, ANXA3, ANTXR2, USO1, RASGEF1B, G3BP2, HNRNPD, RCHY1 \n",
      "\t   PARM1, HNRNPDL, ENOPH1, AREG, SEC31A, EREG, THAP9-AS1, MTHFD2L, COPS4, CXCL2 \n",
      "PC_ 3 \n",
      "Positive:  LAMTOR3, DAPP1, DNAJB14, ADH1C, H2AFZ, ADH1B, PPP3CA, ADH5, FLJ20021, METAP1 \n",
      "\t   SLC39A8, UBE2D3, NFKB1, CISD2, MANBA, BDH2, EIF4E, PPA2, AIMP1, TSPAN5 \n",
      "\t   PAPSS1, SGMS2, RAP1GDS1, HADH, PDLIM5, RPL34, OSTC, GPRIN3, CCDC109B, CASP6 \n",
      "Negative:  PPP5C, CALM3, IGFL2, PRKD2, FOXA3, STRN4, SYMPK, SLC1A5, SNRPD2, EML2 \n",
      "\t   AP2S1, OPA3, ARHGAP35, VASP, TMEM160, FOSB, SAE1, ERCC1, BBC3, CD3EAP \n",
      "\t   INAFM1, C5AR1, TRAPPC6A, NAPA, PPP1R37, GLTSCR2, GEMIN7, SEPW1, CLASRP, CARD8 \n",
      "PC_ 4 \n",
      "Positive:  RRP7A, LINC01315, POLDIP3, RNU12, CYB5R3, TCF20, ARFGAP3, PACSIN2, NDUFA6, BIK \n",
      "\t   MCAT, TSPO, TTLL12, SAMM50, PARVB, PRR5, NUP50, KIAA0930, FAM118A, FBLN1 \n",
      "\t   ATXN10, PRR34-AS1, TTC38, CERK, CTA-29F11.1, SMARCB1, DERL3, MMP11, MIF, CHCHD10 \n",
      "Negative:  C14orf142, TMEM251, BTBD7, MOAP1, DDX24, ITPK1, LGMN, IFI27L1, CPSF2, IFI27 \n",
      "\t   NDUFB1, IFI27L2, ATXN3, TRIP11, SERPINA1, TC2N, PPP4R3A, CCDC88C, CALM1, RNF115 \n",
      "\t   DICER1, RBM8A, CH17-373J23.1, RP11-14N7.2, SEC22B, POLR3GL, PSMC1, REG4, TXNIP, HMGCS2 \n",
      "PC_ 5 \n",
      "Positive:  PIH1D1, AKT1S1, CD37, TBC1D17, IL4I1, NUP62, TRPM4, ATF5, VRK3, MYH14 \n",
      "\t   SNRNP70, NR1H2, EMC10, RUVBL2, JOSD2, SF3B4, MTMR11, HIST2H2BE, APH1A, PLEKHO1 \n",
      "\t   ANP32E, LINC00869, C1orf54, MRPS21, RPRD2, CLEC11A, CH17-189H20.1, FTL, ECM1, PDE4DIP \n",
      "Negative:  SIVA1, INF2, AKT1, C14orf2, CEP170B, ZFYVE21, APOPT1, KLC1, BAG5, TRMT61A \n",
      "\t   CKB, MARK3, EIF5, TNFAIP2, CDC42BPB, AMN, TRAF3, RCOR1, ANKRD9, CINP \n",
      "\t   HSP90AA1, DYNC1H1, IGLC3, IGLC6, IGLC2, IGLC7, IGLL5, VPREB3, CHCHD10, IGLV3-1 \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2023-08-08 16:03:13] define_signif_tumor_subclusters(), tumor: malignant_SMC09\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  RPS21, ADRM1, OSBPL2, MTG2, PSMA7, SLCO4A1, PRELID3B, ATP5E, MRGBP, CTSZ \n",
      "\t   OGFR, NELFCD, GNAS, STX16, VAPB, COL9A3, RAB22A, PMEPA1, PCK1, RBM38 \n",
      "\t   RAE1, DIDO1, RTFDC1, AURKA, FAM210B, PFDN4, GID8, ZNF217, TSHZ2, SLC17A9 \n",
      "Negative:  OLFML3, HIPK1, TRIM33, BCL2L15, BCAS2, NRAS, PTPN22, CSDE1, SIKE1, RSBN1 \n",
      "\t   ATP1A1, CD58, CD2, SLC16A1, MAN1A2, RHOC, FAM46C, CAPZA1, PHGDH, CTTNBP2NL \n",
      "\t   HMGCS2, RAP1A, C1orf162, REG4, ATP5F1, WDR77, SEC22B, DENND2D, DRAM2, RP11-14N7.2 \n",
      "PC_ 2 \n",
      "Positive:  S100A11, S100A10, THEM4, S100A9, MRPL9, SNX27, S100A8, TUFT1, CGN, S100A6 \n",
      "\t   POGZ, PSMB4, S100A4, SELENBP1, S100A2, PI4KB, S100A16, PSMD4, S100A14, VPS72 \n",
      "\t   S100A13, SCNM1, CCL4, CCL3, CCL3L3, CHTOP, CCL4L2, CCL18, CDC42SE1, ZNHIT3 \n",
      "Negative:  DDX6, ARCN1, KMT2A, ATP5L, UBE4A, CD3G, CD3D, GZMA, GZMK, NDUFS4 \n",
      "\t   DHX29, MOCS2, SKIV2L2, PLPP1, ITGA2, CD3E, IL6ST, PELO, GPBP1, ITGA1 \n",
      "\t   PLK2, PDE4D, PARP8, NDUFAF2, EMB, SMIM15, MRPS30, MPZL2, ZSWIM6, PAIP1 \n",
      "PC_ 3 \n",
      "Positive:  ADAR, UBE2Q1, HAX1, PMVK, UBAP2L, PBXIP1, C1orf43, TPM3, SHC1, RPS27 \n",
      "\t   CKS1B, RAB13, JTB, FLAD1, SLC39A1, ZBTB7B, GATAD2B, ADAM15, ILF2, EFNA4 \n",
      "\t   SNAPIN, EFNA3, CHTOP, EFNA1, S100A13, SLC50A1, S100A14, DPM3, S100A16, KRTCAP2 \n",
      "Negative:  SEPT11, CCNI, SHROOM3, CCNG2, CXCL13, SCARB2, CNOT6L, NUP54, MRPL1, CXCL10 \n",
      "\t   CXCL9, ANXA3, SDAD1, NAAA, ANTXR2, USO1, RASGEF1B, G3BP2, HNRNPD, RCHY1 \n",
      "\t   PARM1, HNRNPDL, AREG, ENOPH1, EREG, MTHFD2L, GRSF1, DCK, UTP3, ANKRD17 \n",
      "PC_ 4 \n",
      "Positive:  HMGN1, BRWD1, WRB, PSMG1, BACE2, ETS2, DYRK1A, FAM3B, DSCR3, TTC3 \n",
      "\t   MX2, PIGP, MX1, CBR1, TMPRSS2, RUNX1, RCAN1, ABCG1, SLC5A3, TFF3 \n",
      "\t   TFF2, MRPS6, TFF1, ATP5O, SLC37A1, CRYZL1, SON, NDUFV3, GART, PKNOX1 \n",
      "Negative:  FUT2, BCAT2, RPL18, PPP1R15A, NUCB1, SULT2B1, BAX, FTL, LMTK3, RUVBL2 \n",
      "\t   CYTH2, GRWD1, SNRNP70, KDELR1, EMP3, TRPM4, CARD8, SEPW1, CD37, GLTSCR2 \n",
      "\t   NAPA, PIH1D1, C5AR1, INAFM1, BBC3, SAE1, TMEM160, ARHGAP35, AP2S1, SLC1A5 \n",
      "PC_ 5 \n",
      "Positive:  TFF2, TFF3, ABCG1, TFF1, SLC37A1, NDUFV3, TMPRSS2, PKNOX1, U2AF1, MX1 \n",
      "\t   RRP1B, MX2, PDXK, CSTB, RRP1, AGPAT3, PFKL, C21orf2, FAM3B, UBE2G2 \n",
      "\t   SUMO3, PTTG1IP, ITGB2, BACE2, FAM207A, COL18A1, WRB, COL6A1, COL6A2, SPATC1L \n",
      "Negative:  KPNA6, TXLNA, KHDRBS1, HACD3, EIF3I, HNRNPA1, COPZ1, PTP4A2, ITGA5, BLOC1S1 \n",
      "\t   CBX5, DPP8, LCK, MARCKSL1, COL16A1, CD63, HDAC1, BSDC1, SMUG1, ORMDL2 \n",
      "\t   CLPX, PEF1, SCAMP2, CALCOCO1, ZBTB8OS, PYM1, TINAGL1, PDCD7, ATP5G2, COX5A \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2023-08-08 16:03:23] define_signif_tumor_subclusters(), tumor: malignant_SMC10\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  C16orf91, GNPTG, TSR3, UBE2I, TPSAB1, TPSB2, RPUSD1, HAGHL, FAM173A, METRN \n",
      "\t   JMJD8, STUB1, RHOT2, MSLN, FAM195A, C16orf13, CAPN15, NOTCH1, DECR2, EGFL7 \n",
      "\t   AGPAT2, NME4, SNHG7, LCN15, TMEM8A, TMEM141, MRPL28, RABL6, PHPT1, AXIN1 \n",
      "Negative:  SEPT11, CCNI, SHROOM3, CCNG2, SCARB2, CXCL13, NUP54, CNOT6L, CXCL10, CXCL9 \n",
      "\t   MRPL1, SDAD1, ANXA3, ANTXR2, NAAA, RASGEF1B, USO1, G3BP2, HNRNPD, RCHY1 \n",
      "\t   HNRNPDL, CCL3, CCL18, PARM1, CCL15, CCL4, AREG, CCL5, ENOPH1, TAF15 \n",
      "PC_ 2 \n",
      "Positive:  RP11-138A9.2, LINC-PINT, RP11-138A9.1, MKLN1, AC058791.1, PODXL, AC016831.7, CHCHD3, EXOC4, MEST \n",
      "\t   AKR1B1, KLHDC10, BPGM, CALD1, UBE2H, CNOT4, TNPO3, C7orf73, MTPN, ATP6V1F \n",
      "\t   CALU, PTN, METTL2B, AC090498.1, CREB3L2, HILPDA, TRIM24, IMPDH1, ZC3HAV1, RBM28 \n",
      "Negative:  IGFL2, FOXA3, SYMPK, PPP5C, SNRPD2, EML2, OPA3, CALM3, PRKD2, VASP \n",
      "\t   STRN4, FOSB, ERCC1, SLC1A5, CD3EAP, AP2S1, TRAPPC6A, PPP1R37, GEMIN7, ARHGAP35 \n",
      "\t   CLASRP, PVRL2, TOMM40, BCAM, CBLC, APOE, TMEM160, BCL3, RELB, APOC1 \n",
      "PC_ 3 \n",
      "Positive:  RABAC1, ARHGEF1, DEDD2, CD79A, GSK3A, RPS19, ERF, CEACAM6, CEACAM5, PAFAH1B3 \n",
      "\t   CEACAM7, ATP5SL, CEACAM1, EXOSC5, B9D2, ETHE1, TGFB1, HNRNPUL1, ZNF428, AXL \n",
      "\t   PLAUR, CYP2S1, MALAT1, KCNN4, SCYL1, NEAT1, LTBP3, EGLN2, PVR, SSSCA1 \n",
      "Negative:  RPL34, OSTC, CCDC109B, CASP6, HADH, SGMS2, PLA2G12A, PAPSS1, GAR1, TIFA \n",
      "\t   AIMP1, LARP7, CAMK2D, SNHG8, PPA2, SEC24D, BDH2, USP53, CISD2, C4orf3 \n",
      "\t   UBE2D3, FABP2, MANBA, MAD2L1, NFKB1, TNIP3, SLC39A8, ANXA5, FLJ20021, PPP3CA \n",
      "PC_ 4 \n",
      "Positive:  CLEC11A, C19orf48, KLK1, JOSD2, KLK6, EMC10, NR1H2, KLK10, MYH14, ETFB \n",
      "\t   VRK3, ATF5, NKG7, NUP62, PPP2R1A, IL4I1, ZNF331, TBC1D17, MYADM, AKT1S1 \n",
      "\t   NDUFA3, TFPT, PRPF31, CNOT3, LENG1, TMC4, MBOAT7, TSEN34, RPS9, LAIR1 \n",
      "Negative:  ACP5, ZNF791, ELOF1, WDR83OS, DHPS, C19orf43, ECSIT, ASNA1, HOOK2, JUNB \n",
      "\t   PRKCSH, PRDX2, RNASEH2A, TMEM205, DNASE2, LDLR, GCDH, SMARCA4, FARSA, C19orf52 \n",
      "\t   YIPF2, CALR, CARM1, RAD23A, TMED1, GADD45GIP1, DNM2, NFIX, QTRT1, TRMT1 \n",
      "PC_ 5 \n",
      "Positive:  GZMH, GZMB, SDR39U1, DHRS1, STXBP6, RABGGTA, SCFD1, TINF2, STRN3, GMPR2 \n",
      "\t   HECTD1, NEDD8, ARHGAP5, CHMP4A, EGLN3, IRF9, SPTSSA, PSME2, EAPP, EMC9 \n",
      "\t   SNX6, PSME1, BAZ1A, DCAF11, SRP54, FAM177A1, PCK2, PPP2R3C, DHRS4L2, PSMA6 \n",
      "Negative:  IMPDH1, RBM28, HILPDA, SND1, METTL2B, CALU, ARF5, CALD1, BPGM, ZNF800 \n",
      "\t   CNOT4, ATP6V1F, AKR1B1, C7orf73, WASL, TNPO3, EXOC4, MTPN, NDUFA5, UBE2H \n",
      "\t   FAM3C, ING3, PTN, TSPAN12, CHCHD3, LSM8, KLHDC10, AC090498.1, CFTR, PODXL \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2023-08-08 16:03:31] define_signif_tumor_subclusters(), tumor: malignant_SMC11\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  AHCY, ITCH, DYNLRB1, MAP1LC3A, PIGU, EIF2S2, NCOA6, RALY, TP53INP2, CHMP4B \n",
      "\t   ACSS2, GSS, NECAB3, TRPC4AP, CBFA2T2, EDEM2, PROCR, CDK5RAP1, MMP24-AS1, EIF6 \n",
      "\t   MAPRE1, UQCC1, COMMD7, ERGIC3, NOL4L, SPAG4, ASXL1, CPNE1, PMEPA1, RAB22A \n",
      "Negative:  CH17-189H20.1, PDE4DIP, LINC00869, CHD1L, HIST2H2BE, TXNIP, SF3B4, MTMR11, POLR3GL, PLEKHO1 \n",
      "\t   RBM8A, ANP32E, RNF115, APH1A, CH17-373J23.1, C1orf54, RP11-14N7.2, MRPS21, SEC22B, REG4 \n",
      "\t   RPRD2, SETD2, CCDC12, TDGF1, KIF9, CCRL2, CXCR6, HMGCS2, KLHL18, ECM1 \n",
      "PC_ 2 \n",
      "Positive:  HMGN2, RPS6KA1, ARID1A, CD52, SFN, GPN2, NUDC, SH3BGRL3, SLC9A1, TMEM222 \n",
      "\t   SYTL1, ZNF593, WASF2, IFI6, STMN1, STX12, MTFR1L, PPP1R8, SEPN1, THEMIS2 \n",
      "\t   LDLRAP1, RPA2, TMEM57, DNAJC8, TMEM50A, ATPIF1, RSRP1, SESN2, SYF2, PHACTR4 \n",
      "Negative:  DERL3, MIF, DDT, SMARCB1, MMP11, CABIN1, GUCD1, CHCHD10, SNRPD3, KIAA1671 \n",
      "\t   VPREB3, HPS4, IGLC7, IGLV3-1, IGLC6, IGLV6-57, TPST2, IGLL5, MAPK1, IGLC3 \n",
      "\t   IGLC2, SDF2L1, PITPNB, YDJC, HSCB, UBE2L3, THAP7, XBP1, CRKL, SNAP29 \n",
      "PC_ 3 \n",
      "Positive:  CAPN1, CDC42EP2, DPF2, MRPL49, FRMD8, SYVN1, NEAT1, FAU, MALAT1, TM7SF2 \n",
      "\t   SCYL1, LTBP3, VPS51, SSSCA1, ZFPL1, FAM89B, SAC3D1, EHBP1L1, SIPA1, ARL2 \n",
      "\t   MAP3K11, ATG2A, RELA, EHD1, KAT5, RNASEH2C, SF1, OVOL1, RPS6KA4, CFL1 \n",
      "Negative:  ADH5, ADH1B, EIF4E, ADH1C, METAP1, DAPP1, TSPAN5, RAP1GDS1, PPM1K, LAMTOR3 \n",
      "\t   PDLIM5, SPP1, DNAJB14, PYURF, GPRIN3, SPARCL1, H2AFZ, FAM13A, NUDT9, PPP3CA \n",
      "\t   FLJ20021, HSD17B11, SLC39A8, CDS1, NFKB1, MANBA, MRPS18C, UBE2D3, CISD2, PAPSS1 \n",
      "PC_ 4 \n",
      "Positive:  CD37, TRPM4, PIH1D1, SNRNP70, RUVBL2, FTL, BAX, NUCB1, PPP1R15A, BCAT2 \n",
      "\t   FUT2, RPL18, SULT2B1, LMTK3, CYTH2, GRWD1, KDELR1, EMP3, CARD8, SEPW1 \n",
      "\t   GLTSCR2, NAPA, C5AR1, AKT1S1, INAFM1, TBC1D17, BBC3, IL4I1, SAE1, NUP62 \n",
      "Negative:  C14orf1, BATF, IFT43, GPATCH2L, PGF, EIF2B2, FOS, TMED10, ACYP1, GSTZ1 \n",
      "\t   DLST, YLPM1, AHSA1, ISCA2, SPTLC2, NPC2, ABCD4, SLIRP, ELMSAN1, SNW1 \n",
      "\t   NUMB, PSEN1, SEL1L, RBM25, GPR65, SIPA1L1, MED6, SYNJ2BP, COX16, SRSF5 \n",
      "PC_ 5 \n",
      "Positive:  RPL19, LASP1, RPL23, CWC25, DDX52, MRPL45, SYNRG, MLLT6, CISD3, PSMB3 \n",
      "\t   DUSP14, AATF, DHRS11, GGNBP2, MYO19, ZNHIT3, CCL4L2, CCL3L3, CCL4, CCL3 \n",
      "\t   THRA, CCL18, MSL1, CCL15, CASC3, CCL5, TAF15, AP2B1, WIPF2, RP11-1094M14.11 \n",
      "Negative:  SERF2, PDIA3, MFAP1, CKMT1A, CASC4, CKMT1B, EIF3J, CCNDBP1, HAUS2, B2M \n",
      "\t   TRIM69, SNAP23, SORD, ZNF106, TMEM87A, EHD4, DUOX2, RTF1, NUSAP1, DUOXA2 \n",
      "\t   CHP1, RHOV, C15orf48, BLOC1S6, SPINT1, SQRDL, DUT, RP11-532F12.5, FBN1, PPP1R14D \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2023-08-08 16:03:37] define_signif_tumor_subclusters(), tumor: malignant_SMC14\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  ACTN4, LGALS4, ECH1, EIF3K, HNRNPL, FAM98C, SIRT2, NFKBIB, PSMD8, MRPS12 \n",
      "\t   KCNK6, PAK4, GMFG, YIF1B, SAMD4B, C19orf33, PAF1, PPP1R14A, MED29, ZFP36 \n",
      "\t   SPINT2, RPS16, ZNF146, SUPT5H, COX7A1, TIMM50, CAPNS1, FBL, TBCB, FCGBP \n",
      "Negative:  TSPAN5, ADH1B, EIF4E, ADH5, RAP1GDS1, METAP1, PDLIM5, ADH1C, GPRIN3, DAPP1 \n",
      "\t   FAM13A, SLC39A8, FLJ20021, NFKB1, PPP3CA, LAMTOR3, DNAJB14, MANBA, H2AFZ, PYURF \n",
      "\t   UBE2D3, CISD2, PPM1K, BDH2, PPA2, SPP1, AIMP1, PAPSS1, SPARCL1, SGMS2 \n",
      "PC_ 2 \n",
      "Positive:  SCFD1, STXBP6, STRN3, GZMB, HECTD1, GZMH, ARHGAP5, SDR39U1, EGLN3, DHRS1 \n",
      "\t   SPTSSA, RABGGTA, EAPP, TINF2, SNX6, GMPR2, BAZ1A, NEDD8, SRP54, FAM177A1 \n",
      "\t   CHMP4A, PPP2R3C, IRF9, PSMA6, PSME2, NFKBIA, RALGAPA1, EMC9, MBIP, SEC23A \n",
      "Negative:  CXCL2, CXCL3, MTHFD2L, CXCL5, PPBP, EREG, CXCL1, AREG, CXCL8, PARM1 \n",
      "\t   RASSF6, RCHY1, ANKRD17, G3BP2, USO1, DCK, NAAA, GRSF1, SDAD1, UTP3 \n",
      "\t   CXCL9, CXCL10, JCHAIN, NUP54, YTHDC1, SCARB2, UBA6, YOD1, PFKFB2, C4BPB \n",
      "PC_ 3 \n",
      "Positive:  GNG2, TMX1, SAV1, MAP4K5, ATP5S, C14orf166, ARF6, ERO1A, NEMF, PSMC6 \n",
      "\t   GNPNAT1, KLHDC2, BMP4, DNAAF2, CDKN3, MGAT2, CNIH1, RPL36AL, GMFB, LRR1 \n",
      "\t   GCH1, RPS29, MAPK1IP1L, MIS18BP1, LGALS3, FKBP3, FBXO34, FBXO33, LAMTOR5, CD53 \n",
      "Negative:  TMEM160, ARHGAP35, SAE1, AP2S1, BBC3, SLC1A5, INAFM1, C5AR1, STRN4, NAPA \n",
      "\t   GLTSCR2, PRKD2, SEPW1, CARD8, EMP3, CALM3, PPP5C, IGFL2, KDELR1, FOXA3 \n",
      "\t   GRWD1, SYMPK, CYTH2, SNRPD2, LMTK3, EML2, SULT2B1, OPA3, RPL18, VASP \n",
      "PC_ 4 \n",
      "Positive:  RPL18, SULT2B1, FUT2, LMTK3, BCAT2, PPP1R15A, BAX, NUCB1, FTL, RUVBL2 \n",
      "\t   SNRNP70, CYTH2, TRPM4, CD37, PIH1D1, GRWD1, AKT1S1, KDELR1, TBC1D17, EMP3 \n",
      "\t   IL4I1, CARD8, NUP62, SEPW1, SAE1, ATF5, ARHGAP35, TMEM160, SLC1A5, BBC3 \n",
      "Negative:  IP6K1, GMPPB, RBM6, APEH, RBM5, DAG1, TCTA, GNAI2, RHOA, SEMA3B \n",
      "\t   GPX1, LAMB2, IFRD2, QRICH1, IMPDH2, HYAL2, NDUFAF3, DALRD3, WDR6, P4HTM \n",
      "\t   TUSC2, ARIH2, PRKAR2A, RASSF1, IP6K2, NPRL2, SLC26A6, UQCRC1, CYB561D2, SHISA5 \n",
      "PC_ 5 \n",
      "Positive:  EIF2B2, PGF, ACYP1, TMED10, DLST, FOS, YLPM1, ISCA2, BATF, NPC2 \n",
      "\t   C14orf1, IFT43, ABCD4, GPATCH2L, ELMSAN1, GSTZ1, NUMB, AHSA1, PSEN1, SPTLC2 \n",
      "\t   RBM25, SLIRP, SIPA1L1, SNW1, MED6, SYNJ2BP, SEL1L, COX16, GPR65, SRSF5 \n",
      "Negative:  BAZ1A, SRP54, SNX6, EAPP, FAM177A1, SPTSSA, PPP2R3C, EGLN3, ARHGAP5, PSMA6 \n",
      "\t   HECTD1, STRN3, NFKBIA, SCFD1, RALGAPA1, STXBP6, MBIP, GTPBP6, SLC25A6, GZMB \n",
      "\t   ASMTL, AKAP17A, SEC23A, DHRSX, ZBED1, GZMH, CD99, PNN, ARSD, ARSE \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2023-08-08 16:03:43] define_signif_tumor_subclusters(), tumor: malignant_SMC15\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  ADAMDEC1, PPP2R2A, ADAM28, BNIP3L, SLC25A37, DPYSL2, PTK2B, R3HCC1, EPHX2, KLK1 \n",
      "\t   KLK6, C19orf48, TNFRSF10B, CLEC11A, KLK10, JOSD2, ETFB, EMC10, NR1H2, NKG7 \n",
      "\t   MYH14, CLU, VRK3, PPP2R1A, EGR3, ATF5, NUP62, IL4I1, ZNF331, TBC1D17 \n",
      "Negative:  COX6C, STK3, POLR2K, HRSP12, RPL30, MATN2, LAPTM4B, MTDH, CPQ, SDC2 \n",
      "\t   PTDSS1, UQCRB, NDUFB9, PLEKHF2, SQLE, TATDN1, ANXA13, FAM91A1, RNF139, KIAA0196 \n",
      "\t   FBXO32, ATAD2, ZHX1, WDYHV1, C8orf76, NDUFAF6, NSMCE2, DERL1, MRPL13, ENPP2 \n",
      "PC_ 2 \n",
      "Positive:  PKN1, GIPC1, NDUFB7, DDX39A, ILVBL, TECR, BRD4, DNAJB1, ADGRE5, AKAP8 \n",
      "\t   SAMD1, AKAP8L, IL27RA, TPM4, CC2D1A, CTD-3252C9.4, C19orf53, CCDC130, RAB8A, IER2 \n",
      "\t   STX10, HSH2D, NACC1, TRMT1, FAM32A, NFIX, AP1M1, GADD45GIP1, KLF2, RAD23A \n",
      "Negative:  PTPN22, RSBN1, BCL2L15, SLC16A1, RHOC, HIPK1, CAPZA1, CTTNBP2NL, OLFML3, RAP1A \n",
      "\t   C1orf162, TRIM33, BCAS2, ATP5F1, NRAS, WDR77, CSDE1, DENND2D, SIKE1, DRAM2 \n",
      "\t   ATP1A1, LRIF1, CD58, CD53, CD2, LAMTOR5, AHCYL1, MAN1A2, EPS8L3, GSTM3 \n",
      "PC_ 3 \n",
      "Positive:  CFL1, OVOL1, RNASEH2C, KAT5, RELA, MAP3K11, MUS81, SIPA1, EHBP1L1, FAM89B \n",
      "\t   SSSCA1, DCXR, GPS1, STRA13, ASPSCR1, DUS1L, PYCR1, LTBP3, FASN, MAFG \n",
      "\t   SCYL1, SIRT7, CCDC57, MALAT1, PCYT2, SLC16A3, NEAT1, ANAPC11, FRMD8, DPF2 \n",
      "Negative:  EFNB2, TEX30, ARGLU1, TNFSF13B, ABHD13, RP11-54H7.4, TPP2, IRS2, COL4A1, COL4A2 \n",
      "\t   PCCA, RAB20, RP11-12G12.7, CARKD, TM9SF2, CARS2, GPR183, ING1, UBAC2, ANKRD10 \n",
      "\t   STK24, FARP1, ARHGEF7, IPO5, ATP11A, PCID2, RAP2A, CUL4A, MBNL2, LAMP1 \n",
      "PC_ 4 \n",
      "Positive:  S100A16, S100A14, S100A13, S100A2, S100A4, CHTOP, S100A6, SNAPIN, S100A8, ILF2 \n",
      "\t   S100A9, S100A11, S100A10, GATAD2B, THEM4, MRPL9, SLC39A1, SNX27, TUFT1, JTB \n",
      "\t   CGN, RAB13, POGZ, RPS27, PSMB4, TPM3, C1orf43, SELENBP1, UBAP2L, PI4KB \n",
      "Negative:  FAU, TM7SF2, SYVN1, VPS51, MRPL49, ZFPL1, CAPN1, SAC3D1, CDC42EP2, DPF2 \n",
      "\t   ARL2, FRMD8, ATG2A, NEAT1, MALAT1, EHD1, SCYL1, LTBP3, SF1, SSSCA1 \n",
      "\t   RPS6KA4, ORMDL3, PSMD3, FAM89B, MIEN1, ERBB2, MED24, THRA, STARD3, PPP1R1B \n",
      "PC_ 5 \n",
      "Positive:  OXLD1, NPLOC4, CCDC137, ARL16, ACTG1, HGS, SLC38A10, MRPL12, C17orf89, SLC25A10 \n",
      "\t   BAIAP2, FAM195B, CHMP6, P4HB, ENDOV, ARHGDIA, RNF213, ALYREF, EIF4A3, ANAPC11 \n",
      "\t   GAA, PCYT2, CBX4, SIRT7, ENGASE, MAFG, PYCR1, CANT1, ASPSCR1, LGALS3BP \n",
      "Negative:  NCSTN, CD48, IGSF8, COPA, LINC01133, PEA15, DCAF8, TAGLN2, SLAMF7, DUSP23 \n",
      "\t   LY9, ACKR1, IFI16, ITLN1, ETV3, SH2D2A, PRCC, HDGF, MRPL24, F11R \n",
      "\t   ISG20L2, GPATCH4, APOA1BP, TSTD1, MEF2D, ARHGAP30, PFDN2, UFC1, B4GALT3, NDUFS2 \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2023-08-08 16:03:49] define_signif_tumor_subclusters(), tumor: malignant_SMC16\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  CLK3, UBL7, CSK, ISLR, SCAMP2, PML, LOXL1, COX5A, CD276, NPTN \n",
      "\t   ADPGK, RPP25, ARIH1, HEXA, PKM, GLCE, UACA, TLE3, ANP32A, RPLP1 \n",
      "\t   COMMD4, CLN6, PIAS1, C15orf61, MAN2C1, AAGAB, SIN3A, SMAD3, RPL4, SNUPN \n",
      "Negative:  NEAT1, FRMD8, DPF2, MALAT1, CDC42EP2, CAPN1, MRPL49, SCYL1, SYVN1, FAU \n",
      "\t   LTBP3, TM7SF2, VPS51, SSSCA1, ZFPL1, COL18A1, PTTG1IP, COL6A1, ITGB2, FAM207A \n",
      "\t   COL6A2, SUMO3, SPATC1L, UBE2G2, LSS, SAC3D1, C21orf2, YBEY, PRMT2, PFKL \n",
      "PC_ 2 \n",
      "Positive:  ZNF791, WDR83OS, DHPS, C19orf43, ACP5, ASNA1, HOOK2, ELOF1, RNASEH2A, GCDH \n",
      "\t   PRDX2, DNASE2, JUNB, FARSA, CALR, ECSIT, RAD23A, PRKCSH, GADD45GIP1, TMEM205 \n",
      "\t   NFIX, LDLR, TRMT1, SMARCA4, C19orf52, NACC1, YIPF2, STX10, CARM1, TMED1 \n",
      "Negative:  PLAC8, COQ2, MRPS18C, COPS4, CDS1, HSD17B11, THAP9-AS1, NUDT9, SEC31A, SPARCL1 \n",
      "\t   ENOPH1, SPP1, HNRNPDL, HNRNPD, RASGEF1B, ANTXR2, ANXA3, PPM1K, MRPL1, CNOT6L \n",
      "\t   CXCL13, PYURF, CCNG2, CCNI, FAM13A, CD24, SEPT11, QRSL1, C6orf203, GPRIN3 \n",
      "PC_ 3 \n",
      "Positive:  CHD1L, PDE4DIP, CH17-189H20.1, TXNIP, LINC00869, POLR3GL, RBM8A, HIST2H2BE, RNF115, SF3B4 \n",
      "\t   CH17-373J23.1, MTMR11, RP11-14N7.2, SEC22B, PLEKHO1, HMGCS2, REG4, PHGDH, FAM46C, ANP32E \n",
      "\t   MAN1A2, CD2, CD58, APH1A, ATP1A1, SIKE1, CSDE1, NRAS, BCAS2, C1orf54 \n",
      "Negative:  TRIB1, FAM84B, NSMCE2, MYC, KIAA0196, SQLE, NDUFB9, TATDN1, RNF139, FAM91A1 \n",
      "\t   ANXA13, FBXO32, WDYHV1, ATAD2, ZHX1, C8orf76, DERL1, MRPL13, ENPP2, MAL2 \n",
      "\t   MED30, RAD21, UTP23, EIF3H, EBAG9, ENY2, POLR2K, COX6C, STK3, HRSP12 \n",
      "PC_ 4 \n",
      "Positive:  TSPAN3, RCN2, IDH3A, DNAJA4, ETFA, WDR61, IREB2, FBXO22, PSMA4, UBE2Q2 \n",
      "\t   MORF4L1, CTSH, IMP3, TMED3, SNUPN, MTHFS, SIN3A, BCL2A1, MAN2C1, ZFAND6 \n",
      "\t   COMMD4, FAH, RPP25, ABHD17C, COX5A, CEMIP, SCAMP2, MESDC2, CSK, MESDC1 \n",
      "Negative:  PTN, AC090498.1, MTPN, C7orf73, CREB3L2, CNOT4, TRIM24, CALD1, ZC3HAV1, BPGM \n",
      "\t   C7orf55, AKR1B1, LUC7L2, HIPK2, EXOC4, TBXAS1, CHCHD3, PARP12, PODXL, MKLN1 \n",
      "\t   KDM7A, SLC37A3, MKRN1, ADCK2, NDUFB2, MEST, MRPS33, KLHDC10, SSBP1, TRBC1 \n",
      "PC_ 5 \n",
      "Positive:  HIST1H2AC, HIST1H4C, HIST1H2BD, HIST1H1C, HIST1H2BG, HIST1H2AE, TRIM38, BTN3A2, LRRC16A, BTN2A1 \n",
      "\t   ABT1, GMNN, RP1-313I6.12, C6orf62, ZNF165, ACOT13, ZSCAN16-AS1, TDP2, TRIM27, MRS2 \n",
      "\t   HLA-F, SOX4, HLA-A, ID4, JMJD6, MXRA7, METTL23, ZNRD1, ST6GALNAC1, CYGB \n",
      "Negative:  MED1, RPL19, CDK12, PPP1R1B, LASP1, STARD3, RPL23, ERBB2, CWC25, MIEN1 \n",
      "\t   PSMB3, ORMDL3, CISD3, PSMD3, MLLT6, MED24, MRPL45, DDX52, THRA, SYNRG \n",
      "\t   MSL1, DUSP14, CASC3, AATF, DHRS11, WIPF2, GGNBP2, RARA, MYO19, TOP2A \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2023-08-08 16:03:58] define_signif_tumor_subclusters(), tumor: malignant_SMC17\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  MRPS12, NFKBIB, SIRT2, HNRNPL, PAK4, ECH1, GMFG, LGALS4, ACTN4, SAMD4B \n",
      "\t   EIF3K, PAF1, FAM98C, MED29, PSMD8, ZFP36, KCNK6, RPS16, YIF1B, C19orf33 \n",
      "\t   SUPT5H, PIH1D1, PPP1R14A, TIMM50, SPINT2, CD37, ZNF146, FBL, COX7A1, TRPM4 \n",
      "Negative:  CCL3, CCL18, CCL15, CCL4, CCL5, TAF15, AP2B1, CCL3L3, RP11-1094M14.11, NLE1 \n",
      "\t   RFFL, ZNF830, CCL4L2, CCL13, ZNHIT3, CCL8, CCL11, MYO19, CCL2, TMEM98 \n",
      "\t   GGNBP2, MYO1D, DHRS11, PSMD11, SEPT11, ZNF207, SHROOM3, CCNI, SCARB2, CCNG2 \n",
      "PC_ 2 \n",
      "Positive:  TRIOBP, LGALS1, GCAT, H1F0, EIF3L, POLR2F, SH3BP1, MAFF, TMEM184B, CSNK1E \n",
      "\t   GGA1, KDELR3, DDX17, CBY1, LGALS2, TOMM22, JOSD1, GTPBP1, SUN2, CDC42EP1 \n",
      "\t   CBX6, APOBEC3C, RAC2, APOBEC3G, IL2RB, RPL3, MPST, MIEF1, TST, ATF4 \n",
      "Negative:  TGFB1, HNRNPUL1, AXL, CYP2S1, EGLN2, RAB4B, SNRPA, ITPKC, LTBP4, SHKBP1 \n",
      "\t   BLVRB, SERTAD3, SERTAD1, PLD3, AKT2, PSMC4, FCGBP, FBL, TIMM50, SUPT5H \n",
      "\t   RPS16, ZFP36, MED29, PAF1, SAMD4B, GMFG, PAK4, MRPS12, NFKBIB, SIRT2 \n",
      "PC_ 3 \n",
      "Positive:  HIST1H2AC, HIST1H4C, HIST1H2BD, HIST1H2BG, HIST1H1C, HIST1H2AE, TRIM38, BTN3A2, LRRC16A, BTN2A1 \n",
      "\t   GMNN, C6orf62, ABT1, ACOT13, RP1-313I6.12, TDP2, ZNF165, MRS2, ZSCAN16-AS1, SOX4 \n",
      "\t   TRIM27, HLA-F, ID4, HLA-A, DEK, ZNRD1, TPMT, PPP1R11, TRIM31, NUP153 \n",
      "Negative:  RPL41, PA2G4, ERBB3, RPS26, ESYT1, RAB5B, MYL6B, PYM1, MYL6, SMARCC2 \n",
      "\t   ORMDL2, RNF41, CD63, NABP2, BLOC1S1, SLC39A5, ITGA5, COPZ1, CS, HNRNPA1 \n",
      "\t   CBX5, CNPY2, SMUG1, IL23A, CALCOCO1, STAT2, ATP5G2, ATF7, BAZ2A, TARBP2 \n",
      "PC_ 4 \n",
      "Positive:  EIF3J, B2M, CASC4, MFAP1, TRIM69, SERF2, PDIA3, CKMT1A, CKMT1B, CCNDBP1 \n",
      "\t   SORD, HAUS2, SNAP23, DUOX2, ZNF106, DUOXA2, TMEM87A, EHD4, C15orf48, BLOC1S6 \n",
      "\t   SQRDL, RTF1, NUSAP1, CHP1, RHOV, SPINT1, RP11-532F12.5, PPP1R14D, GCHFR, RMDN3 \n",
      "Negative:  NDUFS7, MUM1, ATP5D, C19orf24, MIDN, CIRBP, DAZAP1, STK11, RPS15, C19orf25 \n",
      "\t   GPX4, MBD3, POLR2E, HMHA1, CNN2, TMEM259, WDR18, ARID3A, R3HDM4, MED16 \n",
      "\t   CFD, PTBP1, MISP, FSTL3, RNF126, RHOT2, MSLN, STUB1, FAM195A, JMJD8 \n",
      "PC_ 5 \n",
      "Positive:  TFPI, ITGAV, COL3A1, ZC3H15, COL5A2, NCKAP1, SLC40A1, FRZB, DNAJC10, ASNSD1 \n",
      "\t   SSFA2, ORMDL1, HIBCH, UBE2E3, INPP1, CWC22, MFSD6, SESTD1, NAB1, PLEKHA3 \n",
      "\t   GLS, PRKRA, STAT1, STAT4, MYO1B, NABP1, SLC39A10, STK17B, RP11-347P5.1, COL4A1 \n",
      "Negative:  PLCB3, BAD, PPP1R14B, GPR137, FKBP2, ESRRA, VEGFB, TRMT112, PRDX5, DNAJC4 \n",
      "\t   AP003774.1, NUDT22, CCDC88B, TRPT1, RPS6KA4, FERMT3, SF1, STIP1, EHD1, MACROD1 \n",
      "\t   ATG2A, OTUB1, ARL2, COX8A, ARHGAP5, HECTD1, EGLN3, STRN3, SPTSSA, EAPP \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2023-08-08 16:04:03] define_signif_tumor_subclusters(), tumor: malignant_SMC18\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  GPX3, TNIP1, ANXA6, CCDC69, GM2A, SPARC, ATOX1, G3BP1, LARP1, CNOT8 \n",
      "\t   MRPL22, CYFIP2, ADAM19, THG1L, CLINT1, RNF145, UBLCP1, TTC1, SLU7, PTTG1 \n",
      "\t   ATP10B, CCNG1, NUDCD2, MAT2B, GIT1, SSH2, NSRP1, RARS, CPD, GOSR1 \n",
      "Negative:  C20orf27, ITPA, DDRGK1, MRPS26, PTPRA, IDH3B, NOP56, SNRPB, NSFL1C, FKBP1A \n",
      "\t   SDCBP2, PSMF1, MACC1, TWISTNB, CTA-293F17.1, SNX13, ITGB8, AHR, HNRNPA2B1, CBX3 \n",
      "\t   CDCA7L, AGR3, NFE2L3, CYCS, SNX10, TRA2A, AGR2, SKAP2, IL6, TSPAN13 \n",
      "PC_ 2 \n",
      "Positive:  NT5C3A, FKBP9, KBTBD2, AVL9, LSM5, AQP1, GARS, GGCT, PRR15, CREB5 \n",
      "\t   CPVL, TAX1BP1, HIBADH, HOXA10, HOXA9, HOTAIRM1, SKAP2, SNX10, CBX3, HNRNPA2B1 \n",
      "\t   NFE2L3, ICA1, CYCS, RPA3, NDUFA4, C1GALT1, TRA2A, C7orf26, CCT6A, MALSU1 \n",
      "Negative:  AGO2, CHRAC1, ST3GAL1, PTK2, NDRG1, PTP4A3, SLA, THEM6, PHF20L1, LY6E \n",
      "\t   EFR3A, TOP1MT, ASAP1, GSDMD, FAM49B, NAPRT, PVT1, EEF1D, MYC, TSTA3 \n",
      "\t   FAM84B, FAM83H, TRIB1, NSMCE2, SCRIB, KIAA0196, PUF60, SQLE, NDUFB9, EPPK1 \n",
      "PC_ 3 \n",
      "Positive:  ATP5S, ARF6, MAP4K5, NEMF, SAV1, TMX1, KLHDC2, GNG2, DNAAF2, C14orf166 \n",
      "\t   MGAT2, RPL36AL, ERO1A, LRR1, PSMC6, RPS29, GNPNAT1, MIS18BP1, BMP4, FKBP3 \n",
      "\t   FBXO33, CDKN3, PNN, CNIH1, SEC23A, MBIP, GMFB, RALGAPA1, NFKBIA, GCH1 \n",
      "Negative:  PIH1D1, CD37, TRPM4, SNRNP70, RUVBL2, FTL, RPL18, FUT2, NUCB1, BAX \n",
      "\t   BCAT2, PPP1R15A, SULT2B1, LMTK3, CYTH2, GRWD1, AKT1S1, TBC1D17, KDELR1, IL4I1 \n",
      "\t   EMP3, NUP62, CARD8, ATF5, VRK3, BBC3, SEPW1, SAE1, INAFM1, TMEM160 \n",
      "PC_ 4 \n",
      "Positive:  ENOPH1, HNRNPDL, HNRNPD, SEC31A, RASGEF1B, ANTXR2, THAP9-AS1, ANXA3, COPS4, MRPL1 \n",
      "\t   CNOT6L, PLAC8, CXCL13, CCNG2, COQ2, CCNI, SEPT11, MRPS18C, SHROOM3, SCARB2 \n",
      "\t   CDS1, NUP54, CXCL10, CXCL9, HSD17B11, SDAD1, NAAA, USO1, NUDT9, G3BP2 \n",
      "Negative:  RCC1, TRNAU1AP, PHACTR4, SNHG12, TAF12, SESN2, YTHDF2, ATPIF1, EPB41, DNAJC8 \n",
      "\t   SRSF4, LAPTM5, RPA2, PUM1, THEMIS2, SNRNP40, ZCCHC17, PPP1R8, SERINC2, STX12 \n",
      "\t   TINAGL1, PEF1, IFI6, COL16A1, WASF2, PTP4A2, SYTL1, KHDRBS1, KPNA6, TMEM222 \n",
      "PC_ 5 \n",
      "Positive:  RASSF6, ANKRD17, CXCL8, DCK, JCHAIN, UTP3, CXCL1, GRSF1, PPBP, YTHDC1 \n",
      "\t   CXCL5, UBA6, CXCL3, CXCL2, CENPC, MTHFD2L, IGFBP7, G3BP2, EREG, NAAA \n",
      "\t   USO1, RCHY1, SDAD1, PARM1, AREG, CXCL9, POLR2B, CXCL10, NUP54, NOA1 \n",
      "Negative:  KLK1, KLK6, C19orf48, CLEC11A, KLK10, JOSD2, RPS9, TSEN34, EMC10, ETFB \n",
      "\t   MBOAT7, NR1H2, LAIR1, TMC4, MYH14, LENG1, CNOT3, NKG7, PRPF31, PPP2R1A \n",
      "\t   TFPT, VRK3, NDUFA3, LENG8, ZNF331, MYADM, ATF5, CDC42EP5, NUP62, LILRB4 \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2023-08-08 16:04:11] define_signif_tumor_subclusters(), tumor: malignant_SMC19\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  PKNOX1, NDUFV3, U2AF1, RRP1B, SLC37A1, PDXK, CSTB, TFF1, TFF2, TFF3 \n",
      "\t   RRP1, ABCG1, TMPRSS2, AGPAT3, MX1, PFKL, MX2, C21orf2, FAM3B, UBE2G2 \n",
      "\t   BACE2, SUMO3, PTTG1IP, WRB, ITGB2, HMGN1, FAM207A, BRWD1, COL18A1, PSMG1 \n",
      "Negative:  HINT1, LYRM7, CDC42SE2, ISOC1, LINC01184, SLC12A2, FNIP1, PRRC1, P4HA2, MARCH3 \n",
      "\t   PDLIM4, C5orf56, LMNB1, IRF1, PHAX, RAD50, SEPT8, ALDH7A1, UQCRQ, AFF4 \n",
      "\t   CSNK1G3, ZCCHC10, HSPA4, PPIC, C5orf15, VDAC1, SNX2, TCF7, SKP1, HSD17B4 \n",
      "PC_ 2 \n",
      "Positive:  NUP88, RPAIN, CTC-524C5.2, C1QBP, DERL2, TXNDC17, MED31, XAF1, TM4SF5, ACADVL \n",
      "\t   CXCL16, MED11, PHF23, ARRB2, PELP1, GABARAP, GGT6, SPNS2, CTDNEP1, UBE2G1 \n",
      "\t   ELP5, ATP2A3, CLDN7, NCBP3, EIF5A, HDAC10, PLXNB2, PIM3, TRABD, ITGAE \n",
      "Negative:  SDC2, PTDSS1, CPQ, UQCRB, PLEKHF2, MTDH, LAPTM4B, NDUFAF6, MATN2, ESRP1 \n",
      "\t   RPL30, HRSP12, KIAA1429, STK3, COX6C, GEM, POLR2K, CDH17, PDP1, FAM92A1 \n",
      "\t   OTUD6B-AS1, DECR1, NBN, RIPK2, CPNE3, ENY2, RMDN1, WWP1, CA2, C8orf59 \n",
      "PC_ 3 \n",
      "Positive:  FKBP4, RHNO1, ADIPOR2, WNK1, TEAD4, KDM5A, CCND2, TIGAR, DYRK4, NDUFA9 \n",
      "\t   VWF, CD9, PLEKHG6, TNFRSF1A, SCNN1A, LTBR, CD27, DHRS4, PABPN1, AP1G2 \n",
      "\t   NGDN, C14orf119, TAPBPL, DHRS4L2, ACIN1, MRPL51, PCK2, PSMB5, DCAF11, PRMT5 \n",
      "Negative:  CD37, PIH1D1, TRPM4, SNRNP70, RUVBL2, FTL, BAX, NUCB1, FUT2, BCAT2 \n",
      "\t   PPP1R15A, RPL18, SULT2B1, LMTK3, CYTH2, GRWD1, KDELR1, EMP3, CARD8, SEPW1 \n",
      "\t   GLTSCR2, NAPA, C5AR1, INAFM1, AKT1S1, BBC3, TBC1D17, SAE1, IL4I1, NUP62 \n",
      "PC_ 4 \n",
      "Positive:  UBE2G1, SPNS2, ATP2A3, NCBP3, MNT, TSR1, CLUH, PAFAH1B1, TAX1BP3, GGT6 \n",
      "\t   ITGAE, EMC6, RPA1, PELP1, SERPINF1, ARRB2, MIR22HG, MED11, PRPF8, CXCL16 \n",
      "\t   RILP, TM4SF5, SLC43A2, PITPNA, PITPNA-AS1, MYO1C, CRK, YWHAE, NUP88, ABR \n",
      "Negative:  MRPL45, MLLT6, CISD3, DDX52, PSMB3, SYNRG, CWC25, DUSP14, RPL23, AATF \n",
      "\t   LASP1, DHRS11, GGNBP2, RPL19, MYO19, MED1, ZNHIT3, CCL4L2, CDK12, CCL3L3 \n",
      "\t   CCL4, PPP1R1B, CCL3, STARD3, CCL18, CCL15, ERBB2, CCL5, MIEN1, TAF15 \n",
      "PC_ 5 \n",
      "Positive:  PIGT, WFDC2, SYS1, DNTTIP1, SDC4, UBE2C, SLPI, TNNC2, PI3, STK4 \n",
      "\t   TOMM34, ACOT8, YWHAB, CTSA, ADA, PKIG, SERINC3, PLTP, HNF4A, OSER1 \n",
      "\t   PCIF1, MYBL2, IFT52, MMP9, SRSF6, CHD6, TOP1, MAFB, NCOA3, RALGAPB \n",
      "Negative:  CISD3, MLLT6, PSMB3, MRPL45, CWC25, RPL23, DDX52, LASP1, SYNRG, RPL19 \n",
      "\t   DUSP14, AATF, MED1, DHRS11, CDK12, GGNBP2, PPP1R1B, MYO19, ZNHIT3, STARD3 \n",
      "\t   CCL4L2, ERBB2, CCL3L3, CCL4, MIEN1, CCL3, CCL18, CCL15, ORMDL3, CCL5 \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2023-08-08 16:04:15] define_signif_tumor_subclusters(), tumor: malignant_SMC20\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  GABPB1, GABPB1-AS1, USP8, DTWD1, FGF7, SPPL2A, GALK2, LYSMD2, COPS2, EID1 \n",
      "\t   TMOD3, THG1L, FBN1, ADAM19, MAPK6, CLINT1, CYFIP2, MRPL22, DUT, RNF145 \n",
      "\t   CNOT8, LARP1, ARPP19, UBLCP1, SPARC, ATOX1, G3BP1, SQRDL, TTC1, RSL24D1 \n",
      "Negative:  SLCO4A1, MRGBP, RPS21, OGFR, COL9A3, DIDO1, ADRM1, GID8, SLC17A9, OSBPL2 \n",
      "\t   YTHDF1, ARFGAP1, MTG2, PPDPF, PSMA7, PRELID3B, PTK6, ATP5E, CTSZ, HELZ2 \n",
      "\t   NELFCD, GNAS, STX16, VAPB, RAB22A, STMN3, PMEPA1, PCK1, ARFRP1, RBM38 \n",
      "PC_ 2 \n",
      "Positive:  LAPTM4A, SDC1, TTC32, RDH14, PUM2, SMC6, RHOB, VSNL1, UBXN2A, SF3B6 \n",
      "\t   FAM49A, TP53I3, DDX1, ITSN2, FAM84A, NCOA1, PTRHD1, LPIN1, ASXL2, ROCK2 \n",
      "\t   RAB10, HADHA, PQLC3, HADHB, SLC35F6, PDIA6, TMEM214, NOL10, OST4, ODC1 \n",
      "Negative:  CXCL2, CXCL3, MTHFD2L, EREG, AREG, PARM1, CCL3, CCL4, RCHY1, CXCL5 \n",
      "\t   CCL3L3, CCL18, G3BP2, PPBP, CCL15, CCL4L2, USO1, CXCL1, CCL5, ZNHIT3 \n",
      "\t   NAAA, CXCL8, TAF15, SDAD1, MYO19, RASSF6, AP2B1, CXCL9, GGNBP2, CXCL10 \n",
      "PC_ 3 \n",
      "Positive:  PDP1, FAM92A1, NDUFAF6, OTUD6B-AS1, DECR1, PLEKHF2, UQCRB, NBN, PTDSS1, RIPK2 \n",
      "\t   SDC2, CPNE3, CPQ, MTDH, RMDN1, LAPTM4B, MATN2, WWP1, RPL30, HRSP12 \n",
      "\t   CA2, STK3, ATP5G2, ATF7, CALCOCO1, SMUG1, TARBP2, COX6C, C8orf59, CBX5 \n",
      "Negative:  SIGIRR, ANO9, PKP3, IFITM2, IFITM1, IFITM3, RNH1, PSMD13, HRAS, RASSF7 \n",
      "\t   RIC8A, IRF7, BET1L, CDHR5, EPS8L2, TALDO1, RPLP2, HLA-F, HLA-A, TRIM27 \n",
      "\t   ZSCAN16-AS1, ZNRD1, ZNF165, PPP1R11, RP1-313I6.12, ABT1, BTN2A1, BTN3A2, DPP7, SSNA1 \n",
      "PC_ 4 \n",
      "Positive:  CALCOCO1, SMUG1, ATP5G2, ATF7, CBX5, TARBP2, PCBP2, HNRNPA1, PRR13, COPZ1 \n",
      "\t   SP1, AAAS, ITGA5, C12orf10, BLOC1S1, PFDN5, ITGB7, CD63, ORMDL2, IGFBP6 \n",
      "\t   PYM1, EIF4B, RAB5B, KRT18, RPS26, ERBB3, KRT8, PA2G4, RPL41, KRT6A \n",
      "Negative:  SDC1, RDH14, VSNL1, TTC32, SMC6, LAPTM4A, PUM2, FAM49A, RHOB, DDX1 \n",
      "\t   UBXN2A, FAM84A, SF3B6, LPIN1, TP53I3, ROCK2, ITSN2, PQLC3, NCOA1, PTRHD1 \n",
      "\t   PDIA6, ASXL2, NOL10, RAB10, ODC1, HADHA, HPCAL1, HADHB, RRM2, SLC35F6 \n",
      "PC_ 5 \n",
      "Positive:  NDFIP2, RBM26, SPRY2, MYCBP2, DNAJC3, MBNL2, FBXL3, RAP2A, IPO5, CLN5 \n",
      "\t   FARP1, STK24, UBAC2, GPR183, KCTD12, TM9SF2, RP11-12G12.7, PCCA, LMO7, TPP2 \n",
      "\t   TEX30, UCHL3, EFNB2, ARGLU1, COMMD6, ABHD13, TBC1D4, TNFSF13B, KLF5, RP11-54H7.4 \n",
      "Negative:  SEPW1, GLTSCR2, NAPA, CARD8, C5AR1, EMP3, INAFM1, KDELR1, BBC3, GRWD1 \n",
      "\t   SAE1, CYTH2, LMTK3, SULT2B1, TMEM160, RPL18, FUT2, BCAT2, ARHGAP35, PPP1R15A \n",
      "\t   NUCB1, BAX, AP2S1, FTL, RUVBL2, SLC1A5, SNRNP70, STRN4, TRPM4, PRKD2 \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2023-08-08 16:04:23] define_signif_tumor_subclusters(), tumor: malignant_SMC21\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  NFE2L3, HNRNPA2B1, CBX3, SNX10, SKAP2, HOTAIRM1, HOXA9, NOP56, SNRPB, NSFL1C \n",
      "\t   FKBP1A, SDCBP2, PSMF1, THOC2, C1GALT1C1, XIAP, MCTS1, STAG2, LAMP2, SH2D1A \n",
      "\t   ZDHHC9, NKAP, NDUFA1, UTP14A, RNF113A, AIFM1, RBMX2, UPF3B, STK26, RPL39 \n",
      "Negative:  STRN3, HECTD1, SCFD1, ARHGAP5, STXBP6, EGLN3, SPTSSA, GZMB, GZMH, EAPP \n",
      "\t   SDR39U1, DHRS1, SNX6, RABGGTA, BAZ1A, TINF2, SRP54, GMPR2, FAM177A1, NEDD8 \n",
      "\t   PPP2R3C, CHMP4A, PSMA6, IRF9, NFKBIA, PSME2, RALGAPA1, EMC9, MBIP, SEC23A \n",
      "PC_ 2 \n",
      "Positive:  TERF1, RPL7, LACTB2, TRAM1, RDH10, UBE2W, SULF1, TCEB1, TMEM70, LY96 \n",
      "\t   PEX2, MRPS28, TPD52, ZBTB10, PAG1, FABP5, ZFAND1, CHMP4C, LRRCC1, C8orf59 \n",
      "\t   CA2, RIPK2, NBN, WWP1, CPNE3, RMDN1, DECR1, OTUD6B-AS1, FAM92A1, PDP1 \n",
      "Negative:  CDC42EP2, DPF2, CAPN1, FRMD8, MRPL49, NEAT1, MALAT1, SCYL1, SYVN1, LTBP3 \n",
      "\t   SSSCA1, FAU, FAM89B, TM7SF2, EHBP1L1, VPS51, ZFPL1, SIPA1, SAC3D1, MAP3K11 \n",
      "\t   ARL2, ATG2A, RELA, EHD1, KAT5, SF1, RNASEH2C, RPS6KA4, CCDC88B, OVOL1 \n",
      "PC_ 3 \n",
      "Positive:  BBC3, INAFM1, SAE1, C5AR1, NAPA, GLTSCR2, ARHGAP35, TMEM160, AP2S1, SLC1A5 \n",
      "\t   STRN4, PRKD2, SEPW1, CALM3, PPP5C, IGFL2, CARD8, FOXA3, EMP3, SYMPK \n",
      "\t   SNRPD2, KDELR1, EML2, GRWD1, OPA3, CYTH2, VASP, FOSB, LMTK3, ERCC1 \n",
      "Negative:  BCAP31, IDH3G, SSR4, ARHGAP4, NAA10, HCFC1, IRAK1, MECP2, FLNA, EMD \n",
      "\t   RPL10, TAZ, ATP6AP1, GDI1, FAM50A, PLXNA3, LAGE3, UBL4A, FAM127A, SLC10A3 \n",
      "\t   IKBKG, G6PD, DKC1, FAM3A, MPP1, F8A1, MOSPD1, FUNDC2, BRCC3, VBP1 \n",
      "PC_ 4 \n",
      "Positive:  ZFPL1, SAC3D1, VPS51, ARL2, TM7SF2, ATG2A, FAU, EHD1, SYVN1, MRPL49 \n",
      "\t   SF1, CAPN1, RPS6KA4, CDC42EP2, CCDC88B, DPF2, AP003774.1, PRDX5, FRMD8, TRMT112 \n",
      "\t   NEAT1, ESRRA, GPR137, MALAT1, BAD, SCYL1, PLCB3, LTBP3, PPP1R14B, SSSCA1 \n",
      "Negative:  DNM2, QTRT1, TMED1, ILF3, CARM1, ILF3-AS1, YIPF2, SLC44A2, C19orf52, SMARCA4 \n",
      "\t   AP1M2, LDLR, CDKN2D, TMEM205, PRKCSH, ECSIT, ATG4D, ELOF1, KEAP1, ACP5 \n",
      "\t   ZNF791, PDE4A, WDR83OS, CDC37, DHPS, ICAM3, C19orf43, ICAM1, ASNA1, MRPL4 \n",
      "PC_ 5 \n",
      "Positive:  MRPL45, MLLT6, CISD3, DDX52, PSMB3, SYNRG, CWC25, DUSP14, RPL23, LASP1 \n",
      "\t   AATF, RPL19, DHRS11, MED1, GGNBP2, CDK12, MYO19, PPP1R1B, ZNHIT3, CCL4L2 \n",
      "\t   STARD3, CCL3L3, ERBB2, CCL4, CCL3, MIEN1, CCL18, ORMDL3, CCL15, CCL5 \n",
      "Negative:  FAM53C, HNRNPA0, KDM3B, SMAD5, EGR1, TGFBI, ETF1, CXCL14, HSPA9, H2AFY \n",
      "\t   PITX1, CTNNA1, TXNDC15, SIL1, C5orf24, PAIP2, DDX46, MZB1, CAMLG, TMEM173 \n",
      "\t   SEC24A, UBE2D2, SAR1B, CXXC5, CDKN2AIPNL, PURA, UBE2B, CYSTM1, PPP2CA, SKP1 \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2023-08-08 16:04:30] define_signif_tumor_subclusters(), tumor: malignant_SMC22\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  FAM3B, MX2, BACE2, MX1, TMPRSS2, WRB, ABCG1, TFF3, HMGN1, TFF2 \n",
      "\t   RRP1B, PDXK, U2AF1, PKNOX1, NDUFV3, CSTB, SLC37A1, TFF1, BRWD1, RRP1 \n",
      "\t   PSMG1, AGPAT3, PANK2, MAVS, PFKL, ETS2, CDC25B, C21orf2, CENPB, DYRK1A \n",
      "Negative:  CCL13, ZNF830, CCL11, CCL8, CCL2, RFFL, TMEM98, MYO1D, NLE1, RP11-1094M14.11 \n",
      "\t   PSMD11, AP2B1, ZNF207, TAF15, RHOT1, CCL5, SUZ12, UTP6, CCL15, COPRS \n",
      "\t   EVI2A, CCL18, EVI2B, CCL3, CCL4, NF1, CCL3L3, CCL4L2, ZNHIT3, GOSR1 \n",
      "PC_ 2 \n",
      "Positive:  NEAT1, MALAT1, SCYL1, FRMD8, LTBP3, DPF2, SSSCA1, CDC42EP2, FAM89B, CAPN1 \n",
      "\t   EHBP1L1, MRPL49, SIPA1, SYVN1, FAU, MAP3K11, TM7SF2, RELA, VPS51, ZFPL1 \n",
      "\t   KAT5, SAC3D1, RNASEH2C, ARL2, OVOL1, ATG2A, CFL1, EHD1, MUS81, SF1 \n",
      "Negative:  UBE2D3, MANBA, CISD2, NFKB1, BDH2, SLC39A8, PPA2, FLJ20021, AIMP1, PPP3CA \n",
      "\t   PAPSS1, H2AFZ, DNAJB14, LAMTOR3, DAPP1, SGMS2, HADH, RPL34, OSTC, CCDC109B \n",
      "\t   CASP6, PLA2G12A, GAR1, TIFA, LARP7, CAMK2D, SNHG8, SEC24D, USP53, C4orf3 \n",
      "PC_ 3 \n",
      "Positive:  MRPL28, AXIN1, FAM234A, TMEM8A, LUC7L, NME4, HBA2, NPRL3, DECR2, SNRNP25 \n",
      "\t   MPG, POLR3K, CAPN15, C16orf13, FAM195A, NDUFS7, MUM1, DAZAP1, C19orf24, MSLN \n",
      "\t   CIRBP, RPS15, MIDN, RHOT2, ATP5D, ILF3-AS1, AP1M2, SLC44A2, ILF3, STUB1 \n",
      "Negative:  CEACAM6, CD79A, RPS19, ARHGEF1, RABAC1, DEDD2, CEACAM1, GSK3A, ERF, CEACAM5 \n",
      "\t   PAFAH1B3, ETHE1, S100A8, CEACAM7, ZNF428, S100A9, S100A6, S100A11, S100A10, S100A4 \n",
      "\t   THEM4, PLAUR, S100A2, ATP5SL, MRPL9, SNX27, S100A16, KCNN4, TUFT1, EXOSC5 \n",
      "PC_ 4 \n",
      "Positive:  AKT1S1, TBC1D17, IL4I1, NUP62, PIH1D1, ATF5, VRK3, CD37, MYH14, TRPM4 \n",
      "\t   NR1H2, SNRNP70, EMC10, RUVBL2, JOSD2, CLEC11A, FTL, C19orf48, BAX, KLK1 \n",
      "\t   NUCB1, KLK6, PPP1R15A, BCAT2, KLK10, FUT2, ETFB, RPL18, NKG7, SULT2B1 \n",
      "Negative:  MXRA7, ST6GALNAC1, CYGB, JMJD6, RHBDF2, SPHK1, METTL23, PRPSAP1, SRSF2, UBALD2 \n",
      "\t   MFSD11, EXOC7, SEC14L1, SRP68, SEPT9, ACOX1, TRIM47, TMC6, WBP2, TMC8 \n",
      "\t   UNK, H3F3B, GALK1, SYNGR2, ITGB4, SAP30BP, MYO15B, TK1, LLGL2, TSEN54 \n",
      "PC_ 5 \n",
      "Positive:  C1orf43, UBAP2L, TPM3, HAX1, RPS27, UBE2Q1, ADAR, RAB13, PMVK, JTB \n",
      "\t   PBXIP1, SLC39A1, SHC1, GATAD2B, CKS1B, ILF2, FLAD1, SNAPIN, ZBTB7B, ADAM15 \n",
      "\t   CHTOP, EFNA4, S100A13, EFNA3, S100A14, EFNA1, SLC50A1, S100A16, DPM3, KRTCAP2 \n",
      "Negative:  CLTC, YPEL2, PTRH2, SKA2, VMP1, RAD51C, RPS6KB1, RNF43, CA4, APPBP2 \n",
      "\t   SUPT4H1, NACA2, MED13, CYB561, DCAF7, TMEM41A, LIPH, MAP3K13, SENP2, MAGEF1 \n",
      "\t   IGF2BP2, EPHB3, TRA2B, TACO1, DNAJB11, POLR2H, EIF4A2, ST6GAL1, EIF4G1, RPL39L \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2023-08-08 16:04:38] define_signif_tumor_subclusters(), tumor: malignant_SMC23\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  GFER, NOXO1, ZNF598, NPW, TBL3, SLC9A3R2, SNHG9, NTHL1, RPS2, TSC2 \n",
      "\t   SNHG19, TRAF7, NDUFB10, MLST8, RRBP1, SNX5, MGME1, PET117, PGP, SEC23B \n",
      "\t   MSRB1, ECI1, LINC00493, DTD1, RP1-122P22.4, RNPS1, RIN2, DSTN, PDPK1, NAA20 \n",
      "Negative:  MORF4L1, CTSH, PSMA4, TMED3, IREB2, WDR61, MTHFS, DNAJA4, IDH3A, TSPAN3 \n",
      "\t   BCL2A1, RCN2, ETFA, ZFAND6, FBXO22, UBE2Q2, FAH, IMP3, SNUPN, SIN3A \n",
      "\t   MAN2C1, ABHD17C, COMMD4, CEMIP, RPP25, MESDC2, COX5A, SCAMP2, MESDC1, RPS17 \n",
      "PC_ 2 \n",
      "Positive:  LTBP3, SCYL1, SSSCA1, FAM89B, MALAT1, EHBP1L1, SIPA1, MAP3K11, RELA, NEAT1 \n",
      "\t   KAT5, RNASEH2C, FRMD8, OVOL1, DPF2, CFL1, MUS81, CDC42EP2, CAPN1, MRPL49 \n",
      "\t   SYVN1, FAU, TM7SF2, VPS51, ZFPL1, EIF1AD, SAC3D1, ARL2, BANF1, ATG2A \n",
      "Negative:  RIPK2, NBN, CPNE3, DECR1, RMDN1, OTUD6B-AS1, FAM92A1, WWP1, PDP1, CA2 \n",
      "\t   CDH17, C8orf59, GEM, KIAA1429, ESRP1, NDUFAF6, LRRCC1, PLEKHF2, UQCRB, CHMP4C \n",
      "\t   PTDSS1, SDC2, ZFAND1, CPQ, MTDH, LAPTM4B, FABP5, MATN2, PAG1, RPL30 \n",
      "PC_ 3 \n",
      "Positive:  RAB5B, RPS26, ERBB3, PA2G4, PYM1, RPL41, ESYT1, ORMDL2, MYL6B, CD63 \n",
      "\t   BLOC1S1, MYL6, ITGA5, COPZ1, SMARCC2, HNRNPA1, CBX5, RNF41, SMUG1, NABP2 \n",
      "\t   CALCOCO1, ATP5G2, SLC39A5, RPS29, MIS18BP1, LRR1, FKBP3, RPL36AL, ATF7, FBXO33 \n",
      "Negative:  CXCL9, CXCL10, SDAD1, NUP54, NAAA, SCARB2, USO1, G3BP2, SHROOM3, RCHY1 \n",
      "\t   SEPT11, PARM1, CCNI, AREG, CCNG2, EREG, MTHFD2L, CXCL13, CXCL2, CNOT6L \n",
      "\t   CXCL3, CXCL5, MRPL1, PPBP, ANXA3, CXCL1, ANTXR2, CXCL8, RASGEF1B, RASSF6 \n",
      "PC_ 4 \n",
      "Positive:  AP003774.1, PRDX5, CCDC88B, RPS6KA4, TRMT112, SF1, EHD1, ATG2A, ESRRA, ARL2 \n",
      "\t   SAC3D1, GPR137, ZFPL1, BAD, VPS51, TM7SF2, PLCB3, FAU, PPP1R14B, SYVN1 \n",
      "\t   FKBP2, MRPL49, VEGFB, CAPN1, DNAJC4, NUDT22, CDC42EP2, TRPT1, DPF2, FERMT3 \n",
      "Negative:  TRIM38, LRRC16A, HIST1H1C, GMNN, HIST1H4C, C6orf62, HIST1H2AC, ACOT13, HIST1H2BD, TDP2 \n",
      "\t   HIST1H2BG, MRS2, HIST1H2AE, SOX4, ID4, BTN3A2, DEK, TPMT, NUP153, BTN2A1 \n",
      "\t   ATXN1, MYLIP, DTNBP1, JARID2, ABT1, CD83, MCUR1, RP1-313I6.12, RANBP9, NOL7 \n",
      "PC_ 5 \n",
      "Positive:  SULT2B1, LMTK3, CYTH2, GRWD1, KDELR1, EMP3, CARD8, GLTSCR2, SEPW1, NAPA \n",
      "\t   C5AR1, INAFM1, BBC3, SAE1, TMEM160, ARHGAP35, AP2S1, SLC1A5, STRN4, PRKD2 \n",
      "\t   CALM3, PPP5C, IGFL2, LLNLR-245B6.1, A1BG, TRAPPC2P1, CTC-444N24.11, ZNF444, ZNF787, MUM1 \n",
      "Negative:  TMEM106B, ARL4A, PHF14, BZW2, NDUFA4, ICA1, TSPAN13, RPA3, C1GALT1, AGR2 \n",
      "\t   C7orf26, ZDHHC4, AGR3, KDELR2, RAC1, AHR, USP42, SNX13, EIF2AK1, TWISTNB \n",
      "\t   AIMP2, MACC1, CTA-293F17.1, CCZ1, ITGB8, RNF216, CDCA7L, EHD1, ATG2A, SF1 \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2023-08-08 16:04:43] define_signif_tumor_subclusters(), tumor: malignant_SMC24\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  CCL3L3, CCL4, CCL3, CCL4L2, CCL18, ZNHIT3, CCL15, MYO19, CCL5, TAF15 \n",
      "\t   AP2B1, GGNBP2, RP11-1094M14.11, NLE1, DHRS11, RFFL, ZNF830, CCL13, AATF, USO1 \n",
      "\t   NAAA, G3BP2, SDAD1, RCHY1, CXCL9, PARM1, CCL8, CXCL10, EREG, AREG \n",
      "Negative:  RPS19BP1, ATF4, MIEF1, RPL3, APOBEC3G, AIP, CDK2AP2, TMEM134, GSTP1, CORO1B \n",
      "\t   APOBEC3C, RPS6KB2, TBC1D10C, PPP1CA, CBX6, POLD4, RAD9A, DHPS, C19orf43, WDR83OS \n",
      "\t   ASNA1, ANKRD13D, SUN2, ZNF791, HOOK2, ADRBK1, ACP5, JUNB, ELOF1, KDM2A \n",
      "PC_ 2 \n",
      "Positive:  LAMC2, ARPC5, LAMC1, SMG7, TSEN15, C1orf21, DHX9, RGS16, EDEM3, GLUL \n",
      "\t   FAM129A, IER5, ACBD6, QSOX1, CEP350, TOR1AIP1, RNF2, TOR1AIP2, IVNS1ABP, ABL2 \n",
      "\t   TOR3A, TPR, RALGPS2, RFWD2, C1orf27, KIAA0040, MRPS14, PTGS2, CACYBP, RABGAP1L \n",
      "Negative:  SULT2B1, LMTK3, RPL18, CYTH2, GRWD1, KDELR1, FUT2, EMP3, BCAT2, CARD8 \n",
      "\t   PPP1R15A, NUCB1, FTL, RUVBL2, BAX, SNRNP70, SEPW1, TRPM4, CD37, PIH1D1 \n",
      "\t   GLTSCR2, SLC7A11, ELF2, NDUFC1, PGRMC2, LARP1B, NAA15, SPRY1, NAPA, RP11-83A24.2 \n",
      "PC_ 3 \n",
      "Positive:  CD96, C3orf52, ATG3, CCDC80, GTPBP8, C3orf17, NAA50, ATP6V1A, TIGIT, ZBTB20 \n",
      "\t   B4GALT4, TMEM39A, NDUFC1, ELF2, NAA15, RP11-83A24.2, TIMMDC1, SLC7A11, ZNF330, SCOC \n",
      "\t   SETD7, USP38, MGST2, SMARCA5, ANAPC10, ABCE1, OTUD4, PGRMC2, LSM6, RPS3A \n",
      "Negative:  APOC1, CLPTM1, APOE, RELB, CLASRP, TOMM40, GEMIN7, PPP1R37, PVRL2, TRAPPC6A \n",
      "\t   BCAM, CD3EAP, ERCC1, CBLC, FOSB, BCL3, VASP, PVR, OPA3, EML2 \n",
      "\t   KCNN4, SNRPD2, PLAUR, SYMPK, FOXA3, ZNF428, IGFL2, PPP5C, ETHE1, CALM3 \n",
      "PC_ 4 \n",
      "Positive:  KRTCAP2, MUC1, DPM3, MTX1, SLC50A1, GBA, SCAMP3, EFNA1, FDPS, RUSC1 \n",
      "\t   ASH1L, YY1AP1, DAP3, GON4L, RIT1, KIAA0907, ARHGEF2, SSR2, LAMTOR2, RAB25 \n",
      "\t   LMNA, SEMA4A, SLC25A44, PMF1, SMG5, GLMP, CCT3, MEF2D, SLC39A1, APOA1BP \n",
      "Negative:  LRBA, RPS3A, PRMT9, SH3D19, LSM6, FBXW7, OTUD4, ARFIP1, ABCE1, TRIM2 \n",
      "\t   ANAPC10, SFRP2, PLRG1, SMARCA5, PPID, USP38, FNIP2, RAPGEF2, NAF1, ZNF330 \n",
      "\t   TMA16, SCOC, LINC01207, MGST2, TMEM192, ZNF331, MYADM, PPP2R1A, NDUFA3, NKG7 \n",
      "PC_ 5 \n",
      "Positive:  B4GALT3, NDUFS2, UFC1, FCER1G, PFDN2, SDHC, ARHGAP30, FCGR2A, TSTD1, HSPA6 \n",
      "\t   FCGR3A, F11R, FCGR2B, ITLN1, DUSP12, LY9, ATF6, UHMK1, SLAMF7, UAP1 \n",
      "\t   CD48, HSD17B7, NCSTN, RGS5, PBX1, COPA, MGST3, DCAF8, PEA15, IGSF8 \n",
      "Negative:  OXLD1, CCDC137, NPLOC4, ARL16, ACTG1, HGS, SLC38A10, MRPL12, SLC25A10, C17orf89 \n",
      "\t   FAM195B, BAIAP2, P4HB, CHMP6, ARHGDIA, ALYREF, ENDOV, ANAPC11, RNF213, PCYT2 \n",
      "\t   EIF4A3, SIRT7, GAA, CBX4, MAFG, ENGASE, PYCR1, CANT1, ASPSCR1, LGALS3BP \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2023-08-08 16:04:48] define_signif_tumor_subclusters(), tumor: malignant_SMC25\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  RSBN1, PTPN22, SLC16A1, RHOC, BCL2L15, CAPZA1, CTTNBP2NL, HIPK1, RAP1A, C1orf162 \n",
      "\t   OLFML3, ATP5F1, WDR77, DENND2D, TRIM33, DRAM2, LRIF1, CD53, BCAS2, LAMTOR5 \n",
      "\t   AHCYL1, EPS8L3, NRAS, GSTM3, CSDE1, GSTM4, CD52, ARHGAP5, HECTD1, EGLN3 \n",
      "Negative:  RBM12, CPNE1, NFS1, SPAG4, ROMO1, ERGIC3, UQCC1, RBM39, EIF6, PHF20 \n",
      "\t   SCAND1, MMP24-AS1, CNBD2, LINC00657, EPB41L1, PROCR, AAR2, EDEM2, DLGAP4, TRPC4AP \n",
      "\t   MYL9, GSS, TGIF2, ACSS2, C20orf24, NDRG3, TP53INP2, SAMHD1, NCOA6, RPN2 \n",
      "PC_ 2 \n",
      "Positive:  ANXA3, MRPL1, ANTXR2, CNOT6L, CXCL13, RASGEF1B, CCNG2, CCNI, HNRNPD, HNRNPDL \n",
      "\t   SEPT11, ENOPH1, SEC31A, SHROOM3, THAP9-AS1, COPS4, SCARB2, PLAC8, COQ2, NUP54 \n",
      "\t   MRPS18C, CXCL10, CDS1, CXCL9, HSD17B11, SDAD1, NUDT9, NAAA, SPARCL1, USO1 \n",
      "Negative:  TMEM205, LDLR, PRKCSH, ECSIT, SMARCA4, ELOF1, C19orf52, ACP5, YIPF2, ZNF791 \n",
      "\t   CARM1, WDR83OS, TMED1, DHPS, DNM2, C19orf43, ASNA1, QTRT1, HOOK2, ILF3 \n",
      "\t   JUNB, ILF3-AS1, PRDX2, SLC44A2, RNASEH2A, AP1M2, DNASE2, CDKN2D, GCDH, ATG4D \n",
      "PC_ 3 \n",
      "Positive:  STK17A, COA1, MRPL32, PSMA2, BLVRA, INHBA, MPLKIP, UBE2D4, CDK13, DBNL \n",
      "\t   RALA, YAE1D1, VPS41, AEBP1, TRGC2, POLD2, STARD3NL, YKT6, SFRP4, NUDCD3 \n",
      "\t   EPDR1, AOAH, DDX56, EEPD1, TMED4, SEPT7, GCNT3, MYO1E, CCNB2, GTF2A2 \n",
      "Negative:  MRPL1, CNOT6L, ANXA3, CXCL13, ANTXR2, CCNG2, SEPT11, SHROOM3, RASGEF1B, SCARB2 \n",
      "\t   CCNI, NUP54, HNRNPD, CXCL10, HNRNPDL, ENOPH1, SEC31A, CXCL9, THAP9-AS1, COPS4 \n",
      "\t   SDAD1, PLAC8, NAAA, COQ2, USO1, G3BP2, MRPS18C, RCHY1, ANKRD17, DCK \n",
      "PC_ 4 \n",
      "Positive:  VPS51, ZFPL1, TM7SF2, MRPL49, CAPN1, SAC3D1, CDC42EP2, SYVN1, FAU, DPF2 \n",
      "\t   ARL2, FRMD8, ATG2A, NEAT1, EHD1, MALAT1, SF1, SCYL1, RPS6KA4, LTBP3 \n",
      "\t   SSSCA1, CCDC88B, FAM89B, AP003774.1, EHBP1L1, SIPA1, PRDX5, MAP3K11, RELA, TRMT112 \n",
      "Negative:  HSD17B11, CDS1, NUDT9, MRPS18C, SPARCL1, COQ2, SPP1, PLAC8, PPM1K, COPS4 \n",
      "\t   PYURF, THAP9-AS1, FAM13A, SCNM1, VPS72, GPRIN3, PSMD4, CDC42SE1, SEC31A, PI4KB \n",
      "\t   C1orf56, PDLIM5, SELENBP1, PSMB4, POGZ, CERS2, RAP1GDS1, CGN, ENOPH1, TSPAN5 \n",
      "PC_ 5 \n",
      "Positive:  STARD3NL, SFRP4, TRGC2, EPDR1, VPS41, AOAH, YAE1D1, EEPD1, RALA, SEPT7 \n",
      "\t   CDK13, HERPUD2, MPLKIP, INHBA, RP9, PSMA2, MRPL32, NT5C3A, STK17A, FKBP9 \n",
      "\t   COA1, KBTBD2, BLVRA, AVL9, UBE2D4, LSM5, DBNL, AQP1, GARS, AEBP1 \n",
      "Negative:  APH1A, ANP32E, PLEKHO1, C1orf54, MTMR11, MRPS21, RPRD2, SF3B4, ECM1, HIST2H2BE \n",
      "\t   MCL1, ENSA, LINC00869, GOLPH3L, CH17-189H20.1, CTSS, PDE4DIP, CTSK, CERS2, CHD1L \n",
      "\t   C1orf56, TXNIP, CDC42SE1, SCNM1, POLR3GL, VPS72, RBM8A, PSMD4, RNF115, PI4KB \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2023-08-08 16:04:54] define_signif_tumor_subclusters(), tumor: T cells\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  HADH, RPL34, SGMS2, PAPSS1, OSTC, AIMP1, CCDC109B, PPA2, CASP6, BDH2 \n",
      "\t   PLA2G12A, CISD2, GAR1, UBE2D3, MANBA, TIFA, NFKB1, SLC39A8, LARP7, FLJ20021 \n",
      "\t   CAMK2D, PPP3CA, H2AFZ, SNHG8, DNAJB14, LAMTOR3, DAPP1, SEC24D, ADH1C, ADH1B \n",
      "Negative:  MUS81, CFL1, OVOL1, RNASEH2C, KAT5, RELA, EIF1AD, MAP3K11, BANF1, SF3B2 \n",
      "\t   SIPA1, RAB1B, EHBP1L1, YIF1A, CD248, FAM89B, BRMS1, SSSCA1, MRPL11, LTBP3 \n",
      "\t   DPP3, SCYL1, ZDHHC24, MALAT1, CTSF, CCS, NEAT1, FRMD8, RBM14, DPF2 \n",
      "PC_ 2 \n",
      "Positive:  LDLR, TMEM205, SMARCA4, PRKCSH, C19orf52, ECSIT, YIPF2, ELOF1, CARM1, ACP5 \n",
      "\t   TMED1, ZNF791, DNM2, WDR83OS, QTRT1, DHPS, C19orf43, ILF3, ASNA1, ILF3-AS1 \n",
      "\t   HOOK2, SLC44A2, JUNB, AP1M2, PRDX2, CDKN2D, RNASEH2A, ATG4D, DNASE2, INAFM1 \n",
      "Negative:  GPR65, ZC3H14, SEL1L, SNW1, FOXN3, SLIRP, PSMC1, SPTLC2, AHSA1, CALM1 \n",
      "\t   GSTZ1, GPATCH2L, IFT43, CCDC88C, C14orf1, BATF, PPP4R3A, FOS, TMED10, ACYP1 \n",
      "\t   EIF2B2, PGF, TC2N, DLST, YLPM1, TRIP11, ISCA2, ATXN3, ELMSAN1, NUMB \n",
      "PC_ 3 \n",
      "Positive:  SLC17A9, GID8, YTHDF1, DIDO1, COL9A3, OGFR, ARFGAP1, MRGBP, PPDPF, SLCO4A1 \n",
      "\t   PTK6, RPS21, ADRM1, OSBPL2, HELZ2, MTG2, PSMA7, STMN3, PRELID3B, ARFRP1 \n",
      "\t   ATP5E, CTSZ, SLC2A4RG, NELFCD, TPD52L2, GNAS, DNAJC5, UCKL1, STX16, VAPB \n",
      "Negative:  RNASEH2A, DNASE2, PRDX2, GCDH, JUNB, HOOK2, FARSA, ASNA1, CALR, C19orf43 \n",
      "\t   DHPS, RAD23A, WDR83OS, GADD45GIP1, NFIX, ZNF791, TRMT1, ACP5, NACC1, ELOF1 \n",
      "\t   STX10, IER2, ECSIT, CCDC130, PRKCSH, C19orf53, TMEM205, CTD-3252C9.4, CC2D1A, LDLR \n",
      "PC_ 4 \n",
      "Positive:  RBM38, RAE1, PCK1, RTFDC1, PMEPA1, AURKA, RAB22A, FAM210B, VAPB, STX16 \n",
      "\t   PFDN4, GNAS, ZNF217, TSHZ2, NELFCD, ATP9A, CTSZ, NFATC2, ATP5E, DPM1 \n",
      "\t   PRELID3B, ADNP, PSMA7, MTG2, BCAS4, OSBPL2, PARD6B, ADRM1, PTPN1, RPS21 \n",
      "Negative:  IFI16, ACKR1, ETV3, DUSP23, SH2D2A, TAGLN2, LINC01133, PRCC, IGSF8, HDGF \n",
      "\t   PEA15, MRPL24, ISG20L2, DCAF8, GPATCH4, COPA, APOA1BP, MEF2D, NCSTN, CCT3 \n",
      "\t   CD48, GLMP, SLAMF7, SMG5, PMF1, LY9, SLC25A44, ITLN1, SEMA4A, LMNA \n",
      "PC_ 5 \n",
      "Positive:  ESRRA, GPR137, TRMT112, BAD, PRDX5, PLCB3, AP003774.1, PPP1R14B, CCDC88B, RPS6KA4 \n",
      "\t   FKBP2, SF1, VEGFB, EHD1, DNAJC4, ATG2A, NUDT22, ARL2, TRPT1, SAC3D1 \n",
      "\t   FERMT3, ZFPL1, STIP1, VPS51, MACROD1, TM7SF2, OTUB1, COX8A, FAU, MARK2 \n",
      "Negative:  LY9, SLAMF7, ITLN1, CD48, F11R, NCSTN, TSTD1, COPA, ARHGAP30, DCAF8 \n",
      "\t   PFDN2, PEA15, UFC1, IGSF8, B4GALT3, LINC01133, NDUFS2, FCER1G, TAGLN2, DUSP23 \n",
      "\t   SDHC, ACKR1, IFI16, FCGR2A, ETV3, HSPA6, SH2D2A, FCGR3A, PRCC, FCGR2B \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2023-08-08 16:05:15] define_signif_tumor_subclusters(), tumor: B cells\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  CST7, APMAP, CST3, NXT1, CD93, THBD, FOXA2, XRN2, RALGAPA2, CRNKL1 \n",
      "\t   NAA20, RIN2, RP1-122P22.4, DTD1, LINC00493, SEC23B, PET117, MGME1, SNX5, RRBP1 \n",
      "\t   FASTK, TMUB1, DSTN, AGAP3, ABCF2, SNRPB2, CHPF2, KIF16B, NUB1, RHEB \n",
      "Negative:  PIH1D1, CD37, TRPM4, SNRNP70, AKT1S1, TBC1D17, IL4I1, RUVBL2, NUP62, ATF5 \n",
      "\t   VRK3, FTL, MYH14, NR1H2, BAX, EMC10, JOSD2, NUCB1, CLEC11A, PPP1R15A \n",
      "\t   C19orf48, BCAT2, KLK1, FUT2, RPL18, KLK6, KLK10, SULT2B1, ETFB, LMTK3 \n",
      "PC_ 2 \n",
      "Positive:  CHMP4C, ZFAND1, LRRCC1, FABP5, C8orf59, PAG1, CA2, WWP1, ZBTB10, RMDN1 \n",
      "\t   TPD52, CPNE3, MRPS28, RIPK2, PEX2, NBN, LY96, TMEM70, DECR1, MYL6 \n",
      "\t   TCEB1, OTUD6B-AS1, FAM92A1, SMARCC2, PDP1, UBE2W, CDH17, RNF41, RDH10, GEM \n",
      "Negative:  ZSCAN16-AS1, ZNF165, RP1-313I6.12, TRIM27, ABT1, HLA-F, BTN2A1, HLA-A, ZNRD1, BTN3A2 \n",
      "\t   PPP1R11, TRIM31, TRIM15, HIST1H2AE, TRIM26, HCG18, HIST1H2BG, RPP21, HLA-E, HIST1H2BD \n",
      "\t   GNL1, ABCF1, HIST1H2AC, PPP1R10, HIST1H4C, MRPS18B, HIST1H1C, TRIM38, LRRC16A, GMNN \n",
      "PC_ 3 \n",
      "Positive:  NFIX, TRMT1, NACC1, GADD45GIP1, RAD23A, STX10, CALR, RNASEH2A, PRDX2, GCDH \n",
      "\t   DNASE2, IER2, JUNB, WDR83OS, ASNA1, DHPS, FARSA, HOOK2, C19orf43, ZNF791 \n",
      "\t   ACP5, CCDC130, ELOF1, C19orf53, ECSIT, CTD-3252C9.4, PRKCSH, CC2D1A, IL27RA, TMEM205 \n",
      "Negative:  PIGT, SYS1, SDC4, SLPI, PI3, STK4, TOMM34, MMP9, YWHAB, CD40 \n",
      "\t   ADA, SLC35C2, TP53RK, PKIG, ZMYND8, NCOA3, SERINC3, SULF2, HNF4A, ARFGEF2 \n",
      "\t   OSER1, CSE1L, MYBL2, STAU1, IFT52, DDX27, SRSF6, ZNFX1, CHD6, ZFAS1 \n",
      "PC_ 4 \n",
      "Positive:  HSPD1, HSPE1, COQ10B, MOB4, SF3B1, SATB2, ANKRD44, FCGR3A, FCGR2B, HSPA6 \n",
      "\t   DUSP12, PGAP1, C2orf69, ATF6, RP11-347P5.1, UHMK1, STK17B, UAP1, C2orf47, FCGR2A \n",
      "\t   SLC39A10, HSD17B7, NABP1, SPATS2L, RGS5, MYO1B, SDHC, PBX1, STAT4, BZW1 \n",
      "Negative:  VPS51, ZFPL1, TM7SF2, SAC3D1, SYVN1, MRPL49, FAU, ARL2, CAPN1, ATG2A \n",
      "\t   CDC42EP2, DPF2, EHD1, FRMD8, NEAT1, SF1, SCYL1, MALAT1, LTBP3, SSSCA1 \n",
      "\t   FAM89B, EHBP1L1, SIPA1, RPS6KA4, MAP3K11, RELA, KAT5, RNASEH2C, OVOL1, CCDC88B \n",
      "PC_ 5 \n",
      "Positive:  YIPF2, C19orf52, CARM1, SMARCA4, TMED1, LDLR, DNM2, TMEM205, QTRT1, PRKCSH \n",
      "\t   ILF3, ILF3-AS1, ECSIT, SLC44A2, ELOF1, AP1M2, ACP5, CDKN2D, ZNF791, ATG4D \n",
      "\t   WDR83OS, DHPS, KEAP1, C19orf43, MXRA7, ST6GALNAC1, JMJD6, METTL23, PDE4A, ASNA1 \n",
      "Negative:  WDFY1, MRPL44, SERPINE2, ACSL3, CUL3, FARSB, MFF, CHPF, AGFG1, GMPPA \n",
      "\t   CCL20, DNPEP, PID1, TRIP12, DNAJB2, SP140L, TUBA4A, STK16, ANKZF1, SP100 \n",
      "\t   ZFAND2B, FAM134A, CNPPD1, CAB39, BCS1L, ITM2C, RQCD1, VIL1, PSMD1, CTDSP1 \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2023-08-08 16:05:29] define_signif_tumor_subclusters(), tumor: Stromal cells\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  SERPINE2, MRPL44, WDFY1, CUL3, ACSL3, S100P, MFF, FARSB, MRFAP1, CHPF \n",
      "\t   AGFG1, GMPPA, STX18, CCL20, LYAR, DNPEP, DNAJB2, PID1, TMEM128, TUBA4A \n",
      "\t   LRPAP1, TRIP12, STK16, HTT, ANKZF1, NOP14, SP140L, ZFAND2B, MFSD10, FAM134A \n",
      "Negative:  CDC37, PDE4A, KEAP1, ATG4D, ICAM3, TRIM31, PPP1R11, CDKN2D, ZNRD1, TRIM15 \n",
      "\t   HLA-A, TRIM26, HLA-F, ICAM1, AP1M2, HCG18, RPP21, TRIM27, HLA-E, SLC44A2 \n",
      "\t   MRPL4, ZSCAN16-AS1, GNL1, ILF3-AS1, DNMT1, ZNF165, ABCF1, ILF3, RP1-313I6.12, EIF3G \n",
      "PC_ 2 \n",
      "Positive:  FOSB, EML2, OPA3, VASP, SNRPD2, ERCC1, SYMPK, CD3EAP, FOXA3, TRAPPC6A \n",
      "\t   IGFL2, PPP1R37, PPP5C, APOE, APOC1, CALM3, GEMIN7, PRKD2, CLPTM1, CLASRP \n",
      "\t   RELB, STRN4, TOMM40, SLC1A5, AP2S1, PVRL2, ARHGAP35, TMEM160, BCAM, SAE1 \n",
      "Negative:  SCAF4, MIS18A, C21orf59, IFNAR2, IL10RB, IFNAR1, IFNGR2, TMEM50B, GART, SON \n",
      "\t   CRYZL1, ATP5O, MRPS6, TRMT6, PRNP, C20orf196, CDS2, SLC5A3, CRLS1, PCNA \n",
      "\t   GPCPD1, TMEM230, FERMT1, BMP2, SMOX, TMX4, RCAN1, PLCB4, RNF24, MKKS \n",
      "PC_ 3 \n",
      "Positive:  PKIB, SMPDL3A, SERINC1, TPD52L1, MAN1A1, HDDC2, NCOA7, ASF1A, HINT3, NUS1 \n",
      "\t   CENPW, GOPC, RWDD1, FAM26F, ECHDC1, TSPYL1, PTPRK, DSE, ARHGAP18, NT5DC1 \n",
      "\t   EPB41L2, HDAC2, MARCKS, CTGF, STX7, RPS12, TCF21, TBPL1, NSRP1, SSH2 \n",
      "Negative:  MUS81, CFL1, OVOL1, RNASEH2C, KAT5, RELA, MAP3K11, SIPA1, EHBP1L1, FAM89B \n",
      "\t   SSSCA1, LTBP3, EIF1AD, FAU, SYVN1, MRPL49, TM7SF2, CAPN1, VPS51, CDC42EP2 \n",
      "\t   SCYL1, ZFPL1, DPF2, BANF1, FRMD8, SAC3D1, MALAT1, NEAT1, ARL2, SF3B2 \n",
      "PC_ 4 \n",
      "Positive:  ATL3, PLA2G16, RTN3, MARK2, COX8A, OTUB1, MACROD1, STIP1, STX5, NXF1 \n",
      "\t   FERMT3, TMEM179B, TRPT1, TMEM223, NUDT22, POLR2G, DNAJC4, TTC9C, VEGFB, HNRNPUL2 \n",
      "\t   UBXN1, FKBP2, UQCC3, METTL12, PPP1R14B, GANAB, B3GAT3, PLCB3, EML3, COL9A3 \n",
      "Negative:  FAM134A, CNPPD1, ZFAND2B, BCS1L, ANKZF1, RQCD1, STK16, TUBA4A, VIL1, DNAJB2 \n",
      "\t   DNPEP, CTDSP1, GMPPA, TMBIM1, PNKD, CHPF, AAMP, FARSB, ARPC2, IGFBP5 \n",
      "\t   IGFBP2, ACSL3, RPL37A, WDFY1, XRCC5, MRPL44, FN1, SERPINE2, ATIC, SPAG16 \n",
      "PC_ 5 \n",
      "Positive:  MTHFS, TMED3, CTSH, BCL2A1, MORF4L1, ZFAND6, PSMA4, FAH, ABHD17C, IREB2 \n",
      "\t   CEMIP, WDR61, PRELID3B, ATP5E, CTSZ, PSMA7, NELFCD, MESDC2, MTG2, GNAS \n",
      "\t   DNAJA4, STX16, OSBPL2, VAPB, MESDC1, RAB22A, PMEPA1, ADRM1, RPS17, IDH3A \n",
      "Negative:  SAC3D1, ARL2, ZFPL1, ATG2A, VPS51, TM7SF2, EHD1, FAU, SF1, SYVN1 \n",
      "\t   MRPL49, RPS6KA4, CAPN1, CDC42EP2, CCDC88B, DPF2, AP003774.1, FRMD8, PRDX5, NEAT1 \n",
      "\t   TRMT112, MALAT1, ESRRA, SCYL1, GPR137, BAD, LTBP3, PLCB3, SSSCA1, PPP1R14B \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2023-08-08 16:05:40] define_signif_tumor_subclusters(), tumor: Epithelial cells\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  RPL10A, RPL30, FKBP5, HRSP12, PPARD, STK3, MATN2, SRPK1, DEF6, COX6C \n",
      "\t   LAPTM4B, POLR2K, TAF11, SNRPC, MTDH, MAPK14, C6orf106, RPS10, NUDT3, CPQ \n",
      "\t   MAPK13, C6orf1, SDC2, RPL7, RDH10, C6orf222, UBE2W, PTDSS1, TCEB1, TERF1 \n",
      "Negative:  ZSCAN16-AS1, TRIM27, ZNF165, HLA-F, RP1-313I6.12, ABT1, HLA-A, BTN2A1, ZNRD1, BTN3A2 \n",
      "\t   PPP1R11, HIST1H2AE, TRIM31, HIST1H2BG, HIST1H2BD, TRIM15, HIST1H2AC, HIST1H4C, TRIM26, HIST1H1C \n",
      "\t   HCG18, TRIM38, RPP21, HLA-E, LRRC16A, GNL1, ABCF1, GMNN, PPP1R10, MRPS18B \n",
      "PC_ 2 \n",
      "Positive:  IL4I1, NUP62, TBC1D17, ATF5, AKT1S1, VRK3, MYH14, NR1H2, EMC10, JOSD2 \n",
      "\t   CLEC11A, C19orf48, KLK10, KLK6, ETFB, KLK1, NKG7, PPP2R1A, ZNF331, MYADM \n",
      "\t   NDUFA3, TFPT, PRPF31, CNOT3, LENG1, TMC4, MBOAT7, TSEN34, RPS9, LAIR1 \n",
      "Negative:  RARS2, SNX14, NT5E, CYB5R4, ME1, PGM3, IBTK, FAM46A, BCKDHB, PLSCR1 \n",
      "\t   GYG1, TM4SF1, TM4SF4, SH3BGRL2, WWTR1, COMMD2, RNF13, PFN2, TSC22D2, SERP1 \n",
      "\t   EIF2A, SELT, SIAH2, GPR171, MBNL1, RAP2B, DHX36, GMPS, SSR3, TIPARP \n",
      "PC_ 3 \n",
      "Positive:  BLOC1S1, CD63, ORMDL2, ITGA5, PYM1, COPZ1, RAB5B, SMUG1, CBX5, HNRNPA1 \n",
      "\t   CALCOCO1, ATF7, ATP5G2, RPS26, TARBP2, PCBP2, PRR13, ERBB3, SP1, PA2G4 \n",
      "\t   AAAS, RPL41, C12orf10, ESYT1, PFDN5, MYL6B, MYL6, ITGB7, SMARCC2, RNF41 \n",
      "Negative:  HES1, ATP13A3, OPA1, FAM43A, ACAP2, CCDC50, PPP1R2, CLDN1, APOD, LINC00969 \n",
      "\t   LPP, MUC4, RPL39L, TNK2, ST6GAL1, DNASE2, EIF4A2, RNASEH2A, GCDH, PRDX2 \n",
      "\t   JUNB, HOOK2, FARSA, ASNA1, CALR, DNAJB11, C19orf43, DHPS, WDR83OS, RAD23A \n",
      "PC_ 4 \n",
      "Positive:  FERMT3, TRPT1, STIP1, NUDT22, MACROD1, OTUB1, COX8A, MARK2, DNAJC4, RTN3 \n",
      "\t   VEGFB, ATL3, FKBP2, PLA2G16, PPP1R14B, STX5, PLCB3, NXF1, TMEM179B, TMEM223 \n",
      "\t   BAD, POLR2G, TTC9C, GPR137, HNRNPUL2, ESRRA, DDX17, UBXN1, KDELR3, CBY1 \n",
      "Negative:  DNASE2, GCDH, RNASEH2A, FARSA, PRDX2, CALR, JUNB, RAD23A, HOOK2, WSB2 \n",
      "\t   FBXO21, SDS, VSIG10, SDSL, ASNA1, RITA1, TESC, C12orf49, MED13L, GADD45GIP1 \n",
      "\t   TBX3, TAOK3, DDX54, PEBP1, SUDS3, NFIX, OAS3, C19orf43, PRKAB1, TRMT1 \n",
      "PC_ 5 \n",
      "Positive:  TM7SF2, FAU, VPS51, ZFPL1, SAC3D1, ARL2, SYVN1, ATG2A, EHD1, MRPL49 \n",
      "\t   SF1, CAPN1, RPS6KA4, CDC42EP2, CCDC88B, AP003774.1, DPF2, PRDX5, FRMD8, TRMT112 \n",
      "\t   NEAT1, ESRRA, MALAT1, GPR137, SCYL1, BAD, LTBP3, PLCB3, SSSCA1, PPP1R14B \n",
      "Negative:  GIMAP4, TMEM176B, TMEM176A, AOC1, CDK5, SLC4A2, FASTK, TMUB1, AGAP3, ABCF2 \n",
      "\t   CHPF2, NUB1, FAM96A, HACD3, DPP8, KIAA0101, VWA9, OAZ2, CLPX, SPG21 \n",
      "\t   MTFMT, SNX1, PDCD7, DENND4A, RHEB, PPIB, RAB11A, PRKAG2, TSPAN3, RCN2 \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2023-08-08 16:05:45] define_signif_tumor_subclusters(), tumor: Myeloids\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  CORO1B, RPS6KB2, TMEM134, TBC1D10C, AIP, CDK2AP2, PPP1CA, GSTP1, POLD4, RAD9A \n",
      "\t   ANKRD13D, RCE1, RBM4, NDUFV1, LRFN4, RBM14, RHOD, ADRBK1, KDM2A, CCS \n",
      "\t   NUDT8, CTSF, ZDHHC24, DPP3, UNC93B1, MRPL11, NDUFS8, BRMS1, CD248, YIF1A \n",
      "Negative:  AP2B1, TAF15, RP11-1094M14.11, CCL5, NLE1, CCL15, RFFL, CCL18, ZNF830, CCL13 \n",
      "\t   CCL3, CCL8, CCL11, CCL4, CCL2, TMEM98, MYO1D, CCL3L3, PSMD11, CCL4L2 \n",
      "\t   ZNF207, ZNHIT3, RHOT1, MYO19, SUZ12, GGNBP2, UTP6, DHRS11, COPRS, AATF \n",
      "PC_ 2 \n",
      "Positive:  THAP9-AS1, COPS4, PLAC8, SEC31A, ENOPH1, COQ2, HNRNPDL, ANXA3, HNRNPD, MRPL1 \n",
      "\t   CNOT6L, ANTXR2, RASGEF1B, CXCL13, CCNG2, MRPS18C, CCNI, SEPT11, SHROOM3, CDS1 \n",
      "\t   SCARB2, NUP54, CXCL10, CXCL9, HSD17B11, SDAD1, NAAA, USO1, NUDT9, G3BP2 \n",
      "Negative:  CTD-2336O2.1, AGPAT5, CLN8, FBXO25, DEFA6, ERICH1, DEFA5, CLDN23, MFHAS1, MSRA \n",
      "\t   FDFT1, CTSB, CNOT7, SDS, SDSL, RITA1, TBX3, DDX54, MED13L, C12orf49 \n",
      "\t   OAS3, PAXIP1-AS1, LINC01003, TESC, INSIG1, KMT2C, GALNT11, FBXO21, PRKAG2, RBM33 \n",
      "PC_ 3 \n",
      "Positive:  UQCRB, PLEKHF2, NDUFAF6, PTDSS1, ESRP1, SDC2, CPQ, KIAA1429, GEM, MTDH \n",
      "\t   LAPTM4B, CDH17, MATN2, PDP1, RPL30, FAM92A1, HRSP12, OTUD6B-AS1, STK3, DECR1 \n",
      "\t   NBN, COX6C, RIPK2, CPNE3, POLR2K, RMDN1, WWP1, RAC2, CDC42EP1, IL2RB \n",
      "Negative:  MALAT1, NEAT1, SCYL1, FRMD8, LTBP3, SESN2, PHACTR4, ATPIF1, DPF2, RCC1 \n",
      "\t   DNAJC8, TRNAU1AP, SSSCA1, SNHG12, TAF12, CDC42EP2, RPA2, YTHDF2, FAM89B, EPB41 \n",
      "\t   THEMIS2, CAPN1, SRSF4, EHBP1L1, LAPTM5, PPP1R8, MRPL49, PUM1, SIPA1, SNRNP40 \n",
      "PC_ 4 \n",
      "Positive:  PPP1R8, THEMIS2, STX12, RPA2, IFI6, WASF2, DNAJC8, SYTL1, ATPIF1, TMEM222 \n",
      "\t   SLC9A1, SESN2, NUDC, GPN2, SFN, ARID1A, PHACTR4, RPS6KA1, RCC1, HMGN2 \n",
      "\t   TRNAU1AP, SNHG12, CD52, TAF12, YTHDF2, EPB41, SH3BGRL3, SRSF4, ZNF593, LAPTM5 \n",
      "Negative:  ARL14, NMD3, KPNA4, PDCD10, GOLIM4, MECOM, SMC4, SEC62, MFSD1, GPR160 \n",
      "\t   RARRES1, PRKCI, LXN, SKIL, GFM1, RPL22L1, TNIK, MLF1, FNDC3B, RSRC1 \n",
      "\t   TNFSF10, CCNL1, NCEH1, TIPARP, ECT2, TBL1XR1, SSR3, ZMAT3, GMPS, ACTL6A \n",
      "PC_ 5 \n",
      "Positive:  IMPDH2, QRICH1, NDUFAF3, LAMB2, GPX1, DALRD3, RHOA, TCTA, WDR6, DAG1 \n",
      "\t   P4HTM, GMPPB, APEH, IP6K1, RBM6, RBM5, ARIH2, GNAI2, PRKAR2A, SEMA3B \n",
      "\t   IP6K2, IFRD2, SLC26A6, HYAL2, UQCRC1, TUSC2, SHISA5, RASSF1, TMA7, NPRL2 \n",
      "Negative:  CLDN1, LPP, CCDC50, OPA1, HES1, RPL39L, ATP13A3, ST6GAL1, FAM43A, EIF4A2 \n",
      "\t   ACAP2, DNAJB11, TRA2B, PPP1R2, APOD, IGF2BP2, LINC00969, SENP2, MUC4, TNK2 \n",
      "\t   LIPH, TMEM41A, MAP3K13, MAGEF1, EPHB3, POLR2H, EIF4G1, PSMD2, RAP1GDS1, TSPAN5 \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2023-08-08 16:05:47] define_signif_tumor_subclusters(), tumor: Mast cells\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  COX6C, POLR2K, STK3, HRSP12, RPL30, ENY2, MATN2, EBAG9, EIF3H, ENPP2 \n",
      "\t   MAL2, MRPL13, MED30, UTP23, RAD21, LAPTM4B, DERL1, C8orf76, MTDH, ZHX1 \n",
      "\t   ATAD2, CPQ, WDYHV1, SDC2, FBXO32, PTDSS1, UQCRB, ANXA13, PLEKHF2, NDUFAF6 \n",
      "Negative:  ADNP, BCAS4, PARD6B, DPM1, NFATC2, ATP9A, PTPN1, TSHZ2, ZNF217, CEBPB \n",
      "\t   PFDN4, TMEM189, FAM210B, AURKA, SNAI1, GABARAP, PHF23, ACADVL, XAF1, CTDNEP1 \n",
      "\t   RTFDC1, RNF114, ELP5, MED31, CLDN7, RAE1, SPATA2, EIF5A, TXNDC17, B4GALT5 \n",
      "PC_ 2 \n",
      "Positive:  U2SURP, ATP1B3, PLOD2, RNF7, PLSCR1, RASA2, GYG1, ZBTB38, TM4SF1, SLC25A36 \n",
      "\t   RBP1, TM4SF4, COPB2, MRPS22, WWTR1, COMMD2, ANAPC13, RNF13, AMOTL2, PFN2 \n",
      "\t   RYK, TSC22D2, SRPRB, SERP1, EIF2A, SELT, SIAH2, GPR171, CDV3, MBNL1 \n",
      "Negative:  TRPM4, CD37, SNRNP70, PIH1D1, RUVBL2, BAX, FTL, NUCB1, PPP1R15A, BCAT2 \n",
      "\t   FUT2, RPL18, SULT2B1, LMTK3, CYTH2, GRWD1, KDELR1, EMP3, CARD8, SEPW1 \n",
      "\t   NPRL2, CYB561D2, RASSF1, TMEM115, GLTSCR2, RRP9, TUSC2, MAPKAPK3, PCBP4, TEX264 \n",
      "PC_ 3 \n",
      "Positive:  TTC19, ZSWIM7, TRIM16, PMP22, ELAC2, LINC00675, SCO1, STX8, NDEL1, RPL26 \n",
      "\t   RANGRF, PPP1R14B, PLCB3, BAD, FKBP2, RASD1, GPR137, VEGFB, TMEM107, ESRRA \n",
      "\t   DNAJC4, TRMT112, NUDT22, PEMT, VAMP2, TRPT1, PRDX5, FERMT3, STIP1, PER1 \n",
      "Negative:  JUNB, PRDX2, HOOK2, RNASEH2A, ASNA1, DNASE2, C19orf43, DHPS, GCDH, ELOF1 \n",
      "\t   WDR83OS, ACP5, ECSIT, ZNF791, PRKCSH, FARSA, TMEM205, CALR, LDLR, RAD23A \n",
      "\t   GADD45GIP1, SMARCA4, NFIX, C19orf52, TRMT1, NACC1, YIPF2, STX10, CARM1, TMED1 \n",
      "PC_ 4 \n",
      "Positive:  SPATA2, RNF114, B4GALT5, ZFAS1, SNAI1, ZNFX1, DDX27, STAU1, TMEM189, CSE1L \n",
      "\t   CEBPB, ARFGEF2, PTPN1, SULF2, PARD6B, NCOA3, ZMYND8, BCAS4, TP53RK, ADNP \n",
      "\t   SLC35C2, CD40, DPM1, NFATC2, MMP9, ATP9A, TSHZ2, PCIF1, SERINC3, PKIG \n",
      "Negative:  GGT6, PELP1, SPNS2, ARRB2, MED11, CXCL16, UBE2G1, ATP2A3, TM4SF5, PSMB6 \n",
      "\t   NCBP3, MINK1, SLC25A11, ITGAE, RNF167, EMC6, PFN1, TAX1BP3, SPAG7, CLUH \n",
      "\t   KIF1C, RABEP1, PAFAH1B1, NUP88, MNT, RPAIN, TSR1, RPA1, CTC-524C5.2, SERPINF1 \n",
      "PC_ 5 \n",
      "Positive:  KRT20, KRT23, TMEM99, KRT19, KRT10, KRT17, SMARCE1, CCR7, IGFBP4, EIF1 \n",
      "\t   TOP2A, JUP, RARA, P3H4, WIPF2, CASC3, SPTLC2, AHSA1, MSL1, SLIRP \n",
      "\t   GSTZ1, GPATCH2L, SNW1, IFT43, THRA, SEL1L, C14orf1, GPR65, GSN, RAB14 \n",
      "Negative:  FDPS, RUSC1, ASH1L, YY1AP1, SCAMP3, DAP3, GON4L, GBA, RIT1, MTX1 \n",
      "\t   KIAA0907, MUC1, ARHGEF2, KRTCAP2, SSR2, DPM3, SLC50A1, LAMTOR2, EFNA1, EFNA3 \n",
      "\t   EFNA4, RAB25, ADAM15, ZBTB7B, LMNA, FLAD1, SEMA4A, CKS1B, SLC25A44, PMF1 \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2023-08-08 16:05:48] -mirroring for hspike\n",
      "INFO [2023-08-08 16:05:48] define_signif_tumor_subclusters(p_val=0.1\n",
      "INFO [2023-08-08 16:05:48] define_signif_tumor_subclusters(), tumor: spike_tumor_cell_T cells\n",
      "INFO [2023-08-08 16:05:48] cut tree into: 1 groups\n",
      "INFO [2023-08-08 16:05:48] -processing spike_tumor_cell_T cells,spike_tumor_cell_T cells_s1\n",
      "INFO [2023-08-08 16:05:48] define_signif_tumor_subclusters(), tumor: spike_tumor_cell_B cells\n",
      "INFO [2023-08-08 16:05:48] cut tree into: 1 groups\n",
      "INFO [2023-08-08 16:05:48] -processing spike_tumor_cell_B cells,spike_tumor_cell_B cells_s1\n",
      "INFO [2023-08-08 16:05:48] define_signif_tumor_subclusters(), tumor: spike_tumor_cell_Stromal cells\n",
      "INFO [2023-08-08 16:05:49] cut tree into: 1 groups\n",
      "INFO [2023-08-08 16:05:49] -processing spike_tumor_cell_Stromal cells,spike_tumor_cell_Stromal cells_s1\n",
      "INFO [2023-08-08 16:05:49] define_signif_tumor_subclusters(), tumor: spike_tumor_cell_Epithelial cells\n",
      "INFO [2023-08-08 16:05:49] cut tree into: 1 groups\n",
      "INFO [2023-08-08 16:05:49] -processing spike_tumor_cell_Epithelial cells,spike_tumor_cell_Epithelial cells_s1\n",
      "INFO [2023-08-08 16:05:49] define_signif_tumor_subclusters(), tumor: spike_tumor_cell_Myeloids\n",
      "INFO [2023-08-08 16:05:49] cut tree into: 1 groups\n",
      "INFO [2023-08-08 16:05:49] -processing spike_tumor_cell_Myeloids,spike_tumor_cell_Myeloids_s1\n",
      "INFO [2023-08-08 16:05:49] define_signif_tumor_subclusters(), tumor: spike_tumor_cell_Mast cells\n",
      "INFO [2023-08-08 16:05:49] cut tree into: 1 groups\n",
      "INFO [2023-08-08 16:05:49] -processing spike_tumor_cell_Mast cells,spike_tumor_cell_Mast cells_s1\n",
      "INFO [2023-08-08 16:05:49] define_signif_tumor_subclusters(), tumor: simnorm_cell_T cells\n",
      "INFO [2023-08-08 16:05:49] cut tree into: 1 groups\n",
      "INFO [2023-08-08 16:05:49] -processing simnorm_cell_T cells,simnorm_cell_T cells_s1\n",
      "INFO [2023-08-08 16:05:49] define_signif_tumor_subclusters(), tumor: simnorm_cell_B cells\n",
      "INFO [2023-08-08 16:05:49] cut tree into: 1 groups\n",
      "INFO [2023-08-08 16:05:49] -processing simnorm_cell_B cells,simnorm_cell_B cells_s1\n",
      "INFO [2023-08-08 16:05:49] define_signif_tumor_subclusters(), tumor: simnorm_cell_Stromal cells\n",
      "INFO [2023-08-08 16:05:49] cut tree into: 1 groups\n",
      "INFO [2023-08-08 16:05:49] -processing simnorm_cell_Stromal cells,simnorm_cell_Stromal cells_s1\n",
      "INFO [2023-08-08 16:05:49] define_signif_tumor_subclusters(), tumor: simnorm_cell_Epithelial cells\n",
      "INFO [2023-08-08 16:05:49] cut tree into: 1 groups\n",
      "INFO [2023-08-08 16:05:49] -processing simnorm_cell_Epithelial cells,simnorm_cell_Epithelial cells_s1\n",
      "INFO [2023-08-08 16:05:49] define_signif_tumor_subclusters(), tumor: simnorm_cell_Myeloids\n",
      "INFO [2023-08-08 16:05:49] cut tree into: 1 groups\n",
      "INFO [2023-08-08 16:05:49] -processing simnorm_cell_Myeloids,simnorm_cell_Myeloids_s1\n",
      "INFO [2023-08-08 16:05:49] define_signif_tumor_subclusters(), tumor: simnorm_cell_Mast cells\n",
      "INFO [2023-08-08 16:05:49] cut tree into: 1 groups\n",
      "INFO [2023-08-08 16:05:49] -processing simnorm_cell_Mast cells,simnorm_cell_Mast cells_s1\n",
      "INFO [2023-08-08 16:09:57] ::plot_cnv:Start\n",
      "INFO [2023-08-08 16:09:57] ::plot_cnv:Current data dimensions (r,c)=7077,63689 Total=451182872.676612 Min=0.758515889139591 Max=5.40416897818636.\n",
      "INFO [2023-08-08 16:10:02] ::plot_cnv:Depending on the size of the matrix this may take a moment.\n",
      "INFO [2023-08-08 16:10:29] plot_cnv(): auto thresholding at: (0.917621 , 1.084401)\n",
      "INFO [2023-08-08 16:12:28] plot_cnv_observation:Start\n",
      "INFO [2023-08-08 16:12:28] Observation data size: Cells= 47285 Genes= 7077\n",
      "INFO [2023-08-08 16:12:28] clustering observations via method: ward.D\n",
      "INFO [2023-08-08 16:12:28] Number of cells in group(1) is 294\n",
      "INFO [2023-08-08 16:12:28] group size being clustered:  294,7077\n",
      "INFO [2023-08-08 16:12:29] Number of cells in group(2) is 293\n",
      "INFO [2023-08-08 16:12:29] group size being clustered:  293,7077\n",
      "INFO [2023-08-08 16:12:30] Number of cells in group(3) is 205\n",
      "INFO [2023-08-08 16:12:30] group size being clustered:  205,7077\n",
      "INFO [2023-08-08 16:12:30] Number of cells in group(4) is 169\n",
      "INFO [2023-08-08 16:12:30] group size being clustered:  169,7077\n",
      "INFO [2023-08-08 16:12:30] Number of cells in group(5) is 155\n",
      "INFO [2023-08-08 16:12:30] group size being clustered:  155,7077\n",
      "INFO [2023-08-08 16:12:30] Number of cells in group(6) is 154\n",
      "INFO [2023-08-08 16:12:30] group size being clustered:  154,7077\n",
      "INFO [2023-08-08 16:12:31] Number of cells in group(7) is 153\n",
      "INFO [2023-08-08 16:12:31] group size being clustered:  153,7077\n",
      "INFO [2023-08-08 16:12:31] Number of cells in group(8) is 114\n",
      "INFO [2023-08-08 16:12:31] group size being clustered:  114,7077\n",
      "INFO [2023-08-08 16:12:31] Number of cells in group(9) is 110\n",
      "INFO [2023-08-08 16:12:31] group size being clustered:  110,7077\n",
      "INFO [2023-08-08 16:12:31] Number of cells in group(10) is 93\n",
      "INFO [2023-08-08 16:12:31] group size being clustered:  93,7077\n",
      "INFO [2023-08-08 16:12:31] Number of cells in group(11) is 93\n",
      "INFO [2023-08-08 16:12:31] group size being clustered:  93,7077\n",
      "INFO [2023-08-08 16:12:31] Number of cells in group(12) is 90\n",
      "INFO [2023-08-08 16:12:31] group size being clustered:  90,7077\n",
      "INFO [2023-08-08 16:12:31] Number of cells in group(13) is 86\n",
      "INFO [2023-08-08 16:12:31] group size being clustered:  86,7077\n",
      "INFO [2023-08-08 16:12:31] Number of cells in group(14) is 86\n",
      "INFO [2023-08-08 16:12:31] group size being clustered:  86,7077\n",
      "INFO [2023-08-08 16:12:31] Number of cells in group(15) is 78\n",
      "INFO [2023-08-08 16:12:31] group size being clustered:  78,7077\n",
      "INFO [2023-08-08 16:12:31] Number of cells in group(16) is 66\n",
      "INFO [2023-08-08 16:12:31] group size being clustered:  66,7077\n",
      "INFO [2023-08-08 16:12:31] Number of cells in group(17) is 59\n",
      "INFO [2023-08-08 16:12:31] group size being clustered:  59,7077\n",
      "INFO [2023-08-08 16:12:31] Number of cells in group(18) is 53\n",
      "INFO [2023-08-08 16:12:31] group size being clustered:  53,7077\n",
      "INFO [2023-08-08 16:12:31] Number of cells in group(19) is 53\n",
      "INFO [2023-08-08 16:12:31] group size being clustered:  53,7077\n",
      "INFO [2023-08-08 16:12:31] Number of cells in group(20) is 53\n",
      "INFO [2023-08-08 16:12:31] group size being clustered:  53,7077\n",
      "INFO [2023-08-08 16:12:31] Number of cells in group(21) is 48\n",
      "INFO [2023-08-08 16:12:31] group size being clustered:  48,7077\n",
      "INFO [2023-08-08 16:12:31] Number of cells in group(22) is 47\n",
      "INFO [2023-08-08 16:12:31] group size being clustered:  47,7077\n",
      "INFO [2023-08-08 16:12:31] Number of cells in group(23) is 42\n",
      "INFO [2023-08-08 16:12:31] group size being clustered:  42,7077\n",
      "INFO [2023-08-08 16:12:31] Number of cells in group(24) is 39\n",
      "INFO [2023-08-08 16:12:31] group size being clustered:  39,7077\n",
      "INFO [2023-08-08 16:12:31] Number of cells in group(25) is 37\n",
      "INFO [2023-08-08 16:12:31] group size being clustered:  37,7077\n",
      "INFO [2023-08-08 16:12:31] Number of cells in group(26) is 37\n",
      "INFO [2023-08-08 16:12:31] group size being clustered:  37,7077\n",
      "INFO [2023-08-08 16:12:31] Number of cells in group(27) is 20\n",
      "INFO [2023-08-08 16:12:31] group size being clustered:  20,7077\n",
      "INFO [2023-08-08 16:12:31] Number of cells in group(28) is 13\n",
      "INFO [2023-08-08 16:12:31] group size being clustered:  13,7077\n",
      "INFO [2023-08-08 16:12:31] Number of cells in group(29) is 13\n",
      "INFO [2023-08-08 16:12:31] group size being clustered:  13,7077\n",
      "INFO [2023-08-08 16:12:31] Number of cells in group(30) is 7\n",
      "INFO [2023-08-08 16:12:31] group size being clustered:  7,7077\n",
      "INFO [2023-08-08 16:12:31] Number of cells in group(31) is 7\n",
      "INFO [2023-08-08 16:12:31] group size being clustered:  7,7077\n",
      "INFO [2023-08-08 16:12:31] Number of cells in group(32) is 6\n",
      "INFO [2023-08-08 16:12:31] group size being clustered:  6,7077\n",
      "INFO [2023-08-08 16:12:31] Number of cells in group(33) is 4\n",
      "INFO [2023-08-08 16:12:31] group size being clustered:  4,7077\n",
      "INFO [2023-08-08 16:12:31] Number of cells in group(34) is 3\n",
      "INFO [2023-08-08 16:12:31] group size being clustered:  3,7077\n",
      "INFO [2023-08-08 16:12:31] Number of cells in group(35) is 3\n",
      "INFO [2023-08-08 16:12:31] group size being clustered:  3,7077\n",
      "INFO [2023-08-08 16:12:31] Number of cells in group(36) is 1\n",
      "INFO [2023-08-08 16:12:31] Skipping group: 36, since less than 2 entries\n",
      "INFO [2023-08-08 16:12:31] Number of cells in group(37) is 1\n",
      "INFO [2023-08-08 16:12:31] Skipping group: 37, since less than 2 entries\n",
      "INFO [2023-08-08 16:12:31] Number of cells in group(38) is 1\n",
      "INFO [2023-08-08 16:12:31] Skipping group: 38, since less than 2 entries\n",
      "INFO [2023-08-08 16:12:31] Number of cells in group(39) is 1\n",
      "INFO [2023-08-08 16:12:31] Skipping group: 39, since less than 2 entries\n",
      "INFO [2023-08-08 16:12:31] Number of cells in group(40) is 1\n",
      "INFO [2023-08-08 16:12:31] Skipping group: 40, since less than 2 entries\n",
      "INFO [2023-08-08 16:12:31] Number of cells in group(41) is 1\n",
      "INFO [2023-08-08 16:12:31] Skipping group: 41, since less than 2 entries\n",
      "INFO [2023-08-08 16:12:31] Number of cells in group(42) is 1\n",
      "INFO [2023-08-08 16:12:32] Skipping group: 42, since less than 2 entries\n",
      "INFO [2023-08-08 16:12:32] Number of cells in group(43) is 1\n",
      "INFO [2023-08-08 16:12:32] Skipping group: 43, since less than 2 entries\n",
      "INFO [2023-08-08 16:12:32] Number of cells in group(44) is 332\n",
      "INFO [2023-08-08 16:12:32] group size being clustered:  332,7077\n",
      "INFO [2023-08-08 16:12:33] Number of cells in group(45) is 255\n",
      "INFO [2023-08-08 16:12:33] group size being clustered:  255,7077\n",
      "INFO [2023-08-08 16:12:33] Number of cells in group(46) is 213\n",
      "INFO [2023-08-08 16:12:33] group size being clustered:  213,7077\n",
      "INFO [2023-08-08 16:12:34] Number of cells in group(47) is 149\n",
      "INFO [2023-08-08 16:12:34] group size being clustered:  149,7077\n",
      "INFO [2023-08-08 16:12:34] Number of cells in group(48) is 114\n",
      "INFO [2023-08-08 16:12:34] group size being clustered:  114,7077\n",
      "INFO [2023-08-08 16:12:34] Number of cells in group(49) is 106\n",
      "INFO [2023-08-08 16:12:34] group size being clustered:  106,7077\n",
      "INFO [2023-08-08 16:12:34] Number of cells in group(50) is 99\n",
      "INFO [2023-08-08 16:12:34] group size being clustered:  99,7077\n",
      "INFO [2023-08-08 16:12:34] Number of cells in group(51) is 97\n",
      "INFO [2023-08-08 16:12:34] group size being clustered:  97,7077\n",
      "INFO [2023-08-08 16:12:34] Number of cells in group(52) is 93\n",
      "INFO [2023-08-08 16:12:34] group size being clustered:  93,7077\n",
      "INFO [2023-08-08 16:12:34] Number of cells in group(53) is 81\n",
      "INFO [2023-08-08 16:12:34] group size being clustered:  81,7077\n",
      "INFO [2023-08-08 16:12:34] Number of cells in group(54) is 64\n",
      "INFO [2023-08-08 16:12:34] group size being clustered:  64,7077\n",
      "INFO [2023-08-08 16:12:34] Number of cells in group(55) is 58\n",
      "INFO [2023-08-08 16:12:34] group size being clustered:  58,7077\n",
      "INFO [2023-08-08 16:12:34] Number of cells in group(56) is 10\n",
      "INFO [2023-08-08 16:12:34] group size being clustered:  10,7077\n",
      "INFO [2023-08-08 16:12:34] Number of cells in group(57) is 9\n",
      "INFO [2023-08-08 16:12:34] group size being clustered:  9,7077\n",
      "INFO [2023-08-08 16:12:34] Number of cells in group(58) is 3\n",
      "INFO [2023-08-08 16:12:34] group size being clustered:  3,7077\n",
      "INFO [2023-08-08 16:12:34] Number of cells in group(59) is 3\n",
      "INFO [2023-08-08 16:12:34] group size being clustered:  3,7077\n",
      "INFO [2023-08-08 16:12:34] Number of cells in group(60) is 1\n",
      "INFO [2023-08-08 16:12:34] Skipping group: 60, since less than 2 entries\n",
      "INFO [2023-08-08 16:12:34] Number of cells in group(61) is 1\n",
      "INFO [2023-08-08 16:12:34] Skipping group: 61, since less than 2 entries\n",
      "INFO [2023-08-08 16:12:34] Number of cells in group(62) is 1\n",
      "INFO [2023-08-08 16:12:34] Skipping group: 62, since less than 2 entries\n",
      "INFO [2023-08-08 16:12:34] Number of cells in group(63) is 285\n",
      "INFO [2023-08-08 16:12:34] group size being clustered:  285,7077\n",
      "INFO [2023-08-08 16:12:35] Number of cells in group(64) is 241\n",
      "INFO [2023-08-08 16:12:35] group size being clustered:  241,7077\n",
      "INFO [2023-08-08 16:12:36] Number of cells in group(65) is 225\n",
      "INFO [2023-08-08 16:12:36] group size being clustered:  225,7077\n",
      "INFO [2023-08-08 16:12:36] Number of cells in group(66) is 223\n",
      "INFO [2023-08-08 16:12:36] group size being clustered:  223,7077\n",
      "INFO [2023-08-08 16:12:37] Number of cells in group(67) is 213\n",
      "INFO [2023-08-08 16:12:37] group size being clustered:  213,7077\n",
      "INFO [2023-08-08 16:12:37] Number of cells in group(68) is 168\n",
      "INFO [2023-08-08 16:12:37] group size being clustered:  168,7077\n",
      "INFO [2023-08-08 16:12:37] Number of cells in group(69) is 164\n",
      "INFO [2023-08-08 16:12:37] group size being clustered:  164,7077\n",
      "INFO [2023-08-08 16:12:37] Number of cells in group(70) is 161\n",
      "INFO [2023-08-08 16:12:37] group size being clustered:  161,7077\n",
      "INFO [2023-08-08 16:12:38] Number of cells in group(71) is 156\n",
      "INFO [2023-08-08 16:12:38] group size being clustered:  156,7077\n",
      "INFO [2023-08-08 16:12:38] Number of cells in group(72) is 106\n",
      "INFO [2023-08-08 16:12:38] group size being clustered:  106,7077\n",
      "INFO [2023-08-08 16:12:38] Number of cells in group(73) is 69\n",
      "INFO [2023-08-08 16:12:38] group size being clustered:  69,7077\n",
      "INFO [2023-08-08 16:12:38] Number of cells in group(74) is 40\n",
      "INFO [2023-08-08 16:12:38] group size being clustered:  40,7077\n",
      "INFO [2023-08-08 16:12:38] Number of cells in group(75) is 34\n",
      "INFO [2023-08-08 16:12:38] group size being clustered:  34,7077\n",
      "INFO [2023-08-08 16:12:38] Number of cells in group(76) is 9\n",
      "INFO [2023-08-08 16:12:38] group size being clustered:  9,7077\n",
      "INFO [2023-08-08 16:12:38] Number of cells in group(77) is 6\n",
      "INFO [2023-08-08 16:12:38] group size being clustered:  6,7077\n",
      "INFO [2023-08-08 16:12:38] Number of cells in group(78) is 4\n",
      "INFO [2023-08-08 16:12:38] group size being clustered:  4,7077\n",
      "INFO [2023-08-08 16:12:38] Number of cells in group(79) is 4\n",
      "INFO [2023-08-08 16:12:38] group size being clustered:  4,7077\n",
      "INFO [2023-08-08 16:12:38] Number of cells in group(80) is 3\n",
      "INFO [2023-08-08 16:12:38] group size being clustered:  3,7077\n",
      "INFO [2023-08-08 16:12:38] Number of cells in group(81) is 3\n",
      "INFO [2023-08-08 16:12:38] group size being clustered:  3,7077\n",
      "INFO [2023-08-08 16:12:38] Number of cells in group(82) is 2\n",
      "INFO [2023-08-08 16:12:38] group size being clustered:  2,7077\n",
      "INFO [2023-08-08 16:12:38] Number of cells in group(83) is 2\n",
      "INFO [2023-08-08 16:12:38] group size being clustered:  2,7077\n",
      "INFO [2023-08-08 16:12:38] Number of cells in group(84) is 1\n",
      "INFO [2023-08-08 16:12:38] Skipping group: 84, since less than 2 entries\n",
      "INFO [2023-08-08 16:12:38] Number of cells in group(85) is 1\n",
      "INFO [2023-08-08 16:12:38] Skipping group: 85, since less than 2 entries\n",
      "INFO [2023-08-08 16:12:38] Number of cells in group(86) is 1\n",
      "INFO [2023-08-08 16:12:38] Skipping group: 86, since less than 2 entries\n",
      "INFO [2023-08-08 16:12:38] Number of cells in group(87) is 1\n",
      "INFO [2023-08-08 16:12:38] Skipping group: 87, since less than 2 entries\n",
      "INFO [2023-08-08 16:12:38] Number of cells in group(88) is 1\n",
      "INFO [2023-08-08 16:12:38] Skipping group: 88, since less than 2 entries\n",
      "INFO [2023-08-08 16:12:38] Number of cells in group(89) is 351\n",
      "INFO [2023-08-08 16:12:38] group size being clustered:  351,7077\n",
      "INFO [2023-08-08 16:12:40] Number of cells in group(90) is 234\n",
      "INFO [2023-08-08 16:12:40] group size being clustered:  234,7077\n",
      "INFO [2023-08-08 16:12:40] Number of cells in group(91) is 192\n",
      "INFO [2023-08-08 16:12:40] group size being clustered:  192,7077\n",
      "INFO [2023-08-08 16:12:40] Number of cells in group(92) is 164\n",
      "INFO [2023-08-08 16:12:40] group size being clustered:  164,7077\n",
      "INFO [2023-08-08 16:12:41] Number of cells in group(93) is 159\n",
      "INFO [2023-08-08 16:12:41] group size being clustered:  159,7077\n",
      "INFO [2023-08-08 16:12:41] Number of cells in group(94) is 155\n",
      "INFO [2023-08-08 16:12:41] group size being clustered:  155,7077\n",
      "INFO [2023-08-08 16:12:41] Number of cells in group(95) is 151\n",
      "INFO [2023-08-08 16:12:41] group size being clustered:  151,7077\n",
      "INFO [2023-08-08 16:12:41] Number of cells in group(96) is 146\n",
      "INFO [2023-08-08 16:12:41] group size being clustered:  146,7077\n",
      "INFO [2023-08-08 16:12:41] Number of cells in group(97) is 113\n",
      "INFO [2023-08-08 16:12:41] group size being clustered:  113,7077\n",
      "INFO [2023-08-08 16:12:41] Number of cells in group(98) is 110\n",
      "INFO [2023-08-08 16:12:41] group size being clustered:  110,7077\n",
      "INFO [2023-08-08 16:12:41] Number of cells in group(99) is 30\n",
      "INFO [2023-08-08 16:12:41] group size being clustered:  30,7077\n",
      "INFO [2023-08-08 16:12:41] Number of cells in group(100) is 15\n",
      "INFO [2023-08-08 16:12:41] group size being clustered:  15,7077\n",
      "INFO [2023-08-08 16:12:41] Number of cells in group(101) is 8\n",
      "INFO [2023-08-08 16:12:41] group size being clustered:  8,7077\n",
      "INFO [2023-08-08 16:12:41] Number of cells in group(102) is 7\n",
      "INFO [2023-08-08 16:12:41] group size being clustered:  7,7077\n",
      "INFO [2023-08-08 16:12:41] Number of cells in group(103) is 2\n",
      "INFO [2023-08-08 16:12:41] group size being clustered:  2,7077\n",
      "INFO [2023-08-08 16:12:41] Number of cells in group(104) is 2\n",
      "INFO [2023-08-08 16:12:41] group size being clustered:  2,7077\n",
      "INFO [2023-08-08 16:12:41] Number of cells in group(105) is 2\n",
      "INFO [2023-08-08 16:12:41] group size being clustered:  2,7077\n",
      "INFO [2023-08-08 16:12:41] Number of cells in group(106) is 294\n",
      "INFO [2023-08-08 16:12:41] group size being clustered:  294,7077\n",
      "INFO [2023-08-08 16:12:42] Number of cells in group(107) is 188\n",
      "INFO [2023-08-08 16:12:42] group size being clustered:  188,7077\n",
      "INFO [2023-08-08 16:12:43] Number of cells in group(108) is 100\n",
      "INFO [2023-08-08 16:12:43] group size being clustered:  100,7077\n",
      "INFO [2023-08-08 16:12:43] Number of cells in group(109) is 56\n",
      "INFO [2023-08-08 16:12:43] group size being clustered:  56,7077\n",
      "INFO [2023-08-08 16:12:43] Number of cells in group(110) is 13\n",
      "INFO [2023-08-08 16:12:43] group size being clustered:  13,7077\n",
      "INFO [2023-08-08 16:12:43] Number of cells in group(111) is 9\n",
      "INFO [2023-08-08 16:12:43] group size being clustered:  9,7077\n",
      "INFO [2023-08-08 16:12:43] Number of cells in group(112) is 190\n",
      "INFO [2023-08-08 16:12:43] group size being clustered:  190,7077\n",
      "INFO [2023-08-08 16:12:43] Number of cells in group(113) is 183\n",
      "INFO [2023-08-08 16:12:43] group size being clustered:  183,7077\n",
      "INFO [2023-08-08 16:12:43] Number of cells in group(114) is 180\n",
      "INFO [2023-08-08 16:12:43] group size being clustered:  180,7077\n",
      "INFO [2023-08-08 16:12:43] Number of cells in group(115) is 160\n",
      "INFO [2023-08-08 16:12:43] group size being clustered:  160,7077\n",
      "INFO [2023-08-08 16:12:44] Number of cells in group(116) is 139\n",
      "INFO [2023-08-08 16:12:44] group size being clustered:  139,7077\n",
      "INFO [2023-08-08 16:12:44] Number of cells in group(117) is 130\n",
      "INFO [2023-08-08 16:12:44] group size being clustered:  130,7077\n",
      "INFO [2023-08-08 16:12:44] Number of cells in group(118) is 110\n",
      "INFO [2023-08-08 16:12:44] group size being clustered:  110,7077\n",
      "INFO [2023-08-08 16:12:44] Number of cells in group(119) is 91\n",
      "INFO [2023-08-08 16:12:44] group size being clustered:  91,7077\n",
      "INFO [2023-08-08 16:12:44] Number of cells in group(120) is 70\n",
      "INFO [2023-08-08 16:12:44] group size being clustered:  70,7077\n",
      "INFO [2023-08-08 16:12:44] Number of cells in group(121) is 66\n",
      "INFO [2023-08-08 16:12:44] group size being clustered:  66,7077\n",
      "INFO [2023-08-08 16:12:44] Number of cells in group(122) is 66\n",
      "INFO [2023-08-08 16:12:44] group size being clustered:  66,7077\n",
      "INFO [2023-08-08 16:12:44] Number of cells in group(123) is 56\n",
      "INFO [2023-08-08 16:12:44] group size being clustered:  56,7077\n",
      "INFO [2023-08-08 16:12:44] Number of cells in group(124) is 46\n",
      "INFO [2023-08-08 16:12:44] group size being clustered:  46,7077\n",
      "INFO [2023-08-08 16:12:44] Number of cells in group(125) is 36\n",
      "INFO [2023-08-08 16:12:44] group size being clustered:  36,7077\n",
      "INFO [2023-08-08 16:12:44] Number of cells in group(126) is 11\n",
      "INFO [2023-08-08 16:12:44] group size being clustered:  11,7077\n",
      "INFO [2023-08-08 16:12:44] Number of cells in group(127) is 4\n",
      "INFO [2023-08-08 16:12:44] group size being clustered:  4,7077\n",
      "INFO [2023-08-08 16:12:44] Number of cells in group(128) is 1\n",
      "INFO [2023-08-08 16:12:44] Skipping group: 128, since less than 2 entries\n",
      "INFO [2023-08-08 16:12:44] Number of cells in group(129) is 1\n",
      "INFO [2023-08-08 16:12:44] Skipping group: 129, since less than 2 entries\n",
      "INFO [2023-08-08 16:12:44] Number of cells in group(130) is 1\n",
      "INFO [2023-08-08 16:12:44] Skipping group: 130, since less than 2 entries\n",
      "INFO [2023-08-08 16:12:44] Number of cells in group(131) is 286\n",
      "INFO [2023-08-08 16:12:44] group size being clustered:  286,7077\n",
      "INFO [2023-08-08 16:12:45] Number of cells in group(132) is 235\n",
      "INFO [2023-08-08 16:12:45] group size being clustered:  235,7077\n",
      "INFO [2023-08-08 16:12:46] Number of cells in group(133) is 205\n",
      "INFO [2023-08-08 16:12:46] group size being clustered:  205,7077\n",
      "INFO [2023-08-08 16:12:46] Number of cells in group(134) is 180\n",
      "INFO [2023-08-08 16:12:46] group size being clustered:  180,7077\n",
      "INFO [2023-08-08 16:12:46] Number of cells in group(135) is 170\n",
      "INFO [2023-08-08 16:12:46] group size being clustered:  170,7077\n",
      "INFO [2023-08-08 16:12:46] Number of cells in group(136) is 150\n",
      "INFO [2023-08-08 16:12:46] group size being clustered:  150,7077\n",
      "INFO [2023-08-08 16:12:46] Number of cells in group(137) is 125\n",
      "INFO [2023-08-08 16:12:46] group size being clustered:  125,7077\n",
      "INFO [2023-08-08 16:12:47] Number of cells in group(138) is 105\n",
      "INFO [2023-08-08 16:12:47] group size being clustered:  105,7077\n",
      "INFO [2023-08-08 16:12:47] Number of cells in group(139) is 101\n",
      "INFO [2023-08-08 16:12:47] group size being clustered:  101,7077\n",
      "INFO [2023-08-08 16:12:47] Number of cells in group(140) is 100\n",
      "INFO [2023-08-08 16:12:47] group size being clustered:  100,7077\n",
      "INFO [2023-08-08 16:12:47] Number of cells in group(141) is 99\n",
      "INFO [2023-08-08 16:12:47] group size being clustered:  99,7077\n",
      "INFO [2023-08-08 16:12:47] Number of cells in group(142) is 91\n",
      "INFO [2023-08-08 16:12:47] group size being clustered:  91,7077\n",
      "INFO [2023-08-08 16:12:47] Number of cells in group(143) is 81\n",
      "INFO [2023-08-08 16:12:47] group size being clustered:  81,7077\n",
      "INFO [2023-08-08 16:12:47] Number of cells in group(144) is 81\n",
      "INFO [2023-08-08 16:12:47] group size being clustered:  81,7077\n",
      "INFO [2023-08-08 16:12:47] Number of cells in group(145) is 77\n",
      "INFO [2023-08-08 16:12:47] group size being clustered:  77,7077\n",
      "INFO [2023-08-08 16:12:47] Number of cells in group(146) is 77\n",
      "INFO [2023-08-08 16:12:47] group size being clustered:  77,7077\n",
      "INFO [2023-08-08 16:12:47] Number of cells in group(147) is 77\n",
      "INFO [2023-08-08 16:12:47] group size being clustered:  77,7077\n",
      "INFO [2023-08-08 16:12:47] Number of cells in group(148) is 76\n",
      "INFO [2023-08-08 16:12:47] group size being clustered:  76,7077\n",
      "INFO [2023-08-08 16:12:47] Number of cells in group(149) is 75\n",
      "INFO [2023-08-08 16:12:47] group size being clustered:  75,7077\n",
      "INFO [2023-08-08 16:12:47] Number of cells in group(150) is 73\n",
      "INFO [2023-08-08 16:12:47] group size being clustered:  73,7077\n",
      "INFO [2023-08-08 16:12:47] Number of cells in group(151) is 72\n",
      "INFO [2023-08-08 16:12:47] group size being clustered:  72,7077\n",
      "INFO [2023-08-08 16:12:47] Number of cells in group(152) is 71\n",
      "INFO [2023-08-08 16:12:47] group size being clustered:  71,7077\n",
      "INFO [2023-08-08 16:12:47] Number of cells in group(153) is 67\n",
      "INFO [2023-08-08 16:12:47] group size being clustered:  67,7077\n",
      "INFO [2023-08-08 16:12:47] Number of cells in group(154) is 67\n",
      "INFO [2023-08-08 16:12:47] group size being clustered:  67,7077\n",
      "INFO [2023-08-08 16:12:47] Number of cells in group(155) is 65\n",
      "INFO [2023-08-08 16:12:47] group size being clustered:  65,7077\n",
      "INFO [2023-08-08 16:12:47] Number of cells in group(156) is 64\n",
      "INFO [2023-08-08 16:12:47] group size being clustered:  64,7077\n",
      "INFO [2023-08-08 16:12:47] Number of cells in group(157) is 61\n",
      "INFO [2023-08-08 16:12:47] group size being clustered:  61,7077\n",
      "INFO [2023-08-08 16:12:47] Number of cells in group(158) is 59\n",
      "INFO [2023-08-08 16:12:47] group size being clustered:  59,7077\n",
      "INFO [2023-08-08 16:12:47] Number of cells in group(159) is 58\n",
      "INFO [2023-08-08 16:12:47] group size being clustered:  58,7077\n",
      "INFO [2023-08-08 16:12:47] Number of cells in group(160) is 57\n",
      "INFO [2023-08-08 16:12:47] group size being clustered:  57,7077\n",
      "INFO [2023-08-08 16:12:47] Number of cells in group(161) is 55\n",
      "INFO [2023-08-08 16:12:47] group size being clustered:  55,7077\n",
      "INFO [2023-08-08 16:12:48] Number of cells in group(162) is 52\n",
      "INFO [2023-08-08 16:12:48] group size being clustered:  52,7077\n",
      "INFO [2023-08-08 16:12:48] Number of cells in group(163) is 51\n",
      "INFO [2023-08-08 16:12:48] group size being clustered:  51,7077\n",
      "INFO [2023-08-08 16:12:48] Number of cells in group(164) is 48\n",
      "INFO [2023-08-08 16:12:48] group size being clustered:  48,7077\n",
      "INFO [2023-08-08 16:12:48] Number of cells in group(165) is 47\n",
      "INFO [2023-08-08 16:12:48] group size being clustered:  47,7077\n",
      "INFO [2023-08-08 16:12:48] Number of cells in group(166) is 44\n",
      "INFO [2023-08-08 16:12:48] group size being clustered:  44,7077\n",
      "INFO [2023-08-08 16:12:48] Number of cells in group(167) is 42\n",
      "INFO [2023-08-08 16:12:48] group size being clustered:  42,7077\n",
      "INFO [2023-08-08 16:12:48] Number of cells in group(168) is 40\n",
      "INFO [2023-08-08 16:12:48] group size being clustered:  40,7077\n",
      "INFO [2023-08-08 16:12:48] Number of cells in group(169) is 39\n",
      "INFO [2023-08-08 16:12:48] group size being clustered:  39,7077\n",
      "INFO [2023-08-08 16:12:48] Number of cells in group(170) is 38\n",
      "INFO [2023-08-08 16:12:48] group size being clustered:  38,7077\n",
      "INFO [2023-08-08 16:12:48] Number of cells in group(171) is 38\n",
      "INFO [2023-08-08 16:12:48] group size being clustered:  38,7077\n",
      "INFO [2023-08-08 16:12:48] Number of cells in group(172) is 37\n",
      "INFO [2023-08-08 16:12:48] group size being clustered:  37,7077\n",
      "INFO [2023-08-08 16:12:48] Number of cells in group(173) is 34\n",
      "INFO [2023-08-08 16:12:48] group size being clustered:  34,7077\n",
      "INFO [2023-08-08 16:12:48] Number of cells in group(174) is 34\n",
      "INFO [2023-08-08 16:12:48] group size being clustered:  34,7077\n",
      "INFO [2023-08-08 16:12:48] Number of cells in group(175) is 34\n",
      "INFO [2023-08-08 16:12:48] group size being clustered:  34,7077\n",
      "INFO [2023-08-08 16:12:48] Number of cells in group(176) is 32\n",
      "INFO [2023-08-08 16:12:48] group size being clustered:  32,7077\n",
      "INFO [2023-08-08 16:12:48] Number of cells in group(177) is 30\n",
      "INFO [2023-08-08 16:12:48] group size being clustered:  30,7077\n",
      "INFO [2023-08-08 16:12:48] Number of cells in group(178) is 25\n",
      "INFO [2023-08-08 16:12:48] group size being clustered:  25,7077\n",
      "INFO [2023-08-08 16:12:48] Number of cells in group(179) is 25\n",
      "INFO [2023-08-08 16:12:48] group size being clustered:  25,7077\n",
      "INFO [2023-08-08 16:12:48] Number of cells in group(180) is 24\n",
      "INFO [2023-08-08 16:12:48] group size being clustered:  24,7077\n",
      "INFO [2023-08-08 16:12:48] Number of cells in group(181) is 24\n",
      "INFO [2023-08-08 16:12:48] group size being clustered:  24,7077\n",
      "INFO [2023-08-08 16:12:48] Number of cells in group(182) is 24\n",
      "INFO [2023-08-08 16:12:48] group size being clustered:  24,7077\n",
      "INFO [2023-08-08 16:12:48] Number of cells in group(183) is 21\n",
      "INFO [2023-08-08 16:12:48] group size being clustered:  21,7077\n",
      "INFO [2023-08-08 16:12:48] Number of cells in group(184) is 20\n",
      "INFO [2023-08-08 16:12:48] group size being clustered:  20,7077\n",
      "INFO [2023-08-08 16:12:48] Number of cells in group(185) is 19\n",
      "INFO [2023-08-08 16:12:48] group size being clustered:  19,7077\n",
      "INFO [2023-08-08 16:12:48] Number of cells in group(186) is 17\n",
      "INFO [2023-08-08 16:12:48] group size being clustered:  17,7077\n",
      "INFO [2023-08-08 16:12:48] Number of cells in group(187) is 17\n",
      "INFO [2023-08-08 16:12:48] group size being clustered:  17,7077\n",
      "INFO [2023-08-08 16:12:48] Number of cells in group(188) is 12\n",
      "INFO [2023-08-08 16:12:48] group size being clustered:  12,7077\n",
      "INFO [2023-08-08 16:12:48] Number of cells in group(189) is 11\n",
      "INFO [2023-08-08 16:12:48] group size being clustered:  11,7077\n",
      "INFO [2023-08-08 16:12:48] Number of cells in group(190) is 6\n",
      "INFO [2023-08-08 16:12:48] group size being clustered:  6,7077\n",
      "INFO [2023-08-08 16:12:48] Number of cells in group(191) is 5\n",
      "INFO [2023-08-08 16:12:48] group size being clustered:  5,7077\n",
      "INFO [2023-08-08 16:12:48] Number of cells in group(192) is 4\n",
      "INFO [2023-08-08 16:12:48] group size being clustered:  4,7077\n",
      "INFO [2023-08-08 16:12:48] Number of cells in group(193) is 4\n",
      "INFO [2023-08-08 16:12:48] group size being clustered:  4,7077\n",
      "INFO [2023-08-08 16:12:48] Number of cells in group(194) is 2\n",
      "INFO [2023-08-08 16:12:48] group size being clustered:  2,7077\n",
      "INFO [2023-08-08 16:12:48] Number of cells in group(195) is 1\n",
      "INFO [2023-08-08 16:12:48] Skipping group: 195, since less than 2 entries\n",
      "INFO [2023-08-08 16:12:48] Number of cells in group(196) is 1\n",
      "INFO [2023-08-08 16:12:48] Skipping group: 196, since less than 2 entries\n",
      "INFO [2023-08-08 16:12:48] Number of cells in group(197) is 1\n",
      "INFO [2023-08-08 16:12:48] Skipping group: 197, since less than 2 entries\n",
      "INFO [2023-08-08 16:12:48] Number of cells in group(198) is 1\n",
      "INFO [2023-08-08 16:12:48] Skipping group: 198, since less than 2 entries\n",
      "INFO [2023-08-08 16:12:48] Number of cells in group(199) is 1\n",
      "INFO [2023-08-08 16:12:48] Skipping group: 199, since less than 2 entries\n",
      "INFO [2023-08-08 16:12:48] Number of cells in group(200) is 1\n",
      "INFO [2023-08-08 16:12:48] Skipping group: 200, since less than 2 entries\n",
      "INFO [2023-08-08 16:12:48] Number of cells in group(201) is 1\n",
      "INFO [2023-08-08 16:12:48] Skipping group: 201, since less than 2 entries\n",
      "INFO [2023-08-08 16:12:48] Number of cells in group(202) is 1\n",
      "INFO [2023-08-08 16:12:48] Skipping group: 202, since less than 2 entries\n",
      "INFO [2023-08-08 16:12:48] Number of cells in group(203) is 1\n",
      "INFO [2023-08-08 16:12:48] Skipping group: 203, since less than 2 entries\n",
      "INFO [2023-08-08 16:12:48] Number of cells in group(204) is 1\n",
      "INFO [2023-08-08 16:12:48] Skipping group: 204, since less than 2 entries\n",
      "INFO [2023-08-08 16:12:48] Number of cells in group(205) is 303\n",
      "INFO [2023-08-08 16:12:48] group size being clustered:  303,7077\n",
      "INFO [2023-08-08 16:12:49] Number of cells in group(206) is 170\n",
      "INFO [2023-08-08 16:12:49] group size being clustered:  170,7077\n",
      "INFO [2023-08-08 16:12:49] Number of cells in group(207) is 169\n",
      "INFO [2023-08-08 16:12:49] group size being clustered:  169,7077\n",
      "INFO [2023-08-08 16:12:50] Number of cells in group(208) is 164\n",
      "INFO [2023-08-08 16:12:50] group size being clustered:  164,7077\n",
      "INFO [2023-08-08 16:12:50] Number of cells in group(209) is 158\n",
      "INFO [2023-08-08 16:12:50] group size being clustered:  158,7077\n",
      "INFO [2023-08-08 16:12:50] Number of cells in group(210) is 153\n",
      "INFO [2023-08-08 16:12:50] group size being clustered:  153,7077\n",
      "INFO [2023-08-08 16:12:50] Number of cells in group(211) is 126\n",
      "INFO [2023-08-08 16:12:50] group size being clustered:  126,7077\n",
      "INFO [2023-08-08 16:12:50] Number of cells in group(212) is 111\n",
      "INFO [2023-08-08 16:12:50] group size being clustered:  111,7077\n",
      "INFO [2023-08-08 16:12:50] Number of cells in group(213) is 95\n",
      "INFO [2023-08-08 16:12:50] group size being clustered:  95,7077\n",
      "INFO [2023-08-08 16:12:50] Number of cells in group(214) is 94\n",
      "INFO [2023-08-08 16:12:50] group size being clustered:  94,7077\n",
      "INFO [2023-08-08 16:12:50] Number of cells in group(215) is 91\n",
      "INFO [2023-08-08 16:12:50] group size being clustered:  91,7077\n",
      "INFO [2023-08-08 16:12:50] Number of cells in group(216) is 88\n",
      "INFO [2023-08-08 16:12:50] group size being clustered:  88,7077\n",
      "INFO [2023-08-08 16:12:50] Number of cells in group(217) is 76\n",
      "INFO [2023-08-08 16:12:50] group size being clustered:  76,7077\n",
      "INFO [2023-08-08 16:12:50] Number of cells in group(218) is 67\n",
      "INFO [2023-08-08 16:12:50] group size being clustered:  67,7077\n",
      "INFO [2023-08-08 16:12:50] Number of cells in group(219) is 61\n",
      "INFO [2023-08-08 16:12:50] group size being clustered:  61,7077\n",
      "INFO [2023-08-08 16:12:50] Number of cells in group(220) is 50\n",
      "INFO [2023-08-08 16:12:50] group size being clustered:  50,7077\n",
      "INFO [2023-08-08 16:12:51] Number of cells in group(221) is 46\n",
      "INFO [2023-08-08 16:12:51] group size being clustered:  46,7077\n",
      "INFO [2023-08-08 16:12:51] Number of cells in group(222) is 26\n",
      "INFO [2023-08-08 16:12:51] group size being clustered:  26,7077\n",
      "INFO [2023-08-08 16:12:51] Number of cells in group(223) is 17\n",
      "INFO [2023-08-08 16:12:51] group size being clustered:  17,7077\n",
      "INFO [2023-08-08 16:12:51] Number of cells in group(224) is 14\n",
      "INFO [2023-08-08 16:12:51] group size being clustered:  14,7077\n",
      "INFO [2023-08-08 16:12:51] Number of cells in group(225) is 5\n",
      "INFO [2023-08-08 16:12:51] group size being clustered:  5,7077\n",
      "INFO [2023-08-08 16:12:51] Number of cells in group(226) is 1\n",
      "INFO [2023-08-08 16:12:51] Skipping group: 226, since less than 2 entries\n",
      "INFO [2023-08-08 16:12:51] Number of cells in group(227) is 1\n",
      "INFO [2023-08-08 16:12:51] Skipping group: 227, since less than 2 entries\n",
      "INFO [2023-08-08 16:12:51] Number of cells in group(228) is 1\n",
      "INFO [2023-08-08 16:12:51] Skipping group: 228, since less than 2 entries\n",
      "INFO [2023-08-08 16:12:51] Number of cells in group(229) is 1\n",
      "INFO [2023-08-08 16:12:51] Skipping group: 229, since less than 2 entries\n",
      "INFO [2023-08-08 16:12:51] Number of cells in group(230) is 1\n",
      "INFO [2023-08-08 16:12:51] Skipping group: 230, since less than 2 entries\n",
      "INFO [2023-08-08 16:12:51] Number of cells in group(231) is 217\n",
      "INFO [2023-08-08 16:12:51] group size being clustered:  217,7077\n",
      "INFO [2023-08-08 16:12:51] Number of cells in group(232) is 196\n",
      "INFO [2023-08-08 16:12:51] group size being clustered:  196,7077\n",
      "INFO [2023-08-08 16:12:51] Number of cells in group(233) is 183\n",
      "INFO [2023-08-08 16:12:51] group size being clustered:  183,7077\n",
      "INFO [2023-08-08 16:12:52] Number of cells in group(234) is 173\n",
      "INFO [2023-08-08 16:12:52] group size being clustered:  173,7077\n",
      "INFO [2023-08-08 16:12:52] Number of cells in group(235) is 162\n",
      "INFO [2023-08-08 16:12:52] group size being clustered:  162,7077\n",
      "INFO [2023-08-08 16:12:52] Number of cells in group(236) is 150\n",
      "INFO [2023-08-08 16:12:52] group size being clustered:  150,7077\n",
      "INFO [2023-08-08 16:12:52] Number of cells in group(237) is 144\n",
      "INFO [2023-08-08 16:12:52] group size being clustered:  144,7077\n",
      "INFO [2023-08-08 16:12:52] Number of cells in group(238) is 139\n",
      "INFO [2023-08-08 16:12:52] group size being clustered:  139,7077\n",
      "INFO [2023-08-08 16:12:52] Number of cells in group(239) is 137\n",
      "INFO [2023-08-08 16:12:52] group size being clustered:  137,7077\n",
      "INFO [2023-08-08 16:12:53] Number of cells in group(240) is 136\n",
      "INFO [2023-08-08 16:12:53] group size being clustered:  136,7077\n",
      "INFO [2023-08-08 16:12:53] Number of cells in group(241) is 134\n",
      "INFO [2023-08-08 16:12:53] group size being clustered:  134,7077\n",
      "INFO [2023-08-08 16:12:53] Number of cells in group(242) is 127\n",
      "INFO [2023-08-08 16:12:53] group size being clustered:  127,7077\n",
      "INFO [2023-08-08 16:12:53] Number of cells in group(243) is 124\n",
      "INFO [2023-08-08 16:12:53] group size being clustered:  124,7077\n",
      "INFO [2023-08-08 16:12:53] Number of cells in group(244) is 116\n",
      "INFO [2023-08-08 16:12:53] group size being clustered:  116,7077\n",
      "INFO [2023-08-08 16:12:53] Number of cells in group(245) is 116\n",
      "INFO [2023-08-08 16:12:53] group size being clustered:  116,7077\n",
      "INFO [2023-08-08 16:12:53] Number of cells in group(246) is 112\n",
      "INFO [2023-08-08 16:12:53] group size being clustered:  112,7077\n",
      "INFO [2023-08-08 16:12:53] Number of cells in group(247) is 111\n",
      "INFO [2023-08-08 16:12:53] group size being clustered:  111,7077\n",
      "INFO [2023-08-08 16:12:53] Number of cells in group(248) is 100\n",
      "INFO [2023-08-08 16:12:53] group size being clustered:  100,7077\n",
      "INFO [2023-08-08 16:12:53] Number of cells in group(249) is 90\n",
      "INFO [2023-08-08 16:12:53] group size being clustered:  90,7077\n",
      "INFO [2023-08-08 16:12:53] Number of cells in group(250) is 88\n",
      "INFO [2023-08-08 16:12:53] group size being clustered:  88,7077\n",
      "INFO [2023-08-08 16:12:54] Number of cells in group(251) is 87\n",
      "INFO [2023-08-08 16:12:54] group size being clustered:  87,7077\n",
      "INFO [2023-08-08 16:12:54] Number of cells in group(252) is 86\n",
      "INFO [2023-08-08 16:12:54] group size being clustered:  86,7077\n",
      "INFO [2023-08-08 16:12:54] Number of cells in group(253) is 84\n",
      "INFO [2023-08-08 16:12:54] group size being clustered:  84,7077\n",
      "INFO [2023-08-08 16:12:54] Number of cells in group(254) is 83\n",
      "INFO [2023-08-08 16:12:54] group size being clustered:  83,7077\n",
      "INFO [2023-08-08 16:12:54] Number of cells in group(255) is 80\n",
      "INFO [2023-08-08 16:12:54] group size being clustered:  80,7077\n",
      "INFO [2023-08-08 16:12:54] Number of cells in group(256) is 70\n",
      "INFO [2023-08-08 16:12:54] group size being clustered:  70,7077\n",
      "INFO [2023-08-08 16:12:54] Number of cells in group(257) is 61\n",
      "INFO [2023-08-08 16:12:54] group size being clustered:  61,7077\n",
      "INFO [2023-08-08 16:12:54] Number of cells in group(258) is 59\n",
      "INFO [2023-08-08 16:12:54] group size being clustered:  59,7077\n",
      "INFO [2023-08-08 16:12:54] Number of cells in group(259) is 56\n",
      "INFO [2023-08-08 16:12:54] group size being clustered:  56,7077\n",
      "INFO [2023-08-08 16:12:54] Number of cells in group(260) is 52\n",
      "INFO [2023-08-08 16:12:54] group size being clustered:  52,7077\n",
      "INFO [2023-08-08 16:12:54] Number of cells in group(261) is 50\n",
      "INFO [2023-08-08 16:12:54] group size being clustered:  50,7077\n",
      "INFO [2023-08-08 16:12:54] Number of cells in group(262) is 43\n",
      "INFO [2023-08-08 16:12:54] group size being clustered:  43,7077\n",
      "INFO [2023-08-08 16:12:54] Number of cells in group(263) is 41\n",
      "INFO [2023-08-08 16:12:54] group size being clustered:  41,7077\n",
      "INFO [2023-08-08 16:12:54] Number of cells in group(264) is 34\n",
      "INFO [2023-08-08 16:12:54] group size being clustered:  34,7077\n",
      "INFO [2023-08-08 16:12:54] Number of cells in group(265) is 33\n",
      "INFO [2023-08-08 16:12:54] group size being clustered:  33,7077\n",
      "INFO [2023-08-08 16:12:54] Number of cells in group(266) is 32\n",
      "INFO [2023-08-08 16:12:54] group size being clustered:  32,7077\n",
      "INFO [2023-08-08 16:12:54] Number of cells in group(267) is 20\n",
      "INFO [2023-08-08 16:12:54] group size being clustered:  20,7077\n",
      "INFO [2023-08-08 16:12:54] Number of cells in group(268) is 13\n",
      "INFO [2023-08-08 16:12:54] group size being clustered:  13,7077\n",
      "INFO [2023-08-08 16:12:54] Number of cells in group(269) is 9\n",
      "INFO [2023-08-08 16:12:54] group size being clustered:  9,7077\n",
      "INFO [2023-08-08 16:12:54] Number of cells in group(270) is 7\n",
      "INFO [2023-08-08 16:12:54] group size being clustered:  7,7077\n",
      "INFO [2023-08-08 16:12:54] Number of cells in group(271) is 6\n",
      "INFO [2023-08-08 16:12:54] group size being clustered:  6,7077\n",
      "INFO [2023-08-08 16:12:54] Number of cells in group(272) is 5\n",
      "INFO [2023-08-08 16:12:54] group size being clustered:  5,7077\n",
      "INFO [2023-08-08 16:12:54] Number of cells in group(273) is 4\n",
      "INFO [2023-08-08 16:12:54] group size being clustered:  4,7077\n",
      "INFO [2023-08-08 16:12:54] Number of cells in group(274) is 3\n",
      "INFO [2023-08-08 16:12:54] group size being clustered:  3,7077\n",
      "INFO [2023-08-08 16:12:54] Number of cells in group(275) is 3\n",
      "INFO [2023-08-08 16:12:54] group size being clustered:  3,7077\n",
      "INFO [2023-08-08 16:12:54] Number of cells in group(276) is 3\n",
      "INFO [2023-08-08 16:12:54] group size being clustered:  3,7077\n",
      "INFO [2023-08-08 16:12:54] Number of cells in group(277) is 2\n",
      "INFO [2023-08-08 16:12:54] group size being clustered:  2,7077\n",
      "INFO [2023-08-08 16:12:54] Number of cells in group(278) is 1\n",
      "INFO [2023-08-08 16:12:54] Skipping group: 278, since less than 2 entries\n",
      "INFO [2023-08-08 16:12:54] Number of cells in group(279) is 1\n",
      "INFO [2023-08-08 16:12:54] Skipping group: 279, since less than 2 entries\n",
      "INFO [2023-08-08 16:12:54] Number of cells in group(280) is 1\n",
      "INFO [2023-08-08 16:12:54] Skipping group: 280, since less than 2 entries\n",
      "INFO [2023-08-08 16:12:54] Number of cells in group(281) is 1\n",
      "INFO [2023-08-08 16:12:54] Skipping group: 281, since less than 2 entries\n",
      "INFO [2023-08-08 16:12:54] Number of cells in group(282) is 1\n",
      "INFO [2023-08-08 16:12:54] Skipping group: 282, since less than 2 entries\n",
      "INFO [2023-08-08 16:12:54] Number of cells in group(283) is 1\n",
      "INFO [2023-08-08 16:12:54] Skipping group: 283, since less than 2 entries\n",
      "INFO [2023-08-08 16:12:54] Number of cells in group(284) is 1\n",
      "INFO [2023-08-08 16:12:54] Skipping group: 284, since less than 2 entries\n",
      "INFO [2023-08-08 16:12:54] Number of cells in group(285) is 1\n",
      "INFO [2023-08-08 16:12:54] Skipping group: 285, since less than 2 entries\n",
      "INFO [2023-08-08 16:12:54] Number of cells in group(286) is 374\n",
      "INFO [2023-08-08 16:12:54] group size being clustered:  374,7077\n",
      "INFO [2023-08-08 16:12:56] Number of cells in group(287) is 267\n",
      "INFO [2023-08-08 16:12:56] group size being clustered:  267,7077\n",
      "INFO [2023-08-08 16:12:57] Number of cells in group(288) is 187\n",
      "INFO [2023-08-08 16:12:57] group size being clustered:  187,7077\n",
      "INFO [2023-08-08 16:12:57] Number of cells in group(289) is 175\n",
      "INFO [2023-08-08 16:12:57] group size being clustered:  175,7077\n",
      "INFO [2023-08-08 16:12:57] Number of cells in group(290) is 147\n",
      "INFO [2023-08-08 16:12:57] group size being clustered:  147,7077\n",
      "INFO [2023-08-08 16:12:57] Number of cells in group(291) is 139\n",
      "INFO [2023-08-08 16:12:57] group size being clustered:  139,7077\n",
      "INFO [2023-08-08 16:12:57] Number of cells in group(292) is 122\n",
      "INFO [2023-08-08 16:12:57] group size being clustered:  122,7077\n",
      "INFO [2023-08-08 16:12:57] Number of cells in group(293) is 97\n",
      "INFO [2023-08-08 16:12:58] group size being clustered:  97,7077\n",
      "INFO [2023-08-08 16:12:58] Number of cells in group(294) is 94\n",
      "INFO [2023-08-08 16:12:58] group size being clustered:  94,7077\n",
      "INFO [2023-08-08 16:12:58] Number of cells in group(295) is 89\n",
      "INFO [2023-08-08 16:12:58] group size being clustered:  89,7077\n",
      "INFO [2023-08-08 16:12:58] Number of cells in group(296) is 74\n",
      "INFO [2023-08-08 16:12:58] group size being clustered:  74,7077\n",
      "INFO [2023-08-08 16:12:58] Number of cells in group(297) is 71\n",
      "INFO [2023-08-08 16:12:58] group size being clustered:  71,7077\n",
      "INFO [2023-08-08 16:12:58] Number of cells in group(298) is 62\n",
      "INFO [2023-08-08 16:12:58] group size being clustered:  62,7077\n",
      "INFO [2023-08-08 16:12:58] Number of cells in group(299) is 58\n",
      "INFO [2023-08-08 16:12:58] group size being clustered:  58,7077\n",
      "INFO [2023-08-08 16:12:58] Number of cells in group(300) is 57\n",
      "INFO [2023-08-08 16:12:58] group size being clustered:  57,7077\n",
      "INFO [2023-08-08 16:12:58] Number of cells in group(301) is 57\n",
      "INFO [2023-08-08 16:12:58] group size being clustered:  57,7077\n",
      "INFO [2023-08-08 16:12:58] Number of cells in group(302) is 52\n",
      "INFO [2023-08-08 16:12:58] group size being clustered:  52,7077\n",
      "INFO [2023-08-08 16:12:58] Number of cells in group(303) is 20\n",
      "INFO [2023-08-08 16:12:58] group size being clustered:  20,7077\n",
      "INFO [2023-08-08 16:12:58] Number of cells in group(304) is 17\n",
      "INFO [2023-08-08 16:12:58] group size being clustered:  17,7077\n",
      "INFO [2023-08-08 16:12:58] Number of cells in group(305) is 15\n",
      "INFO [2023-08-08 16:12:58] group size being clustered:  15,7077\n",
      "INFO [2023-08-08 16:12:58] Number of cells in group(306) is 14\n",
      "INFO [2023-08-08 16:12:58] group size being clustered:  14,7077\n",
      "INFO [2023-08-08 16:12:58] Number of cells in group(307) is 4\n",
      "INFO [2023-08-08 16:12:58] group size being clustered:  4,7077\n",
      "INFO [2023-08-08 16:12:58] Number of cells in group(308) is 3\n",
      "INFO [2023-08-08 16:12:58] group size being clustered:  3,7077\n",
      "INFO [2023-08-08 16:12:58] Number of cells in group(309) is 1\n",
      "INFO [2023-08-08 16:12:58] Skipping group: 309, since less than 2 entries\n",
      "INFO [2023-08-08 16:12:58] Number of cells in group(310) is 1\n",
      "INFO [2023-08-08 16:12:58] Skipping group: 310, since less than 2 entries\n",
      "INFO [2023-08-08 16:12:58] Number of cells in group(311) is 1\n",
      "INFO [2023-08-08 16:12:58] Skipping group: 311, since less than 2 entries\n",
      "INFO [2023-08-08 16:12:58] Number of cells in group(312) is 1\n",
      "INFO [2023-08-08 16:12:58] Skipping group: 312, since less than 2 entries\n",
      "INFO [2023-08-08 16:12:58] Number of cells in group(313) is 1\n",
      "INFO [2023-08-08 16:12:58] Skipping group: 313, since less than 2 entries\n",
      "INFO [2023-08-08 16:12:58] Number of cells in group(314) is 288\n",
      "INFO [2023-08-08 16:12:58] group size being clustered:  288,7077\n",
      "INFO [2023-08-08 16:12:59] Number of cells in group(315) is 211\n",
      "INFO [2023-08-08 16:12:59] group size being clustered:  211,7077\n",
      "INFO [2023-08-08 16:12:59] Number of cells in group(316) is 150\n",
      "INFO [2023-08-08 16:12:59] group size being clustered:  150,7077\n",
      "INFO [2023-08-08 16:12:59] Number of cells in group(317) is 149\n",
      "INFO [2023-08-08 16:12:59] group size being clustered:  149,7077\n",
      "INFO [2023-08-08 16:13:00] Number of cells in group(318) is 124\n",
      "INFO [2023-08-08 16:13:00] group size being clustered:  124,7077\n",
      "INFO [2023-08-08 16:13:00] Number of cells in group(319) is 123\n",
      "INFO [2023-08-08 16:13:00] group size being clustered:  123,7077\n",
      "INFO [2023-08-08 16:13:00] Number of cells in group(320) is 120\n",
      "INFO [2023-08-08 16:13:00] group size being clustered:  120,7077\n",
      "INFO [2023-08-08 16:13:00] Number of cells in group(321) is 103\n",
      "INFO [2023-08-08 16:13:00] group size being clustered:  103,7077\n",
      "INFO [2023-08-08 16:13:00] Number of cells in group(322) is 92\n",
      "INFO [2023-08-08 16:13:00] group size being clustered:  92,7077\n",
      "INFO [2023-08-08 16:13:00] Number of cells in group(323) is 91\n",
      "INFO [2023-08-08 16:13:00] group size being clustered:  91,7077\n",
      "INFO [2023-08-08 16:13:00] Number of cells in group(324) is 82\n",
      "INFO [2023-08-08 16:13:00] group size being clustered:  82,7077\n",
      "INFO [2023-08-08 16:13:00] Number of cells in group(325) is 66\n",
      "INFO [2023-08-08 16:13:00] group size being clustered:  66,7077\n",
      "INFO [2023-08-08 16:13:00] Number of cells in group(326) is 32\n",
      "INFO [2023-08-08 16:13:00] group size being clustered:  32,7077\n",
      "INFO [2023-08-08 16:13:00] Number of cells in group(327) is 26\n",
      "INFO [2023-08-08 16:13:00] group size being clustered:  26,7077\n",
      "INFO [2023-08-08 16:13:00] Number of cells in group(328) is 17\n",
      "INFO [2023-08-08 16:13:00] group size being clustered:  17,7077\n",
      "INFO [2023-08-08 16:13:00] Number of cells in group(329) is 9\n",
      "INFO [2023-08-08 16:13:00] group size being clustered:  9,7077\n",
      "INFO [2023-08-08 16:13:00] Number of cells in group(330) is 5\n",
      "INFO [2023-08-08 16:13:00] group size being clustered:  5,7077\n",
      "INFO [2023-08-08 16:13:00] Number of cells in group(331) is 2\n",
      "INFO [2023-08-08 16:13:00] group size being clustered:  2,7077\n",
      "INFO [2023-08-08 16:13:00] Number of cells in group(332) is 2\n",
      "INFO [2023-08-08 16:13:00] group size being clustered:  2,7077\n",
      "INFO [2023-08-08 16:13:00] Number of cells in group(333) is 2\n",
      "INFO [2023-08-08 16:13:00] group size being clustered:  2,7077\n",
      "INFO [2023-08-08 16:13:00] Number of cells in group(334) is 1\n",
      "INFO [2023-08-08 16:13:00] Skipping group: 334, since less than 2 entries\n",
      "INFO [2023-08-08 16:13:00] Number of cells in group(335) is 1\n",
      "INFO [2023-08-08 16:13:00] Skipping group: 335, since less than 2 entries\n",
      "INFO [2023-08-08 16:13:00] Number of cells in group(336) is 1\n",
      "INFO [2023-08-08 16:13:00] Skipping group: 336, since less than 2 entries\n",
      "INFO [2023-08-08 16:13:00] Number of cells in group(337) is 1\n",
      "INFO [2023-08-08 16:13:00] Skipping group: 337, since less than 2 entries\n",
      "INFO [2023-08-08 16:13:00] Number of cells in group(338) is 260\n",
      "INFO [2023-08-08 16:13:00] group size being clustered:  260,7077\n",
      "INFO [2023-08-08 16:13:01] Number of cells in group(339) is 211\n",
      "INFO [2023-08-08 16:13:01] group size being clustered:  211,7077\n",
      "INFO [2023-08-08 16:13:01] Number of cells in group(340) is 197\n",
      "INFO [2023-08-08 16:13:01] group size being clustered:  197,7077\n",
      "INFO [2023-08-08 16:13:02] Number of cells in group(341) is 182\n",
      "INFO [2023-08-08 16:13:02] group size being clustered:  182,7077\n",
      "INFO [2023-08-08 16:13:02] Number of cells in group(342) is 129\n",
      "INFO [2023-08-08 16:13:02] group size being clustered:  129,7077\n",
      "INFO [2023-08-08 16:13:02] Number of cells in group(343) is 117\n",
      "INFO [2023-08-08 16:13:02] group size being clustered:  117,7077\n",
      "INFO [2023-08-08 16:13:02] Number of cells in group(344) is 112\n",
      "INFO [2023-08-08 16:13:02] group size being clustered:  112,7077\n",
      "INFO [2023-08-08 16:13:02] Number of cells in group(345) is 109\n",
      "INFO [2023-08-08 16:13:02] group size being clustered:  109,7077\n",
      "INFO [2023-08-08 16:13:02] Number of cells in group(346) is 102\n",
      "INFO [2023-08-08 16:13:02] group size being clustered:  102,7077\n",
      "INFO [2023-08-08 16:13:02] Number of cells in group(347) is 92\n",
      "INFO [2023-08-08 16:13:02] group size being clustered:  92,7077\n",
      "INFO [2023-08-08 16:13:02] Number of cells in group(348) is 68\n",
      "INFO [2023-08-08 16:13:02] group size being clustered:  68,7077\n",
      "INFO [2023-08-08 16:13:02] Number of cells in group(349) is 68\n",
      "INFO [2023-08-08 16:13:02] group size being clustered:  68,7077\n",
      "INFO [2023-08-08 16:13:02] Number of cells in group(350) is 57\n",
      "INFO [2023-08-08 16:13:02] group size being clustered:  57,7077\n",
      "INFO [2023-08-08 16:13:02] Number of cells in group(351) is 14\n",
      "INFO [2023-08-08 16:13:02] group size being clustered:  14,7077\n",
      "INFO [2023-08-08 16:13:02] Number of cells in group(352) is 10\n",
      "INFO [2023-08-08 16:13:02] group size being clustered:  10,7077\n",
      "INFO [2023-08-08 16:13:02] Number of cells in group(353) is 7\n",
      "INFO [2023-08-08 16:13:02] group size being clustered:  7,7077\n",
      "INFO [2023-08-08 16:13:02] Number of cells in group(354) is 4\n",
      "INFO [2023-08-08 16:13:02] group size being clustered:  4,7077\n",
      "INFO [2023-08-08 16:13:02] Number of cells in group(355) is 4\n",
      "INFO [2023-08-08 16:13:02] group size being clustered:  4,7077\n",
      "INFO [2023-08-08 16:13:02] Number of cells in group(356) is 2\n",
      "INFO [2023-08-08 16:13:02] group size being clustered:  2,7077\n",
      "INFO [2023-08-08 16:13:02] Number of cells in group(357) is 1\n",
      "INFO [2023-08-08 16:13:02] Skipping group: 357, since less than 2 entries\n",
      "INFO [2023-08-08 16:13:02] Number of cells in group(358) is 1\n",
      "INFO [2023-08-08 16:13:02] Skipping group: 358, since less than 2 entries\n",
      "INFO [2023-08-08 16:13:02] Number of cells in group(359) is 1\n",
      "INFO [2023-08-08 16:13:02] Skipping group: 359, since less than 2 entries\n",
      "INFO [2023-08-08 16:13:02] Number of cells in group(360) is 1\n",
      "INFO [2023-08-08 16:13:02] Skipping group: 360, since less than 2 entries\n",
      "INFO [2023-08-08 16:13:02] Number of cells in group(361) is 284\n",
      "INFO [2023-08-08 16:13:02] group size being clustered:  284,7077\n",
      "INFO [2023-08-08 16:13:03] Number of cells in group(362) is 251\n",
      "INFO [2023-08-08 16:13:03] group size being clustered:  251,7077\n",
      "INFO [2023-08-08 16:13:04] Number of cells in group(363) is 159\n",
      "INFO [2023-08-08 16:13:04] group size being clustered:  159,7077\n",
      "INFO [2023-08-08 16:13:04] Number of cells in group(364) is 143\n",
      "INFO [2023-08-08 16:13:04] group size being clustered:  143,7077\n",
      "INFO [2023-08-08 16:13:04] Number of cells in group(365) is 111\n",
      "INFO [2023-08-08 16:13:04] group size being clustered:  111,7077\n",
      "INFO [2023-08-08 16:13:04] Number of cells in group(366) is 109\n",
      "INFO [2023-08-08 16:13:04] group size being clustered:  109,7077\n",
      "INFO [2023-08-08 16:13:04] Number of cells in group(367) is 105\n",
      "INFO [2023-08-08 16:13:04] group size being clustered:  105,7077\n",
      "INFO [2023-08-08 16:13:04] Number of cells in group(368) is 104\n",
      "INFO [2023-08-08 16:13:04] group size being clustered:  104,7077\n",
      "INFO [2023-08-08 16:13:04] Number of cells in group(369) is 102\n",
      "INFO [2023-08-08 16:13:04] group size being clustered:  102,7077\n",
      "INFO [2023-08-08 16:13:04] Number of cells in group(370) is 84\n",
      "INFO [2023-08-08 16:13:05] group size being clustered:  84,7077\n",
      "INFO [2023-08-08 16:13:05] Number of cells in group(371) is 78\n",
      "INFO [2023-08-08 16:13:05] group size being clustered:  78,7077\n",
      "INFO [2023-08-08 16:13:05] Number of cells in group(372) is 66\n",
      "INFO [2023-08-08 16:13:05] group size being clustered:  66,7077\n",
      "INFO [2023-08-08 16:13:05] Number of cells in group(373) is 65\n",
      "INFO [2023-08-08 16:13:05] group size being clustered:  65,7077\n",
      "INFO [2023-08-08 16:13:05] Number of cells in group(374) is 65\n",
      "INFO [2023-08-08 16:13:05] group size being clustered:  65,7077\n",
      "INFO [2023-08-08 16:13:05] Number of cells in group(375) is 58\n",
      "INFO [2023-08-08 16:13:05] group size being clustered:  58,7077\n",
      "INFO [2023-08-08 16:13:05] Number of cells in group(376) is 43\n",
      "INFO [2023-08-08 16:13:05] group size being clustered:  43,7077\n",
      "INFO [2023-08-08 16:13:05] Number of cells in group(377) is 42\n",
      "INFO [2023-08-08 16:13:05] group size being clustered:  42,7077\n",
      "INFO [2023-08-08 16:13:05] Number of cells in group(378) is 40\n",
      "INFO [2023-08-08 16:13:05] group size being clustered:  40,7077\n",
      "INFO [2023-08-08 16:13:05] Number of cells in group(379) is 26\n",
      "INFO [2023-08-08 16:13:05] group size being clustered:  26,7077\n",
      "INFO [2023-08-08 16:13:05] Number of cells in group(380) is 24\n",
      "INFO [2023-08-08 16:13:05] group size being clustered:  24,7077\n",
      "INFO [2023-08-08 16:13:05] Number of cells in group(381) is 17\n",
      "INFO [2023-08-08 16:13:05] group size being clustered:  17,7077\n",
      "INFO [2023-08-08 16:13:05] Number of cells in group(382) is 13\n",
      "INFO [2023-08-08 16:13:05] group size being clustered:  13,7077\n",
      "INFO [2023-08-08 16:13:05] Number of cells in group(383) is 13\n",
      "INFO [2023-08-08 16:13:05] group size being clustered:  13,7077\n",
      "INFO [2023-08-08 16:13:05] Number of cells in group(384) is 11\n",
      "INFO [2023-08-08 16:13:05] group size being clustered:  11,7077\n",
      "INFO [2023-08-08 16:13:05] Number of cells in group(385) is 4\n",
      "INFO [2023-08-08 16:13:05] group size being clustered:  4,7077\n",
      "INFO [2023-08-08 16:13:05] Number of cells in group(386) is 4\n",
      "INFO [2023-08-08 16:13:05] group size being clustered:  4,7077\n",
      "INFO [2023-08-08 16:13:05] Number of cells in group(387) is 3\n",
      "INFO [2023-08-08 16:13:05] group size being clustered:  3,7077\n",
      "INFO [2023-08-08 16:13:05] Number of cells in group(388) is 2\n",
      "INFO [2023-08-08 16:13:05] group size being clustered:  2,7077\n",
      "INFO [2023-08-08 16:13:05] Number of cells in group(389) is 1\n",
      "INFO [2023-08-08 16:13:05] Skipping group: 389, since less than 2 entries\n",
      "INFO [2023-08-08 16:13:05] Number of cells in group(390) is 255\n",
      "INFO [2023-08-08 16:13:05] group size being clustered:  255,7077\n",
      "INFO [2023-08-08 16:13:06] Number of cells in group(391) is 253\n",
      "INFO [2023-08-08 16:13:06] group size being clustered:  253,7077\n",
      "INFO [2023-08-08 16:13:06] Number of cells in group(392) is 142\n",
      "INFO [2023-08-08 16:13:06] group size being clustered:  142,7077\n",
      "INFO [2023-08-08 16:13:06] Number of cells in group(393) is 135\n",
      "INFO [2023-08-08 16:13:06] group size being clustered:  135,7077\n",
      "INFO [2023-08-08 16:13:06] Number of cells in group(394) is 122\n",
      "INFO [2023-08-08 16:13:06] group size being clustered:  122,7077\n",
      "INFO [2023-08-08 16:13:07] Number of cells in group(395) is 115\n",
      "INFO [2023-08-08 16:13:07] group size being clustered:  115,7077\n",
      "INFO [2023-08-08 16:13:07] Number of cells in group(396) is 113\n",
      "INFO [2023-08-08 16:13:07] group size being clustered:  113,7077\n",
      "INFO [2023-08-08 16:13:07] Number of cells in group(397) is 109\n",
      "INFO [2023-08-08 16:13:07] group size being clustered:  109,7077\n",
      "INFO [2023-08-08 16:13:07] Number of cells in group(398) is 108\n",
      "INFO [2023-08-08 16:13:07] group size being clustered:  108,7077\n",
      "INFO [2023-08-08 16:13:07] Number of cells in group(399) is 107\n",
      "INFO [2023-08-08 16:13:07] group size being clustered:  107,7077\n",
      "INFO [2023-08-08 16:13:07] Number of cells in group(400) is 106\n",
      "INFO [2023-08-08 16:13:07] group size being clustered:  106,7077\n",
      "INFO [2023-08-08 16:13:07] Number of cells in group(401) is 103\n",
      "INFO [2023-08-08 16:13:07] group size being clustered:  103,7077\n",
      "INFO [2023-08-08 16:13:07] Number of cells in group(402) is 103\n",
      "INFO [2023-08-08 16:13:07] group size being clustered:  103,7077\n",
      "INFO [2023-08-08 16:13:07] Number of cells in group(403) is 96\n",
      "INFO [2023-08-08 16:13:07] group size being clustered:  96,7077\n",
      "INFO [2023-08-08 16:13:07] Number of cells in group(404) is 94\n",
      "INFO [2023-08-08 16:13:07] group size being clustered:  94,7077\n",
      "INFO [2023-08-08 16:13:07] Number of cells in group(405) is 81\n",
      "INFO [2023-08-08 16:13:07] group size being clustered:  81,7077\n",
      "INFO [2023-08-08 16:13:07] Number of cells in group(406) is 71\n",
      "INFO [2023-08-08 16:13:07] group size being clustered:  71,7077\n",
      "INFO [2023-08-08 16:13:07] Number of cells in group(407) is 63\n",
      "INFO [2023-08-08 16:13:07] group size being clustered:  63,7077\n",
      "INFO [2023-08-08 16:13:07] Number of cells in group(408) is 62\n",
      "INFO [2023-08-08 16:13:07] group size being clustered:  62,7077\n",
      "INFO [2023-08-08 16:13:07] Number of cells in group(409) is 58\n",
      "INFO [2023-08-08 16:13:07] group size being clustered:  58,7077\n",
      "INFO [2023-08-08 16:13:07] Number of cells in group(410) is 54\n",
      "INFO [2023-08-08 16:13:07] group size being clustered:  54,7077\n",
      "INFO [2023-08-08 16:13:07] Number of cells in group(411) is 54\n",
      "INFO [2023-08-08 16:13:07] group size being clustered:  54,7077\n",
      "INFO [2023-08-08 16:13:07] Number of cells in group(412) is 51\n",
      "INFO [2023-08-08 16:13:07] group size being clustered:  51,7077\n",
      "INFO [2023-08-08 16:13:07] Number of cells in group(413) is 44\n",
      "INFO [2023-08-08 16:13:07] group size being clustered:  44,7077\n",
      "INFO [2023-08-08 16:13:08] Number of cells in group(414) is 43\n",
      "INFO [2023-08-08 16:13:08] group size being clustered:  43,7077\n",
      "INFO [2023-08-08 16:13:08] Number of cells in group(415) is 42\n",
      "INFO [2023-08-08 16:13:08] group size being clustered:  42,7077\n",
      "INFO [2023-08-08 16:13:08] Number of cells in group(416) is 40\n",
      "INFO [2023-08-08 16:13:08] group size being clustered:  40,7077\n",
      "INFO [2023-08-08 16:13:08] Number of cells in group(417) is 39\n",
      "INFO [2023-08-08 16:13:08] group size being clustered:  39,7077\n",
      "INFO [2023-08-08 16:13:08] Number of cells in group(418) is 39\n",
      "INFO [2023-08-08 16:13:08] group size being clustered:  39,7077\n",
      "INFO [2023-08-08 16:13:08] Number of cells in group(419) is 38\n",
      "INFO [2023-08-08 16:13:08] group size being clustered:  38,7077\n",
      "INFO [2023-08-08 16:13:08] Number of cells in group(420) is 31\n",
      "INFO [2023-08-08 16:13:08] group size being clustered:  31,7077\n",
      "INFO [2023-08-08 16:13:08] Number of cells in group(421) is 31\n",
      "INFO [2023-08-08 16:13:08] group size being clustered:  31,7077\n",
      "INFO [2023-08-08 16:13:08] Number of cells in group(422) is 28\n",
      "INFO [2023-08-08 16:13:08] group size being clustered:  28,7077\n",
      "INFO [2023-08-08 16:13:08] Number of cells in group(423) is 23\n",
      "INFO [2023-08-08 16:13:08] group size being clustered:  23,7077\n",
      "INFO [2023-08-08 16:13:08] Number of cells in group(424) is 15\n",
      "INFO [2023-08-08 16:13:08] group size being clustered:  15,7077\n",
      "INFO [2023-08-08 16:13:08] Number of cells in group(425) is 15\n",
      "INFO [2023-08-08 16:13:08] group size being clustered:  15,7077\n",
      "INFO [2023-08-08 16:13:08] Number of cells in group(426) is 15\n",
      "INFO [2023-08-08 16:13:08] group size being clustered:  15,7077\n",
      "INFO [2023-08-08 16:13:08] Number of cells in group(427) is 10\n",
      "INFO [2023-08-08 16:13:08] group size being clustered:  10,7077\n",
      "INFO [2023-08-08 16:13:08] Number of cells in group(428) is 10\n",
      "INFO [2023-08-08 16:13:08] group size being clustered:  10,7077\n",
      "INFO [2023-08-08 16:13:08] Number of cells in group(429) is 6\n",
      "INFO [2023-08-08 16:13:08] group size being clustered:  6,7077\n",
      "INFO [2023-08-08 16:13:08] Number of cells in group(430) is 5\n",
      "INFO [2023-08-08 16:13:08] group size being clustered:  5,7077\n",
      "INFO [2023-08-08 16:13:08] Number of cells in group(431) is 5\n",
      "INFO [2023-08-08 16:13:08] group size being clustered:  5,7077\n",
      "INFO [2023-08-08 16:13:08] Number of cells in group(432) is 5\n",
      "INFO [2023-08-08 16:13:08] group size being clustered:  5,7077\n",
      "INFO [2023-08-08 16:13:08] Number of cells in group(433) is 3\n",
      "INFO [2023-08-08 16:13:08] group size being clustered:  3,7077\n",
      "INFO [2023-08-08 16:13:08] Number of cells in group(434) is 3\n",
      "INFO [2023-08-08 16:13:08] group size being clustered:  3,7077\n",
      "INFO [2023-08-08 16:13:08] Number of cells in group(435) is 2\n",
      "INFO [2023-08-08 16:13:08] group size being clustered:  2,7077\n",
      "INFO [2023-08-08 16:13:08] Number of cells in group(436) is 2\n",
      "INFO [2023-08-08 16:13:08] group size being clustered:  2,7077\n",
      "INFO [2023-08-08 16:13:08] Number of cells in group(437) is 1\n",
      "INFO [2023-08-08 16:13:08] Skipping group: 437, since less than 2 entries\n",
      "INFO [2023-08-08 16:13:08] Number of cells in group(438) is 1\n",
      "INFO [2023-08-08 16:13:08] Skipping group: 438, since less than 2 entries\n",
      "INFO [2023-08-08 16:13:08] Number of cells in group(439) is 1\n",
      "INFO [2023-08-08 16:13:08] Skipping group: 439, since less than 2 entries\n",
      "INFO [2023-08-08 16:13:08] Number of cells in group(440) is 1\n",
      "INFO [2023-08-08 16:13:08] Skipping group: 440, since less than 2 entries\n",
      "INFO [2023-08-08 16:13:08] Number of cells in group(441) is 1\n",
      "INFO [2023-08-08 16:13:08] Skipping group: 441, since less than 2 entries\n",
      "INFO [2023-08-08 16:13:08] Number of cells in group(442) is 1\n",
      "INFO [2023-08-08 16:13:08] Skipping group: 442, since less than 2 entries\n",
      "INFO [2023-08-08 16:13:08] Number of cells in group(443) is 1\n",
      "INFO [2023-08-08 16:13:08] Skipping group: 443, since less than 2 entries\n",
      "INFO [2023-08-08 16:13:08] Number of cells in group(444) is 307\n",
      "INFO [2023-08-08 16:13:08] group size being clustered:  307,7077\n",
      "INFO [2023-08-08 16:13:09] Number of cells in group(445) is 265\n",
      "INFO [2023-08-08 16:13:09] group size being clustered:  265,7077\n",
      "INFO [2023-08-08 16:13:10] Number of cells in group(446) is 235\n",
      "INFO [2023-08-08 16:13:10] group size being clustered:  235,7077\n",
      "INFO [2023-08-08 16:13:10] Number of cells in group(447) is 154\n",
      "INFO [2023-08-08 16:13:10] group size being clustered:  154,7077\n",
      "INFO [2023-08-08 16:13:10] Number of cells in group(448) is 153\n",
      "INFO [2023-08-08 16:13:10] group size being clustered:  153,7077\n",
      "INFO [2023-08-08 16:13:10] Number of cells in group(449) is 74\n",
      "INFO [2023-08-08 16:13:10] group size being clustered:  74,7077\n",
      "INFO [2023-08-08 16:13:10] Number of cells in group(450) is 66\n",
      "INFO [2023-08-08 16:13:11] group size being clustered:  66,7077\n",
      "INFO [2023-08-08 16:13:11] Number of cells in group(451) is 52\n",
      "INFO [2023-08-08 16:13:11] group size being clustered:  52,7077\n",
      "INFO [2023-08-08 16:13:11] Number of cells in group(452) is 29\n",
      "INFO [2023-08-08 16:13:11] group size being clustered:  29,7077\n",
      "INFO [2023-08-08 16:13:11] Number of cells in group(453) is 2\n",
      "INFO [2023-08-08 16:13:11] group size being clustered:  2,7077\n",
      "INFO [2023-08-08 16:13:11] Number of cells in group(454) is 1\n",
      "INFO [2023-08-08 16:13:11] Skipping group: 454, since less than 2 entries\n",
      "INFO [2023-08-08 16:13:11] Number of cells in group(455) is 1\n",
      "INFO [2023-08-08 16:13:11] Skipping group: 455, since less than 2 entries\n",
      "INFO [2023-08-08 16:13:11] Number of cells in group(456) is 1\n",
      "INFO [2023-08-08 16:13:11] Skipping group: 456, since less than 2 entries\n",
      "INFO [2023-08-08 16:13:11] Number of cells in group(457) is 1\n",
      "INFO [2023-08-08 16:13:11] Skipping group: 457, since less than 2 entries\n",
      "INFO [2023-08-08 16:13:11] Number of cells in group(458) is 1\n",
      "INFO [2023-08-08 16:13:11] Skipping group: 458, since less than 2 entries\n",
      "INFO [2023-08-08 16:13:11] Number of cells in group(459) is 1\n",
      "INFO [2023-08-08 16:13:11] Skipping group: 459, since less than 2 entries\n",
      "INFO [2023-08-08 16:13:11] Number of cells in group(460) is 286\n",
      "INFO [2023-08-08 16:13:11] group size being clustered:  286,7077\n",
      "INFO [2023-08-08 16:13:11] Number of cells in group(461) is 253\n",
      "INFO [2023-08-08 16:13:12] group size being clustered:  253,7077\n",
      "INFO [2023-08-08 16:13:12] Number of cells in group(462) is 238\n",
      "INFO [2023-08-08 16:13:12] group size being clustered:  238,7077\n",
      "INFO [2023-08-08 16:13:13] Number of cells in group(463) is 236\n",
      "INFO [2023-08-08 16:13:13] group size being clustered:  236,7077\n",
      "INFO [2023-08-08 16:13:13] Number of cells in group(464) is 230\n",
      "INFO [2023-08-08 16:13:13] group size being clustered:  230,7077\n",
      "INFO [2023-08-08 16:13:14] Number of cells in group(465) is 175\n",
      "INFO [2023-08-08 16:13:14] group size being clustered:  175,7077\n",
      "INFO [2023-08-08 16:13:14] Number of cells in group(466) is 148\n",
      "INFO [2023-08-08 16:13:14] group size being clustered:  148,7077\n",
      "INFO [2023-08-08 16:13:14] Number of cells in group(467) is 104\n",
      "INFO [2023-08-08 16:13:14] group size being clustered:  104,7077\n",
      "INFO [2023-08-08 16:13:14] Number of cells in group(468) is 73\n",
      "INFO [2023-08-08 16:13:14] group size being clustered:  73,7077\n",
      "INFO [2023-08-08 16:13:14] Number of cells in group(469) is 67\n",
      "INFO [2023-08-08 16:13:14] group size being clustered:  67,7077\n",
      "INFO [2023-08-08 16:13:14] Number of cells in group(470) is 35\n",
      "INFO [2023-08-08 16:13:14] group size being clustered:  35,7077\n",
      "INFO [2023-08-08 16:13:14] Number of cells in group(471) is 28\n",
      "INFO [2023-08-08 16:13:14] group size being clustered:  28,7077\n",
      "INFO [2023-08-08 16:13:14] Number of cells in group(472) is 21\n",
      "INFO [2023-08-08 16:13:14] group size being clustered:  21,7077\n",
      "INFO [2023-08-08 16:13:14] Number of cells in group(473) is 12\n",
      "INFO [2023-08-08 16:13:14] group size being clustered:  12,7077\n",
      "INFO [2023-08-08 16:13:14] Number of cells in group(474) is 5\n",
      "INFO [2023-08-08 16:13:14] group size being clustered:  5,7077\n",
      "INFO [2023-08-08 16:13:14] Number of cells in group(475) is 5\n",
      "INFO [2023-08-08 16:13:14] group size being clustered:  5,7077\n",
      "INFO [2023-08-08 16:13:14] Number of cells in group(476) is 5\n",
      "INFO [2023-08-08 16:13:14] group size being clustered:  5,7077\n",
      "INFO [2023-08-08 16:13:14] Number of cells in group(477) is 3\n",
      "INFO [2023-08-08 16:13:14] group size being clustered:  3,7077\n",
      "INFO [2023-08-08 16:13:14] Number of cells in group(478) is 2\n",
      "INFO [2023-08-08 16:13:14] group size being clustered:  2,7077\n",
      "INFO [2023-08-08 16:13:14] Number of cells in group(479) is 2\n",
      "INFO [2023-08-08 16:13:14] group size being clustered:  2,7077\n",
      "INFO [2023-08-08 16:13:14] Number of cells in group(480) is 2\n",
      "INFO [2023-08-08 16:13:14] group size being clustered:  2,7077\n",
      "INFO [2023-08-08 16:13:14] Number of cells in group(481) is 1\n",
      "INFO [2023-08-08 16:13:14] Skipping group: 481, since less than 2 entries\n",
      "INFO [2023-08-08 16:13:14] Number of cells in group(482) is 1\n",
      "INFO [2023-08-08 16:13:14] Skipping group: 482, since less than 2 entries\n",
      "INFO [2023-08-08 16:13:14] Number of cells in group(483) is 1\n",
      "INFO [2023-08-08 16:13:14] Skipping group: 483, since less than 2 entries\n",
      "INFO [2023-08-08 16:13:14] Number of cells in group(484) is 1\n",
      "INFO [2023-08-08 16:13:14] Skipping group: 484, since less than 2 entries\n",
      "INFO [2023-08-08 16:13:14] Number of cells in group(485) is 1\n",
      "INFO [2023-08-08 16:13:14] Skipping group: 485, since less than 2 entries\n",
      "INFO [2023-08-08 16:13:14] Number of cells in group(486) is 1\n",
      "INFO [2023-08-08 16:13:14] Skipping group: 486, since less than 2 entries\n",
      "INFO [2023-08-08 16:13:14] Number of cells in group(487) is 201\n",
      "INFO [2023-08-08 16:13:14] group size being clustered:  201,7077\n",
      "INFO [2023-08-08 16:13:15] Number of cells in group(488) is 186\n",
      "INFO [2023-08-08 16:13:15] group size being clustered:  186,7077\n",
      "INFO [2023-08-08 16:13:15] Number of cells in group(489) is 181\n",
      "INFO [2023-08-08 16:13:15] group size being clustered:  181,7077\n",
      "INFO [2023-08-08 16:13:15] Number of cells in group(490) is 176\n",
      "INFO [2023-08-08 16:13:15] group size being clustered:  176,7077\n",
      "INFO [2023-08-08 16:13:15] Number of cells in group(491) is 134\n",
      "INFO [2023-08-08 16:13:15] group size being clustered:  134,7077\n",
      "INFO [2023-08-08 16:13:15] Number of cells in group(492) is 113\n",
      "INFO [2023-08-08 16:13:15] group size being clustered:  113,7077\n",
      "INFO [2023-08-08 16:13:16] Number of cells in group(493) is 104\n",
      "INFO [2023-08-08 16:13:16] group size being clustered:  104,7077\n",
      "INFO [2023-08-08 16:13:16] Number of cells in group(494) is 101\n",
      "INFO [2023-08-08 16:13:16] group size being clustered:  101,7077\n",
      "INFO [2023-08-08 16:13:16] Number of cells in group(495) is 97\n",
      "INFO [2023-08-08 16:13:16] group size being clustered:  97,7077\n",
      "INFO [2023-08-08 16:13:16] Number of cells in group(496) is 75\n",
      "INFO [2023-08-08 16:13:16] group size being clustered:  75,7077\n",
      "INFO [2023-08-08 16:13:16] Number of cells in group(497) is 55\n",
      "INFO [2023-08-08 16:13:16] group size being clustered:  55,7077\n",
      "INFO [2023-08-08 16:13:16] Number of cells in group(498) is 32\n",
      "INFO [2023-08-08 16:13:16] group size being clustered:  32,7077\n",
      "INFO [2023-08-08 16:13:16] Number of cells in group(499) is 30\n",
      "INFO [2023-08-08 16:13:16] group size being clustered:  30,7077\n",
      "INFO [2023-08-08 16:13:16] Number of cells in group(500) is 28\n",
      "INFO [2023-08-08 16:13:16] group size being clustered:  28,7077\n",
      "INFO [2023-08-08 16:13:16] Number of cells in group(501) is 20\n",
      "INFO [2023-08-08 16:13:16] group size being clustered:  20,7077\n",
      "INFO [2023-08-08 16:13:16] Number of cells in group(502) is 5\n",
      "INFO [2023-08-08 16:13:16] group size being clustered:  5,7077\n",
      "INFO [2023-08-08 16:13:16] Number of cells in group(503) is 1\n",
      "INFO [2023-08-08 16:13:16] Skipping group: 503, since less than 2 entries\n",
      "INFO [2023-08-08 16:13:16] Number of cells in group(504) is 1\n",
      "INFO [2023-08-08 16:13:16] Skipping group: 504, since less than 2 entries\n",
      "INFO [2023-08-08 16:13:16] Number of cells in group(505) is 330\n",
      "INFO [2023-08-08 16:13:16] group size being clustered:  330,7077\n",
      "INFO [2023-08-08 16:13:17] Number of cells in group(506) is 270\n",
      "INFO [2023-08-08 16:13:17] group size being clustered:  270,7077\n",
      "INFO [2023-08-08 16:13:18] Number of cells in group(507) is 264\n",
      "INFO [2023-08-08 16:13:18] group size being clustered:  264,7077\n",
      "INFO [2023-08-08 16:13:19] Number of cells in group(508) is 249\n",
      "INFO [2023-08-08 16:13:19] group size being clustered:  249,7077\n",
      "INFO [2023-08-08 16:13:19] Number of cells in group(509) is 220\n",
      "INFO [2023-08-08 16:13:19] group size being clustered:  220,7077\n",
      "INFO [2023-08-08 16:13:20] Number of cells in group(510) is 111\n",
      "INFO [2023-08-08 16:13:20] group size being clustered:  111,7077\n",
      "INFO [2023-08-08 16:13:20] Number of cells in group(511) is 104\n",
      "INFO [2023-08-08 16:13:20] group size being clustered:  104,7077\n",
      "INFO [2023-08-08 16:13:20] Number of cells in group(512) is 82\n",
      "INFO [2023-08-08 16:13:20] group size being clustered:  82,7077\n",
      "INFO [2023-08-08 16:13:20] Number of cells in group(513) is 22\n",
      "INFO [2023-08-08 16:13:20] group size being clustered:  22,7077\n",
      "INFO [2023-08-08 16:13:20] Number of cells in group(514) is 15\n",
      "INFO [2023-08-08 16:13:20] group size being clustered:  15,7077\n",
      "INFO [2023-08-08 16:13:20] Number of cells in group(515) is 6\n",
      "INFO [2023-08-08 16:13:20] group size being clustered:  6,7077\n",
      "INFO [2023-08-08 16:13:20] Number of cells in group(516) is 3\n",
      "INFO [2023-08-08 16:13:20] group size being clustered:  3,7077\n",
      "INFO [2023-08-08 16:13:20] Number of cells in group(517) is 2\n",
      "INFO [2023-08-08 16:13:20] group size being clustered:  2,7077\n",
      "INFO [2023-08-08 16:13:20] Number of cells in group(518) is 1\n",
      "INFO [2023-08-08 16:13:20] Skipping group: 518, since less than 2 entries\n",
      "INFO [2023-08-08 16:13:20] Number of cells in group(519) is 1\n",
      "INFO [2023-08-08 16:13:20] Skipping group: 519, since less than 2 entries\n",
      "INFO [2023-08-08 16:13:20] Number of cells in group(520) is 1\n",
      "INFO [2023-08-08 16:13:20] Skipping group: 520, since less than 2 entries\n",
      "INFO [2023-08-08 16:13:20] Number of cells in group(521) is 216\n",
      "INFO [2023-08-08 16:13:20] group size being clustered:  216,7077\n",
      "INFO [2023-08-08 16:13:20] Number of cells in group(522) is 209\n",
      "INFO [2023-08-08 16:13:20] group size being clustered:  209,7077\n",
      "INFO [2023-08-08 16:13:21] Number of cells in group(523) is 205\n",
      "INFO [2023-08-08 16:13:21] group size being clustered:  205,7077\n",
      "INFO [2023-08-08 16:13:21] Number of cells in group(524) is 202\n",
      "INFO [2023-08-08 16:13:21] group size being clustered:  202,7077\n",
      "INFO [2023-08-08 16:13:21] Number of cells in group(525) is 188\n",
      "INFO [2023-08-08 16:13:21] group size being clustered:  188,7077\n",
      "INFO [2023-08-08 16:13:22] Number of cells in group(526) is 182\n",
      "INFO [2023-08-08 16:13:22] group size being clustered:  182,7077\n",
      "INFO [2023-08-08 16:13:22] Number of cells in group(527) is 151\n",
      "INFO [2023-08-08 16:13:22] group size being clustered:  151,7077\n",
      "INFO [2023-08-08 16:13:22] Number of cells in group(528) is 142\n",
      "INFO [2023-08-08 16:13:22] group size being clustered:  142,7077\n",
      "INFO [2023-08-08 16:13:22] Number of cells in group(529) is 124\n",
      "INFO [2023-08-08 16:13:22] group size being clustered:  124,7077\n",
      "INFO [2023-08-08 16:13:22] Number of cells in group(530) is 118\n",
      "INFO [2023-08-08 16:13:22] group size being clustered:  118,7077\n",
      "INFO [2023-08-08 16:13:22] Number of cells in group(531) is 113\n",
      "INFO [2023-08-08 16:13:22] group size being clustered:  113,7077\n",
      "INFO [2023-08-08 16:13:22] Number of cells in group(532) is 90\n",
      "INFO [2023-08-08 16:13:22] group size being clustered:  90,7077\n",
      "INFO [2023-08-08 16:13:22] Number of cells in group(533) is 71\n",
      "INFO [2023-08-08 16:13:22] group size being clustered:  71,7077\n",
      "INFO [2023-08-08 16:13:22] Number of cells in group(534) is 60\n",
      "INFO [2023-08-08 16:13:22] group size being clustered:  60,7077\n",
      "INFO [2023-08-08 16:13:23] Number of cells in group(535) is 56\n",
      "INFO [2023-08-08 16:13:23] group size being clustered:  56,7077\n",
      "INFO [2023-08-08 16:13:23] Number of cells in group(536) is 48\n",
      "INFO [2023-08-08 16:13:23] group size being clustered:  48,7077\n",
      "INFO [2023-08-08 16:13:23] Number of cells in group(537) is 48\n",
      "INFO [2023-08-08 16:13:23] group size being clustered:  48,7077\n",
      "INFO [2023-08-08 16:13:23] Number of cells in group(538) is 43\n",
      "INFO [2023-08-08 16:13:23] group size being clustered:  43,7077\n",
      "INFO [2023-08-08 16:13:23] Number of cells in group(539) is 26\n",
      "INFO [2023-08-08 16:13:23] group size being clustered:  26,7077\n",
      "INFO [2023-08-08 16:13:23] Number of cells in group(540) is 16\n",
      "INFO [2023-08-08 16:13:23] group size being clustered:  16,7077\n",
      "INFO [2023-08-08 16:13:23] Number of cells in group(541) is 12\n",
      "INFO [2023-08-08 16:13:23] group size being clustered:  12,7077\n",
      "INFO [2023-08-08 16:13:23] Number of cells in group(542) is 9\n",
      "INFO [2023-08-08 16:13:23] group size being clustered:  9,7077\n",
      "INFO [2023-08-08 16:13:23] Number of cells in group(543) is 6\n",
      "INFO [2023-08-08 16:13:23] group size being clustered:  6,7077\n",
      "INFO [2023-08-08 16:13:23] Number of cells in group(544) is 4\n",
      "INFO [2023-08-08 16:13:23] group size being clustered:  4,7077\n",
      "INFO [2023-08-08 16:13:23] Number of cells in group(545) is 2\n",
      "INFO [2023-08-08 16:13:23] group size being clustered:  2,7077\n",
      "INFO [2023-08-08 16:13:23] Number of cells in group(546) is 1\n",
      "INFO [2023-08-08 16:13:23] Skipping group: 546, since less than 2 entries\n",
      "INFO [2023-08-08 16:13:23] Number of cells in group(547) is 1\n",
      "INFO [2023-08-08 16:13:23] Skipping group: 547, since less than 2 entries\n",
      "INFO [2023-08-08 16:13:23] Number of cells in group(548) is 1\n",
      "INFO [2023-08-08 16:13:23] Skipping group: 548, since less than 2 entries\n",
      "INFO [2023-08-08 16:13:23] Number of cells in group(549) is 1\n",
      "INFO [2023-08-08 16:13:23] Skipping group: 549, since less than 2 entries\n",
      "INFO [2023-08-08 16:13:23] Number of cells in group(550) is 1\n",
      "INFO [2023-08-08 16:13:23] Skipping group: 550, since less than 2 entries\n",
      "INFO [2023-08-08 16:13:23] Number of cells in group(551) is 1\n",
      "INFO [2023-08-08 16:13:23] Skipping group: 551, since less than 2 entries\n",
      "INFO [2023-08-08 16:13:23] Number of cells in group(552) is 1\n",
      "INFO [2023-08-08 16:13:23] Skipping group: 552, since less than 2 entries\n",
      "INFO [2023-08-08 16:13:23] Number of cells in group(553) is 332\n",
      "INFO [2023-08-08 16:13:23] group size being clustered:  332,7077\n",
      "INFO [2023-08-08 16:13:24] Number of cells in group(554) is 212\n",
      "INFO [2023-08-08 16:13:24] group size being clustered:  212,7077\n",
      "INFO [2023-08-08 16:13:24] Number of cells in group(555) is 183\n",
      "INFO [2023-08-08 16:13:24] group size being clustered:  183,7077\n",
      "INFO [2023-08-08 16:13:25] Number of cells in group(556) is 183\n",
      "INFO [2023-08-08 16:13:25] group size being clustered:  183,7077\n",
      "INFO [2023-08-08 16:13:25] Number of cells in group(557) is 182\n",
      "INFO [2023-08-08 16:13:25] group size being clustered:  182,7077\n",
      "INFO [2023-08-08 16:13:25] Number of cells in group(558) is 138\n",
      "INFO [2023-08-08 16:13:25] group size being clustered:  138,7077\n",
      "INFO [2023-08-08 16:13:25] Number of cells in group(559) is 131\n",
      "INFO [2023-08-08 16:13:25] group size being clustered:  131,7077\n",
      "INFO [2023-08-08 16:13:25] Number of cells in group(560) is 129\n",
      "INFO [2023-08-08 16:13:25] group size being clustered:  129,7077\n",
      "INFO [2023-08-08 16:13:25] Number of cells in group(561) is 113\n",
      "INFO [2023-08-08 16:13:25] group size being clustered:  113,7077\n",
      "INFO [2023-08-08 16:13:26] Number of cells in group(562) is 112\n",
      "INFO [2023-08-08 16:13:26] group size being clustered:  112,7077\n",
      "INFO [2023-08-08 16:13:26] Number of cells in group(563) is 103\n",
      "INFO [2023-08-08 16:13:26] group size being clustered:  103,7077\n",
      "INFO [2023-08-08 16:13:26] Number of cells in group(564) is 95\n",
      "INFO [2023-08-08 16:13:26] group size being clustered:  95,7077\n",
      "INFO [2023-08-08 16:13:26] Number of cells in group(565) is 83\n",
      "INFO [2023-08-08 16:13:26] group size being clustered:  83,7077\n",
      "INFO [2023-08-08 16:13:26] Number of cells in group(566) is 74\n",
      "INFO [2023-08-08 16:13:26] group size being clustered:  74,7077\n",
      "INFO [2023-08-08 16:13:26] Number of cells in group(567) is 70\n",
      "INFO [2023-08-08 16:13:26] group size being clustered:  70,7077\n",
      "INFO [2023-08-08 16:13:26] Number of cells in group(568) is 66\n",
      "INFO [2023-08-08 16:13:26] group size being clustered:  66,7077\n",
      "INFO [2023-08-08 16:13:26] Number of cells in group(569) is 64\n",
      "INFO [2023-08-08 16:13:26] group size being clustered:  64,7077\n",
      "INFO [2023-08-08 16:13:26] Number of cells in group(570) is 28\n",
      "INFO [2023-08-08 16:13:26] group size being clustered:  28,7077\n",
      "INFO [2023-08-08 16:13:26] Number of cells in group(571) is 24\n",
      "INFO [2023-08-08 16:13:26] group size being clustered:  24,7077\n",
      "INFO [2023-08-08 16:13:26] Number of cells in group(572) is 8\n",
      "INFO [2023-08-08 16:13:26] group size being clustered:  8,7077\n",
      "INFO [2023-08-08 16:13:26] Number of cells in group(573) is 8\n",
      "INFO [2023-08-08 16:13:26] group size being clustered:  8,7077\n",
      "INFO [2023-08-08 16:13:26] Number of cells in group(574) is 6\n",
      "INFO [2023-08-08 16:13:26] group size being clustered:  6,7077\n",
      "INFO [2023-08-08 16:13:26] Number of cells in group(575) is 6\n",
      "INFO [2023-08-08 16:13:26] group size being clustered:  6,7077\n",
      "INFO [2023-08-08 16:13:26] Number of cells in group(576) is 5\n",
      "INFO [2023-08-08 16:13:26] group size being clustered:  5,7077\n",
      "INFO [2023-08-08 16:13:26] Number of cells in group(577) is 4\n",
      "INFO [2023-08-08 16:13:26] group size being clustered:  4,7077\n",
      "INFO [2023-08-08 16:13:26] Number of cells in group(578) is 2\n",
      "INFO [2023-08-08 16:13:26] group size being clustered:  2,7077\n",
      "INFO [2023-08-08 16:13:26] Number of cells in group(579) is 1\n",
      "INFO [2023-08-08 16:13:26] Skipping group: 579, since less than 2 entries\n",
      "INFO [2023-08-08 16:13:26] Number of cells in group(580) is 1\n",
      "INFO [2023-08-08 16:13:26] Skipping group: 580, since less than 2 entries\n",
      "INFO [2023-08-08 16:13:26] Number of cells in group(581) is 233\n",
      "INFO [2023-08-08 16:13:26] group size being clustered:  233,7077\n",
      "INFO [2023-08-08 16:13:27] Number of cells in group(582) is 225\n",
      "INFO [2023-08-08 16:13:27] group size being clustered:  225,7077\n",
      "INFO [2023-08-08 16:13:27] Number of cells in group(583) is 223\n",
      "INFO [2023-08-08 16:13:27] group size being clustered:  223,7077\n",
      "INFO [2023-08-08 16:13:27] Number of cells in group(584) is 196\n",
      "INFO [2023-08-08 16:13:27] group size being clustered:  196,7077\n",
      "INFO [2023-08-08 16:13:28] Number of cells in group(585) is 182\n",
      "INFO [2023-08-08 16:13:28] group size being clustered:  182,7077\n",
      "INFO [2023-08-08 16:13:28] Number of cells in group(586) is 153\n",
      "INFO [2023-08-08 16:13:28] group size being clustered:  153,7077\n",
      "INFO [2023-08-08 16:13:28] Number of cells in group(587) is 125\n",
      "INFO [2023-08-08 16:13:28] group size being clustered:  125,7077\n",
      "INFO [2023-08-08 16:13:28] Number of cells in group(588) is 104\n",
      "INFO [2023-08-08 16:13:28] group size being clustered:  104,7077\n",
      "INFO [2023-08-08 16:13:28] Number of cells in group(589) is 93\n",
      "INFO [2023-08-08 16:13:28] group size being clustered:  93,7077\n",
      "INFO [2023-08-08 16:13:28] Number of cells in group(590) is 89\n",
      "INFO [2023-08-08 16:13:28] group size being clustered:  89,7077\n",
      "INFO [2023-08-08 16:13:28] Number of cells in group(591) is 10\n",
      "INFO [2023-08-08 16:13:28] group size being clustered:  10,7077\n",
      "INFO [2023-08-08 16:13:28] Number of cells in group(592) is 2\n",
      "INFO [2023-08-08 16:13:28] group size being clustered:  2,7077\n",
      "INFO [2023-08-08 16:13:28] Number of cells in group(593) is 2\n",
      "INFO [2023-08-08 16:13:28] group size being clustered:  2,7077\n",
      "INFO [2023-08-08 16:13:28] Number of cells in group(594) is 1\n",
      "INFO [2023-08-08 16:13:28] Skipping group: 594, since less than 2 entries\n",
      "INFO [2023-08-08 16:13:28] Number of cells in group(595) is 1\n",
      "INFO [2023-08-08 16:13:28] Skipping group: 595, since less than 2 entries\n",
      "INFO [2023-08-08 16:13:28] Number of cells in group(596) is 283\n",
      "INFO [2023-08-08 16:13:28] group size being clustered:  283,7077\n",
      "INFO [2023-08-08 16:13:29] Number of cells in group(597) is 232\n",
      "INFO [2023-08-08 16:13:29] group size being clustered:  232,7077\n",
      "INFO [2023-08-08 16:13:30] Number of cells in group(598) is 200\n",
      "INFO [2023-08-08 16:13:30] group size being clustered:  200,7077\n",
      "INFO [2023-08-08 16:13:30] Number of cells in group(599) is 138\n",
      "INFO [2023-08-08 16:13:30] group size being clustered:  138,7077\n",
      "INFO [2023-08-08 16:13:30] Number of cells in group(600) is 116\n",
      "INFO [2023-08-08 16:13:30] group size being clustered:  116,7077\n",
      "INFO [2023-08-08 16:13:30] Number of cells in group(601) is 108\n",
      "INFO [2023-08-08 16:13:30] group size being clustered:  108,7077\n",
      "INFO [2023-08-08 16:13:30] Number of cells in group(602) is 82\n",
      "INFO [2023-08-08 16:13:30] group size being clustered:  82,7077\n",
      "INFO [2023-08-08 16:13:30] Number of cells in group(603) is 73\n",
      "INFO [2023-08-08 16:13:30] group size being clustered:  73,7077\n",
      "INFO [2023-08-08 16:13:30] Number of cells in group(604) is 69\n",
      "INFO [2023-08-08 16:13:30] group size being clustered:  69,7077\n",
      "INFO [2023-08-08 16:13:30] Number of cells in group(605) is 55\n",
      "INFO [2023-08-08 16:13:30] group size being clustered:  55,7077\n",
      "INFO [2023-08-08 16:13:30] Number of cells in group(606) is 53\n",
      "INFO [2023-08-08 16:13:30] group size being clustered:  53,7077\n",
      "INFO [2023-08-08 16:13:30] Number of cells in group(607) is 46\n",
      "INFO [2023-08-08 16:13:30] group size being clustered:  46,7077\n",
      "INFO [2023-08-08 16:13:30] Number of cells in group(608) is 38\n",
      "INFO [2023-08-08 16:13:30] group size being clustered:  38,7077\n",
      "INFO [2023-08-08 16:13:30] Number of cells in group(609) is 34\n",
      "INFO [2023-08-08 16:13:30] group size being clustered:  34,7077\n",
      "INFO [2023-08-08 16:13:30] Number of cells in group(610) is 31\n",
      "INFO [2023-08-08 16:13:30] group size being clustered:  31,7077\n",
      "INFO [2023-08-08 16:13:30] Number of cells in group(611) is 8\n",
      "INFO [2023-08-08 16:13:30] group size being clustered:  8,7077\n",
      "INFO [2023-08-08 16:13:30] Number of cells in group(612) is 5\n",
      "INFO [2023-08-08 16:13:31] group size being clustered:  5,7077\n",
      "INFO [2023-08-08 16:13:31] Number of cells in group(613) is 4\n",
      "INFO [2023-08-08 16:13:31] group size being clustered:  4,7077\n",
      "INFO [2023-08-08 16:13:31] Number of cells in group(614) is 1\n",
      "INFO [2023-08-08 16:13:31] Skipping group: 614, since less than 2 entries\n",
      "INFO [2023-08-08 16:13:31] Number of cells in group(615) is 1\n",
      "INFO [2023-08-08 16:13:31] Skipping group: 615, since less than 2 entries\n",
      "INFO [2023-08-08 16:13:31] Number of cells in group(616) is 1\n",
      "INFO [2023-08-08 16:13:31] Skipping group: 616, since less than 2 entries\n",
      "INFO [2023-08-08 16:13:31] Number of cells in group(617) is 1\n",
      "INFO [2023-08-08 16:13:31] Skipping group: 617, since less than 2 entries\n",
      "INFO [2023-08-08 16:13:31] Number of cells in group(618) is 1\n",
      "INFO [2023-08-08 16:13:31] Skipping group: 618, since less than 2 entries\n",
      "INFO [2023-08-08 16:13:31] Number of cells in group(619) is 266\n",
      "INFO [2023-08-08 16:13:31] group size being clustered:  266,7077\n",
      "INFO [2023-08-08 16:13:31] Number of cells in group(620) is 233\n",
      "INFO [2023-08-08 16:13:31] group size being clustered:  233,7077\n",
      "INFO [2023-08-08 16:13:32] Number of cells in group(621) is 165\n",
      "INFO [2023-08-08 16:13:32] group size being clustered:  165,7077\n",
      "INFO [2023-08-08 16:13:32] Number of cells in group(622) is 155\n",
      "INFO [2023-08-08 16:13:32] group size being clustered:  155,7077\n",
      "INFO [2023-08-08 16:13:32] Number of cells in group(623) is 147\n",
      "INFO [2023-08-08 16:13:32] group size being clustered:  147,7077\n",
      "INFO [2023-08-08 16:13:32] Number of cells in group(624) is 120\n",
      "INFO [2023-08-08 16:13:32] group size being clustered:  120,7077\n",
      "INFO [2023-08-08 16:13:32] Number of cells in group(625) is 108\n",
      "INFO [2023-08-08 16:13:32] group size being clustered:  108,7077\n",
      "INFO [2023-08-08 16:13:32] Number of cells in group(626) is 76\n",
      "INFO [2023-08-08 16:13:32] group size being clustered:  76,7077\n",
      "INFO [2023-08-08 16:13:32] Number of cells in group(627) is 72\n",
      "INFO [2023-08-08 16:13:32] group size being clustered:  72,7077\n",
      "INFO [2023-08-08 16:13:32] Number of cells in group(628) is 72\n",
      "INFO [2023-08-08 16:13:32] group size being clustered:  72,7077\n",
      "INFO [2023-08-08 16:13:32] Number of cells in group(629) is 70\n",
      "INFO [2023-08-08 16:13:32] group size being clustered:  70,7077\n",
      "INFO [2023-08-08 16:13:33] Number of cells in group(630) is 54\n",
      "INFO [2023-08-08 16:13:33] group size being clustered:  54,7077\n",
      "INFO [2023-08-08 16:13:33] Number of cells in group(631) is 39\n",
      "INFO [2023-08-08 16:13:33] group size being clustered:  39,7077\n",
      "INFO [2023-08-08 16:13:33] Number of cells in group(632) is 34\n",
      "INFO [2023-08-08 16:13:33] group size being clustered:  34,7077\n",
      "INFO [2023-08-08 16:13:33] Number of cells in group(633) is 17\n",
      "INFO [2023-08-08 16:13:33] group size being clustered:  17,7077\n",
      "INFO [2023-08-08 16:13:33] Number of cells in group(634) is 2\n",
      "INFO [2023-08-08 16:13:33] group size being clustered:  2,7077\n",
      "INFO [2023-08-08 16:13:33] Number of cells in group(635) is 1\n",
      "INFO [2023-08-08 16:13:33] Skipping group: 635, since less than 2 entries\n",
      "INFO [2023-08-08 16:13:33] Number of cells in group(636) is 1\n",
      "INFO [2023-08-08 16:13:33] Skipping group: 636, since less than 2 entries\n",
      "INFO [2023-08-08 16:13:57] plot_cnv_observation:Writing observation groupings/color.\n",
      "INFO [2023-08-08 16:13:57] plot_cnv_observation:Done writing observation groupings/color.\n",
      "INFO [2023-08-08 16:14:00] plot_cnv_observation:Writing observation heatmap thresholds.\n",
      "INFO [2023-08-08 16:14:00] plot_cnv_observation:Done writing observation heatmap thresholds.\n",
      "INFO [2023-08-08 16:15:45] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
      "INFO [2023-08-08 16:15:45] Quantiles of plotted data range: 0.917621182053342,0.990328802927306,1,1.00663399754825,1.0844014153194\n",
      "INFO [2023-08-08 16:16:43] plot_cnv_references:Start\n",
      "INFO [2023-08-08 16:16:43] Reference data size: Cells= 16404 Genes= 7077\n",
      "INFO [2023-08-08 16:16:56] plot_cnv_references:Number reference groups= 327\n",
      "INFO [2023-08-08 16:16:58] plot_cnv_references:Plotting heatmap.\n",
      "INFO [2023-08-08 16:17:30] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
      "INFO [2023-08-08 16:17:30] Quantiles of plotted data range: 0.917621182053342,0.992944690961636,1,1.00289965324484,1.0844014153194\n",
      "INFO [2023-08-08 16:21:51] ::plot_cnv:Start\n",
      "INFO [2023-08-08 16:21:51] ::plot_cnv:Current data dimensions (r,c)=7077,63689 Total=451182872.676612 Min=0.758515889139591 Max=5.40416897818636.\n",
      "INFO [2023-08-08 16:21:56] ::plot_cnv:Depending on the size of the matrix this may take a moment.\n",
      "INFO [2023-08-08 16:22:23] plot_cnv(): auto thresholding at: (0.917621 , 1.084401)\n",
      "INFO [2023-08-08 16:22:46] plot_cnv_observation:Start\n",
      "INFO [2023-08-08 16:22:46] Observation data size: Cells= 47285 Genes= 7077\n",
      "INFO [2023-08-08 16:22:57] plot_cnv_observation:Writing observation groupings/color.\n",
      "INFO [2023-08-08 16:22:57] plot_cnv_observation:Done writing observation groupings/color.\n",
      "INFO [2023-08-08 16:23:00] plot_cnv_observation:Writing observation heatmap thresholds.\n",
      "INFO [2023-08-08 16:23:00] plot_cnv_observation:Done writing observation heatmap thresholds.\n",
      "INFO [2023-08-08 16:24:45] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
      "INFO [2023-08-08 16:24:45] Quantiles of plotted data range: 0.917621182053342,0.990328802927306,1,1.00663399754825,1.0844014153194\n",
      "INFO [2023-08-08 16:25:46] plot_cnv_references:Start\n",
      "INFO [2023-08-08 16:25:46] Reference data size: Cells= 16404 Genes= 7077\n",
      "INFO [2023-08-08 16:54:35] plot_cnv_references:Number reference groups= 6\n",
      "INFO [2023-08-08 16:54:37] plot_cnv_references:Plotting heatmap.\n",
      "INFO [2023-08-08 16:55:11] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
      "INFO [2023-08-08 16:55:11] Quantiles of plotted data range: 0.917621182053342,0.992944690961636,1,1.00289965324484,1.0844014153194\n"
     ]
    }
   ],
   "source": [
    "# perform infercnv operations to reveal cnv signal\n",
    "infercnv_obj = infercnv::run(infercnv_obj,\n",
    "                             cutoff=0.1,  # use 1 for smart-seq, 0.1 for 10x-genomics\n",
    "                             out_dir=\"outputs/infercnv/crc/\",  # dir is auto-created for storing outputs\n",
    "                             cluster_by_groups=T,   # cluster\n",
    "                             denoise=T,\n",
    "                             HMM=T,\n",
    "                            analysis_mode='subclusters',\n",
    "                                       leiden_resolution=0.01, #initial run returned too many subclusters\n",
    "                             )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c84e2a00-a4ac-4e88-8473-48cac33c34e0",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2023-08-08 18:49:46] ::process_data:Start\n",
      "INFO [2023-08-08 18:49:46] Checking for saved results.\n",
      "INFO [2023-08-08 18:49:46] Trying to reload from step 14\n",
      "INFO [2023-08-08 18:50:18] Trying to reload from step 12\n",
      "INFO [2023-08-08 18:50:49] Trying to reload from step 11\n",
      "INFO [2023-08-08 18:51:24] Trying to reload from step 10\n",
      "INFO [2023-08-08 18:51:59] Trying to reload from step 9\n",
      "INFO [2023-08-08 18:52:33] Trying to reload from step 8\n",
      "INFO [2023-08-08 18:53:06] Trying to reload from step 4\n",
      "INFO [2023-08-08 18:53:24] Using backup from step 4\n",
      "INFO [2023-08-08 18:53:24] \n",
      "\n",
      "\tSTEP 1: incoming data\n",
      "\n",
      "INFO [2023-08-08 18:53:24] \n",
      "\n",
      "\tSTEP 02: Removing lowly expressed genes\n",
      "\n",
      "INFO [2023-08-08 18:53:24] \n",
      "\n",
      "\tSTEP 03: normalization by sequencing depth\n",
      "\n",
      "INFO [2023-08-08 18:53:24] \n",
      "\n",
      "\tSTEP 04: log transformation of data\n",
      "\n",
      "INFO [2023-08-08 18:53:24] \n",
      "\n",
      "\tSTEP 08: removing average of reference data (before smoothing)\n",
      "\n",
      "INFO [2023-08-08 18:53:24] ::subtract_ref_expr_from_obs:Start inv_log=FALSE, use_bounds=TRUE\n",
      "INFO [2023-08-08 18:53:24] subtracting mean(normal) per gene per cell across all data\n",
      "INFO [2023-08-08 18:53:45] -subtracting expr per gene, use_bounds=TRUE\n",
      "INFO [2023-08-08 18:54:20] -mirroring for hspike\n",
      "INFO [2023-08-08 18:54:20] ::subtract_ref_expr_from_obs:Start inv_log=FALSE, use_bounds=TRUE\n",
      "INFO [2023-08-08 18:54:20] subtracting mean(normal) per gene per cell across all data\n",
      "INFO [2023-08-08 18:54:24] -subtracting expr per gene, use_bounds=TRUE\n",
      "INFO [2023-08-08 18:58:08] \n",
      "\n",
      "\tSTEP 09: apply max centered expression threshold: 3\n",
      "\n",
      "INFO [2023-08-08 18:58:08] ::process_data:setting max centered expr, threshold set to: +/-:  3\n",
      "INFO [2023-08-08 18:58:23] -mirroring for hspike\n",
      "INFO [2023-08-08 18:58:23] ::process_data:setting max centered expr, threshold set to: +/-:  3\n",
      "INFO [2023-08-08 19:02:04] \n",
      "\n",
      "\tSTEP 10: Smoothing data per cell by chromosome\n",
      "\n",
      "INFO [2023-08-08 19:02:04] smooth_by_chromosome: chr: 1\n",
      "INFO [2023-08-08 19:03:25] smooth_by_chromosome: chr: 2\n",
      "INFO [2023-08-08 19:04:37] smooth_by_chromosome: chr: 3\n",
      "INFO [2023-08-08 19:05:45] smooth_by_chromosome: chr: 4\n",
      "INFO [2023-08-08 19:06:45] smooth_by_chromosome: chr: 5\n",
      "INFO [2023-08-08 19:07:49] smooth_by_chromosome: chr: 6\n",
      "INFO [2023-08-08 19:08:55] smooth_by_chromosome: chr: 7\n",
      "INFO [2023-08-08 19:10:03] smooth_by_chromosome: chr: X\n",
      "INFO [2023-08-08 19:11:02] smooth_by_chromosome: chr: 8\n",
      "INFO [2023-08-08 19:12:02] smooth_by_chromosome: chr: 9\n",
      "INFO [2023-08-08 19:13:05] smooth_by_chromosome: chr: 11\n",
      "INFO [2023-08-08 19:14:14] smooth_by_chromosome: chr: 10\n",
      "INFO [2023-08-08 19:15:15] smooth_by_chromosome: chr: 12\n",
      "INFO [2023-08-08 19:16:23] smooth_by_chromosome: chr: 13\n",
      "INFO [2023-08-08 19:17:18] smooth_by_chromosome: chr: 14\n",
      "INFO [2023-08-08 19:18:19] smooth_by_chromosome: chr: 15\n",
      "INFO [2023-08-08 19:19:18] smooth_by_chromosome: chr: 16\n",
      "INFO [2023-08-08 19:20:22] smooth_by_chromosome: chr: 17\n",
      "INFO [2023-08-08 19:21:32] smooth_by_chromosome: chr: 18\n",
      "INFO [2023-08-08 19:22:13] smooth_by_chromosome: chr: 20\n",
      "INFO [2023-08-08 19:23:12] smooth_by_chromosome: chr: 19\n",
      "INFO [2023-08-08 19:24:23] smooth_by_chromosome: chr: Y\n",
      "INFO [2023-08-08 19:24:30] smooth_by_chromosome: chr: 22\n",
      "INFO [2023-08-08 19:25:25] smooth_by_chromosome: chr: 21\n",
      "INFO [2023-08-08 19:26:00] -mirroring for hspike\n",
      "INFO [2023-08-08 19:26:00] smooth_by_chromosome: chr: chrA\n",
      "INFO [2023-08-08 19:26:01] smooth_by_chromosome: chr: chr_0\n",
      "INFO [2023-08-08 19:26:02] smooth_by_chromosome: chr: chr_B\n",
      "INFO [2023-08-08 19:26:04] smooth_by_chromosome: chr: chr_0pt5\n",
      "INFO [2023-08-08 19:26:05] smooth_by_chromosome: chr: chr_C\n",
      "INFO [2023-08-08 19:26:06] smooth_by_chromosome: chr: chr_1pt5\n",
      "INFO [2023-08-08 19:26:07] smooth_by_chromosome: chr: chr_D\n",
      "INFO [2023-08-08 19:26:08] smooth_by_chromosome: chr: chr_2pt0\n",
      "INFO [2023-08-08 19:26:10] smooth_by_chromosome: chr: chr_E\n",
      "INFO [2023-08-08 19:26:11] smooth_by_chromosome: chr: chr_3pt0\n",
      "INFO [2023-08-08 19:26:12] smooth_by_chromosome: chr: chr_F\n",
      "INFO [2023-08-08 19:30:12] \n",
      "\n",
      "\tSTEP 11: re-centering data across chromosome after smoothing\n",
      "\n",
      "INFO [2023-08-08 19:30:12] ::center_smooth across chromosomes per cell\n",
      "INFO [2023-08-08 19:31:37] -mirroring for hspike\n",
      "INFO [2023-08-08 19:31:37] ::center_smooth across chromosomes per cell\n",
      "INFO [2023-08-08 19:35:49] \n",
      "\n",
      "\tSTEP 12: removing average of reference data (after smoothing)\n",
      "\n",
      "INFO [2023-08-08 19:35:49] ::subtract_ref_expr_from_obs:Start inv_log=FALSE, use_bounds=TRUE\n",
      "INFO [2023-08-08 19:35:49] subtracting mean(normal) per gene per cell across all data\n",
      "INFO [2023-08-08 19:36:10] -subtracting expr per gene, use_bounds=TRUE\n",
      "INFO [2023-08-08 19:36:47] -mirroring for hspike\n",
      "INFO [2023-08-08 19:36:47] ::subtract_ref_expr_from_obs:Start inv_log=FALSE, use_bounds=TRUE\n",
      "INFO [2023-08-08 19:36:47] subtracting mean(normal) per gene per cell across all data\n",
      "INFO [2023-08-08 19:36:50] -subtracting expr per gene, use_bounds=TRUE\n",
      "INFO [2023-08-08 19:40:07] \n",
      "\n",
      "\tSTEP 14: invert log2(FC) to FC\n",
      "\n",
      "INFO [2023-08-08 19:40:07] invert_log2(), computing 2^x\n",
      "INFO [2023-08-08 19:40:17] -mirroring for hspike\n",
      "INFO [2023-08-08 19:40:17] invert_log2(), computing 2^x\n",
      "INFO [2023-08-08 19:44:27] \n",
      "\n",
      "\tSTEP 15: Clustering samples (not defining tumor subclusters)\n",
      "\n",
      "INFO [2023-08-08 19:44:27] define_signif_tumor_subclusters(p_val=0.1\n",
      "INFO [2023-08-08 19:44:34] define_signif_tumor_subclusters(), tumor: malignant_SMC01\n",
      "INFO [2023-08-08 19:46:52] cut tree into: 1 groups\n",
      "INFO [2023-08-08 19:46:52] -processing malignant_SMC01,malignant_SMC01_s1\n",
      "INFO [2023-08-08 19:46:52] define_signif_tumor_subclusters(), tumor: malignant_SMC02\n",
      "INFO [2023-08-08 19:47:38] cut tree into: 1 groups\n",
      "INFO [2023-08-08 19:47:38] -processing malignant_SMC02,malignant_SMC02_s1\n",
      "INFO [2023-08-08 19:47:38] define_signif_tumor_subclusters(), tumor: malignant_SMC03\n",
      "INFO [2023-08-08 19:49:05] cut tree into: 1 groups\n",
      "INFO [2023-08-08 19:49:05] -processing malignant_SMC03,malignant_SMC03_s1\n",
      "INFO [2023-08-08 19:49:05] define_signif_tumor_subclusters(), tumor: malignant_SMC04\n",
      "INFO [2023-08-08 19:50:08] cut tree into: 1 groups\n",
      "INFO [2023-08-08 19:50:08] -processing malignant_SMC04,malignant_SMC04_s1\n",
      "INFO [2023-08-08 19:50:08] define_signif_tumor_subclusters(), tumor: malignant_SMC05\n",
      "INFO [2023-08-08 19:50:15] cut tree into: 1 groups\n",
      "INFO [2023-08-08 19:50:15] -processing malignant_SMC05,malignant_SMC05_s1\n",
      "INFO [2023-08-08 19:50:15] define_signif_tumor_subclusters(), tumor: malignant_SMC06\n",
      "INFO [2023-08-08 19:50:58] cut tree into: 1 groups\n",
      "INFO [2023-08-08 19:50:58] -processing malignant_SMC06,malignant_SMC06_s1\n",
      "INFO [2023-08-08 19:50:58] define_signif_tumor_subclusters(), tumor: malignant_SMC07\n",
      "INFO [2023-08-08 19:57:52] cut tree into: 1 groups\n",
      "INFO [2023-08-08 19:57:52] -processing malignant_SMC07,malignant_SMC07_s1\n",
      "INFO [2023-08-08 19:57:52] define_signif_tumor_subclusters(), tumor: malignant_SMC08\n",
      "INFO [2023-08-08 19:59:19] cut tree into: 1 groups\n",
      "INFO [2023-08-08 19:59:19] -processing malignant_SMC08,malignant_SMC08_s1\n",
      "INFO [2023-08-08 19:59:19] define_signif_tumor_subclusters(), tumor: malignant_SMC09\n",
      "INFO [2023-08-08 20:04:37] cut tree into: 1 groups\n",
      "INFO [2023-08-08 20:04:37] -processing malignant_SMC09,malignant_SMC09_s1\n",
      "INFO [2023-08-08 20:04:37] define_signif_tumor_subclusters(), tumor: malignant_SMC10\n",
      "INFO [2023-08-08 20:06:12] cut tree into: 1 groups\n",
      "INFO [2023-08-08 20:06:12] -processing malignant_SMC10,malignant_SMC10_s1\n",
      "INFO [2023-08-08 20:06:12] define_signif_tumor_subclusters(), tumor: malignant_SMC11\n",
      "INFO [2023-08-08 20:07:06] cut tree into: 1 groups\n",
      "INFO [2023-08-08 20:07:06] -processing malignant_SMC11,malignant_SMC11_s1\n",
      "INFO [2023-08-08 20:07:06] define_signif_tumor_subclusters(), tumor: malignant_SMC14\n",
      "INFO [2023-08-08 20:08:03] cut tree into: 1 groups\n",
      "INFO [2023-08-08 20:08:03] -processing malignant_SMC14,malignant_SMC14_s1\n",
      "INFO [2023-08-08 20:08:03] define_signif_tumor_subclusters(), tumor: malignant_SMC15\n",
      "INFO [2023-08-08 20:09:20] cut tree into: 1 groups\n",
      "INFO [2023-08-08 20:09:20] -processing malignant_SMC15,malignant_SMC15_s1\n",
      "INFO [2023-08-08 20:09:20] define_signif_tumor_subclusters(), tumor: malignant_SMC16\n",
      "INFO [2023-08-08 20:12:24] cut tree into: 1 groups\n",
      "INFO [2023-08-08 20:12:24] -processing malignant_SMC16,malignant_SMC16_s1\n",
      "INFO [2023-08-08 20:12:24] define_signif_tumor_subclusters(), tumor: malignant_SMC17\n",
      "INFO [2023-08-08 20:12:56] cut tree into: 1 groups\n",
      "INFO [2023-08-08 20:12:56] -processing malignant_SMC17,malignant_SMC17_s1\n",
      "INFO [2023-08-08 20:12:56] define_signif_tumor_subclusters(), tumor: malignant_SMC18\n",
      "INFO [2023-08-08 20:14:10] cut tree into: 1 groups\n",
      "INFO [2023-08-08 20:14:10] -processing malignant_SMC18,malignant_SMC18_s1\n",
      "INFO [2023-08-08 20:14:10] define_signif_tumor_subclusters(), tumor: malignant_SMC19\n",
      "INFO [2023-08-08 20:14:54] cut tree into: 1 groups\n",
      "INFO [2023-08-08 20:14:54] -processing malignant_SMC19,malignant_SMC19_s1\n",
      "INFO [2023-08-08 20:14:54] define_signif_tumor_subclusters(), tumor: malignant_SMC20\n",
      "INFO [2023-08-08 20:15:46] cut tree into: 1 groups\n",
      "INFO [2023-08-08 20:15:46] -processing malignant_SMC20,malignant_SMC20_s1\n",
      "INFO [2023-08-08 20:15:46] define_signif_tumor_subclusters(), tumor: malignant_SMC21\n",
      "INFO [2023-08-08 20:17:35] cut tree into: 1 groups\n",
      "INFO [2023-08-08 20:17:35] -processing malignant_SMC21,malignant_SMC21_s1\n",
      "INFO [2023-08-08 20:17:35] define_signif_tumor_subclusters(), tumor: malignant_SMC22\n",
      "INFO [2023-08-08 20:19:25] cut tree into: 1 groups\n",
      "INFO [2023-08-08 20:19:25] -processing malignant_SMC22,malignant_SMC22_s1\n",
      "INFO [2023-08-08 20:19:25] define_signif_tumor_subclusters(), tumor: malignant_SMC23\n",
      "INFO [2023-08-08 20:20:15] cut tree into: 1 groups\n",
      "INFO [2023-08-08 20:20:15] -processing malignant_SMC23,malignant_SMC23_s1\n",
      "INFO [2023-08-08 20:20:15] define_signif_tumor_subclusters(), tumor: malignant_SMC24\n",
      "INFO [2023-08-08 20:21:01] cut tree into: 1 groups\n",
      "INFO [2023-08-08 20:21:01] -processing malignant_SMC24,malignant_SMC24_s1\n",
      "INFO [2023-08-08 20:21:01] define_signif_tumor_subclusters(), tumor: malignant_SMC25\n",
      "INFO [2023-08-08 20:21:54] cut tree into: 1 groups\n",
      "INFO [2023-08-08 20:21:54] -processing malignant_SMC25,malignant_SMC25_s1\n",
      "INFO [2023-08-08 20:21:54] define_signif_tumor_subclusters(), tumor: T cells\n",
      "INFO [2023-08-08 20:39:50] cut tree into: 1 groups\n",
      "INFO [2023-08-08 20:39:50] -processing T cells,T cells_s1\n",
      "INFO [2023-08-08 20:39:50] define_signif_tumor_subclusters(), tumor: B cells\n",
      "INFO [2023-08-08 20:51:35] cut tree into: 1 groups\n",
      "INFO [2023-08-08 20:51:35] -processing B cells,B cells_s1\n",
      "INFO [2023-08-08 20:51:35] define_signif_tumor_subclusters(), tumor: Stromal cells\n",
      "INFO [2023-08-08 20:55:17] cut tree into: 1 groups\n",
      "INFO [2023-08-08 20:55:17] -processing Stromal cells,Stromal cells_s1\n",
      "INFO [2023-08-08 20:55:17] define_signif_tumor_subclusters(), tumor: Epithelial cells\n",
      "INFO [2023-08-08 20:55:38] cut tree into: 1 groups\n",
      "INFO [2023-08-08 20:55:38] -processing Epithelial cells,Epithelial cells_s1\n",
      "INFO [2023-08-08 20:55:38] define_signif_tumor_subclusters(), tumor: Myeloids\n",
      "INFO [2023-08-08 20:55:40] cut tree into: 1 groups\n",
      "INFO [2023-08-08 20:55:40] -processing Myeloids,Myeloids_s1\n",
      "INFO [2023-08-08 20:55:40] define_signif_tumor_subclusters(), tumor: Mast cells\n",
      "INFO [2023-08-08 20:55:40] cut tree into: 1 groups\n",
      "INFO [2023-08-08 20:55:40] -processing Mast cells,Mast cells_s1\n",
      "INFO [2023-08-08 20:55:40] -mirroring for hspike\n",
      "INFO [2023-08-08 20:55:40] define_signif_tumor_subclusters(p_val=0.1\n",
      "INFO [2023-08-08 20:55:40] define_signif_tumor_subclusters(), tumor: spike_tumor_cell_T cells\n",
      "INFO [2023-08-08 20:55:40] cut tree into: 1 groups\n",
      "INFO [2023-08-08 20:55:40] -processing spike_tumor_cell_T cells,spike_tumor_cell_T cells_s1\n",
      "INFO [2023-08-08 20:55:41] define_signif_tumor_subclusters(), tumor: spike_tumor_cell_B cells\n",
      "INFO [2023-08-08 20:55:41] cut tree into: 1 groups\n",
      "INFO [2023-08-08 20:55:41] -processing spike_tumor_cell_B cells,spike_tumor_cell_B cells_s1\n",
      "INFO [2023-08-08 20:55:41] define_signif_tumor_subclusters(), tumor: spike_tumor_cell_Stromal cells\n",
      "INFO [2023-08-08 20:55:41] cut tree into: 1 groups\n",
      "INFO [2023-08-08 20:55:41] -processing spike_tumor_cell_Stromal cells,spike_tumor_cell_Stromal cells_s1\n",
      "INFO [2023-08-08 20:55:41] define_signif_tumor_subclusters(), tumor: spike_tumor_cell_Epithelial cells\n",
      "INFO [2023-08-08 20:55:41] cut tree into: 1 groups\n",
      "INFO [2023-08-08 20:55:41] -processing spike_tumor_cell_Epithelial cells,spike_tumor_cell_Epithelial cells_s1\n",
      "INFO [2023-08-08 20:55:41] define_signif_tumor_subclusters(), tumor: spike_tumor_cell_Myeloids\n",
      "INFO [2023-08-08 20:55:41] cut tree into: 1 groups\n",
      "INFO [2023-08-08 20:55:41] -processing spike_tumor_cell_Myeloids,spike_tumor_cell_Myeloids_s1\n",
      "INFO [2023-08-08 20:55:41] define_signif_tumor_subclusters(), tumor: spike_tumor_cell_Mast cells\n",
      "INFO [2023-08-08 20:55:41] cut tree into: 1 groups\n",
      "INFO [2023-08-08 20:55:41] -processing spike_tumor_cell_Mast cells,spike_tumor_cell_Mast cells_s1\n",
      "INFO [2023-08-08 20:55:41] define_signif_tumor_subclusters(), tumor: simnorm_cell_T cells\n",
      "INFO [2023-08-08 20:55:41] cut tree into: 1 groups\n",
      "INFO [2023-08-08 20:55:41] -processing simnorm_cell_T cells,simnorm_cell_T cells_s1\n",
      "INFO [2023-08-08 20:55:41] define_signif_tumor_subclusters(), tumor: simnorm_cell_B cells\n",
      "INFO [2023-08-08 20:55:41] cut tree into: 1 groups\n",
      "INFO [2023-08-08 20:55:41] -processing simnorm_cell_B cells,simnorm_cell_B cells_s1\n",
      "INFO [2023-08-08 20:55:41] define_signif_tumor_subclusters(), tumor: simnorm_cell_Stromal cells\n",
      "INFO [2023-08-08 20:55:41] cut tree into: 1 groups\n",
      "INFO [2023-08-08 20:55:41] -processing simnorm_cell_Stromal cells,simnorm_cell_Stromal cells_s1\n",
      "INFO [2023-08-08 20:55:41] define_signif_tumor_subclusters(), tumor: simnorm_cell_Epithelial cells\n",
      "INFO [2023-08-08 20:55:41] cut tree into: 1 groups\n",
      "INFO [2023-08-08 20:55:41] -processing simnorm_cell_Epithelial cells,simnorm_cell_Epithelial cells_s1\n",
      "INFO [2023-08-08 20:55:41] define_signif_tumor_subclusters(), tumor: simnorm_cell_Myeloids\n",
      "INFO [2023-08-08 20:55:41] cut tree into: 1 groups\n",
      "INFO [2023-08-08 20:55:41] -processing simnorm_cell_Myeloids,simnorm_cell_Myeloids_s1\n",
      "INFO [2023-08-08 20:55:41] define_signif_tumor_subclusters(), tumor: simnorm_cell_Mast cells\n",
      "INFO [2023-08-08 20:55:41] cut tree into: 1 groups\n",
      "INFO [2023-08-08 20:55:41] -processing simnorm_cell_Mast cells,simnorm_cell_Mast cells_s1\n",
      "INFO [2023-08-08 21:04:02] ::plot_cnv:Start\n",
      "INFO [2023-08-08 21:04:02] ::plot_cnv:Current data dimensions (r,c)=7077,63689 Total=451182872.676612 Min=0.758515889139591 Max=5.40416897818636.\n",
      "INFO [2023-08-08 21:04:07] ::plot_cnv:Depending on the size of the matrix this may take a moment.\n",
      "INFO [2023-08-08 21:04:34] plot_cnv(): auto thresholding at: (0.917621 , 1.084401)\n",
      "INFO [2023-08-08 21:04:58] plot_cnv_observation:Start\n",
      "INFO [2023-08-08 21:04:58] Observation data size: Cells= 47285 Genes= 7077\n",
      "INFO [2023-08-08 21:05:08] plot_cnv_observation:Writing observation groupings/color.\n",
      "INFO [2023-08-08 21:05:08] plot_cnv_observation:Done writing observation groupings/color.\n",
      "INFO [2023-08-08 21:05:11] plot_cnv_observation:Writing observation heatmap thresholds.\n",
      "INFO [2023-08-08 21:05:11] plot_cnv_observation:Done writing observation heatmap thresholds.\n",
      "INFO [2023-08-08 21:06:58] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
      "INFO [2023-08-08 21:06:58] Quantiles of plotted data range: 0.917621182053342,0.990328802927306,1,1.00663399754825,1.0844014153194\n",
      "INFO [2023-08-08 21:07:59] plot_cnv_references:Start\n",
      "INFO [2023-08-08 21:07:59] Reference data size: Cells= 16404 Genes= 7077\n",
      "INFO [2023-08-08 21:37:36] plot_cnv_references:Number reference groups= 6\n",
      "INFO [2023-08-08 21:37:37] plot_cnv_references:Plotting heatmap.\n",
      "INFO [2023-08-08 21:38:12] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
      "INFO [2023-08-08 21:38:12] Quantiles of plotted data range: 0.917621182053342,0.992944690961636,1,1.00289965324484,1.0844014153194\n",
      "INFO [2023-08-08 21:38:23] \n",
      "\n",
      "\tSTEP 17: HMM-based CNV prediction\n",
      "\n",
      "INFO [2023-08-08 21:38:23] predict_CNV_via_HMM_on_whole_tumor_samples\n",
      "INFO [2023-08-08 21:38:45] -done predicting CNV based on initial tumor subclusters\n",
      "INFO [2023-08-08 21:38:45] get_predicted_CNV_regions(subcluster)\n",
      "INFO [2023-08-08 21:38:45] -processing cell_group_name: malignant_SMC01.malignant_SMC01_s1, size: 2791\n",
      "INFO [2023-08-08 21:39:07] -processing cell_group_name: malignant_SMC02.malignant_SMC02_s1, size: 1689\n",
      "INFO [2023-08-08 21:39:22] -processing cell_group_name: malignant_SMC03.malignant_SMC03_s1, size: 2123\n",
      "INFO [2023-08-08 21:39:40] -processing cell_group_name: malignant_SMC04.malignant_SMC04_s1, size: 1841\n",
      "INFO [2023-08-08 21:39:57] -processing cell_group_name: malignant_SMC05.malignant_SMC05_s1, size: 660\n",
      "INFO [2023-08-08 21:40:07] -processing cell_group_name: malignant_SMC06.malignant_SMC06_s1, size: 1541\n",
      "INFO [2023-08-08 21:40:22] -processing cell_group_name: malignant_SMC07.malignant_SMC07_s1, size: 4070\n",
      "INFO [2023-08-08 21:40:51] -processing cell_group_name: malignant_SMC08.malignant_SMC08_s1, size: 2089\n",
      "INFO [2023-08-08 21:41:09] -processing cell_group_name: malignant_SMC09.malignant_SMC09_s1, size: 3789\n",
      "INFO [2023-08-08 21:41:36] -processing cell_group_name: malignant_SMC10.malignant_SMC10_s1, size: 2200\n",
      "INFO [2023-08-08 21:41:55] -processing cell_group_name: malignant_SMC11.malignant_SMC11_s1, size: 1698\n",
      "INFO [2023-08-08 21:42:11] -processing cell_group_name: malignant_SMC14.malignant_SMC14_s1, size: 1749\n",
      "INFO [2023-08-08 21:42:27] -processing cell_group_name: malignant_SMC15.malignant_SMC15_s1, size: 2027\n",
      "INFO [2023-08-08 21:42:44] -processing cell_group_name: malignant_SMC16.malignant_SMC16_s1, size: 2956\n",
      "INFO [2023-08-08 21:43:07] -processing cell_group_name: malignant_SMC17.malignant_SMC17_s1, size: 1343\n",
      "INFO [2023-08-08 21:43:21] -processing cell_group_name: malignant_SMC18.malignant_SMC18_s1, size: 1936\n",
      "INFO [2023-08-08 21:43:38] -processing cell_group_name: malignant_SMC19.malignant_SMC19_s1, size: 1540\n",
      "INFO [2023-08-08 21:43:53] -processing cell_group_name: malignant_SMC20.malignant_SMC20_s1, size: 1681\n",
      "INFO [2023-08-08 21:44:10] -processing cell_group_name: malignant_SMC21.malignant_SMC21_s1, size: 2348\n",
      "INFO [2023-08-08 21:44:29] -processing cell_group_name: malignant_SMC22.malignant_SMC22_s1, size: 2363\n",
      "INFO [2023-08-08 21:44:48] -processing cell_group_name: malignant_SMC23.malignant_SMC23_s1, size: 1639\n",
      "INFO [2023-08-08 21:45:04] -processing cell_group_name: malignant_SMC24.malignant_SMC24_s1, size: 1580\n",
      "INFO [2023-08-08 21:45:19] -processing cell_group_name: malignant_SMC25.malignant_SMC25_s1, size: 1632\n",
      "INFO [2023-08-08 21:45:34] -processing cell_group_name: T cells.T cells_s1, size: 6376\n",
      "INFO [2023-08-08 21:46:16] -processing cell_group_name: B cells.B cells_s1, size: 5208\n",
      "INFO [2023-08-08 21:46:50] -processing cell_group_name: Stromal cells.Stromal cells_s1, size: 3197\n",
      "INFO [2023-08-08 21:47:15] -processing cell_group_name: Epithelial cells.Epithelial cells_s1, size: 1070\n",
      "INFO [2023-08-08 21:47:27] -processing cell_group_name: Myeloids.Myeloids_s1, size: 369\n",
      "INFO [2023-08-08 21:47:36] -processing cell_group_name: Mast cells.Mast cells_s1, size: 184\n",
      "INFO [2023-08-08 21:47:43] -writing cell clusters file: outputs/infercnv/crc//17_HMM_predHMMi6.hmm_mode-samples.cell_groupings\n",
      "INFO [2023-08-08 21:47:43] -writing cnv regions file: outputs/infercnv/crc//17_HMM_predHMMi6.hmm_mode-samples.pred_cnv_regions.dat\n",
      "INFO [2023-08-08 21:47:44] -writing per-gene cnv report: outputs/infercnv/crc//17_HMM_predHMMi6.hmm_mode-samples.pred_cnv_genes.dat\n",
      "INFO [2023-08-08 21:47:44] -writing gene ordering info: outputs/infercnv/crc//17_HMM_predHMMi6.hmm_mode-samples.genes_used.dat\n",
      "INFO [2023-08-08 21:49:19] ::plot_cnv:Start\n",
      "INFO [2023-08-08 21:49:19] ::plot_cnv:Current data dimensions (r,c)=7077,63689 Total=1365387817 Min=1 Max=6.\n",
      "INFO [2023-08-08 21:49:24] ::plot_cnv:Depending on the size of the matrix this may take a moment.\n",
      "INFO [2023-08-08 21:49:47] plot_cnv_observation:Start\n",
      "INFO [2023-08-08 21:49:47] Observation data size: Cells= 47285 Genes= 7077\n",
      "INFO [2023-08-08 21:49:57] plot_cnv_observation:Writing observation groupings/color.\n",
      "INFO [2023-08-08 21:49:57] plot_cnv_observation:Done writing observation groupings/color.\n",
      "INFO [2023-08-08 21:50:00] plot_cnv_observation:Writing observation heatmap thresholds.\n",
      "INFO [2023-08-08 21:50:00] plot_cnv_observation:Done writing observation heatmap thresholds.\n",
      "INFO [2023-08-08 21:51:46] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
      "INFO [2023-08-08 21:51:46] Quantiles of plotted data range: 1,3,3,3,6\n",
      "INFO [2023-08-08 21:52:45] plot_cnv_references:Start\n",
      "INFO [2023-08-08 21:52:45] Reference data size: Cells= 16404 Genes= 7077\n",
      "INFO [2023-08-08 22:22:46] plot_cnv_references:Number reference groups= 6\n",
      "INFO [2023-08-08 22:22:48] plot_cnv_references:Plotting heatmap.\n"
     ]
    }
   ],
   "source": [
    "# perform infercnv operations to reveal cnv signal\n",
    "infercnv_obj = infercnv::run(infercnv_obj,\n",
    "                             cutoff=0.1,  # use 1 for smart-seq, 0.1 for 10x-genomics\n",
    "                             out_dir=\"outputs/infercnv/crc/\",  # dir is auto-created for storing outputs\n",
    "                             cluster_by_groups=T,   # cluster\n",
    "                             denoise=T,\n",
    "                             HMM=T,\n",
    "                            analysis_mode='samples',\n",
    "                                       leiden_resolution=0.01, #initial run returned too many subclusters\n",
    "                             )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "56e476df-471f-4cff-9573-a82feb66588f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "infercnv_obj = readRDS(\"outputs/infercnv/crc/17_HMM_predHMMi6.hmm_mode-samples.infercnv_obj\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "e8930791-dba8-48a1-bc11-fe73c25ab763",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2023-08-09 15:05:05] ::process_data:Start\n",
      "INFO [2023-08-09 15:05:05] Checking for saved results.\n",
      "INFO [2023-08-09 15:05:05] Trying to reload from step 17\n",
      "INFO [2023-08-09 15:05:23] Using backup HMM from step 17\n",
      "INFO [2023-08-09 15:05:23] Trying to reload from step 15\n",
      "INFO [2023-08-09 15:05:54] Using backup from step 15\n",
      "INFO [2023-08-09 15:05:54] \n",
      "\n",
      "\tSTEP 1: incoming data\n",
      "\n",
      "INFO [2023-08-09 15:05:54] \n",
      "\n",
      "\tSTEP 02: Removing lowly expressed genes\n",
      "\n",
      "INFO [2023-08-09 15:05:54] \n",
      "\n",
      "\tSTEP 03: normalization by sequencing depth\n",
      "\n",
      "INFO [2023-08-09 15:05:54] \n",
      "\n",
      "\tSTEP 04: log transformation of data\n",
      "\n",
      "INFO [2023-08-09 15:05:54] \n",
      "\n",
      "\tSTEP 08: removing average of reference data (before smoothing)\n",
      "\n",
      "INFO [2023-08-09 15:05:54] \n",
      "\n",
      "\tSTEP 09: apply max centered expression threshold: 3\n",
      "\n",
      "INFO [2023-08-09 15:05:54] \n",
      "\n",
      "\tSTEP 10: Smoothing data per cell by chromosome\n",
      "\n",
      "INFO [2023-08-09 15:05:54] \n",
      "\n",
      "\tSTEP 11: re-centering data across chromosome after smoothing\n",
      "\n",
      "INFO [2023-08-09 15:05:54] \n",
      "\n",
      "\tSTEP 12: removing average of reference data (after smoothing)\n",
      "\n",
      "INFO [2023-08-09 15:05:54] \n",
      "\n",
      "\tSTEP 14: invert log2(FC) to FC\n",
      "\n",
      "INFO [2023-08-09 15:05:54] \n",
      "\n",
      "\tSTEP 15: Clustering samples (not defining tumor subclusters)\n",
      "\n",
      "INFO [2023-08-09 15:06:14] \n",
      "\n",
      "\tSTEP 18: Run Bayesian Network Model on HMM predicted CNVs\n",
      "\n",
      "INFO [2023-08-09 15:06:14] Creating the following Directory:  outputs/infercnv/crc//BayesNetOutput.HMMi6.hmm_mode-samples\n",
      "INFO [2023-08-09 15:06:14] Initializing new MCM InferCNV Object.\n",
      "INFO [2023-08-09 15:06:14] validating infercnv_obj\n",
      "INFO [2023-08-09 15:06:15] Total CNV's:  113\n",
      "INFO [2023-08-09 15:06:15] Loading BUGS Model.\n",
      "INFO [2023-08-09 15:06:15] Running Sampling Using Parallel with  4 Cores\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in parallel::mclapply(seq_along(obj@cell_gene), FUN = par_func, :\n",
      "“scheduled core 4 did not deliver a result, all values of the job will be affected”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2023-08-09 19:22:19] Obtaining probabilities post-sampling\n"
     ]
    },
    {
     "ename": "ERROR",
     "evalue": "Error in do.call(rbind, mcmc[[j]]): second argument must be a list\n",
     "output_type": "error",
     "traceback": [
      "Error in do.call(rbind, mcmc[[j]]): second argument must be a list\nTraceback:\n",
      "1. infercnv::run(infercnv_obj, cutoff = 0.1, out_dir = \"outputs/infercnv/crc/\", \n .     cluster_by_groups = T, denoise = T, HMM = T, analysis_mode = \"samples\", \n .     leiden_resolution = 0.01, )",
      "2. infercnv::inferCNVBayesNet(infercnv_obj = infercnv_obj, HMM_states = hmm.infercnv_obj@expr.data, \n .     file_dir = out_dir, no_plot = no_plot, postMcmcMethod = \"removeCNV\", \n .     out_dir = file.path(out_dir, sprintf(\"BayesNetOutput.%s\", \n .         hmm_resume_file_token)), resume_file_token = hmm_resume_file_token, \n .     quietly = TRUE, CORES = num_threads, plotingProbs = plot_probabilities, \n .     diagnostics = diagnostics, HMM_type = HMM_type, k_obs_groups = k_obs_groups, \n .     cluster_by_groups = cluster_by_groups, reassignCNVs = reassignCNVs)",
      "3. runMCMC(MCMC_inferCNV_obj, diagnostics)",
      "4. runMCMC(MCMC_inferCNV_obj, diagnostics)",
      "5. getProbabilities(obj, mcmc)",
      "6. getProbabilities(obj, mcmc)",
      "7. do.call(rbind, mcmc[[j]])",
      "8. stop(\"second argument must be a list\")"
     ]
    }
   ],
   "source": [
    "# perform infercnv operations to reveal cnv signal\n",
    "infercnv_obj = infercnv::run(infercnv_obj,\n",
    "                             cutoff=0.1,  # use 1 for smart-seq, 0.1 for 10x-genomics\n",
    "                             out_dir=\"outputs/infercnv/crc/\",  # dir is auto-created for storing outputs\n",
    "                             cluster_by_groups=T,   # cluster\n",
    "                             denoise=T,\n",
    "                             HMM=T,\n",
    "                            analysis_mode='samples',\n",
    "                                       leiden_resolution=0.01, #initial run returned too many subclusters\n",
    "                             )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5aad73de-46ae-4dab-9ae9-56eb7c3faf8d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R (infercnv-env)",
   "language": "R",
   "name": "ir_infercnv"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.2.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
