{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7afc6593-e264-4297-9c38-26eaec325593",
   "metadata": {},
   "outputs": [],
   "source": [
    "# this library was built against GRCh37 / hg19"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "2047c7b8-0553-4ddb-89c1-329948bfe470",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "The legacy packages maptools, rgdal, and rgeos, underpinning this package\n",
      "will retire shortly. Please refer to R-spatial evolution reports on\n",
      "https://r-spatial.org/r/2023/05/15/evolution4.html for details.\n",
      "This package is now running under evolution status 0 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(infercnv)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "60ff9356-10b0-460b-b806-0e9917519f79",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(rtracklayer)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "0d9316f3-7215-4180-abae-debef5172c55",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "── \u001b[1mAttaching core tidyverse packages\u001b[22m ──────────────────────── tidyverse 2.0.0 ──\n",
      "\u001b[32m✔\u001b[39m \u001b[34mdplyr    \u001b[39m 1.1.2     \u001b[32m✔\u001b[39m \u001b[34mreadr    \u001b[39m 2.1.4\n",
      "\u001b[32m✔\u001b[39m \u001b[34mforcats  \u001b[39m 1.0.0     \u001b[32m✔\u001b[39m \u001b[34mstringr  \u001b[39m 1.5.0\n",
      "\u001b[32m✔\u001b[39m \u001b[34mggplot2  \u001b[39m 3.4.2     \u001b[32m✔\u001b[39m \u001b[34mtibble   \u001b[39m 3.2.1\n",
      "\u001b[32m✔\u001b[39m \u001b[34mlubridate\u001b[39m 1.9.2     \u001b[32m✔\u001b[39m \u001b[34mtidyr    \u001b[39m 1.3.0\n",
      "\u001b[32m✔\u001b[39m \u001b[34mpurrr    \u001b[39m 1.0.1     \n",
      "── \u001b[1mConflicts\u001b[22m ────────────────────────────────────────── tidyverse_conflicts() ──\n",
      "\u001b[31m✖\u001b[39m \u001b[34mlubridate\u001b[39m::\u001b[32m%within%()\u001b[39m masks \u001b[34mIRanges\u001b[39m::%within%()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mcollapse()\u001b[39m     masks \u001b[34mIRanges\u001b[39m::collapse()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mcombine()\u001b[39m      masks \u001b[34mBiocGenerics\u001b[39m::combine()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mdesc()\u001b[39m         masks \u001b[34mIRanges\u001b[39m::desc()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mtidyr\u001b[39m::\u001b[32mexpand()\u001b[39m       masks \u001b[34mS4Vectors\u001b[39m::expand()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfilter()\u001b[39m       masks \u001b[34mstats\u001b[39m::filter()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfirst()\u001b[39m        masks \u001b[34mS4Vectors\u001b[39m::first()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlag()\u001b[39m          masks \u001b[34mstats\u001b[39m::lag()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mggplot2\u001b[39m::\u001b[32mPosition()\u001b[39m   masks \u001b[34mBiocGenerics\u001b[39m::Position(), \u001b[34mbase\u001b[39m::Position()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mpurrr\u001b[39m::\u001b[32mreduce()\u001b[39m       masks \u001b[34mGenomicRanges\u001b[39m::reduce(), \u001b[34mIRanges\u001b[39m::reduce()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mrename()\u001b[39m       masks \u001b[34mS4Vectors\u001b[39m::rename()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mlubridate\u001b[39m::\u001b[32msecond()\u001b[39m   masks \u001b[34mS4Vectors\u001b[39m::second()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mlubridate\u001b[39m::\u001b[32msecond<-()\u001b[39m masks \u001b[34mS4Vectors\u001b[39m::second<-()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mslice()\u001b[39m        masks \u001b[34mIRanges\u001b[39m::slice()\n",
      "\u001b[36mℹ\u001b[39m Use the conflicted package (\u001b[3m\u001b[34m<http://conflicted.r-lib.org/>\u001b[39m\u001b[23m) to force all conflicts to become errors\n"
     ]
    }
   ],
   "source": [
    "library(tidyverse)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "35d6a756-c4bd-458c-a4d5-83eb06ead780",
   "metadata": {},
   "source": [
    "# Prepare files for infercnv"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19ba92ee-6f48-4ab6-86ff-529caf19730d",
   "metadata": {
    "tags": []
   },
   "source": [
    "### prepare gene position file (one time only, then read from disk)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1a8c149f-6fbb-4d4b-8111-0704cf8fdf03",
   "metadata": {},
   "source": [
    "#### I downloaded the refGene gtf file from https://hgdownload.soe.ucsc.edu/goldenPath/hg19/bigZips/genes/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "760f96ea-0f0a-4d11-b2c1-4143c8da969a",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf <- rtracklayer::import('data/hg19.refGene.gtf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "65707cc3-e948-4baf-b45f-7c9f0cc79bcb",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf_df=as.data.frame(gtf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "a7ba795d-7bea-42de-8dd0-890f4bdb057e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 14</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>seqnames</th><th scope=col>start</th><th scope=col>end</th><th scope=col>width</th><th scope=col>strand</th><th scope=col>source</th><th scope=col>type</th><th scope=col>score</th><th scope=col>phase</th><th scope=col>gene_id</th><th scope=col>transcript_id</th><th scope=col>gene_name</th><th scope=col>exon_number</th><th scope=col>exon_id</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>chr1</td><td>11869</td><td>14362</td><td>2494</td><td>+</td><td>refGene</td><td>transcript</td><td>NA</td><td>NA</td><td>LOC102725121</td><td>NR_148357</td><td>LOC102725121</td><td>NA</td><td>NA         </td></tr>\n",
       "\t<tr><th scope=row>2</th><td>chr1</td><td>11869</td><td>12227</td><td> 359</td><td>+</td><td>refGene</td><td>exon      </td><td>NA</td><td>NA</td><td>LOC102725121</td><td>NR_148357</td><td>LOC102725121</td><td>1 </td><td>NR_148357.1</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>chr1</td><td>12613</td><td>12721</td><td> 109</td><td>+</td><td>refGene</td><td>exon      </td><td>NA</td><td>NA</td><td>LOC102725121</td><td>NR_148357</td><td>LOC102725121</td><td>2 </td><td>NR_148357.2</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>chr1</td><td>13221</td><td>14362</td><td>1142</td><td>+</td><td>refGene</td><td>exon      </td><td>NA</td><td>NA</td><td>LOC102725121</td><td>NR_148357</td><td>LOC102725121</td><td>3 </td><td>NR_148357.3</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>chr1</td><td>11874</td><td>14409</td><td>2536</td><td>+</td><td>refGene</td><td>transcript</td><td>NA</td><td>NA</td><td>DDX11L1     </td><td>NR_046018</td><td>DDX11L1     </td><td>NA</td><td>NA         </td></tr>\n",
       "\t<tr><th scope=row>6</th><td>chr1</td><td>11874</td><td>12227</td><td> 354</td><td>+</td><td>refGene</td><td>exon      </td><td>NA</td><td>NA</td><td>DDX11L1     </td><td>NR_046018</td><td>DDX11L1     </td><td>1 </td><td>NR_046018.1</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 14\n",
       "\\begin{tabular}{r|llllllllllllll}\n",
       "  & seqnames & start & end & width & strand & source & type & score & phase & gene\\_id & transcript\\_id & gene\\_name & exon\\_number & exon\\_id\\\\\n",
       "  & <fct> & <int> & <int> & <int> & <fct> & <fct> & <fct> & <dbl> & <int> & <chr> & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & chr1 & 11869 & 14362 & 2494 & + & refGene & transcript & NA & NA & LOC102725121 & NR\\_148357 & LOC102725121 & NA & NA         \\\\\n",
       "\t2 & chr1 & 11869 & 12227 &  359 & + & refGene & exon       & NA & NA & LOC102725121 & NR\\_148357 & LOC102725121 & 1  & NR\\_148357.1\\\\\n",
       "\t3 & chr1 & 12613 & 12721 &  109 & + & refGene & exon       & NA & NA & LOC102725121 & NR\\_148357 & LOC102725121 & 2  & NR\\_148357.2\\\\\n",
       "\t4 & chr1 & 13221 & 14362 & 1142 & + & refGene & exon       & NA & NA & LOC102725121 & NR\\_148357 & LOC102725121 & 3  & NR\\_148357.3\\\\\n",
       "\t5 & chr1 & 11874 & 14409 & 2536 & + & refGene & transcript & NA & NA & DDX11L1      & NR\\_046018 & DDX11L1      & NA & NA         \\\\\n",
       "\t6 & chr1 & 11874 & 12227 &  354 & + & refGene & exon       & NA & NA & DDX11L1      & NR\\_046018 & DDX11L1      & 1  & NR\\_046018.1\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 14\n",
       "\n",
       "| <!--/--> | seqnames &lt;fct&gt; | start &lt;int&gt; | end &lt;int&gt; | width &lt;int&gt; | strand &lt;fct&gt; | source &lt;fct&gt; | type &lt;fct&gt; | score &lt;dbl&gt; | phase &lt;int&gt; | gene_id &lt;chr&gt; | transcript_id &lt;chr&gt; | gene_name &lt;chr&gt; | exon_number &lt;chr&gt; | exon_id &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | chr1 | 11869 | 14362 | 2494 | + | refGene | transcript | NA | NA | LOC102725121 | NR_148357 | LOC102725121 | NA | NA          |\n",
       "| 2 | chr1 | 11869 | 12227 |  359 | + | refGene | exon       | NA | NA | LOC102725121 | NR_148357 | LOC102725121 | 1  | NR_148357.1 |\n",
       "| 3 | chr1 | 12613 | 12721 |  109 | + | refGene | exon       | NA | NA | LOC102725121 | NR_148357 | LOC102725121 | 2  | NR_148357.2 |\n",
       "| 4 | chr1 | 13221 | 14362 | 1142 | + | refGene | exon       | NA | NA | LOC102725121 | NR_148357 | LOC102725121 | 3  | NR_148357.3 |\n",
       "| 5 | chr1 | 11874 | 14409 | 2536 | + | refGene | transcript | NA | NA | DDX11L1      | NR_046018 | DDX11L1      | NA | NA          |\n",
       "| 6 | chr1 | 11874 | 12227 |  354 | + | refGene | exon       | NA | NA | DDX11L1      | NR_046018 | DDX11L1      | 1  | NR_046018.1 |\n",
       "\n"
      ],
      "text/plain": [
       "  seqnames start end   width strand source  type       score phase gene_id     \n",
       "1 chr1     11869 14362 2494  +      refGene transcript NA    NA    LOC102725121\n",
       "2 chr1     11869 12227  359  +      refGene exon       NA    NA    LOC102725121\n",
       "3 chr1     12613 12721  109  +      refGene exon       NA    NA    LOC102725121\n",
       "4 chr1     13221 14362 1142  +      refGene exon       NA    NA    LOC102725121\n",
       "5 chr1     11874 14409 2536  +      refGene transcript NA    NA    DDX11L1     \n",
       "6 chr1     11874 12227  354  +      refGene exon       NA    NA    DDX11L1     \n",
       "  transcript_id gene_name    exon_number exon_id    \n",
       "1 NR_148357     LOC102725121 NA          NA         \n",
       "2 NR_148357     LOC102725121 1           NR_148357.1\n",
       "3 NR_148357     LOC102725121 2           NR_148357.2\n",
       "4 NR_148357     LOC102725121 3           NR_148357.3\n",
       "5 NR_046018     DDX11L1      NA          NA         \n",
       "6 NR_046018     DDX11L1      1           NR_046018.1"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(gtf_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "f196b0c4-fc52-457c-9d63-0e04e94e0838",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf_infercnv_input = gtf_df %>% filter(type==\"transcript\") %>% select(\"gene_id\", \"seqnames\", \"start\", \"end\") %>% unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "00d0e988-1598-4d83-9b13-dc8b1f02b02b",
   "metadata": {},
   "source": [
    "#### are genes still duplicated here? yes, annotated with slightly different start/stops"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "bfe7ebbc-492b-4da9-ad61-572a48c336e2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "46562"
      ],
      "text/latex": [
       "46562"
      ],
      "text/markdown": [
       "46562"
      ],
      "text/plain": [
       "[1] 46562"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "gtf_infercnv_input %>% nrow()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "602c1fc3-40dd-466d-8648-09652d0f9ea7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "28265"
      ],
      "text/latex": [
       "28265"
      ],
      "text/markdown": [
       "28265"
      ],
      "text/plain": [
       "[1] 28265"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "gtf_infercnv_input %>% select(\"gene_id\") %>% unique() %>% nrow()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "6fef8c30-35c2-4d1e-93e3-d43d68843500",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 18297 × 4</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>gene_id</th><th scope=col>seqnames</th><th scope=col>start</th><th scope=col>end</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>26</th><td>LINC01128  </td><td>chr1          </td><td>   762971</td><td>   794826</td></tr>\n",
       "\t<tr><th scope=row>30</th><td>LINC01128  </td><td>chr1          </td><td>   763178</td><td>   794826</td></tr>\n",
       "\t<tr><th scope=row>42</th><td>MAS1L      </td><td>chr6_apd_hap1 </td><td>   757952</td><td>   759211</td></tr>\n",
       "\t<tr><th scope=row>43</th><td>MAS1L      </td><td>chr6_cox_hap2 </td><td>   973255</td><td>   974514</td></tr>\n",
       "\t<tr><th scope=row>44</th><td>MAS1L      </td><td>chr6_dbb_hap3 </td><td>   757805</td><td>   759064</td></tr>\n",
       "\t<tr><th scope=row>94</th><td>ACYP1      </td><td>chr14         </td><td> 75519946</td><td> 75530727</td></tr>\n",
       "\t<tr><th scope=row>114</th><td>MORF4L2    </td><td>chrX          </td><td>102930426</td><td>102941746</td></tr>\n",
       "\t<tr><th scope=row>132</th><td>MMP23A     </td><td>chr1          </td><td>  1631378</td><td>  1633247</td></tr>\n",
       "\t<tr><th scope=row>148</th><td>PUF60      </td><td>chr8          </td><td>144898514</td><td>144911211</td></tr>\n",
       "\t<tr><th scope=row>166</th><td>HYAL3      </td><td>chr3          </td><td> 50330263</td><td> 50336291</td></tr>\n",
       "\t<tr><th scope=row>177</th><td>MAST4      </td><td>chr5          </td><td> 65892221</td><td> 66282351</td></tr>\n",
       "\t<tr><th scope=row>203</th><td>OR2H2      </td><td>chr6_mcf_hap5 </td><td>   858886</td><td>   859948</td></tr>\n",
       "\t<tr><th scope=row>204</th><td>OR2H2      </td><td>chr6_qbl_hap6 </td><td>   859014</td><td>   860076</td></tr>\n",
       "\t<tr><th scope=row>241</th><td>SRSF10     </td><td>chr1          </td><td> 24290837</td><td> 24306817</td></tr>\n",
       "\t<tr><th scope=row>275</th><td>NRM        </td><td>chr6_mcf_hap5 </td><td>  2037711</td><td>  2040646</td></tr>\n",
       "\t<tr><th scope=row>276</th><td>NRM        </td><td>chr6_cox_hap2 </td><td>  2167899</td><td>  2170834</td></tr>\n",
       "\t<tr><th scope=row>280</th><td>AGER       </td><td>chr6_mann_hap4</td><td>  3491570</td><td>  3494854</td></tr>\n",
       "\t<tr><th scope=row>315</th><td>PMS1       </td><td>chr2          </td><td>190649176</td><td>190671779</td></tr>\n",
       "\t<tr><th scope=row>318</th><td>PMS1       </td><td>chr2          </td><td>190648811</td><td>190671572</td></tr>\n",
       "\t<tr><th scope=row>322</th><td>MDFI       </td><td>chr6          </td><td> 41604756</td><td> 41621982</td></tr>\n",
       "\t<tr><th scope=row>325</th><td>RNASE1     </td><td>chr14         </td><td> 21269387</td><td> 21270523</td></tr>\n",
       "\t<tr><th scope=row>338</th><td>MSANTD3    </td><td>chr9          </td><td>103189711</td><td>103214020</td></tr>\n",
       "\t<tr><th scope=row>340</th><td>RBFOX2     </td><td>chr22         </td><td> 36134781</td><td> 36424585</td></tr>\n",
       "\t<tr><th scope=row>354</th><td>USP15      </td><td>chr12         </td><td> 62654121</td><td> 62730474</td></tr>\n",
       "\t<tr><th scope=row>371</th><td>LINC02606  </td><td>chr1          </td><td>  9485153</td><td>  9489055</td></tr>\n",
       "\t<tr><th scope=row>377</th><td>PIK3CD-AS2 </td><td>chr1          </td><td>  9740903</td><td>  9747632</td></tr>\n",
       "\t<tr><th scope=row>420</th><td>CDCA3      </td><td>chr12         </td><td>  6955889</td><td>  6961094</td></tr>\n",
       "\t<tr><th scope=row>428</th><td>TRPV1      </td><td>chr17         </td><td>  3468740</td><td>  3499125</td></tr>\n",
       "\t<tr><th scope=row>438</th><td>EXOSC10-AS1</td><td>chr1          </td><td> 11160076</td><td> 11162162</td></tr>\n",
       "\t<tr><th scope=row>457</th><td>TPRKB      </td><td>chr2          </td><td> 73956964</td><td> 73964533</td></tr>\n",
       "\t<tr><th scope=row>⋮</th><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><th scope=row>77312</th><td>SNAP91  </td><td>chr6 </td><td> 84262605</td><td> 84419112</td></tr>\n",
       "\t<tr><th scope=row>77324</th><td>SNAP91  </td><td>chr6 </td><td> 84304829</td><td> 84418820</td></tr>\n",
       "\t<tr><th scope=row>77325</th><td>SNAP91  </td><td>chr6 </td><td> 84318152</td><td> 84418820</td></tr>\n",
       "\t<tr><th scope=row>77327</th><td>HP1BP3  </td><td>chr1 </td><td> 21066759</td><td> 21113145</td></tr>\n",
       "\t<tr><th scope=row>77333</th><td>HP1BP3  </td><td>chr1 </td><td> 21066759</td><td> 21113159</td></tr>\n",
       "\t<tr><th scope=row>77337</th><td>HP1BP3  </td><td>chr1 </td><td> 21080549</td><td> 21113145</td></tr>\n",
       "\t<tr><th scope=row>77338</th><td>HP1BP3  </td><td>chr1 </td><td> 21080549</td><td> 21113800</td></tr>\n",
       "\t<tr><th scope=row>77343</th><td>MADD    </td><td>chr11</td><td> 47290739</td><td> 47351582</td></tr>\n",
       "\t<tr><th scope=row>77352</th><td>MADD    </td><td>chr11</td><td> 47292688</td><td> 47351582</td></tr>\n",
       "\t<tr><th scope=row>77353</th><td>CORO1A  </td><td>chr16</td><td> 30194923</td><td> 30200397</td></tr>\n",
       "\t<tr><th scope=row>77401</th><td>SIRT5   </td><td>chr6 </td><td> 13574506</td><td> 13614339</td></tr>\n",
       "\t<tr><th scope=row>77404</th><td>SIRT5   </td><td>chr6 </td><td> 13574537</td><td> 13614339</td></tr>\n",
       "\t<tr><th scope=row>77406</th><td>SIRT5   </td><td>chr6 </td><td> 13574833</td><td> 13605970</td></tr>\n",
       "\t<tr><th scope=row>77408</th><td>SIRT5   </td><td>chr6 </td><td> 13574833</td><td> 13611086</td></tr>\n",
       "\t<tr><th scope=row>77409</th><td>SIRT5   </td><td>chr6 </td><td> 13574833</td><td> 13614339</td></tr>\n",
       "\t<tr><th scope=row>77433</th><td>TXLNA   </td><td>chr1 </td><td> 32645666</td><td> 32663886</td></tr>\n",
       "\t<tr><th scope=row>77434</th><td>ATP6V1G3</td><td>chr1 </td><td>198492352</td><td>198509810</td></tr>\n",
       "\t<tr><th scope=row>77437</th><td>ATL1    </td><td>chr14</td><td> 51026863</td><td> 51099760</td></tr>\n",
       "\t<tr><th scope=row>77438</th><td>TNFSF14 </td><td>chr19</td><td>  6661264</td><td>  6670171</td></tr>\n",
       "\t<tr><th scope=row>77439</th><td>MAPK1   </td><td>chr22</td><td> 22113946</td><td> 22221935</td></tr>\n",
       "\t<tr><th scope=row>77440</th><td>PLXNB2  </td><td>chr22</td><td> 50713408</td><td> 50733221</td></tr>\n",
       "\t<tr><th scope=row>77441</th><td>PLXNB2  </td><td>chr22</td><td> 50713408</td><td> 50738764</td></tr>\n",
       "\t<tr><th scope=row>77453</th><td>PLXNB2  </td><td>chr22</td><td> 50713408</td><td> 50740098</td></tr>\n",
       "\t<tr><th scope=row>77454</th><td>PLXNB2  </td><td>chr22</td><td> 50713408</td><td> 50742250</td></tr>\n",
       "\t<tr><th scope=row>77463</th><td>MBNL1   </td><td>chr3 </td><td>151961421</td><td>152183569</td></tr>\n",
       "\t<tr><th scope=row>77513</th><td>CPNE5   </td><td>chr6 </td><td> 36708553</td><td> 36733572</td></tr>\n",
       "\t<tr><th scope=row>77515</th><td>CPNE5   </td><td>chr6 </td><td> 36708553</td><td> 36807774</td></tr>\n",
       "\t<tr><th scope=row>77516</th><td>CPNE5   </td><td>chr6 </td><td> 36711028</td><td> 36725177</td></tr>\n",
       "\t<tr><th scope=row>77519</th><td>CPNE5   </td><td>chr6 </td><td> 36765162</td><td> 36807774</td></tr>\n",
       "\t<tr><th scope=row>77526</th><td>TMEM181 </td><td>chr6 </td><td>158981123</td><td>159056461</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 18297 × 4\n",
       "\\begin{tabular}{r|llll}\n",
       "  & gene\\_id & seqnames & start & end\\\\\n",
       "  & <chr> & <fct> & <int> & <int>\\\\\n",
       "\\hline\n",
       "\t26 & LINC01128   & chr1           &    762971 &    794826\\\\\n",
       "\t30 & LINC01128   & chr1           &    763178 &    794826\\\\\n",
       "\t42 & MAS1L       & chr6\\_apd\\_hap1  &    757952 &    759211\\\\\n",
       "\t43 & MAS1L       & chr6\\_cox\\_hap2  &    973255 &    974514\\\\\n",
       "\t44 & MAS1L       & chr6\\_dbb\\_hap3  &    757805 &    759064\\\\\n",
       "\t94 & ACYP1       & chr14          &  75519946 &  75530727\\\\\n",
       "\t114 & MORF4L2     & chrX           & 102930426 & 102941746\\\\\n",
       "\t132 & MMP23A      & chr1           &   1631378 &   1633247\\\\\n",
       "\t148 & PUF60       & chr8           & 144898514 & 144911211\\\\\n",
       "\t166 & HYAL3       & chr3           &  50330263 &  50336291\\\\\n",
       "\t177 & MAST4       & chr5           &  65892221 &  66282351\\\\\n",
       "\t203 & OR2H2       & chr6\\_mcf\\_hap5  &    858886 &    859948\\\\\n",
       "\t204 & OR2H2       & chr6\\_qbl\\_hap6  &    859014 &    860076\\\\\n",
       "\t241 & SRSF10      & chr1           &  24290837 &  24306817\\\\\n",
       "\t275 & NRM         & chr6\\_mcf\\_hap5  &   2037711 &   2040646\\\\\n",
       "\t276 & NRM         & chr6\\_cox\\_hap2  &   2167899 &   2170834\\\\\n",
       "\t280 & AGER        & chr6\\_mann\\_hap4 &   3491570 &   3494854\\\\\n",
       "\t315 & PMS1        & chr2           & 190649176 & 190671779\\\\\n",
       "\t318 & PMS1        & chr2           & 190648811 & 190671572\\\\\n",
       "\t322 & MDFI        & chr6           &  41604756 &  41621982\\\\\n",
       "\t325 & RNASE1      & chr14          &  21269387 &  21270523\\\\\n",
       "\t338 & MSANTD3     & chr9           & 103189711 & 103214020\\\\\n",
       "\t340 & RBFOX2      & chr22          &  36134781 &  36424585\\\\\n",
       "\t354 & USP15       & chr12          &  62654121 &  62730474\\\\\n",
       "\t371 & LINC02606   & chr1           &   9485153 &   9489055\\\\\n",
       "\t377 & PIK3CD-AS2  & chr1           &   9740903 &   9747632\\\\\n",
       "\t420 & CDCA3       & chr12          &   6955889 &   6961094\\\\\n",
       "\t428 & TRPV1       & chr17          &   3468740 &   3499125\\\\\n",
       "\t438 & EXOSC10-AS1 & chr1           &  11160076 &  11162162\\\\\n",
       "\t457 & TPRKB       & chr2           &  73956964 &  73964533\\\\\n",
       "\t⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t77312 & SNAP91   & chr6  &  84262605 &  84419112\\\\\n",
       "\t77324 & SNAP91   & chr6  &  84304829 &  84418820\\\\\n",
       "\t77325 & SNAP91   & chr6  &  84318152 &  84418820\\\\\n",
       "\t77327 & HP1BP3   & chr1  &  21066759 &  21113145\\\\\n",
       "\t77333 & HP1BP3   & chr1  &  21066759 &  21113159\\\\\n",
       "\t77337 & HP1BP3   & chr1  &  21080549 &  21113145\\\\\n",
       "\t77338 & HP1BP3   & chr1  &  21080549 &  21113800\\\\\n",
       "\t77343 & MADD     & chr11 &  47290739 &  47351582\\\\\n",
       "\t77352 & MADD     & chr11 &  47292688 &  47351582\\\\\n",
       "\t77353 & CORO1A   & chr16 &  30194923 &  30200397\\\\\n",
       "\t77401 & SIRT5    & chr6  &  13574506 &  13614339\\\\\n",
       "\t77404 & SIRT5    & chr6  &  13574537 &  13614339\\\\\n",
       "\t77406 & SIRT5    & chr6  &  13574833 &  13605970\\\\\n",
       "\t77408 & SIRT5    & chr6  &  13574833 &  13611086\\\\\n",
       "\t77409 & SIRT5    & chr6  &  13574833 &  13614339\\\\\n",
       "\t77433 & TXLNA    & chr1  &  32645666 &  32663886\\\\\n",
       "\t77434 & ATP6V1G3 & chr1  & 198492352 & 198509810\\\\\n",
       "\t77437 & ATL1     & chr14 &  51026863 &  51099760\\\\\n",
       "\t77438 & TNFSF14  & chr19 &   6661264 &   6670171\\\\\n",
       "\t77439 & MAPK1    & chr22 &  22113946 &  22221935\\\\\n",
       "\t77440 & PLXNB2   & chr22 &  50713408 &  50733221\\\\\n",
       "\t77441 & PLXNB2   & chr22 &  50713408 &  50738764\\\\\n",
       "\t77453 & PLXNB2   & chr22 &  50713408 &  50740098\\\\\n",
       "\t77454 & PLXNB2   & chr22 &  50713408 &  50742250\\\\\n",
       "\t77463 & MBNL1    & chr3  & 151961421 & 152183569\\\\\n",
       "\t77513 & CPNE5    & chr6  &  36708553 &  36733572\\\\\n",
       "\t77515 & CPNE5    & chr6  &  36708553 &  36807774\\\\\n",
       "\t77516 & CPNE5    & chr6  &  36711028 &  36725177\\\\\n",
       "\t77519 & CPNE5    & chr6  &  36765162 &  36807774\\\\\n",
       "\t77526 & TMEM181  & chr6  & 158981123 & 159056461\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 18297 × 4\n",
       "\n",
       "| <!--/--> | gene_id &lt;chr&gt; | seqnames &lt;fct&gt; | start &lt;int&gt; | end &lt;int&gt; |\n",
       "|---|---|---|---|---|\n",
       "| 26 | LINC01128   | chr1           |    762971 |    794826 |\n",
       "| 30 | LINC01128   | chr1           |    763178 |    794826 |\n",
       "| 42 | MAS1L       | chr6_apd_hap1  |    757952 |    759211 |\n",
       "| 43 | MAS1L       | chr6_cox_hap2  |    973255 |    974514 |\n",
       "| 44 | MAS1L       | chr6_dbb_hap3  |    757805 |    759064 |\n",
       "| 94 | ACYP1       | chr14          |  75519946 |  75530727 |\n",
       "| 114 | MORF4L2     | chrX           | 102930426 | 102941746 |\n",
       "| 132 | MMP23A      | chr1           |   1631378 |   1633247 |\n",
       "| 148 | PUF60       | chr8           | 144898514 | 144911211 |\n",
       "| 166 | HYAL3       | chr3           |  50330263 |  50336291 |\n",
       "| 177 | MAST4       | chr5           |  65892221 |  66282351 |\n",
       "| 203 | OR2H2       | chr6_mcf_hap5  |    858886 |    859948 |\n",
       "| 204 | OR2H2       | chr6_qbl_hap6  |    859014 |    860076 |\n",
       "| 241 | SRSF10      | chr1           |  24290837 |  24306817 |\n",
       "| 275 | NRM         | chr6_mcf_hap5  |   2037711 |   2040646 |\n",
       "| 276 | NRM         | chr6_cox_hap2  |   2167899 |   2170834 |\n",
       "| 280 | AGER        | chr6_mann_hap4 |   3491570 |   3494854 |\n",
       "| 315 | PMS1        | chr2           | 190649176 | 190671779 |\n",
       "| 318 | PMS1        | chr2           | 190648811 | 190671572 |\n",
       "| 322 | MDFI        | chr6           |  41604756 |  41621982 |\n",
       "| 325 | RNASE1      | chr14          |  21269387 |  21270523 |\n",
       "| 338 | MSANTD3     | chr9           | 103189711 | 103214020 |\n",
       "| 340 | RBFOX2      | chr22          |  36134781 |  36424585 |\n",
       "| 354 | USP15       | chr12          |  62654121 |  62730474 |\n",
       "| 371 | LINC02606   | chr1           |   9485153 |   9489055 |\n",
       "| 377 | PIK3CD-AS2  | chr1           |   9740903 |   9747632 |\n",
       "| 420 | CDCA3       | chr12          |   6955889 |   6961094 |\n",
       "| 428 | TRPV1       | chr17          |   3468740 |   3499125 |\n",
       "| 438 | EXOSC10-AS1 | chr1           |  11160076 |  11162162 |\n",
       "| 457 | TPRKB       | chr2           |  73956964 |  73964533 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| 77312 | SNAP91   | chr6  |  84262605 |  84419112 |\n",
       "| 77324 | SNAP91   | chr6  |  84304829 |  84418820 |\n",
       "| 77325 | SNAP91   | chr6  |  84318152 |  84418820 |\n",
       "| 77327 | HP1BP3   | chr1  |  21066759 |  21113145 |\n",
       "| 77333 | HP1BP3   | chr1  |  21066759 |  21113159 |\n",
       "| 77337 | HP1BP3   | chr1  |  21080549 |  21113145 |\n",
       "| 77338 | HP1BP3   | chr1  |  21080549 |  21113800 |\n",
       "| 77343 | MADD     | chr11 |  47290739 |  47351582 |\n",
       "| 77352 | MADD     | chr11 |  47292688 |  47351582 |\n",
       "| 77353 | CORO1A   | chr16 |  30194923 |  30200397 |\n",
       "| 77401 | SIRT5    | chr6  |  13574506 |  13614339 |\n",
       "| 77404 | SIRT5    | chr6  |  13574537 |  13614339 |\n",
       "| 77406 | SIRT5    | chr6  |  13574833 |  13605970 |\n",
       "| 77408 | SIRT5    | chr6  |  13574833 |  13611086 |\n",
       "| 77409 | SIRT5    | chr6  |  13574833 |  13614339 |\n",
       "| 77433 | TXLNA    | chr1  |  32645666 |  32663886 |\n",
       "| 77434 | ATP6V1G3 | chr1  | 198492352 | 198509810 |\n",
       "| 77437 | ATL1     | chr14 |  51026863 |  51099760 |\n",
       "| 77438 | TNFSF14  | chr19 |   6661264 |   6670171 |\n",
       "| 77439 | MAPK1    | chr22 |  22113946 |  22221935 |\n",
       "| 77440 | PLXNB2   | chr22 |  50713408 |  50733221 |\n",
       "| 77441 | PLXNB2   | chr22 |  50713408 |  50738764 |\n",
       "| 77453 | PLXNB2   | chr22 |  50713408 |  50740098 |\n",
       "| 77454 | PLXNB2   | chr22 |  50713408 |  50742250 |\n",
       "| 77463 | MBNL1    | chr3  | 151961421 | 152183569 |\n",
       "| 77513 | CPNE5    | chr6  |  36708553 |  36733572 |\n",
       "| 77515 | CPNE5    | chr6  |  36708553 |  36807774 |\n",
       "| 77516 | CPNE5    | chr6  |  36711028 |  36725177 |\n",
       "| 77519 | CPNE5    | chr6  |  36765162 |  36807774 |\n",
       "| 77526 | TMEM181  | chr6  | 158981123 | 159056461 |\n",
       "\n"
      ],
      "text/plain": [
       "      gene_id     seqnames       start     end      \n",
       "26    LINC01128   chr1              762971    794826\n",
       "30    LINC01128   chr1              763178    794826\n",
       "42    MAS1L       chr6_apd_hap1     757952    759211\n",
       "43    MAS1L       chr6_cox_hap2     973255    974514\n",
       "44    MAS1L       chr6_dbb_hap3     757805    759064\n",
       "94    ACYP1       chr14           75519946  75530727\n",
       "114   MORF4L2     chrX           102930426 102941746\n",
       "132   MMP23A      chr1             1631378   1633247\n",
       "148   PUF60       chr8           144898514 144911211\n",
       "166   HYAL3       chr3            50330263  50336291\n",
       "177   MAST4       chr5            65892221  66282351\n",
       "203   OR2H2       chr6_mcf_hap5     858886    859948\n",
       "204   OR2H2       chr6_qbl_hap6     859014    860076\n",
       "241   SRSF10      chr1            24290837  24306817\n",
       "275   NRM         chr6_mcf_hap5    2037711   2040646\n",
       "276   NRM         chr6_cox_hap2    2167899   2170834\n",
       "280   AGER        chr6_mann_hap4   3491570   3494854\n",
       "315   PMS1        chr2           190649176 190671779\n",
       "318   PMS1        chr2           190648811 190671572\n",
       "322   MDFI        chr6            41604756  41621982\n",
       "325   RNASE1      chr14           21269387  21270523\n",
       "338   MSANTD3     chr9           103189711 103214020\n",
       "340   RBFOX2      chr22           36134781  36424585\n",
       "354   USP15       chr12           62654121  62730474\n",
       "371   LINC02606   chr1             9485153   9489055\n",
       "377   PIK3CD-AS2  chr1             9740903   9747632\n",
       "420   CDCA3       chr12            6955889   6961094\n",
       "428   TRPV1       chr17            3468740   3499125\n",
       "438   EXOSC10-AS1 chr1            11160076  11162162\n",
       "457   TPRKB       chr2            73956964  73964533\n",
       "⋮     ⋮           ⋮              ⋮         ⋮        \n",
       "77312 SNAP91      chr6            84262605  84419112\n",
       "77324 SNAP91      chr6            84304829  84418820\n",
       "77325 SNAP91      chr6            84318152  84418820\n",
       "77327 HP1BP3      chr1            21066759  21113145\n",
       "77333 HP1BP3      chr1            21066759  21113159\n",
       "77337 HP1BP3      chr1            21080549  21113145\n",
       "77338 HP1BP3      chr1            21080549  21113800\n",
       "77343 MADD        chr11           47290739  47351582\n",
       "77352 MADD        chr11           47292688  47351582\n",
       "77353 CORO1A      chr16           30194923  30200397\n",
       "77401 SIRT5       chr6            13574506  13614339\n",
       "77404 SIRT5       chr6            13574537  13614339\n",
       "77406 SIRT5       chr6            13574833  13605970\n",
       "77408 SIRT5       chr6            13574833  13611086\n",
       "77409 SIRT5       chr6            13574833  13614339\n",
       "77433 TXLNA       chr1            32645666  32663886\n",
       "77434 ATP6V1G3    chr1           198492352 198509810\n",
       "77437 ATL1        chr14           51026863  51099760\n",
       "77438 TNFSF14     chr19            6661264   6670171\n",
       "77439 MAPK1       chr22           22113946  22221935\n",
       "77440 PLXNB2      chr22           50713408  50733221\n",
       "77441 PLXNB2      chr22           50713408  50738764\n",
       "77453 PLXNB2      chr22           50713408  50740098\n",
       "77454 PLXNB2      chr22           50713408  50742250\n",
       "77463 MBNL1       chr3           151961421 152183569\n",
       "77513 CPNE5       chr6            36708553  36733572\n",
       "77515 CPNE5       chr6            36708553  36807774\n",
       "77516 CPNE5       chr6            36711028  36725177\n",
       "77519 CPNE5       chr6            36765162  36807774\n",
       "77526 TMEM181     chr6           158981123 159056461"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "gtf_infercnv_input[gtf_infercnv_input %>% select(\"gene_id\") %>% duplicated(),]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a29d9ac-9d47-4db0-8114-ea8c9de0b02f",
   "metadata": {},
   "source": [
    "#### drop duplicate rows -- they're really similar so can just randomly pick one"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "cf70f3c1-84bc-45de-b82c-ad56af7f1413",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf_infercnv_input = gtf_infercnv_input[!duplicated(gtf_infercnv_input$gene_id),]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e0ecc65e-7302-4ba7-8a53-fe14e50ecb11",
   "metadata": {},
   "source": [
    "#### filter to genes on the 23 chromosomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "6d62d912-e52b-4b6b-8638-b65501f82149",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf_infercnv_input = gtf_infercnv_input %>% filter(seqnames %in% paste0(\"chr\", c(c(1:22), \"X\", \"Y\")))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "18ae6f84-f79b-4d97-b367-fcdd4487fd8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "write.table(gtf_infercnv_input, file=\"data/infercnv_input_files/sys/gene_pos.txt\", col.names=F, row.names=F, quote=F, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "839acfbb-affa-448c-aef3-5edfedd1ae1d",
   "metadata": {},
   "source": [
    "#### prepared the other 2 input files in python, writing from anndata object to disk"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "21a7c8e7-7f10-4045-a81c-a370192b8067",
   "metadata": {},
   "source": [
    "# Run infercnv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "eb63312d-bbe8-43d4-b63e-b3b3d1bdd94c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2023-07-24 16:05:23] Parsing matrix: data/infercnv_input_files/sys/counts.txt\n",
      "INFO [2023-07-24 16:06:22] Parsing gene order file: data/infercnv_input_files/sys/gene_pos.txt\n",
      "INFO [2023-07-24 16:06:22] Parsing cell annotations file: data/infercnv_input_files/sys/cell_annots.txt\n",
      "INFO [2023-07-24 16:06:22] ::order_reduce:Start.\n",
      "INFO [2023-07-24 16:06:23] .order_reduce(): expr and order match.\n",
      "INFO [2023-07-24 16:06:24] ::process_data:order_reduce:Reduction from positional data, new dimensions (r,c) = 23686,7109 Total=2333571590 Min=0 Max=202951.\n",
      "INFO [2023-07-24 16:06:25] num genes removed taking into account provided gene ordering list: 3427 = 14.4684623828422% removed.\n",
      "INFO [2023-07-24 16:06:25] -filtering out cells < 100 or > Inf, removing 0 % of cells\n",
      "WARN [2023-07-24 16:06:26] Please use \"options(scipen = 100)\" before running infercnv if you are using the analysis_mode=\"subclusters\" option or you may encounter an error while the hclust is being generated.\n",
      "INFO [2023-07-24 16:06:30] validating infercnv_obj\n"
     ]
    }
   ],
   "source": [
    "infercnv_obj = CreateInfercnvObject(raw_counts_matrix=\"data/infercnv_input_files/sys/counts.txt\",\n",
    "                                    annotations_file=\"data/infercnv_input_files/sys/cell_annots.txt\",\n",
    "                                    delim=\"\\t\",\n",
    "                                    gene_order_file=\"data/infercnv_input_files/sys/gene_pos.txt\",\n",
    "                                    ref_group_names=c(\"endothelial\",\"macrophage\",\"fibroblast\",\"t.cell\",\"t.cd4\",\"t.cd8\",\"nk\",\"mastocyte\",\"b.cell\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "dcf95983-8f5e-425f-b269-67ff31051724",
   "metadata": {},
   "outputs": [],
   "source": [
    "#if the instructions printed above say to run this, run this\n",
    "options(scipen = 100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "f0f50d10-f2c2-4b61-83c3-1036d95eb546",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IOPub data rate exceeded.\n",
      "The Jupyter server will temporarily stop sending output\n",
      "to the client in order to avoid crashing it.\n",
      "To change this limit, set the config variable\n",
      "`--ServerApp.iopub_data_rate_limit`.\n",
      "\n",
      "Current values:\n",
      "ServerApp.iopub_data_rate_limit=1000000.0 (bytes/sec)\n",
      "ServerApp.rate_limit_window=3.0 (secs)\n",
      "\n",
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  NSUN3, ARL6, CRYBG3, ARL13B, CPOX, CLDND1, PROS1, ST3GAL6, DCBLD2, EPHA3 \n",
      "\t   COL8A1, C3orf38, CMSS1, ZNF654, CGGBP1, FILIP1L, CHMP2B, GBE1, TBC1D23, ROBO1 \n",
      "\t   ROBO2, NIT2, ZNF717, LNP1, FAM86DP, PDZRN3, TMEM45A, EBLN2, TFG, PPP4R2 \n",
      "Negative:  DEAF1, TMEM80, EPS8L2, IRF7, TALDO1, PHRF1, RASSF7, SLC25A22, RPLP2, LRRC56 \n",
      "\t   HRAS, PNPLA2, RNH1, CD151, PTDSS2, POLR2L, ANO9, SIGIRR, TSPAN4, PKP3 \n",
      "\t   CHID1, B4GALNT4, AP2A2, IFITM3, MUC6, MUC5B, IFITM1, TOLLIP, IFITM2, MOB2 \n",
      "PC_ 2 \n",
      "Positive:  CLSTN3, PEX5, CD163L1, C1RL-AS1, CD163, C1RL, NANOG, C1R, SLC2A3, C1S \n",
      "\t   LPCAT3, FOXJ2, EMG1, PHB2, PTPN6, C3AR1, C12orf57, ATN1, ENO2, LRRC23 \n",
      "\t   DSTNP2, NECAP1, RPL13P5, SPSB2, CLEC4A, TPI1, FAM66C, USP5, CDCA3, CLEC4E \n",
      "Negative:  PAICS, SRP72, HOPX, REST, NOA1, POLR2B, IGFBP7, STAP1, SLC39A4, CPSF1 \n",
      "\t   ADCK5, TONSL, SLC52A2, VPS28, CYHR1, KIFC2, DGAT1, HSF1, PPP1R16A, BOP1 \n",
      "\t   UBA6, MFSD3, MROH1, MAF1, RECQL4, SHARPIN, CYC1, LRRC14, GPAA1, SULT1E1 \n",
      "PC_ 3 \n",
      "Positive:  DLL3, EID2B, EID2, DYRK1B, FBL, FCGBP, PSMC4, ZNF546, ZNF780B, ZNF780A \n",
      "\t   AXL, CYP2S1, HNRNPUL1, AKT2, EGLN2, C19orf47, RAB4B, PLD3, PRX, MIA \n",
      "\t   SERTAD1, SERTAD3, BLVRB, SNRPA, SHKBP1, LTBP4, NUMBL, C19orf54, ITPKC, BCAM \n",
      "Negative:  CARD16, CASP1, CASP4, GRIA4, AASDHPPT, MSANTD4, KBTBD3, GUCY1A2, PDGFD, DYNC2H1 \n",
      "\t   CWF19L2, DCUN1D5, ALKBH8, MMP10, ELMOD1, TMEM123, SLC35F2, BIRC2, CUL5, BIRC3 \n",
      "\t   ACAT1, YAP1, TRPC6, NPAT, ARHGAP42, JRKL, ATM, CCDC82, MAML2, MTMR2 \n",
      "PC_ 4 \n",
      "Positive:  DYRK1B, EID2, FBL, EID2B, FCGBP, PSMC4, DLL3, ZNF546, ZNF780B, ZNF780A \n",
      "\t   AKT2, C19orf47, PLD3, PRX, SERTAD1, SERTAD3, BLVRB, SHKBP1, LTBP4, NUMBL \n",
      "\t   ITPKC, C19orf54, SNRPA, BRMS1, RIN1, SLC29A2, CD248, YIF1A, RAB1B, KLC2 \n",
      "Negative:  NIPBL, NUP155, SLC1A3, WDR70, SKP2, GDNF, LMBRD2, IL7R, SPEF2, EGFLAM \n",
      "\t   PRLR, DNAJC21, LIFR, BRIX1, OSMR, RAD1, RAI14, C1QTNF3, RICTOR, AMACR \n",
      "\t   ADAMTS12, DAB2, NPR3, SUB1, PTGER4, ZFR, TTC33, MTMR12, GOLPH3, PDZD2 \n",
      "PC_ 5 \n",
      "Positive:  NUDT4, UBE2N, EEA1, MRPL42, BTG1, DCN, SOCS2, LUM, CRADD, KERA \n",
      "\t   PLXNC1, ATP2B1, TMCC3, POC1B-GALNT4, NDUFA12, GALNT4, POC1B, NR2C1, DUSP6, FGD6 \n",
      "\t   KITLG, VEZT, TMTC3, METAP2, NTN4, CEP290, SNRPF, C12orf29, LTA4H, ELK3 \n",
      "Negative:  ZNF485, ZNF32, CXCL12, RASSF4, ZNF239, AGAP4, SYT15, BMS1P1, ZNF22, ZFAND4 \n",
      "\t   ALOX5, AGAP9, HNRNPF, RASGEF1A, BMS1P2, GDF10, CSGALNACT2, RET, MAPK8, BMS1 \n",
      "\t   ARHGAP22, ZNF33B, WDFY4, ZNF37BP, HSD17B7P2, LOC441666, ZNF37A, LINC00839, ZNF33A, ZNF25 \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2023-07-24 17:12:47] define_signif_tumor_subclusters(), tumor: malignant_SyS10\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  ZNF526, DEDD2, CIC, ERF, POU2F2, PAFAH1B3, ZNF574, PRR19, RABAC1, ARHGEF1 \n",
      "\t   ZNF576, ZNF428, MEGF8, IRGQ, PLAUR, CD79A, PINLYP, RPS19, LIPE, SMG9 \n",
      "\t   XRCC1, BCKDHA, ETHE1, CEACAM1, LYPD3, PHLDB3, KCNN4, EXOSC5, ZNF283, TMEM91 \n",
      "Negative:  FAIM, PIK3CB, CEP70, MRPS22, ESYT3, COPB2, RBP1, CLSTN2, SLC25A36, ZBTB38 \n",
      "\t   MRAS, RASA2, RNF7, ARMC8, ATP1B3, TFDP2, PCOLCE2, DBR1, TRPC1, PLS1 \n",
      "\t   GK5, ATR, XRN1, U2SURP, CHST2, DZIP1L, SLC9A9, PLOD2, NCK1, PLSCR4 \n",
      "PC_ 2 \n",
      "Positive:  CLEC5A, MTRNR2L6, SSBP1, AGK, EPHB6, CUL1, MRPS33, ARHGEF5, TPK1, EPHA1 \n",
      "\t   OR2A7, LOC154761, ZYX, ARHGEF35, EZH2, CASP2, CTAGE4, GSTK1, PDIA4, BRAF \n",
      "\t   NDUFB2, ZNF786, ADCK2, ZNF425, MKRN1, SLC37A3, PARP12, TBXAS1, ZNF398, HIPK2 \n",
      "Negative:  SPAG7, CAMTA2, KIF1C, ENO3, SLC52A1, PFN1, ZFP3, RNF167, ZNF232, SLC25A11 \n",
      "\t   ZNF594, C17orf107, SCIMP, CHRNE, RABEP1, NUP88, MINK1, RPAIN, PLD2, C1QBP \n",
      "\t   ITGAE, CAMKK1, P2RX1, PSMB6, ATP2A3, DHX33, P2RX5, ZZEF1, EMC6, VMO1 \n",
      "PC_ 3 \n",
      "Positive:  SELPLG, TMEM119, FICD, SART3, ISCU, CMKLR1, WSCD2, PRDM4, PWP1, BTBD11 \n",
      "\t   CRY1, RIC8B, POLR3B, TCP11L2, CKAP4, NUAK1, C12orf75, FBXO46, APPL2, ALDH1L2 \n",
      "\t   QPCTL, C12orf45, SLC41A2, SNRPD2, CHST11, EID3, EML2, TXNRD1, NFYB, GPR4 \n",
      "Negative:  EPC1, KIF5B, ARHGAP12, ZEB1, CCDC7, ZEB1-AS1, ITGB1, ZNF438, NRP1, MAP3K8 \n",
      "\t   PARD3, MTPAP, CUL2, SVIL, CREM, BAMBI, CCNY, WAC, WAC-AS1, ZNF791 \n",
      "\t   WDR83, MAN2B1, WDR83OS, ZNF490, ZNF248, DHPS, MPP7, ZNF564, FBXW9, ZNF25 \n",
      "PC_ 4 \n",
      "Positive:  CCDC141, SESTD1, PLEKHA3, TTN-AS1, TTN, CWC22, FKBP7, UBE2E3, PRKRA, ITGA4 \n",
      "\t   OSBPL6, RBM45, CERKL, TTC30A, TTC30B, PDE1A, AGPS, DNAJC10, NFE2L2, FRZB \n",
      "\t   HNRNPA3, NCKAP1, MTX2, DUSP19, HOXD1, NUP35, HOXD3, FSIP2, HOXD4, ZC3H15 \n",
      "Negative:  EML4, PKDCC, COX7A2L, MTA3, SLC8A1, OXER1, THUMPD2, HAAO, ZFP36L2, MAP4K3 \n",
      "\t   THADA, SOS1, PLEKHH2, MORN2, DYNC2LI1, DHX57, LRPPRC, PPM1B, GEMIN6, PREPL \n",
      "\t   CAMKMT, SIX3, SRBD1, PRKCE, EPAS1, RHOQ, PIGF, SRSF7, CRIPT, SOCS5 \n",
      "PC_ 5 \n",
      "Positive:  MDFI, FOXP4, TFEB, TOMM6, USP49, TREM1, MED20, TREM2, ADCY10P1, NFYA \n",
      "\t   IFI27L1, DDX24, CALM1, TMEM251, NRDE2, UBR7, TTC7B, OTUB2, MOAP1, IFI27 \n",
      "\t   RPS6KA5, BTBD7, ITPK1-AS1, GPR68, PSMC1, CCDC88C, TDP1, ITPK1, FOXN3, TC2N \n",
      "Negative:  FBL, DYRK1B, FCGBP, PSMC4, EID2, EID2B, ZNF546, DLL3, ZNF780B, ZNF780A \n",
      "\t   AKT2, C19orf47, PLD3, SARS2, NFKBIB, PRX, SIRT2, RINL, HNRNPL, SERTAD1 \n",
      "\t   ECH1, CAPN12, SERTAD3, ACTN4, EIF3K, MAP4K1, RYR1, BLVRB, RASGRP4, FAM98C \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2023-07-24 17:12:47] define_signif_tumor_subclusters(), tumor: malignant_SyS11\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  ST3GAL6, ARHGAP10, TMEM184C, CPOX, EDNRA, NR3C2, SLC10A7, CLDND1, FILIP1L, DCLK2 \n",
      "\t   TBC1D23, LSM6, LRBA, ZNF827, RPS3A, NIT2, LNP1, SH3D19, MMAA, TMEM45A \n",
      "\t   FAM160A1, TFG, SMAD1, FBXW7, ABI3BP, TMEM154, ARFIP1, OTUD4, ARL13B, TRIM2 \n",
      "Negative:  SGTA, SLC39A3, THOP1, GNG7, GADD45B, ZNF554, LMNB2, ZNF555, TIMM13, ZNF57 \n",
      "\t   SPPL2B, ZNF77, LSM7, TLE2, GNA11, GNA15, OAZ1, S1PR4, NCLN, SF3A2 \n",
      "\t   NFIC, PLEKHJ1, DOHH, DOT1L, FZR1, AP3D1, MFSD12, MOB3A, HMG20B, MKNK2 \n",
      "PC_ 2 \n",
      "Positive:  GNG2, NID2, FRMD6, PTGDR, FRMD6-AS1, PTGER2, TMX1, TXNDC16, GPR137C, PYGL \n",
      "\t   PSMC6, NIN, STYX, SAV1, GNPNAT1, ATL1, FERMT2, MAP4K5, DDHD1, L2HGDH \n",
      "\t   BMP4, SOS2, CDKN3, ARF6, NEMF, KLHDC2, KLHDC1, POLE2, DNAAF2, MGAT2 \n",
      "Negative:  PCDHB13, PCDHB12, PCDHB14, PCDHB11, PCDHB19P, PCDHB10, PCDHB15, PCDHB9, TAF7, PCDHGA2 \n",
      "\t   PCDHB16, PCDHGA3, PCDHGB1, PCDHB7, PCDHGA4, PCDHB6, PCDHGB2, PCDHGA5, PCDHB5, PCDHGB3 \n",
      "\t   PCDHGA6, PCDHB4, PCDHGA7, PCDHGA8, PCDHGB5, PCDHB3, PCDHGA9, PCDHGB6, PCDHB2, PCDHGA10 \n",
      "PC_ 3 \n",
      "Positive:  MORN2, SOS1, MAP4K3, DHX57, THUMPD2, SLC8A1, GEMIN6, PKDCC, SRSF7, GALM \n",
      "\t   EML4, ATL2, CYP1B1, COX7A2L, RMDN2, MTA3, CDC42EP3, OXER1, QPCT, HAAO \n",
      "\t   PRKD3, ZFP36L2, NDUFAF7, THADA, CEBPZ, PLEKHH2, EIF2AK2, DYNC2LI1, GPATCH11, LRPPRC \n",
      "Negative:  HMGB2, HAND2, SCRG1, GALNT7, FBXO8, CBR4, AADAT, SH3RF1, NEK1, PALLD \n",
      "\t   CEP44, HPGD, DDX60L, SPCS3, AGA, TENM3, DDX60, DCTD, SPOCK3, WWC2 \n",
      "\t   CDKN2AIP, TLL1, CPE, MSMO1, KLHL2, TMEM192, SNHG6, COPS5, MCMDC2, FAM218A \n",
      "PC_ 4 \n",
      "Positive:  LSM6, TMEM184C, EDNRA, SLC10A7, ZNF827, ARHGAP10, MMAA, SMAD1, OTUD4, NR3C2 \n",
      "\t   ABCE1, DCLK2, ANAPC10, HHIP, LRBA, HHIP-AS1, RPS3A, GUSBP5, SH3D19, SMARCA5 \n",
      "\t   FAM160A1, GAB1, FBXW7, USP38, TMEM154, INPP4B, ARFIP1, IL15, TRIM2, ZNF330 \n",
      "Negative:  NTNG1, PRMT6, VAV3, AMY2B, SLC25A24, RNPC3, FAM102B, HENMT1, COL11A1, PRPF38B \n",
      "\t   STXBP3, S1PR1, GPSM2, DPH5, SLC30A7, EXTL2, VCAM1, CDC14A, RTCA, ATXN7L2 \n",
      "\t   DBT, CYB561D1, TRMT13, GNAI3, SASS6, GNAT2, AMPD2, SLC35A3, GSTM4, AGL \n",
      "PC_ 5 \n",
      "Positive:  NPHP4, KCNAB2, AJAP1, CHD5, RPL22, C1orf174, RNF207, DFFB, SNHG6, VCPIP1 \n",
      "\t   ICMT, MCMDC2, COPS5, C8orf44, SGK3, MYBL1, CSPP1, ADHFE1, ARFGEF1, CEP104 \n",
      "\t   GPR153, RRS1, PREX2, ACOT7, PDE7A, LRRC47, SULF1, MTFR1, HES2, NCOA2 \n",
      "Negative:  WDR83, MAN2B1, WDR83OS, ZNF490, ZNF564, ZNF709, ZNF791, ZNF443, DHPS, ZNF799 \n",
      "\t   FBXW9, ZNF442, TNPO2, HOOK2, ZNF563, JUNB, ZNF44, PRDX2, ZNF136, RNASEH2A \n",
      "\t   DNASE2, ZNF625, GCDH, FARSA, CALR, ZNF20, RAD23A, GADD45GIP1, ZNF844, DAND5 \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2023-07-24 17:12:51] define_signif_tumor_subclusters(), tumor: malignant_SyS11.met\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  GABRD, GNB1, PRKCZ, WRAP73, TP73-AS1, NADK, TPRG1L, PANK4, SKI, PLCH2 \n",
      "\t   LRRC47, PRDM16, ARHGEF16, MORN1, MEGF6, TNFRSF14, RER1, CDK11A, PEX10, SLC35E2B \n",
      "\t   CEP104, CDK11B, DFFB, MMP23B, MIB2, C1orf174, SSU72, ATAD3A, AJAP1, ATAD3B \n",
      "Negative:  CPOX, ST3GAL6, CLDND1, DCBLD2, COL8A1, CRYBG3, ARL6, CMSS1, NSUN3, FILIP1L \n",
      "\t   ARL13B, TBC1D23, NIT2, LNP1, PROS1, TMEM45A, TFG, EPHA3, ABI3BP, SENP7 \n",
      "\t   C3orf38, TRMT10C, PCNP, ZNF654, ZBTB11, CGGBP1, RPL24, CHMP2B, CEP97, GBE1 \n",
      "PC_ 2 \n",
      "Positive:  CUTC, DNMBP, COX15, ERLIN1, ENTPD7, SLC25A28, CHUK, GOT1, HPS1, LOXL4 \n",
      "\t   R3HCC1L, CWF19L1, CRTAC1, GOLGA7B, BLOC1S2, ZFYVE27, SCD, MARVELD1, SEC31B, AVPI1 \n",
      "\t   NDUFB8, PI4K2A, HIF1AN, MORN4, SEMA4G, MMS19, ZDHHC16, MRPL43, EXOSC1, LZTS2 \n",
      "Negative:  TNFRSF14, PANK4, PLCH2, PEX10, PRDM16, RER1, MORN1, ARHGEF16, SKI, PRKCZ \n",
      "\t   MEGF6, GABRD, TPRG1L, GNB1, WRAP73, NADK, TP73-AS1, CDK11A, LRRC47, SLC35E2B \n",
      "\t   CEP104, CDK11B, DFFB, MMP23B, C1orf174, MIB2, AJAP1, SSU72, NPHP4, ATAD3A \n",
      "PC_ 3 \n",
      "Positive:  ATG3, BTLA, SLC35A5, CD200, CCDC80, CD200R1, C3orf52, GTPBP8, ABHD10, BOC \n",
      "\t   SPICE1, SIDT1, NAA50, BBX, ATP6V1A, CBLB, GRAMD1C, ALCAM, TIGIT, NFKBIZ \n",
      "\t   NXPE3, ZBTB20, LSAMP, CEP97, RPL24, B4GALT4, ARHGAP31, ZBTB11, PCNP, TMEM39A \n",
      "Negative:  CRADD, SOCS2, MRPL42, PLXNC1, UBE2N, TMCC3, NUDT4, NDUFA12, NR2C1, EEA1 \n",
      "\t   FGD6, BTG1, VEZT, DCN, METAP2, LUM, NTN4, KERA, SNRPF, ATP2B1 \n",
      "\t   LTA4H, POC1B-GALNT4, ELK3, GALNT4, CDK17, POC1B, NEDD1, DUSP6, SLC9A7P1, KITLG \n",
      "PC_ 4 \n",
      "Positive:  GINM1, PPIL4, ZC3H12D, TAB2, AKAP12, UST, ZBTB2, SASH1, RMND1, CCDC170 \n",
      "\t   SAMD5, ESR1, STXBP5, SYNE1, RAB32, SHPRH, MYCT1, FBXO30, FBXO5, LOC100507557 \n",
      "\t   EPM2A, UTRN, MTRF1L, STX11, SF3B5, IPCEF1, HYMAI, PLAGL1, EIF1B, EIF1B-AS1 \n",
      "Negative:  ROBO2, ROBO1, GBE1, ZNF717, CHMP2B, CGGBP1, ZNF654, FAM86DP, C3orf38, EPHA3 \n",
      "\t   PDZRN3, PROS1, EBLN2, ARL13B, PPP4R2, NSUN3, ARL6, SHQ1, CRYBG3, RYBP \n",
      "\t   CLDND1, CPOX, PROK2, ST3GAL6, EIF4E3, DCBLD2, COL8A1, FOXP1, CMSS1, MITF \n",
      "PC_ 5 \n",
      "Positive:  TMEM184C, ARHGAP10, EDNRA, NR3C2, SLC10A7, DCLK2, LSM6, LRBA, RPS3A, ZNF827 \n",
      "\t   SH3D19, MMAA, FAM160A1, SMAD1, FBXW7, TMEM154, OTUD4, ARFIP1, ABCE1, TRIM2 \n",
      "\t   TLR2, ANAPC10, SFRP2, HHIP, DCHS2, PLRG1, HHIP-AS1, FGG, GUSBP5, LRAT \n",
      "Negative:  ITGA8, PTER, RSU1, FAM171A1, CUBN, NMT2, TRDMT1, VIM, RPP38, STAM \n",
      "\t   DCLRE1C, MRC1, SUV39H2, CACNB2, HSPA14, FAM107B, NSUN6, ARL5B, FRMD4A, PLXDC2 \n",
      "\t   PRPF18, NEBL, SEPHS1, SKIDA1, MLLT10, DNAJC1, PHYH, COMMD3-BMI1, COMMD3, BMI1 \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2023-07-24 17:12:56] define_signif_tumor_subclusters(), tumor: malignant_SyS12\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  ANO9, SIGIRR, HRAS, RNH1, PTDSS2, LRRC56, PKP3, RASSF7, B4GALNT4, PHRF1 \n",
      "\t   IRF7, IFITM3, DEAF1, IFITM1, TMEM80, EPS8L2, IFITM2, TALDO1, IFITM5, SLC25A22 \n",
      "\t   PSMD13, RPLP2, SIRT3, PNPLA2, RIC8A, BET1L, CD151, POLR2L, TSPAN4, CHID1 \n",
      "Negative:  ZNF552, ZNF586, FKBP1AP1, ZNF776, ZNF587B, ZNF587, ZNF671, ZNF814, ZNF417, ZNF418 \n",
      "\t   ZNF256, ZNF154, IGDCC4, PARP16, CLPX, ZNF606, PDCD7, RASL12, DPP8, MTFMT \n",
      "\t   ZNF135, SLC24A1, ZNF551, SPG21, DENND4A, ZSCAN18, ANKDD1A, ZSCAN4, PLEKHO2, ZNF329 \n",
      "PC_ 2 \n",
      "Positive:  PCDHGA2, PCDHGA3, PCDHGB1, TAF7, PCDHGA4, PCDHGB2, PCDHGA5, PCDHGB3, PCDHB15, PCDHGA6 \n",
      "\t   PCDHB19P, PCDHGA7, PCDHGA8, PCDHB14, PCDHGB5, PCDHB13, PCDHGA9, PCDHB12, PCDHGB6, PCDHB11 \n",
      "\t   PCDHGA10, PCDHB10, PCDHGB7, PCDHB9, PCDHGA11, PCDHB16, PCDHGA12, PCDHB7, PCDHGC3, PCDHB6 \n",
      "Negative:  SMIM7, MED26, SLC35E1, CHERP, SIN3B, C19orf44, OCEL1, ABHD8, CPAMD8, NR2F6 \n",
      "\t   HAUS8, USE1, BABAM1, EPS15L1, MYO9B, KLF2, BST2, AP1M1, MVB12A, FAM32A \n",
      "\t   HSH2D, SLC27A1, PGLS, RAB8A, MAP1S, TPM4, RASAL3, FCHO1, WIZ, JAK3 \n",
      "PC_ 3 \n",
      "Positive:  WDHD1, SOCS4, GCH1, MAPK1IP1L, SAMD4A, CGRRF1, LGALS3, GMFB, CDKN3, DLGAP5 \n",
      "\t   BMP4, DDHD1, FBXO34, FERMT2, GNPNAT1, ATG14, STYX, KTN1, PSMC6, GPR137C \n",
      "\t   PELI2, TXNDC16, OTX2, AP5M1, EXOC5, NAA30, PSMA3, ACTR10, PTGER2, ARID4A \n",
      "Negative:  MEGF6, TPRG1L, ARHGEF16, WRAP73, PRDM16, TP73-AS1, TNFRSF14, LRRC47, PANK4, CEP104 \n",
      "\t   PLCH2, DFFB, PEX10, C1orf174, RER1, AJAP1, MORN1, NPHP4, SKI, KCNAB2 \n",
      "\t   PRKCZ, GABRD, CHD5, GNB1, NADK, RPL22, CDK11A, SLC35E2B, RNF207, ICMT \n",
      "PC_ 4 \n",
      "Positive:  PCDHB9, PCDHB10, PCDHB16, PCDHB11, PCDHB7, PCDHB12, PCDHB6, PCDHB13, PCDHB14, PCDHB5 \n",
      "\t   PCDHB19P, PCDHB15, PCDHB4, TAF7, PCDHGA2, PCDHB3, PCDHGA3, PCDHB2, PCDHGB1, PCDHAC2 \n",
      "\t   PCDHGA4, PCDHA13, PCDHGB2, PCDHGA5, PCDHA12, PCDHGB3, PCDHA11, PCDHGA6, PCDHGA7, PCDHA10 \n",
      "Negative:  ELAC1, SMAD4, MEX3C, ME2, MBD2, POLI, MRO, C18orf54, SKA1, RAB27B \n",
      "\t   CXXC1, TCF4, TXNL1, WDR7, FECH, LOC100505549, ATP8B1, NEDD4L, MALT1, ZNF532 \n",
      "\t   SEC11C, LMAN1, CCBE1, PMAIP1, PIGN, ZCCHC2, STXBP3, GPSM2, PRPF38B, HENMT1 \n",
      "PC_ 5 \n",
      "Positive:  HOMER2, WHAMM, GOLGA6L10, C15orf40, RPS17, BTBD1, TM6SF1, ADAMTSL3, DNM1P41, GOLGA6L9 \n",
      "\t   LOC440300, MEX3B, GOLGA6L4, STARD5, UBE2Q2P1, IL16, ZSCAN2, ARNT2, SCAND2P, FAH \n",
      "\t   WDR73, ZFAND6, SEC11A, BCL2A1, ZNF592, ALPK3, ST20, PDE8A, MTHFS, AKAP13 \n",
      "Negative:  SKI, MORN1, RER1, PRKCZ, PEX10, GABRD, PLCH2, PANK4, TNFRSF14, GNB1 \n",
      "\t   PRDM16, NADK, ARHGEF16, CDK11A, MEGF6, SLC35E2B, TPRG1L, WRAP73, TP73-AS1, LRRC47 \n",
      "\t   CEP104, DFFB, C1orf174, MRPL20, AJAP1, CCNL2, AURKAIP1, NPHP4, MXRA8, KCNAB2 \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2023-07-24 17:13:01] define_signif_tumor_subclusters(), tumor: malignant_SyS12pt\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  RSL24D1, RAB27A, PIGB, CCPG1, NEDD4, RFX7, TEX9, MNS1, ZNF280D, LOC145783 \n",
      "\t   TCF12, CGNL1, MYZAP, SLC27A2, POLR2M, ALDH1A2, AQP9, ADAM10, ATP8B4, HSP90AB4P \n",
      "\t   SLTM, RNF111, CCNB2, DTWD1, MYO1E, GTF2A2, BNIP2, FOXB1, RORA, ANXA2 \n",
      "Negative:  MRPL23, H19, TNNT3, LSP1, IGF2, SYT8, CTSD, INS-IGF2, IFITM10, DUSP8 \n",
      "\t   MOB2, TOLLIP, MUC5B, C11orf21, MUC6, AP2A2, CHID1, TSPAN32, TSPAN4, POLR2L \n",
      "\t   CD151, CD81, PNPLA2, RPLP2, TSSC4, SLC25A22, TALDO1, KCNQ1, EPS8L2, TMEM80 \n",
      "PC_ 2 \n",
      "Positive:  BCL3, PVR, ZNF180, CBLC, ZNF229, BCAM, TOMM40, APOE, ZSWIM4, LOC284454 \n",
      "\t   CC2D1A, C19orf53, MRI1, PODNL1, CCDC130, RFX1, APOC1, IL27RA, TRMT1, IER2 \n",
      "\t   STX10, NFIX, DAND5, PRKACA, GADD45GIP1, CLPTM1, RAD23A, ASF1B, CALR, FARSA \n",
      "Negative:  NXPE3, CEP97, TFG, TMEM45A, ABI3BP, NFKBIZ, LNP1, SENP7, RPL24, NIT2 \n",
      "\t   TRMT10C, PCNP, ZBTB11, TBC1D23, FILIP1L, ALCAM, CMSS1, COL8A1, CBLB, DCBLD2 \n",
      "\t   BBX, ST3GAL6, CD47, CPOX, IFT57, CLDND1, CRYBG3, DZIP3, ARL6, NSUN3 \n",
      "PC_ 3 \n",
      "Positive:  PPM1N, GPR4, VASP, OPA3, FOSB, EML2, ERCC1, CD3EAP, SNRPD2, PPP1R13L \n",
      "\t   QPCTL, ERCC2, FBXO46, KLC3, SIX5, MARK4, DMPK, BLOC1S3, DMWD, SYMPK \n",
      "\t   TRAPPC6A, ARL6, CRYBG3, NSUN3, CLDND1, HIF3A, C3orf38, EPHA3, ARL13B, ZNF654 \n",
      "Negative:  UBE2N, NUDT4, MRPL42, EEA1, SOCS2, BTG1, DCN, CRADD, LUM, KERA \n",
      "\t   ATP2B1, POC1B-GALNT4, GALNT4, PLXNC1, POC1B, DUSP6, TMCC3, KITLG, NDUFA12, TMTC3 \n",
      "\t   NR2C1, CEP290, FGD6, C12orf29, VEZT, MKRN9P, RASSF9, METAP2, LRRIQ1, NTN4 \n",
      "PC_ 4 \n",
      "Positive:  PREX2, ARFGEF1, SULF1, CSPP1, NCOA2, COPS5, SNHG6, TRAM1, MCMDC2, SGK3 \n",
      "\t   LACTB2, C8orf44, XKR9, VCPIP1, EYA1, MYBL1, MSC, ADHFE1, TERF1, RRS1 \n",
      "\t   SBSPON, RPL7, PDE7A, RDH10, MTFR1, STAU2, ARMC1, UBE2W, TMEM70, CYP7B1 \n",
      "Negative:  AMPD2, GNAT2, GNAI3, GSTM4, CYB561D1, GSTM2, ATXN7L2, GSTM1, GSTM5, GSTM3 \n",
      "\t   GPSM2, CSF1, STXBP3, PRPF38B, HENMT1, SULT1A3, SLX1A, CD2BP2, SLX1B, FAM102B \n",
      "\t   AHCYL1, BOLA2, SLC25A24, TBC1D10B, VAV3, ZNF771, NTNG1, STRIP1, DCTPP1, DMPK \n",
      "PC_ 5 \n",
      "Positive:  RAB8A, TPM4, HSH2D, RASAL3, WIZ, FAM32A, AKAP8L, AKAP8, AP1M1, BRD4 \n",
      "\t   NOTCH3, ILVBL, KLF2, ZNF333, NDUFB7, TECR, EPS15L1, DNAJB1, C19orf44, GIPC1 \n",
      "\t   CHERP, PKN1, SLC35E1, MED26, DDX39A, SMIM7, LOC100507373, SIN3B, ASF1B, CPAMD8 \n",
      "Negative:  KIRREL3, ST3GAL4, ETS1, DCPS, FLI1, KCNJ5, TIRAP, ARHGAP32, BARX2, FOXRED1 \n",
      "\t   NFRKB, PRDM10, APLP2, FAM118B, ST14, ZBTB44, ADAMTS15, NTM, IGSF9B, SNX19 \n",
      "\t   JAM3, NCAPD3, RPUSD4, VPS26B, THYN1, ACAD8, GLB1L2, B3GAT1, CDON, DDX25 \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2023-07-24 17:13:03] define_signif_tumor_subclusters(), tumor: malignant_SyS13\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  PCDHGB1, PCDHGA4, PCDHGA3, PCDHGB2, PCDHGA2, PCDHGA5, TAF7, PCDHB15, PCDHGB3, PCDHB19P \n",
      "\t   PCDHB14, PCDHGA6, PCDHB13, PCDHGA7, PCDHB12, PCDHGA8, PCDHGB5, PCDHGA9, PCDHGB6, PCDHB11 \n",
      "\t   PCDHGA10, PCDHB10, PCDHGB7, PCDHGA11, PCDHB9, PCDHGA12, PCDHB16, PCDHGC3, PCDHGC4, PCDHGC5 \n",
      "Negative:  MDH1, UGP2, WDPCP, VPS54, PELI1, LOC100132215, LGALSL, EHBP1, TMEM17, AFTPH \n",
      "\t   B3GNT2, SERTAD2, SLC1A4, COMMD1, CEP68, CCT4, RAB1A, FAM161A, PNO1, ACTR2 \n",
      "\t   WDR92, ETAA1, C1D, SPRED2, MEIS1, XPO1, PPP3R1, CNRIP1, USP34, PLEK \n",
      "PC_ 2 \n",
      "Positive:  CFL1, MUS81, EFEMP2, SNX32, CTSW, OVOL1, FIBP, AP5B1, CCDC85B, FOSL1 \n",
      "\t   RNASEH2C, C11orf68, KAT5, DRAP1, RELA, SART1, SIPA1, MAP3K11, EIF1AD, EHBP1L1 \n",
      "\t   BANF1, FAM89B, SF3B2, LTBP3, AP1M1, KLF2, HSH2D, FAM32A, SCYL1, PACS1 \n",
      "Negative:  RNF14, GNPDA1, PCDH12, NDFIP1, SPRY4, ARHGAP26, PCDH1, NR3C1, FCHSD1, RELL2 \n",
      "\t   YIPF5, HDAC3, PRELID2, DIAPH1, SH3RF2, PCDHGC5, RBM27, TCERG1, PCDHGC4, PPP2R2B \n",
      "\t   PCDHGC3, DPYSL3, JAKMIP2, PCDHGA12, FBXO38, ADRB2, PCDHGA11, ABLIM3, PCDHGB7, AFAP1L1 \n",
      "PC_ 3 \n",
      "Positive:  SLC24A4, CPSF2, NDUFB1, RIN3, ATXN3, LGMN, TRIP11, FBLN5, GOLGA5, TC2N \n",
      "\t   ITPK1, CCDC88C, ITPK1-AS1, GPR68, RPS6KA5, MOAP1, TTC7B, TMEM251, CALM1, UBR7 \n",
      "\t   BTBD7, OTUB2, HEXIM2, FMNL1, PPFIBP2, NRIP3, HEXIM1, TMEM9B, RPL27A, AKIP1 \n",
      "Negative:  SYNGR3, GFER, NPW, TBL3, NTHL1, RPS2, TSC2, PKD1, NDUFB10, RAB26 \n",
      "\t   TRAF7, MLST8, MSRB1, E4F1, FAHD1, ECI1, HAGH, RNPS1, NUBP2, ABCA3 \n",
      "\t   SPSB3, CCNF, MRPS34, TBC1D24, NME3, MAPK8IP3, ATP6V0C, TMEM204, AMDHD2, IFT140 \n",
      "PC_ 4 \n",
      "Positive:  GBP5, GBP4, LRRC8B, GBP2, PKN2, LRRC8D, LRRC8C, GTF2B, LMO4, GBP1 \n",
      "\t   RBMXL1, GBP3, ZNF326, HS2ST1, ZNF644, SH3GLB1, CDC7, ODF2L, COL24A1, ZNHIT6 \n",
      "\t   DDAH1, BCL10, ZNF10, ZNF84, ZNF26, ZNF605, ZNF140, CHFR, GOLGA3, ANKLE2 \n",
      "Negative:  ADAM19, EBF1, CYFIP2, THG1L, CLINT1, RNF145, ITK, UBLCP1, MED7, HAVCR2 \n",
      "\t   TTC1, SGCD, PWWP2A, MRPL22, GEMIN5, C1QTNF2, CNOT8, SLU7, LARP1, PTTG1 \n",
      "\t   HAND1, GABRB2, SAP30L, CCNG1, GALNT10, NUDCD2, MFAP3, HMMR, FAM114A2, GRIA1 \n",
      "PC_ 5 \n",
      "Positive:  WRAP73, MEGF6, TPRG1L, ARHGEF16, PRDM16, TNFRSF14, TP73-AS1, PANK4, PLCH2, LRRC47 \n",
      "\t   PEX10, CEP104, RER1, MORN1, DFFB, SKI, C1orf174, AJAP1, PRKCZ, NPHP4 \n",
      "\t   GABRD, KCNAB2, GNB1, CHD5, NADK, RPL22, CDK11A, RNF207, ICMT, SLC35E2B \n",
      "Negative:  KCNIP1, RANBP17, NPM1, FGF18, HMMR, NUDCD2, CCNG1, FBXW11, STK10, GABRB2 \n",
      "\t   UBTD2, SH3PXD2B, NEURL1B, PTTG1, DUSP1, ERGIC1, SLU7, RPL26L1, ATP6V0E1, C1QTNF2 \n",
      "\t   CREBRF, BNIP1, PWWP2A, STC2, TTC1, UBLCP1, BOD1, RNF145, EBF1, CLINT1 \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2023-07-24 17:13:08] define_signif_tumor_subclusters(), tumor: malignant_SyS14\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  FBLN1, ATXN10, MIRLET7BHG, FAM118A, PPARA, CDPF1, KIAA0930, TTC38, PHF21B, NUP50 \n",
      "\t   ARHGAP8, GTSE1, TRMU, PRR5, CELSR1, PARVG, GRAMD4, CERK, TBC1D22A, C22orf34 \n",
      "\t   SAMM50, BRD1, SULT4A1, ZBED4, ALG12, EFCAB6, CRELD2, PIM3, SCUBE1, MLC1 \n",
      "Negative:  RYK, AMOTL2, ANAPC13, CEP63, EPHB1, SLCO2A1, PPP2R3A, MSL2, RAB6B, SRPRB \n",
      "\t   PCCB, TOPBP1, CDV3, STAG1, NPHP3, SLC35G2, UBA5, NPHP3-ACAD11, NCK1, ACAD11 \n",
      "\t   DNAJC13, DZIP1L, MRPL3, NUDT16, DBR1, NUDT16P1, ARMC8, NEK11, MRAS, ASTE1 \n",
      "PC_ 2 \n",
      "Positive:  CBX7, PDGFB, APOBEC3H, APOBEC3G, RPL3, APOBEC3F, SYNGR1, APOBEC3D, TAB1, APOBEC3C \n",
      "\t   APOBEC3B, ATF4, APOBEC3A, RPS19BP1, CBX6, GRAP2, NPTXR, DNAL4, SUN2, TNRC6B \n",
      "\t   GTPBP1, JOSD1, TOMM22, CBY1, ADSL, FAM227A, DDX17, KDELR3, SGSM3, CSNK1E \n",
      "Negative:  RPS29, LRR1, RPL36AL, MGAT2, DNAAF2, POLE2, KLHDC1, KLHDC2, NEMF, RNMT \n",
      "\t   FAM210A, LDLRAD4, ARF6, CEP192, SEH1L, SOS2, RIOK3, PTPN2, L2HGDH, CEP76 \n",
      "\t   PSMG2, MAP4K5, ATL1, SAV1, SPIRE1, AFG3L2, TUBB6, MPPE1, PCSK1, CHMP1B \n",
      "PC_ 3 \n",
      "Positive:  SMIM7, SIN3B, CPAMD8, MED26, HAUS8, MYO9B, USE1, SLC35E1, OCEL1, CHERP \n",
      "\t   NR2F6, BABAM1, C19orf44, ABHD8, EPS15L1, KLF2, AP1M1, FAM32A, HSH2D, BST2 \n",
      "\t   RAB8A, TPM4, MVB12A, RASAL3, SLC27A1, WIZ, PGLS, AKAP8L, MAP1S, AKAP8 \n",
      "Negative:  ACADSB, BUB3, CPXM2, CHST15, OAT, DLG2, TMEM126B, LHPP, TMEM126A, CREBZF \n",
      "\t   POGLUT1, TMEM39A, TIMMDC1, CD80, MTPAP, FAM53B, ARHGAP31, ADPRH, CCDC89, ATP6V1A \n",
      "\t   GRAMD1C, NAA50, SPICE1, B4GALT4, COX17, SIDT1, TIGIT, ZBTB20, LSAMP, BOC \n",
      "PC_ 4 \n",
      "Positive:  CHTF18, SOX8, LMF1, CACNA1H, RPUSD1, TPSB2, MSLN, HAGHL, TPSAB1, WDR24 \n",
      "\t   JMJD8, TPSD1, STUB1, RHOT2, WDR90, UBE2I, RAB40C, BAIAP3, TSR3, FBXW9 \n",
      "\t   DHPS, HOOK2, WDR83OS, JUNB, PRDX2, GNPTG, WDR83, AXIN1, UNKL, MAN2B1 \n",
      "Negative:  TBL1XR1, ZMAT3, PIK3CA, NAALADL2, NLGN1, KCNMB3, ECT2, ZNF639, NCEH1, MFN1 \n",
      "\t   TNFSF10, GNB4, FNDC3B, ACTL6A, TMEM212, MRPL47, NDUFB5, PLD1, USP13, TNIK \n",
      "\t   TTC14, EIF5A2, FXR1, DNAJC19, RPL22L1, ATP11B, CLDN11, DCUN1D1, MCCC1, SKIL \n",
      "PC_ 5 \n",
      "Positive:  PCDHGB6, PCDHGA10, PCDHGA9, FCHSD1, PCDHGB5, PCDHGB7, HDAC3, RELL2, PCDHGA11, DIAPH1 \n",
      "\t   PCDH1, PCDHGC4, PCDHGA12, PCDHGC3, PCDHGC5, PCDHGA8, PCDHGA7, PCDH12, PCDHGA6, RNF14 \n",
      "\t   PCDHGB3, PCDHGA5, GNPDA1, PCDHGB2, PCDHGA4, NDFIP1, PCDHGB1, PCDHGA3, SPRY4, PCDHGA2 \n",
      "Negative:  PLXNC1, CRADD, SOCS2, TMCC3, MRPL42, NDUFA12, NR2C1, UBE2N, FGD6, VEZT \n",
      "\t   METAP2, NTN4, NUDT4, SNRPF, EEA1, LTA4H, ELK3, BTG1, CDK17, NEDD1 \n",
      "\t   DCN, SLC9A7P1, LUM, TMPO, SLC25A3, KERA, IKBIP, ATP2B1, POC1B-GALNT4, APAF1 \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2023-07-24 17:13:13] define_signif_tumor_subclusters(), tumor: malignant_SyS16\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  LAT2, RFC2, EIF4H, LIMK1, ELN, CLDN4, CLIP2, CLDN3, GTF2IRD1, ABHD11 \n",
      "\t   STX1A, GTF2I, DNAJC30, NCF1, TBL2, BCL7B, GTF2IRD2, BAZ1B, STAG3L2, NSUN5 \n",
      "\t   NCF1B, GTF2IP1, GTF2IRD2B, PMS2P5, STAG3L3, STAG3L1, NSUN5P2, POM121, SPDYE7P, NCF1C \n",
      "Negative:  NFKBIZ, DZIP3, TRAT1, ALCAM, CBLB, CD96, IFT57, CD47, NXPE3, ZBED2 \n",
      "\t   BBX, CEP97, PHLDB2, RPL24, ABHD10, C3orf52, ZBTB11, CD200, PCNP, BTLA \n",
      "\t   TRMT10C, ATG3, ARMC8, DBR1, NCK1, DZIP1L, SLC35G2, MRAS, STAG1, SENP7 \n",
      "PC_ 2 \n",
      "Positive:  ZNF501, KIAA1143, ZNF502, KIF15, ZNF35, TMEM42, ZNF197, ZNF660, ZDHHC3, ZKSCAN7 \n",
      "\t   EXOSC7, ZNF445, CDCP1, TCAIM, LARS2, ABHD5, ANO10, LIMD1, SNRK-AS1, LIMD1-AS1 \n",
      "\t   SNRK, SACM1L, KRBOX1, ZNF662, LZTFL1, HIGD1A, FYCO1, ZBTB47, CXCR6, NKTR \n",
      "Negative:  CNTFR, FAM219A, RPP25L, NUDT2, KIF24, DCTN3, UBAP1, SIGMAR1, DCAF12, GALT \n",
      "\t   UBAP2, IL11RA, UBE2R2, DNAJB5, NOL6, C9orf131, AQP3, NFX1, CHMP5, VCP \n",
      "\t   BAG1, B4GALT1, SMU1, FANCG, DNAJA1, APTX, ACTR2, RAB1A, SPRED2, C1D \n",
      "PC_ 3 \n",
      "Positive:  CDKN3, GMFB, BMP4, CGRRF1, SAMD4A, DDHD1, FERMT2, GCH1, WDHD1, GNPNAT1 \n",
      "\t   SOCS4, MAPK1IP1L, LGALS3, DLGAP5, FBXO34, STYX, ATG14, KTN1, PELI2, OTX2 \n",
      "\t   EXOC5, PSMC6, AP5M1, NAA30, ACTR10, PSMA3, ARID4A, TIMM9, STARD4, STARD4-AS1 \n",
      "Negative:  LRRC56, RASSF7, HRAS, PHRF1, RNH1, IRF7, PTDSS2, DEAF1, ANO9, TMEM80 \n",
      "\t   SIGIRR, EPS8L2, TALDO1, PKP3, SLC25A22, RPLP2, B4GALNT4, PNPLA2, CD151, IFITM3 \n",
      "\t   POLR2L, IFITM1, JAK3, FCHO1, RPL18A, MAP1S, FADS2, FADS1, CCDC124, FADS3 \n",
      "PC_ 4 \n",
      "Positive:  FBXO48, APLF, ARHGAP25, PLEK, ANTXR1, CNRIP1, GFPT1, GTF2IP1, PMS2P5, NFU1 \n",
      "\t   PPP3R1, NSUN5P2, AAK1, NCF1B, STAG3L1, POM121, STAG3L3, PNO1, ANXA4, NSUN5 \n",
      "\t   WDR92, GMCL1, SPDYE7P, SNRNP27, C1D, ETAA1, BAZ1B, MEIS1, SBDSP1, MXD1 \n",
      "Negative:  MUC4, TNK2, MUC20, SDHAP2, SDHAP1, APOD, TFRC, PPP1R2, ACAP2, PCYT1A \n",
      "\t   XXYLT1, TCTEX1D2, FAM43A, UBXN7, LSG1, RNF168, WDR53, TMEM44, FBXO45, ATP13A3 \n",
      "\t   CEP19, PIGX, HES1, PAK2, OPA1, SENP5, MB21D2, NCBP2, PIGZ, CCDC50 \n",
      "PC_ 5 \n",
      "Positive:  PRKCH, TMEM30B, HIF1A, HIF1A-AS2, SNAPC1, LINC00643, RHOJ, PPP2R5E, WDR89, DAAM1 \n",
      "\t   SGPP1, DACT1, SYNE2, KIAA0586, MTHFD1, TIMM9, AKAP5, ARID4A, ZBTB25, PSMA3 \n",
      "\t   ZBTB1, HSPA2, ACTR10, PLEKHG3, CHURC1, NAA30, CHURC1-FNTB, AP5M1, RAB15, MFSD3 \n",
      "Negative:  RPL24, ZBTB11, CEP97, PCNP, NXPE3, TRMT10C, SENP7, NFKBIZ, ABI3BP, ALCAM \n",
      "\t   TFG, CBLB, TMEM45A, BBX, LNP1, CD47, NIT2, IFT57, TBC1D23, DZIP3 \n",
      "\t   FILIP1L, TRAT1, CMSS1, CD96, ZBED2, COL8A1, PHLDB2, ABHD10, DCBLD2, C3orf52 \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2023-07-24 17:13:16] define_signif_tumor_subclusters(), tumor: malignant_SyS2\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  CCDC18, DR1, TMED5, FNBP1L, BCAR3, MTF2, RPL5, DNTTIP2, EVI5, GCLM \n",
      "\t   RPAP2, ARHGAP29, GLMN, ABCD3, BTBD8, F3, TGFBR3, SLC44A3, CDC7, CNN3 \n",
      "\t   RAB27A, RSL24D1, PIGB, CCPG1, ALG14, ZNF644, NEDD4, EXOC5, AP5M1, RWDD3 \n",
      "Negative:  B3GALT6, SDF4, SCNN1D, UBE2J2, ACAP3, PUSL1, TNFRSF4, DVL1, MXRA8, TNFRSF18 \n",
      "\t   AURKAIP1, C1orf159, CCNL2, MRPL20, AGRN, ANKRD65, VWA1, ATAD3B, ATAD3A, SSU72 \n",
      "\t   MIB2, MMP23B, CDK11B, SLC35E2B, CDK11A, NADK, GNB1, GABRD, PRKCZ, SKI \n",
      "PC_ 2 \n",
      "Positive:  TRABD, MLC1, TUBGCP6, HDAC10, MAPK12, PLXNB2, DENND6B, RAPGEF5, CDCA7L, SP4 \n",
      "\t   STEAP1B, ITGB8, IL6, RTEL1, MACC1, TOMM7, FAM126A, TWISTNB, KLHL7, GPNMB \n",
      "\t   RTEL1-TNFRSF6B, TWIST1, MALSU1, IGF2BP3, TRA2A, HDAC9, TNFRSF6B, CCDC126, SNX13, ARFRP1 \n",
      "Negative:  EIF3I, TMEM234, IQCC, MTMR9LP, CCDC28B, LCK, TXLNA, KPNA6, HDAC1, MARCKSL1 \n",
      "\t   TMEM39B, TSSK3, KHDRBS1, BSDC1, PTP4A2, ZBTB8B, COL16A1, PEF1, ZBTB8A, TINAGL1 \n",
      "\t   SERINC2, ZBTB8OS, FABP3, RBBP4, ZCCHC17, SYNC, SNRNP40, KIAA1522, PUM1, S100PBP \n",
      "PC_ 3 \n",
      "Positive:  KIF1B, UBE4B, DFFA, RBP7, PEX14, CASZ1, NMNAT1, TARDBP, SRM, EXOSC10 \n",
      "\t   LZIC, MTOR, ANGPTL7, UBIAD1, CTNNBIP1, FBXO2, FBXO44, FBXO6, MAD2L2, DRAXIN \n",
      "\t   AGTRAP, MTHFR, SLC25A33, CLCN6, SPSB1, KIAA2013, H6PD, PLOD1, SLC2A5, MFN2 \n",
      "Negative:  ZNF250, ZNF16, ZNF252P, COMMD5, ZNF7, RPL8, C8orf33, ZNF34, ZNF251, LRRC14 \n",
      "\t   RECQL4, MFSD3, PPP1R16A, RPS2, NDUFB10, TBL3, MSRB1, FAHD1, KIFC2, GFER \n",
      "\t   HAGH, NUBP2, SYNGR3, SPSB3, CYHR1, MRPS34, NPW, TONSL, NME3, NTHL1 \n",
      "PC_ 4 \n",
      "Positive:  SOSTDC1, ANKMY2, AGMO, BZW2, DGKB, ETV1, TSPAN13, SP4, CDCA7L, ITGB8 \n",
      "\t   RAPGEF5, AHR, ARL4A, STEAP1B, MACC1, SNX13, TWISTNB, HDAC9, TWIST1, IL6 \n",
      "\t   SCIN, TOMM7, TMEM106B, FAM126A, KLHL7, THSD7A, GPNMB, MALSU1, PHF14, IGF2BP3 \n",
      "Negative:  RFC2, LAT2, EIF4H, CLIP2, LIMK1, GTF2IRD1, ELN, CLDN4, GTF2I, CLDN3 \n",
      "\t   NCF1, ABHD11, GTF2IRD2, STX1A, DNAJC30, STAG3L2, TBL2, BCL7B, GTF2IRD2B, BAZ1B \n",
      "\t   NCF1C, NSUN5, NCF1B, NSUN5P1, GTF2IP1, POM121C, PMS2P5, STAG3L3, STAG3L1, NSUN5P2 \n",
      "PC_ 5 \n",
      "Positive:  CDC14A, RTCA, DBT, VCAM1, TRMT13, SASS6, EXTL2, SLC35A3, AGL, SLC30A7 \n",
      "\t   PALMD, DPH5, SNX7, S1PR1, DPYD, COL11A1, PTBP2, RWDD3, RNPC3, ALG14 \n",
      "\t   AMY2B, CNN3, PRMT6, SLC44A3, NTNG1, F3, VAV3, SLC25A24, ABCD3, FAM102B \n",
      "Negative:  BCL11A, FANCL, PAPOLG, VRK2, REL, PUS10, EFEMP1, KIAA1841, PEX13, LOC339803 \n",
      "\t   C2orf74, PNPT1, USP34, XPO1, CCDC88A, FAM161A, PRORSD1P, CCT4, MTIF2, COMMD1 \n",
      "\t   B3GNT2, TMEM17, RPS27A, EHBP1, LOC100132215, WDPCP, CLHC1, MDH1, UGP2, RTN4 \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2023-07-24 17:13:20] define_signif_tumor_subclusters(), tumor: malignant_SyS5\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  GSDMD, EEF1D, TSTA3, ZNF623, ZNF707, MAPK15, SCRIB, PUF60, NRBP2, EPPK1 \n",
      "\t   PLEC, PARP10, GRINA, OPLAH, DGAT1, SLC52A2, HSF1, BOP1, MROH1, ADCK5 \n",
      "\t   EXOSC4, MAF1, CPSF1, SLC39A4, SHARPIN, GPAA1, VPS28, TONSL, CYC1, CYHR1 \n",
      "Negative:  AFTPH, SERTAD2, LGALSL, PELI1, SLC1A4, VPS54, CEP68, UGP2, MDH1, RAB1A \n",
      "\t   WDPCP, ACTR2, LOC100132215, SPRED2, EHBP1, MEIS1, TMEM17, ETAA1, B3GNT2, C1D \n",
      "\t   COMMD1, WDR92, CCT4, PNO1, FAM161A, PPP3R1, XPO1, CNRIP1, USP34, PLEK \n",
      "PC_ 2 \n",
      "Positive:  DLEU2, RNASEH2B, FAM124A, DLEU1, INTS6, TRIM13, WDFY2, SPRYD7, DHRS12, ATP7B \n",
      "\t   KPNA3, ALG11, UTP14C, NEK5, EBPL, NEK3, MRPS31P5, THSD1, TPT1, SLC25A30 \n",
      "\t   TPT1-AS1, LCP1, LRCH1, GTF2F2, COG3, ESD, ZC3H13, NUFIP1, TSC22D1, LACC1 \n",
      "Negative:  SIT1, CD72, CCDC107, TESK1, ARHGEF39, TPM2, RUSC2, TLN1, UNC13B, CREB3 \n",
      "\t   FAM214B, GBA2, STOML2, RGP1, PIGO, NPR2, FANCG, HINT2, VCP, VPS28 \n",
      "\t   SLC39A4, TONSL, CPSF1, CYHR1, KIFC2, TMEM8B, PPP1R16A, ADCK5, MFSD3, SLC52A2 \n",
      "PC_ 3 \n",
      "Positive:  IFT52, TTPAL, SERINC3, L3MBTL1, PKIG, SRSF6, ADA, YWHAB, PTPRT, PABPC1L \n",
      "\t   TOMM34, CHD6, EMILIN3, STK4, LPIN3, ZHX3, SLPI, PLCG1, DHX35, MATN4 \n",
      "\t   TOP1, MAFB, SYS1, DBNDD2, PIGT, WFDC2, DNTTIP1, CTNNBL1, UBE2C, NNAT \n",
      "Negative:  CBX8, RBFOX3, CBX2, CBX4, ENGASE, C1QTNF1, TBC1D16, CANT1, C2CD2, ZBTB21 \n",
      "\t   PRDM15, LGALS3BP, CCDC40, ABCG1, RIPK4, UBASH3A, TIMP2, MX1, GAA, SLC37A1 \n",
      "\t   USP36, MX2, PDE9A, PLAC4, EIF4A3, CYTH1, BACE2, WDR4, PCP4, PGS1 \n",
      "PC_ 4 \n",
      "Positive:  CLHC1, RPS27A, SPTBN1, RTN4, ACYP2, PSME4, GPR75, MTIF2, ERLEC1, CHAC2 \n",
      "\t   PRORSD1P, GPR75-ASB3, ASB3, CCDC88A, NRXN1, PNPT1, STON1, EFEMP1, PPP1R21, FOXN2 \n",
      "\t   VRK2, FBXO11, FANCL, MSH6, MSH2, BCL11A, EPCAM, RABL3, CALM2, NDUFB4 \n",
      "Negative:  BZW1, CLK1, PPIL3, NIF3L1, ORC2, FAM126B, NDUFB3, CFLAR, CASP10, CASP8 \n",
      "\t   TRAK2, STRADB, TMEM237, ALS2, CDK15, FZD7, SUMO1, NOP58, BMPR2, FAM117B \n",
      "\t   ICA1L, WDR12, GJA4, CSMD2, ZSCAN20, PHC2, ZNF362, AK2, DNAJB5, C9orf131 \n",
      "PC_ 5 \n",
      "Positive:  XXYLT1, FAM43A, ACAP2, LSG1, TMEM44, ATP13A3, PPP1R2, HES1, OPA1, APOD \n",
      "\t   MB21D2, CCDC50, SDHAP2, IL1RAP, MUC20, CLDN1, FLJ42393, MUC4, LPP, TNK2 \n",
      "\t   BCL6, SST, SDHAP1, RTP4, TFRC, RPL39L, PCYT1A, ST6GAL1, TCTEX1D2, RFC4 \n",
      "Negative:  ICAM1, ICAM4, MRPL4, S1PR2, ICAM5, DNMT1, EIF3G, RAVER1, P2RY11, PPAN \n",
      "\t   ICAM3, OLFM2, TYK2, PIN1, CDC37, UBL5, FBXL12, ZNF846, ZNF562, ZNF561 \n",
      "\t   ZHX3, LPIN3, EMILIN3, CHD6, SRSF6, PTPRT, L3MBTL1, PLCG1, ZNF121, IFT52 \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2023-07-24 17:13:26] define_signif_tumor_subclusters(), tumor: malignant_SyS7\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  GNB1, NADK, GABRD, CDK11A, PRKCZ, SLC35E2B, CDK11B, SKI, MMP23B, MORN1 \n",
      "\t   MIB2, RER1, SSU72, ATAD3A, PEX10, ATAD3B, VWA1, PLCH2, ANKRD65, PANK4 \n",
      "\t   MRPL20, TNFRSF14, CCNL2, AURKAIP1, PRDM16, MXRA8, ARHGEF16, DVL1, MEGF6, PUSL1 \n",
      "Negative:  SOS1, MAP4K3, MORN2, DHX57, THUMPD2, GEMIN6, SRSF7, GALM, SLC8A1, ATL2 \n",
      "\t   CYP1B1, PKDCC, RMDN2, EML4, CDC42EP3, COX7A2L, QPCT, MTA3, PRKD3, OXER1 \n",
      "\t   NDUFAF7, HAAO, ZFP36L2, CEBPZ, THADA, EIF2AK2, PLEKHH2, GPATCH11, DYNC2LI1, HEATR5B \n",
      "PC_ 2 \n",
      "Positive:  GTF3C2, CAD, MPV17, ATRAID, ZNF512, SLC5A6, PREB, CCDC121, KHK, GPN1 \n",
      "\t   EMILIN1, SUPT7L, SLC4A1AP, OST4, AGBL5, MRPL33, TMEM214, RBKS, MAPRE3, LOC100505716 \n",
      "\t   DPYSL5, FOSL2, SLC35F6, PLB1, HADHB, PPP1CB, HADHA, SPDYA, RAB10, TRMT61B \n",
      "Negative:  EIF3I, MTMR9LP, TMEM234, LCK, IQCC, CCDC28B, HDAC1, TXLNA, KPNA6, COL16A1 \n",
      "\t   PTP4A2, TMEM39B, PEF1, MARCKSL1, KHDRBS1, TINAGL1, SERINC2, TSSK3, FABP3, BSDC1 \n",
      "\t   ZCCHC17, SNRNP40, ZBTB8B, PUM1, SDC3, ZBTB8A, LAPTM5, MATN1-AS1, PTPRU, ZBTB8OS \n",
      "PC_ 3 \n",
      "Positive:  DDX25, PUS3, HYLS1, CDON, RPUSD4, CHEK1, FAM118B, FOXRED1, TIRAP, DCPS \n",
      "\t   STT3A, ST3GAL4, EI24, KIRREL3, ETS1, FLI1, KCNJ5, ARHGAP32, BARX2, NFRKB \n",
      "\t   PRDM10, APLP2, ST14, ZBTB44, ADAMTS15, UNC5C, RAP1GDS1, TSPAN5, BMPR1B, EIF4E \n",
      "Negative:  YPEL5, LBH, ALK, LCLAT1, CLIP4, GALNT14, WDR43, EHD3, TRMT61B, MEMO1 \n",
      "\t   SPDYA, DPY30, PPP1CB, SPAST, PLB1, SLC30A6, FOSL2, LOC100505716, NLRC4, RBKS \n",
      "\t   YIPF4, MRPL33, SLC4A1AP, BIRC6, SUPT7L, GPN1, TTC27, CCDC121, ZNF512, LTBP1 \n",
      "PC_ 4 \n",
      "Positive:  MPP5, ATP6V1D, GPHN, FUT8, EIF2S1, MAX, FNTB, PIGH, RAB15, ARG2 \n",
      "\t   CHURC1-FNTB, VTI1B, CHURC1, RDH11, PLEKHG3, ZFYVE26, HSPA2, RAD51B, ZFP36L1, ZBTB1 \n",
      "\t   ACTN1, DCAF5, ZBTB25, EXD2, GALNT16, AKAP5, ERH, MTHFD1, SLC39A9, LOC100289511 \n",
      "Negative:  PDPN, LRRC38, PRDM2, DHRS3, KAZN, VPS13D, TNFRSF1B, TMEM51, MIIP, MFN2 \n",
      "\t   EFHD2, PLOD1, CASP9, KIAA2013, DNAJC16, DDI2, CLCN6, RSC1A1, MTHFR, PLEKHM2 \n",
      "\t   AGTRAP, SLC25A34, DRAXIN, FBLIM1, MAD2L2, FLJ37453, FBXO6, SPEN, FBXO44, ZBTB17 \n",
      "PC_ 5 \n",
      "Positive:  CAMK2N1, UBXN10, MUL1, OTUD3, PINK1, TMCO4, DDOST, HP1BP3, EIF4G3, NBL1 \n",
      "\t   ECE1, CAPZB, NBPF3, AKR7A2, ALPL, MRTO4, USP48, HSPG2, EMC1, LINC00339 \n",
      "\t   UBR4, IFFO2, ALDH4A1, PAX7, EPHB2, IGSF21, KDM1A, ARHGEF10L, LUZP1, RCC2 \n",
      "Negative:  UBLCP1, TTC1, SLU7, PWWP2A, C1QTNF2, RNF145, PTTG1, EBF1, GABRB2, CCNG1 \n",
      "\t   NUDCD2, CLINT1, HMMR, THG1L, MAT2B, ADAM19, CYFIP2, TENM2, ITK, MED7 \n",
      "\t   HAVCR2, SGCD, MRPL22, GEMIN5, CNOT8, LARP1, HAND1, SAP30L, GALNT10, NPM1 \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2023-07-24 17:13:32] define_signif_tumor_subclusters(), tumor: endothelial\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  FOCAD, MLLT3, KLHL9, MTAP, ACER2, CDKN2A, RPS6, CDKN2B, DENND4C, TUSC1 \n",
      "\t   IFT74, PLAA, CAAP1, PLIN2, TEK, MOB3B, HAUS6, C9orf72, RRAGA, ACO1 \n",
      "\t   ADAMTSL1, DDX58, SH3GL2, TOPORS, CNTLN, BNC2, CCDC171, PSIP1, NDUFB6, SNAPC3 \n",
      "Negative:  JOSD2, EMC10, SPIB, POLD1, NR1H2, NAPSB, ZNF473, VRK3, ATF5, NUP62 \n",
      "\t   IL4I1, TBC1D17, AKT1S1, PNKP, PTOV1, PTOV1-AS1, EIF3K, MED25, ACTN4, FUZ \n",
      "\t   CAPN12, ECH1, AP2A1, HNRNPL, CPT1C, RINL, PRMT1, SLAMF8, CD1E, CD1C \n",
      "PC_ 2 \n",
      "Positive:  GTF3C4, DDX31, TTF1, TSC1, SETX, GTF3C5, RALGDS, NTNG2, GBGT1, MED27 \n",
      "\t   RAPGEF1, SURF6, UCK1, MED22, POMT1, PRRC2B, RPL7A, FAM78A, SURF1, NUP214 \n",
      "\t   SURF2, AIF1L, LAMC3, SURF4, FIBCD1, REXO4, ABL1, ADAMTS13, EXOSC2, CACFD1 \n",
      "Negative:  FAM107A, C3orf67, FHIT, PTPRG, SYNPR, C3orf14, ACOX2, PTPRG-AS1, THOC7, ATXN7 \n",
      "\t   PSMD6-AS2, KCTD6, PSMD6, PDHB, PRICKLE2, PXK, PRICKLE2-AS3, RPP14, ADAMTS9, ABHD6 \n",
      "\t   ADAMTS9-AS2, DNASE1L3, MAGI1, FLNB, SLMAP, SLC25A26, DENND6A, ARF4, LRIG1, PDE12 \n",
      "PC_ 3 \n",
      "Positive:  NCF1, GTF2I, GTF2IRD2, STAG3L2, GTF2IRD1, GTF2IRD2B, NCF1C, NSUN5P1, POM121C, SPDYE5 \n",
      "\t   HIP1, RHBDD2, POR, TMEM120A, STYXL1, MDH2, HSPB1, YWHAG, ZP3, PAM \n",
      "\t   DTX2, GIN1, PPIP5K2, UPK3B, NUDT12, EFNA5, POMZP3, FBXL17, DTX2P1-UPK3BP1-PMS2P11, FER \n",
      "Negative:  NTMT1, ASB6, PTGES, TOR1B, FUBP3, EXOSC2, ABL1, FIBCD1, LAMC3, AIF1L \n",
      "\t   NUP214, FAM78A, APOL1, APOL2, APOL4, APOL3, MYH9, RBFOX2, TXN2, APOL6 \n",
      "\t   PRRC2B, FOXRED2, MCM5, EIF3D, HMOX1, IFT27, POMT1, TOM1, NCF4, CSF2RB \n",
      "PC_ 4 \n",
      "Positive:  CXXC1, SKA1, MRO, ME2, ELAC1, SMAD4, MEX3C, MBD2, POLI, C18orf54 \n",
      "\t   RAB27B, TCF4, TXNL1, WDR7, EPG5, FECH, SETBP1, PIK3C3, KIAA1328, LOC100505549 \n",
      "\t   TPGS2, FHOD3, ATP8B1, ELP2, SLC39A6, NEDD4L, RPRD1A, MALT1, C18orf21, ZNF532 \n",
      "Negative:  PTX3, SHOX2, RSRC1, GFM1, VEPH1, MLF1, LXN, RARRES1, MFSD1, CCNL1 \n",
      "\t   SCHIP1, TIPARP, IFT80, SSR3, SMC4, KCNAB1, GMPS, TRIM59, SLC33A1, KPNA4 \n",
      "\t   C3orf33, PPM1L, DHX36, ARHGEF26, ARHGEF26-AS1, RAP2B, MBNL1, MBNL1-AS1, P2RY12, P2RY13 \n",
      "PC_ 5 \n",
      "Positive:  ZNF706, NCALD, YWHAZ, PABPC1, RRM2B, ANKRD46, UBR5, RNF19A, KLF10, SPAG1 \n",
      "\t   POLR2K, AZIN1, EXOC4, RGS22, CHCHD3, ATP6V1C1, COX6C, VPS13B, FZD6, PLXNA4 \n",
      "\t   OSR2, PELI3, MRPL11, DPP3, CTHRC1, BBS1, ZDHHC24, SLC29A2, CTSF, BRMS1 \n",
      "Negative:  DICER1, GSC, CLMN, SERPINA1, SYNE3, IFI27L2, SNHG10, IFI27, GLRX5, IFI27L1 \n",
      "\t   TCL1A, C14orf132, DDX24, BDKRB2, ATG2B, OTUB2, GSKIP, PAPOLA, VRK1, BTBD7 \n",
      "\t   BCL11B, SETD3, UBR7, CCNK, TMEM251, CCDC85C, EML1, EVL, MOAP1, DEGS2 \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2023-07-24 17:13:32] define_signif_tumor_subclusters(), tumor: macrophage\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  EIF3G, DNMT1, P2RY11, S1PR2, MRPL4, PPAN, ICAM1, OLFM2, ICAM4, ICAM5 \n",
      "\t   PIN1, RAVER1, ICAM3, UBL5, TYK2, CDC37, FBXL12, PDE4A, ZNF846, KEAP1 \n",
      "\t   S1PR5, ZNF562, KRI1, ZNF561, CDKN2D, AP1M2, SLC44A2, ZNF121, ILF3, ZNF426 \n",
      "Negative:  IPO8, TMTC1, CAPRIN2, FAR2, ERGIC2, CCDC91, TSPAN11, PTHLH, KLHL42, DDX11 \n",
      "\t   MRPS35, PPFIBP1, DENND5B, STK38L, MED21, AMN1, TM7SF3, BICD1, FGFR1OP2, FGD4 \n",
      "\t   ITPR2, DNM1L, YARS2, SSPN, PKP2, BHLHE41, ALG10, HIF1A, HIF1A-AS2, PRKCH \n",
      "PC_ 2 \n",
      "Positive:  KERA, ATP2B1, LUM, POC1B-GALNT4, DCN, GALNT4, BTG1, POC1B, EEA1, DUSP6 \n",
      "\t   KITLG, NUDT4, TMTC3, UBE2N, CEP290, C12orf29, MRPL42, MKRN9P, SOCS2, RASSF9 \n",
      "\t   LRRIQ1, CRADD, SLC6A15, PLXNC1, TMTC2, TMCC3, METTL25, NDUFA12, CCDC59, NR2C1 \n",
      "Negative:  MRPS34, SPSB3, NME3, MAPK8IP3, TMEM204, IFT140, NUBP2, TELO2, CLCN7, C16orf91 \n",
      "\t   HAGH, UNKL, FAHD1, GNPTG, TSR3, MSRB1, NDUFB10, BAIAP3, RPS2, UBE2I \n",
      "\t   TPSD1, TBL3, TPSAB1, GFER, TPSB2, SYNGR3, CACNA1H, NPW, SOX8, NTHL1 \n",
      "PC_ 3 \n",
      "Positive:  CCDC134, SREBF2, MEI1, CENPM, NAGA, C22orf46, NDUFA6, XRCC6, TCF20, DESI1 \n",
      "\t   PMM1, NFAM1, CSDC2, POLR3H, RRP7A, ACO2, POLDIP3, PHF5A, CYB5R3, TOB2 \n",
      "\t   ARFGAP3, TEF, ZC3H7B, PACSIN2, RANGAP1, TTLL1, LOC339803, KIAA1841, C2orf74, USP34 \n",
      "Negative:  ZNF433, ZNF844, ZNF763, ZNF700, ZNF20, ZNF439, ZNF625, ZNF440, ZNF136, ZNF44 \n",
      "\t   ZNF491, ZNF563, ZNF441, ZNF442, ZNF823, ZNF799, AKAP12, ZNF443, HNRNPA1P10, ZBTB2 \n",
      "\t   RMND1, CCDC170, ESR1, ZNF709, ZNF627, SYNE1, MYCT1, FBXO5, ACP5, MTRF1L \n",
      "PC_ 4 \n",
      "Positive:  GBA2, RGP1, NPR2, ARHGEF39, CCDC107, CREB3, TPM2, SIT1, HINT2, TLN1 \n",
      "\t   CD72, TMEM8B, RECK, TESK1, GLIPR2, RUSC2, CLTA, UNC13B, GNE, FAM214B \n",
      "\t   STOML2, RNF38, PIGO, MELK, FANCG, VCP, ZCCHC7, C9orf131, GRHPR, DNAJB5 \n",
      "Negative:  CDKN2D, AP1M2, KRI1, S1PR5, SLC44A2, KEAP1, ILF3, QTRT1, PDE4A, DNM2 \n",
      "\t   CDC37, TMED1, TYK2, ICAM3, C19orf38, CARM1, RAVER1, ICAM5, YIPF2, ICAM4 \n",
      "\t   SMARCA4, ICAM1, LDLR, MRPL4, KANK2, S1PR2, DOCK6, DNMT1, RAB3D, EIF3G \n",
      "PC_ 5 \n",
      "Positive:  IPO8, TMTC1, CAPRIN2, ERGIC2, TSPAN11, FAR2, DDX11, CCDC91, PTHLH, DENND5B \n",
      "\t   KLHL42, AMN1, MRPS35, PPFIBP1, BICD1, STK38L, MED21, FGD4, TM7SF3, DNM1L \n",
      "\t   FGFR1OP2, ITPR2, YARS2, SSPN, PKP2, BHLHE41, ALG10, RASSF8, ALG10B, KRAS \n",
      "Negative:  PRORSD1P, MTIF2, RPS27A, CCDC88A, CLHC1, PNPT1, RTN4, EFEMP1, SPTBN1, ACYP2 \n",
      "\t   VRK2, PSME4, FANCL, GPR75, ERLEC1, CHAC2, BCL11A, GPR75-ASB3, ASB3, PAPOLG \n",
      "\t   NRXN1, REL, STON1, PPP1R21, PUS10, FOXN2, PEX13, FBXO11, KIAA1841, MSH6 \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2023-07-24 17:13:38] define_signif_tumor_subclusters(), tumor: fibroblast\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  CYTH1, USP36, PGS1, SOCS3, TIMP2, BIRC5, LGALS3BP, AFMID, CANT1, TK1 \n",
      "\t   C1QTNF1, SYNGR2, ENGASE, RBFOX3, TMC8, CBX2, ANO9, CBX8, SIGIRR, TMC6 \n",
      "\t   PKP3, CBX4, TNRC6C, TBC1D16, B4GALNT4, CCDC40, SEC14L1, IFITM3, GAA, IFITM1 \n",
      "Negative:  STAG1, PCCB, SLC35G2, NCK1, MSL2, PPP2R3A, DZIP1L, EPHB1, DBR1, CEP63 \n",
      "\t   ANAPC13, ARMC8, AMOTL2, MRAS, RYK, SLCO2A1, ESYT3, RAB6B, CEP70, SRPRB \n",
      "\t   FAIM, TOPBP1, CDV3, PIK3CB, NPHP3, MRPS22, UBA5, NPHP3-ACAD11, COPB2, ACAD11 \n",
      "PC_ 2 \n",
      "Positive:  SLC52A1, ZFP3, KIF1C, ZNF232, CAMTA2, SPAG7, ZNF594, ENO3, SCIMP, CC2D1A \n",
      "\t   LOC284454, RABEP1, ZSWIM4, PFN1, NUP88, C19orf53, RPAIN, MRI1, C1QBP, MIS12 \n",
      "\t   DHX33, RNF167, CCDC130, DERL2, LOC728392, NLRP1, IER2, LOC339166, WSCD1, STX10 \n",
      "Negative:  EPB41, MAP3K6, WASF2, SYTL1, GPR3, AHDC1, TMEM222, PPP1R8, FGR, SRSF4 \n",
      "\t   WDTC1, IFI6, STX12, FAM76A, MECR, SLC9A1, PTPRU, NUDC, MATN1-AS1, GPATCH3 \n",
      "\t   LAPTM5, GPN2, ARID1A, PIGV, RPS6KA1, ZDHHC18, HMGN2, SDC3, DHDDS, CD52 \n",
      "PC_ 3 \n",
      "Positive:  EGLN2, STEAP1B, RAPGEF5, RAB4B, IL6, TOMM7, FAM126A, MIA, KLHL7, GPNMB \n",
      "\t   MALSU1, SNRPA, IGF2BP3, C19orf54, ITPKC, NUMBL, PDCD10, LTBP4, ZBBX, SHKBP1 \n",
      "\t   NMD3, BLVRB, B3GALNT1, ZNF707, SERTAD3, PPM1L, SERTAD1, MAPK15, KPNA4, DEGS2 \n",
      "Negative:  PINK1, DDOST, MUL1, CAMK2N1, HP1BP3, EIF4G3, ECE1, UBXN10, NBPF3, OTUD3 \n",
      "\t   ALPL, RAP1GAP, TMCO4, USP48, HSPG2, NBL1, LINC00339, CAPZB, CDC42, AKR7A2 \n",
      "\t   MRTO4, WNT4, ZBTB40, EMC1, C1QA, UBR4, IFFO2, C1QC, ALDH4A1, C1QB \n",
      "PC_ 4 \n",
      "Positive:  PFN1, RNF167, ENO3, SLC25A11, SPAG7, CAMTA2, KIF1C, C17orf107, SLC52A1, CHRNE \n",
      "\t   ZFP3, RABEP1, SCIMP, NUP88, ZNF594, ZNF232, RPAIN, C1QBP, DHX33, DERL2 \n",
      "\t   MIS12, NLRP1, LOC728392, LOC339166, WSCD1, KIAA0753, TXNDC17, MED31, C17orf100, XAF1 \n",
      "Negative:  DPYD, SNX7, PALMD, PTBP2, AGL, RWDD3, SASS6, TRMT13, SLC35A3, DBT \n",
      "\t   RTCA, ALG14, CDC14A, CNN3, VCAM1, SLC44A3, EXTL2, F3, SLC30A7, ABCD3 \n",
      "\t   DPH5, S1PR1, COL11A1, ARHGAP29, RNPC3, AMY2B, PRMT6, GCLM, NTNG1, DNTTIP2 \n",
      "PC_ 5 \n",
      "Positive:  CLHC1, RTN4, SPTBN1, RPS27A, MTIF2, ACYP2, PSME4, PRORSD1P, GPR75, CCDC88A \n",
      "\t   ERLEC1, PNPT1, CHAC2, EFEMP1, VRK2, GPR75-ASB3, FANCL, ASB3, RRAD, CES2 \n",
      "\t   BCL11A, NRXN1, CES4A, PAPOLG, CBFB, STON1, C16orf70, REL, PUS10, PEX13 \n",
      "Negative:  SLC25A46, WDR36, CAMK4, MAN2A1, STARD4, STARD4-AS1, PJA2, NREP, FER, EPB41L4A-AS1 \n",
      "\t   FBXL17, APC, EPB41L4A, SRP19, REEP5, EFNA5, DCP2, NUDT12, MCC, PPIP5K2 \n",
      "\t   YTHDC2, GIN1, TRIM36, PAM, PGGT1B, SLCO4C1, CCDC112, FEM1C, ST8SIA4, TICAM2 \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2023-07-24 17:13:38] define_signif_tumor_subclusters(), tumor: t.cell\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  IFT57, CD47, BBX, DZIP3, CBLB, TRAT1, ALCAM, CD96, NFKBIZ, ZBED2 \n",
      "\t   NXPE3, CEP97, RPL24, PHLDB2, ZBTB11, ABHD10, C3orf52, CD200, PCNP, SENP7 \n",
      "\t   ABI3BP, TRMT10C, BTLA, TFG, TMEM45A, ATG3, SLC35A5, LNP1, CCDC80, NIT2 \n",
      "Negative:  NDUFB7, TECR, ZNF333, DNAJB1, ILVBL, GIPC1, NOTCH3, BRD4, PKN1, AKAP8 \n",
      "\t   AKAP8L, DDX39A, WIZ, RASAL3, LOC100507373, TPM4, RAB8A, ASF1B, HSH2D, FAM32A \n",
      "\t   AP1M1, PRKACA, KLF2, EPS15L1, IL27RA, C19orf44, CHERP, RFX1, SLC35E1, MED26 \n",
      "PC_ 2 \n",
      "Positive:  RNF111, SLTM, CCNB2, HSP90AB4P, MYO1E, ADAM10, GTF2A2, AQP9, BNIP2, FOXB1 \n",
      "\t   ANXA2, ALDH1A2, RORA, POLR2M, TCF12, MYZAP, CGNL1, VPS13C, LOC145783, TLN2 \n",
      "\t   ZNF280D, TPM1, MNS1, LACTB, RPS27L, TEX9, RAB8B, RFX7, APH1B, NEDD4 \n",
      "Negative:  CCDC88A, PRORSD1P, PNPT1, MTIF2, EFEMP1, RPS27A, VRK2, CLHC1, FANCL, BCL11A \n",
      "\t   RTN4, PAPOLG, REL, SPTBN1, PUS10, PEX13, ACYP2, KIAA1841, PSME4, LOC339803 \n",
      "\t   GPR75, C2orf74, USP34, ERLEC1, XPO1, FAM161A, CHAC2, CCT4, GPR75-ASB3, COMMD1 \n",
      "PC_ 3 \n",
      "Positive:  SIRT2, NFKBIB, RINL, HNRNPL, SARS2, ECH1, CAPN12, ACTN4, EIF3K, MAP4K1 \n",
      "\t   RYR1, DLL3, RASGRP4, EID2B, FAM98C, EID2, PSMD8, DYRK1B, KCNK6, FBL \n",
      "\t   FCGBP, C19orf33, PSMC4, YIF1B, ZNF546, ZNF780B, SPINT2, PEX13, PUS10, KIAA1841 \n",
      "Negative:  RIPK4, PRDM15, C2CD2, CERS6-AS1, NOSTRIN, CERS6, SPC25, ZBTB21, ABCG1, SLC37A1 \n",
      "\t   BBS5, UBASH3A, PDE9A, WDR4, NDUFV3, FASTKD1, PKNOX1, PHOSPHO2, KLHL23, PPIG \n",
      "\t   CCDC173, SSB, CBS, METTL5, U2AF1, UBR3, SIK1, GAD1, RRP1B, GORASP2 \n",
      "PC_ 4 \n",
      "Positive:  SH3BP2, NOP14-AS1, NOP14, ADD1, MFSD10, HTT, TNIP2, RGS12, FAM193A, RNF4 \n",
      "\t   LRPAP1, ZFYVE28, MXD4, C4orf48, HAUS3, NELFA, TMEM128, LETM1, LYAR, ZBTB49 \n",
      "\t   NSG1, STX18, PLS1, ATR, TRPC1, XRN1, PCOLCE2, MSX1, GK5, U2SURP \n",
      "Negative:  PCNP, TRMT10C, SENP7, ZBTB11, ABI3BP, RPL24, TFG, TMEM45A, CEP97, LNP1 \n",
      "\t   NIT2, NXPE3, TBC1D23, NFKBIZ, FILIP1L, CMSS1, COL8A1, ALCAM, DCBLD2, ST3GAL6 \n",
      "\t   CBLB, CPOX, BBX, CD47, IFT57, DZIP3, TRAT1, CD96, ZBED2, PHLDB2 \n",
      "PC_ 5 \n",
      "Positive:  ALCAM, NFKBIZ, NXPE3, CBLB, CEP97, BBX, RPL24, CD47, ZBTB11, IFT57 \n",
      "\t   DZIP3, PCNP, TRAT1, TRMT10C, CD96, SENP7, ZBED2, ABI3BP, PHLDB2, TFG \n",
      "\t   ABHD10, TMEM45A, C3orf52, LNP1, CD200, NIT2, BTLA, TBC1D23, ATG3, FILIP1L \n",
      "Negative:  N4BP3, RMND5B, LOC728554, LOC202181, NHP2, COL23A1, HNRNPAB, CLK4, B4GALT7, ZNF354A \n",
      "\t   ZNF354B, ZFP2, TMED9, ZNF454, FAM193B, ZNF879, DDX41, ZNF354C, DOK3, ADAMTS2 \n",
      "\t   PDLIM7, RUFY1, DBN1, HNRNPH1, PRR7, CANX, GRK6, MAML1, F12, LTC4S \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2023-07-24 17:13:40] define_signif_tumor_subclusters(), tumor: t.cd4\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  ADAM10, HSP90AB4P, AQP9, SLTM, RSL24D1, ALDH1A2, RNF111, MYBL1, ADHFE1, VCPIP1 \n",
      "\t   POLR2M, CCNB2, RAB27A, ASPH, RRS1, GGH, CLVS1, PDE7A, C8orf44, MYZAP \n",
      "\t   CHD7, MYO1E, YTHDF3, PIGB, MTFR1, CYP7B1, ARMC1, CGNL1, CCPG1, SGK3 \n",
      "Negative:  LOC100507373, DDX39A, ASF1B, PKN1, PRKACA, IL27RA, GIPC1, DNAJB1, RFX1, TECR \n",
      "\t   PODNL1, NDUFB7, CC2D1A, ZNF333, ILVBL, LOC284454, NOTCH3, ZSWIM4, BRD4, C19orf53 \n",
      "\t   MRI1, AKAP8, CCDC130, AKAP8L, IER2, STX10, WIZ, TRMT1, RASAL3, NFIX \n",
      "PC_ 2 \n",
      "Positive:  PAFAH2, EXTL1, STMN1, PDIK1L, ZNF593, PAQR7, CEP85, UBXN11, SH3BGRL3, CD52 \n",
      "\t   ZNF683, LOC646471, DHDDS, MTFR1L, HMGN2, MAN1C1, RPS6KA1, LDLRAP1, ARID1A, PIGV \n",
      "\t   TMEM50A, ZDHHC18, SYF2, GPN2, RUNX3, CLIC4, GPATCH3, SRRM1, RCAN3, NIPAL3 \n",
      "Negative:  KIAA0753, TXNDC17, WSCD1, MED31, LOC339166, C17orf100, DERL2, NLRP1, MIS12, LOC728392 \n",
      "\t   DHX33, XAF1, C1QBP, RPAIN, NUP88, ALOX12P2, RABEP1, SCIMP, ZNF594, ZNF232 \n",
      "\t   ZFP3, RNASEK, SLC52A1, KIF1C, CAMTA2, C17orf49, SPAG7, ENO3, PFN1, BCL6B \n",
      "PC_ 3 \n",
      "Positive:  PARK7, ERRFI1, RERE, ENO1, SLC2A5, H6PD, SPSB1, SLC25A33, CTNNBIP1, HCLS1 \n",
      "\t   GOLGB1, LZIC, POLQ, NMNAT1, GTF2E1, IQCB1, RBP7, RABL3, UBE4B, KIF1B \n",
      "\t   NDUFB4, FSTL1, DFFA, LRRC58, PEX14, CASZ1, OCEL1, NR2F6, GSK3B, USE1 \n",
      "Negative:  CD72, CCDC107, SIT1, TESK1, ARHGEF39, RUSC2, UNC13B, TPM2, FAM214B, TLN1 \n",
      "\t   STOML2, CREB3, PIGO, GBA2, FANCG, RGP1, VCP, C9orf131, NPR2, DNAJB5 \n",
      "\t   IL11RA, HINT2, GALT, SIGMAR1, TMEM8B, DCTN3, RPP25L, CNTFR, RECK, FAM219A \n",
      "PC_ 4 \n",
      "Positive:  NCOR2, ZNF664, SCARB1, CCDC92, ATP6V0A2, UBC, TCTN2, DHX37, GTF2H3, EIF2B1 \n",
      "\t   AACS, DDX55, TMEM132B, TMED2, TMEM132C, RILPL1, SLC15A4, SNRNP35, RILPL2, FZD10 \n",
      "\t   SBNO1, RIMBP2, CDK2AP1, C12orf65, STX2, MPHOSPH9, RAN, PITPNM2, SFSWAP, ARL6IP4 \n",
      "Negative:  ADAMTSL4, ECM1, MCL1, TARS2, ENSA, GOLPH3L, RPRD2, HORMAD1, PRPF3, CTSS \n",
      "\t   MRPS21, CTSK, C1orf54, ARNT, APH1A, SETDB1, CA14, CERS2, ANP32E, C1orf56 \n",
      "\t   PLEKHO1, CDC42SE1, MLLT11, GABPB2, VPS45, SEMA6C, OTUD7B, TNFAIP8L2, TNFAIP8L2-SCNM1, DHX33 \n",
      "PC_ 5 \n",
      "Positive:  CALM2, EPCAM, MSH6, TTC7A, MSH2, FBXO11, FOXN2, MCFD2, PPP1R21, STON1 \n",
      "\t   SOCS5, NRXN1, ASB3, CRIPT, GPR75-ASB3, CHAC2, GPR75, ERLEC1, PSME4, ACYP2 \n",
      "\t   PIGF, SPTBN1, RTN4, CLHC1, RHOQ, RPS27A, MTIF2, OMA1, TACSTD2, EPAS1 \n",
      "Negative:  CCNB2, RNF111, MYO1E, SLTM, HSP90AB4P, ADAM10, GTF2A2, AQP9, VPS13C, RORA \n",
      "\t   TLN2, BNIP2, FOXB1, ANXA2, ALDH1A2, TPM1, POLR2M, LACTB, MYZAP, RPS27L \n",
      "\t   CGNL1, TCF12, RAB8B, LOC145783, APH1B, ZNF280D, CA12, MNS1, TEX9, USP3 \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2023-07-24 17:13:42] define_signif_tumor_subclusters(), tumor: t.cd8\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  NNAT, CTNNBL1, TTI1, RPRD1B, BLCAP, SRC, TGM2, MANBAL, KIAA1755, RPN2 \n",
      "\t   SNHG11, MROH8, RALGAPB, RBL1, SCIMP, ZNF594, RABEP1, SAMHD1, ZNF232, ACTR5 \n",
      "\t   NUP88, ZFP3, RPAIN, SLC52A1, C1QBP, KIF1C, CAMTA2, SOGA1, DHX33, SPAG7 \n",
      "Negative:  VPS54, PELI1, LGALSL, UGP2, AFTPH, MDH1, SERTAD2, WDPCP, LOC100132215, SLC1A4 \n",
      "\t   EHBP1, CEP68, TMEM17, RAB1A, B3GNT2, ACTR2, SPRED2, COMMD1, MEIS1, ETAA1 \n",
      "\t   CCT4, C1D, FAM161A, WDR92, XPO1, USP34, PNO1, C2orf74, PPP3R1, LOC339803 \n",
      "PC_ 2 \n",
      "Positive:  ACTR2, SPRED2, MEIS1, RAB1A, ETAA1, CEP68, C1D, SLC1A4, WDR92, SERTAD2 \n",
      "\t   AFTPH, PNO1, PELI1, VPS54, LGALSL, UGP2, MDH1, PPP3R1, WDPCP, LOC100132215 \n",
      "\t   CNRIP1, EHBP1, PLEK, TMEM17, FBXO48, APLF, B3GNT2, ARHGAP25, COMMD1, ANTXR1 \n",
      "Negative:  STK4, SLPI, TOMM34, PABPC1L, MATN4, YWHAB, ADA, PKIG, SYS1, DBNDD2 \n",
      "\t   SERINC3, PIGT, WFDC2, TTPAL, DNTTIP1, UBE2C, IFT52, SNX21, L3MBTL1, SRSF6 \n",
      "\t   ACOT8, PTPRT, ZSWIM3, CHD6, ZSWIM1, NEURL2, EMILIN3, CTSA, LPIN3, ZHX3 \n",
      "PC_ 3 \n",
      "Positive:  WDR53, RNF168, FBXO45, UBXN7, CEP19, TCTEX1D2, PIGX, PCYT1A, PAK2, SENP5 \n",
      "\t   TFRC, NCBP2, SDHAP1, PIGZ, DLG1, TNK2, BDH1, MUC4, LOC220729, MUC20 \n",
      "\t   FYTTD1, SDHAP2, LRCH3, APOD, IQCG, PPP1R2, RPL35A, LMLN, ACAP2, XXYLT1 \n",
      "Negative:  LOC339166, NLRP1, WSCD1, KIAA0753, LOC728392, TXNDC17, MIS12, MED31, C17orf100, DERL2 \n",
      "\t   XAF1, ALOX12P2, DHX33, RNASEK, C1QBP, C17orf49, RPAIN, BCL6B, NUP88, CLEC10A \n",
      "\t   RABEP1, DLG4, SCIMP, ACADVL, ZNF594, DVL2, ZNF232, PHF23, ZFP3, GABARAP \n",
      "PC_ 4 \n",
      "Positive:  HMGN2, ARID1A, RPS6KA1, DHDDS, PIGV, ZDHHC18, GPN2, GPATCH3, NUDC, SLC9A1 \n",
      "\t   WDTC1, TMEM222, SYTL1, STMN1, MAP3K6, PAQR7, GPR3, WASF2, LOC646471, AHDC1 \n",
      "\t   MTFR1L, FGR, IFI6, MAN1C1, FAM76A, LDLRAP1, STX12, TMEM50A, PPP1R8, SYF2 \n",
      "Negative:  TTPAL, IFT52, SERINC3, L3MBTL1, SRSF6, PKIG, PTPRT, ADA, YWHAB, CHD6 \n",
      "\t   PABPC1L, EMILIN3, TOMM34, STK4, LPIN3, SLPI, ZHX3, MATN4, PLCG1, TOP1 \n",
      "\t   MAFB, SYS1, DHX35, DBNDD2, FAM83D, PPP1R16B, PIGT, ACTR5, RALGAPB, WFDC2 \n",
      "PC_ 5 \n",
      "Positive:  SH3BP2, TNIP2, ADD1, FAM193A, RNF4, MFSD10, ZFYVE28, MXD4, NOP14-AS1, HAUS3 \n",
      "\t   NOP14, C4orf48, NELFA, HTT, LETM1, RGS12, FGFR3, TACC3, LRPAP1, TMEM128 \n",
      "\t   TMEM129, SLBP, LYAR, UVSSA, CTBP1, MAEA, ZBTB49, LOC100130872, SPON2, RNF212 \n",
      "Negative:  ERG, ETS2, DYRK1A, TTC3, PSMG1, PIGP, BRWD1, HLCS, HMGN1, SIM2 \n",
      "\t   SH3BGR, CHAF1B, PCP4, MORC3, BACE2, PLAC4, CBR1, MX2, SETD4, RUNX1-IT1 \n",
      "\t   MX1, RUNX1, RIPK4, PRDM15, RCAN1, C2CD2, MRPS6, ZBTB21, SLC5A3, ABCG1 \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2023-07-24 17:13:48] define_signif_tumor_subclusters(), tumor: nk\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  KCTD7, LINC00174, RABGEF1, GTF2IRD1P1, TMEM248, TPST1, SBDS, TYW1, CRCP, PMS2P4 \n",
      "\t   ASL, STAG3L4, AUTS2, GUSB, CCT6P1, CALN1, ZNF92, TYW1B, CCT6P3, SBDSP1 \n",
      "\t   ERV3-1, SPDYE7P, ZNF117, POM121, ZNF273, NSUN5P2, ZNF138, STAG3L1, ZNF107, STAG3L3 \n",
      "Negative:  TMEM234, EIF3I, IQCC, CCDC28B, MTMR9LP, TXLNA, LCK, KPNA6, TMEM39B, HDAC1 \n",
      "\t   KHDRBS1, TSSK3, MARCKSL1, BSDC1, PTP4A2, ZBTB8B, COL16A1, ZBTB8A, PEF1, ZBTB8OS \n",
      "\t   TINAGL1, RBBP4, SERINC2, SYNC, FABP3, ZCCHC17, SNRNP40, PUM1, SDC3, KIAA1522 \n",
      "PC_ 2 \n",
      "Positive:  L3MBTL1, SRSF6, IFT52, TTPAL, PTPRT, SERINC3, CHD6, PKIG, EMILIN3, ADA \n",
      "\t   LPIN3, YWHAB, ZHX3, PABPC1L, PLCG1, TOMM34, STK4, TOP1, SLPI, MAFB \n",
      "\t   MATN4, DHX35, SDC4, SYS1, FAM83D, DBNDD2, PIGT, WFDC2, DNTTIP1, UBE2C \n",
      "Negative:  UBL5, FBXL12, ZNF846, ZNF561, PIN1, ZNF121, ZNF562, ZNF426, OLFM2, ZNF266 \n",
      "\t   PPAN, ZNF177, P2RY11, ZNF559-ZNF177, ZNF559, EIF3G, ZNF699, ZNF317, DNMT1, ZNF558 \n",
      "\t   ADAMTS10, S1PR2, MYO1F, HNRNPM, MRPL4, RAB11B, ANGPTL4, ICAM1, ICAM4, ICAM5 \n",
      "PC_ 3 \n",
      "Positive:  DAAM1, DACT1, KIAA0586, TIMM9, ARID4A, PSMA3, ACTR10, TMEM30B, NAA30, PRKCH \n",
      "\t   AP5M1, HIF1A, EXOC5, OTX2, HIF1A-AS2, PELI2, SNAPC1, KTN1, LINC00643, ATG14 \n",
      "\t   RHOJ, FBXO34, DLGAP5, PPP2R5E, LGALS3, WDR89, MAPK1IP1L, SOCS4, SGPP1, WDHD1 \n",
      "Negative:  KIF3C, ASXL2, DTNB, RAB10, RSAD2, CMPK2, HADHA, SOX11, HADHB, RNF144A-AS1 \n",
      "\t   SLC35F6, RPS7, DPYSL5, MAPRE3, TMEM214, AGBL5, RNASEH1, RNF144A, OST4, YWHAQ \n",
      "\t   ADI1, TRAPPC12, IAH1, GRHL1, ADAM17, TAF1B, SLC5A6, PXDN, ATRAID, KLF11 \n",
      "PC_ 4 \n",
      "Positive:  GALNT14, RBKS, LCLAT1, LOC100505716, LBH, MRPL33, FOSL2, PLB1, SPDYA, YPEL5 \n",
      "\t   TRMT61B, PPP1CB, ALK, CLIP4, SLC4A1AP, WDR43, SUPT7L, GPN1, CCDC121, ZNF512 \n",
      "\t   TOP1, MAFB, PLCG1, ZHX3, LPIN3, DHX35, EMILIN3, FAM83D, CHD6, PTPRT \n",
      "Negative:  DDX55, TMED2, EIF2B1, TCTN2, GTF2H3, ATP6V0A2, RILPL1, CCDC92, ZNF664, SNRNP35 \n",
      "\t   NCOR2, RILPL2, SCARB1, SBNO1, UBC, CDK2AP1, DHX37, C12orf65, AACS, TMEM132B \n",
      "\t   MPHOSPH9, TMEM132C, SLC15A4, PITPNM2, FZD10, RIMBP2, ARL6IP4, STX2, RAN, SFSWAP \n",
      "PC_ 5 \n",
      "Positive:  BAG4, DDHD2, LSM1, SKAP1, SNX11, CBX1, HOXB2, NFE2L1, HOXB3, COPZ2 \n",
      "\t   CDK5RAP3, PRR15L, PNPO, SP2, SCRN2, MRPL45P2, NPEPPS, KPNB1, TBKBP1, MRPL10 \n",
      "\t   TBX21, OSBPL7, LPIN1, GREB1, E2F6, SNRNP40, ROCK2, PUM1, ZCCHC17, PDIA6 \n",
      "Negative:  ZFP2, ZNF354B, N4BP3, ZNF354A, ZNF454, CLK4, RMND5B, NHP2, COL23A1, ZNF879 \n",
      "\t   HNRNPAB, ZNF354C, ADAMTS2, RUFY1, HNRNPH1, CANX, MAML1, LTC4S, MGAT4B, SQSTM1 \n",
      "\t   TBC1D9B, RNF130, RGS14, MAPK9, LMAN2, GFPT2, MXD3, CNOT6, VPS37B, ABCB9 \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2023-07-24 17:13:49] define_signif_tumor_subclusters(), tumor: mastocyte\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  ICAM4, ICAM1, MRPL4, ICAM5, S1PR2, DNMT1, EIF3G, P2RY11, PPAN, RAVER1 \n",
      "\t   OLFM2, PIN1, ICAM3, UBL5, TYK2, FBXL12, CDC37, ZNF846, PDE4A, ZNF562 \n",
      "\t   KEAP1, ZNF561, ZNF121, ZNF426, ZNF266, ZNF177, ZNF559-ZNF177, ZNF559, ZNF699, ANGPTL4 \n",
      "Negative:  CHURC1, PLEKHG3, AKAP5, CHURC1-FNTB, ZBTB25, HSPA2, ZBTB1, MTHFD1, RAB15, ARMC1 \n",
      "\t   CYP7B1, FNTB, MTFR1, SYNE2, PDE7A, YTHDF3, MAX, GGH, CHCHD7, PLAG1 \n",
      "\t   SGPP1, IMPAD1, RRS1, RPS20, ASPH, FUT8, CLVS1, FAM110B, ADHFE1, WDR89 \n",
      "PC_ 2 \n",
      "Positive:  IDUA, FGFRL1, SLC26A1, RNF212, SPON2, DGKQ, LOC100130872, CTBP1, TMEM175, MAEA \n",
      "\t   GAK, UVSSA, SLBP, LOC100129917, TMEM129, PCGF3, MYL5, PDE6B, PIGG, ZNF721 \n",
      "\t   ABCA11P, ZNF141, ZNF732, ZNF876P, ZNF595, ZNF718, CYTL1, EVC2, GLYCTK, WDR82 \n",
      "Negative:  FBXO5, MTRF1L, MYCT1, IPCEF1, CNKSR3, SYNE1, SCAF8, ESR1, TIAM2, CCDC170 \n",
      "\t   TFB1M, RMND1, ARID1B, ZBTB2, TMEM242, AKAP12, SNX9, SYNJ2, SERAC1, GTF2H5 \n",
      "\t   TULP4, TMEM181, DYNLT1, SYTL3, EZR, RSPH3, TAGAP, FNDC1, GINM1, SOD2 \n",
      "PC_ 3 \n",
      "Positive:  REEP5, DCP2, MCC, SRP19, NREP, STARD4-AS1, STARD4, APC, YTHDC2, EPB41L4A-AS1 \n",
      "\t   EPB41L4A, CAMK4, TRIM36, WDR36, PGGT1B, SLC25A46, CCDC112, MAN2A1, FEM1C, TICAM2 \n",
      "\t   PJA2, TMED7-TICAM2, FER, TMED7, CDO1, FBXL17, ATG12, AP3S1, EFNA5, COMMD10 \n",
      "Negative:  NAIP, GTF2H2, BDP1, GUSBP9, MCCC2, MAP1B, LOC441081, MRPS27, COX16, SMN2 \n",
      "\t   SMN1, SERF1B, SMOC1, PTCD2, GUSBP3, GTF2H2B, TNPO1, SLC10A1, FCHO2, GTF2H2C \n",
      "\t   TMEM171, FOXD1, BTF3, OCLN, SRSF5, ANKRA2, UTP15, RAD17, LOC100289511, ARHGEF28 \n",
      "PC_ 4 \n",
      "Positive:  FUT8, MAX, FNTB, GPHN, RAB15, MPP5, CHURC1-FNTB, CHURC1, ZBED2, ATP6V1D \n",
      "\t   PHLDB2, EIF2S1, PIGH, CD96, ARG2, ABHD10, PLEKHG3, TRAT1, VTI1B, HSPA2 \n",
      "\t   DZIP3, C3orf52, RDH11, ZBTB1, CD200, ZFYVE26, ZBTB25, RAD51B, BTLA, AKAP5 \n",
      "Negative:  HSD17B4, SRFBP1, TNFAIP8, LOX, DMXL1, SNCAIP, DTWD2, SNX2, SEMA6A, SNX24 \n",
      "\t   COMMD10, PPIC, AP3S1, ATG12, PRDM6, CDO1, TMED7, CEP120, TMED7-TICAM2, TICAM2 \n",
      "\t   CSNK1G3, FEM1C, ZNF608, CCDC112, PGGT1B, ALDH7A1, TRIM36, YTHDC2, PHAX, MCC \n",
      "PC_ 5 \n",
      "Positive:  SBF2, SBF2-AS1, ADM, SWAP70, AMPD3, MTRNR2L8, WEE1, RNF141, LYVE1, ZNF143 \n",
      "\t   MRVI1, CTR9, EIF4G2, ZBED5, LOC644656, GALNT18, USP47, IPO7, TMEM41B, DENND5A \n",
      "\t   SCUBE2, NRIP3, TMEM9B, AKIP1, RPL27A, TRIM66, STK33, RIC3, TUB, STAU2 \n",
      "Negative:  NOLC1, PPRC1, GBF1, LDB1, NFKB2, NPM3, FBXL15, FBXW4, CUEDC2, DPCD \n",
      "\t   ACTR1A, POLL, BTRC, SUFU, KAZALD1, TRIM8, SFXN3, LZTS2, ARL3, MRPL43 \n",
      "\t   SEMA4G, SFXN2, HIF1AN, NDUFB8, WBP1L, SEC31B, SCD, AS3MT, BLOC1S2, CNNM2 \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2023-07-24 17:13:50] define_signif_tumor_subclusters(), tumor: b.cell\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Centering and scaling data matrix\n",
      "\n",
      "PC_ 1 \n",
      "Positive:  ZNF285, ZNF229, ZNF180, PVR, YAP1, TRPC6, BIRC3, ARHGAP42, JRKL, CCDC82 \n",
      "\t   BIRC2, TMEM123, MAML2, IDH1, MMP10, PLEKHM3, PIKFYVE, PTH2R, MAP2, MTMR2 \n",
      "\t   DCUN1D5, FZD5, CEP57, RPE, DYNC2H1, FAM76B, CCNYL1, SESN3, BCL3, PDGFD \n",
      "Negative:  SNRNP27, MXD1, PCBP1-AS1, GMCL1, PCBP1, TIA1, C2orf42, PCYOX1, ANXA4, SNRPG \n",
      "\t   AAK1, FAM136A, NFU1, TGFA, GFPT1, SAMHD1, ADD2, RBL1, SOGA1, MROH8 \n",
      "\t   ANTXR1, CD207, RPN2, DSN1, MANBAL, TEX261, ARHGAP25, NDRG3, NAGK, SLA2 \n",
      "PC_ 2 \n",
      "Positive:  KITLG, TMTC3, DUSP6, CEP290, POC1B, C12orf29, GALNT4, MKRN9P, POC1B-GALNT4, RASSF9 \n",
      "\t   ATP2B1, LRRIQ1, KERA, KIF5B, LUM, SLC6A15, ARHGAP12, DCN, ZEB1, ZEB1-AS1 \n",
      "\t   TMTC2, ZNF438, BTG1, MAP3K8, MTPAP, METTL25, SVIL, CCDC59, BAMBI, PPFIA2 \n",
      "Negative:  WDR83OS, WDR83, MAN2B1, ZNF791, ZNF627, ACP5, ELOF1, HNRNPA1P10, ZNF490, ZNF823 \n",
      "\t   ZNF441, ZNF443, ZNF709, ZNF491, ZNF799, ZNF564, ZNF440, ZNF442, ZNF439, PRDX2 \n",
      "\t   ZNF563, EPOR, RNASEH2A, SWSAP1, CCDC159, DNASE2, TMEM205, RAB3D, DOCK6, SLC12A8 \n",
      "PC_ 3 \n",
      "Positive:  SLCO2A1, RAB6B, RYK, AMOTL2, ANAPC13, EPHB1, CEP63, SRPRB, TOPBP1, PPP2R3A \n",
      "\t   CDV3, MSL2, MRAS, PCCB, ESYT3, NPHP3, CEP70, ARMC8, STAG1, SLC35G2 \n",
      "\t   FAIM, NCK1, PIK3CB, DBR1, DZIP1L, UBA5, MRPS22, COPB2, NPHP3-ACAD11, RBP1 \n",
      "Negative:  CWC15, KDM4D, SESN3, ENDOD1, AMOTL1, SRSF8, FAM76B, PIWIL4, CEP57, FUT4 \n",
      "\t   MTMR2, ANKRD49, GPR83, MAML2, PANX1, MED17, CCDC82, JRKL, ARHGAP42, TRPC6 \n",
      "\t   YAP1, PRSS23, ME3, EED, SYTL2, PICALM, CCDC89, CREBZF, BIRC3, TMEM126A \n",
      "PC_ 4 \n",
      "Positive:  CHURC1-FNTB, CHURC1, PLEKHG3, RAB15, SGPP1, HSPA2, FNTB, SYNE2, FUT8, GPHN \n",
      "\t   MAX, MPP5, ZBTB1, ATP6V1D, MTHFD1, ZBTB25, AKAP5, EIF2S1, PIGH, ARG2 \n",
      "\t   VTI1B, ZFP36L1, RDH11, RAD51B, ACTN1, PRKCH, ZFYVE26, DCAF5, EXD2, TMEM30B \n",
      "Negative:  MXD4, HAUS3, ZFYVE28, C4orf48, RNF4, NELFA, FAM193A, LETM1, FGFR3, TNIP2 \n",
      "\t   TACC3, SH3BP2, TMEM129, ADD1, SLBP, UVSSA, MFSD10, MAEA, CTBP1, NOP14-AS1 \n",
      "\t   LOC100130872, SPON2, RNF212, NOP14, FGFRL1, IDUA, HTT, SLC26A1, DGKQ, RGS12 \n",
      "PC_ 5 \n",
      "Positive:  FZD6, CTHRC1, SLC25A32, DCAF13, LRP12, OXR1, ANGPT1, RSPO2, EIF3E, EMC2 \n",
      "\t   NUDCD1, ENY2, EBAG9, SYBU, TP53INP1, NDUFAF6, MTBP, SNTB1, CCNE2, HAS2 \n",
      "\t   INTS8, ZHX2, DPY19L4, ESRP1, RUNX1, RUNX1-IT1, DMWD, SYMPK, DMPK, HIF3A \n",
      "Negative:  MOCS3, KCNG1, NFATC2, DPM1, ATP9A, ZFP64, TSHZ2, ZNF217, SLC44A2, ILF3 \n",
      "\t   AP1M2, CDKN2D, KRI1, S1PR5, KEAP1, QTRT1, PDE4A, CDC37, TYK2, DNM2 \n",
      "\t   ICAM3, RAVER1, ICAM5, MRPL4, ICAM1, ICAM4, S1PR2, TMED1, DNMT1, EIF3G \n",
      "\n",
      "Computing nearest neighbor graph\n",
      "\n",
      "Computing SNN\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO [2023-07-24 17:13:51] -mirroring for hspike\n",
      "INFO [2023-07-24 17:13:51] define_signif_tumor_subclusters(p_val=0.1\n",
      "INFO [2023-07-24 17:13:52] define_signif_tumor_subclusters(), tumor: spike_tumor_cell_endothelial\n",
      "INFO [2023-07-24 17:13:52] cut tree into: 1 groups\n",
      "INFO [2023-07-24 17:13:52] -processing spike_tumor_cell_endothelial,spike_tumor_cell_endothelial_s1\n",
      "INFO [2023-07-24 17:13:52] define_signif_tumor_subclusters(), tumor: spike_tumor_cell_macrophage\n",
      "INFO [2023-07-24 17:13:52] cut tree into: 1 groups\n",
      "INFO [2023-07-24 17:13:52] -processing spike_tumor_cell_macrophage,spike_tumor_cell_macrophage_s1\n",
      "INFO [2023-07-24 17:13:52] define_signif_tumor_subclusters(), tumor: spike_tumor_cell_fibroblast\n",
      "INFO [2023-07-24 17:13:52] cut tree into: 1 groups\n",
      "INFO [2023-07-24 17:13:52] -processing spike_tumor_cell_fibroblast,spike_tumor_cell_fibroblast_s1\n",
      "INFO [2023-07-24 17:13:52] define_signif_tumor_subclusters(), tumor: spike_tumor_cell_t.cell\n",
      "INFO [2023-07-24 17:13:52] cut tree into: 1 groups\n",
      "INFO [2023-07-24 17:13:52] -processing spike_tumor_cell_t.cell,spike_tumor_cell_t.cell_s1\n",
      "INFO [2023-07-24 17:13:52] define_signif_tumor_subclusters(), tumor: spike_tumor_cell_t.cd4\n",
      "INFO [2023-07-24 17:13:53] cut tree into: 1 groups\n",
      "INFO [2023-07-24 17:13:53] -processing spike_tumor_cell_t.cd4,spike_tumor_cell_t.cd4_s1\n",
      "INFO [2023-07-24 17:13:53] define_signif_tumor_subclusters(), tumor: spike_tumor_cell_t.cd8\n",
      "INFO [2023-07-24 17:13:53] cut tree into: 1 groups\n",
      "INFO [2023-07-24 17:13:53] -processing spike_tumor_cell_t.cd8,spike_tumor_cell_t.cd8_s1\n",
      "INFO [2023-07-24 17:13:53] define_signif_tumor_subclusters(), tumor: spike_tumor_cell_nk\n",
      "INFO [2023-07-24 17:13:53] cut tree into: 1 groups\n",
      "INFO [2023-07-24 17:13:53] -processing spike_tumor_cell_nk,spike_tumor_cell_nk_s1\n",
      "INFO [2023-07-24 17:13:53] define_signif_tumor_subclusters(), tumor: spike_tumor_cell_mastocyte\n",
      "INFO [2023-07-24 17:13:53] cut tree into: 1 groups\n",
      "INFO [2023-07-24 17:13:53] -processing spike_tumor_cell_mastocyte,spike_tumor_cell_mastocyte_s1\n",
      "INFO [2023-07-24 17:13:53] define_signif_tumor_subclusters(), tumor: spike_tumor_cell_b.cell\n",
      "INFO [2023-07-24 17:13:53] cut tree into: 1 groups\n",
      "INFO [2023-07-24 17:13:53] -processing spike_tumor_cell_b.cell,spike_tumor_cell_b.cell_s1\n",
      "INFO [2023-07-24 17:13:53] define_signif_tumor_subclusters(), tumor: simnorm_cell_endothelial\n",
      "INFO [2023-07-24 17:13:53] cut tree into: 1 groups\n",
      "INFO [2023-07-24 17:13:53] -processing simnorm_cell_endothelial,simnorm_cell_endothelial_s1\n",
      "INFO [2023-07-24 17:13:53] define_signif_tumor_subclusters(), tumor: simnorm_cell_macrophage\n",
      "INFO [2023-07-24 17:13:54] cut tree into: 1 groups\n",
      "INFO [2023-07-24 17:13:54] -processing simnorm_cell_macrophage,simnorm_cell_macrophage_s1\n",
      "INFO [2023-07-24 17:13:54] define_signif_tumor_subclusters(), tumor: simnorm_cell_fibroblast\n",
      "INFO [2023-07-24 17:13:54] cut tree into: 1 groups\n",
      "INFO [2023-07-24 17:13:54] -processing simnorm_cell_fibroblast,simnorm_cell_fibroblast_s1\n",
      "INFO [2023-07-24 17:13:54] define_signif_tumor_subclusters(), tumor: simnorm_cell_t.cell\n",
      "INFO [2023-07-24 17:13:54] cut tree into: 1 groups\n",
      "INFO [2023-07-24 17:13:54] -processing simnorm_cell_t.cell,simnorm_cell_t.cell_s1\n",
      "INFO [2023-07-24 17:13:54] define_signif_tumor_subclusters(), tumor: simnorm_cell_t.cd4\n",
      "INFO [2023-07-24 17:13:54] cut tree into: 1 groups\n",
      "INFO [2023-07-24 17:13:54] -processing simnorm_cell_t.cd4,simnorm_cell_t.cd4_s1\n",
      "INFO [2023-07-24 17:13:54] define_signif_tumor_subclusters(), tumor: simnorm_cell_t.cd8\n",
      "INFO [2023-07-24 17:13:54] cut tree into: 1 groups\n",
      "INFO [2023-07-24 17:13:54] -processing simnorm_cell_t.cd8,simnorm_cell_t.cd8_s1\n",
      "INFO [2023-07-24 17:13:54] define_signif_tumor_subclusters(), tumor: simnorm_cell_nk\n",
      "INFO [2023-07-24 17:13:54] cut tree into: 1 groups\n",
      "INFO [2023-07-24 17:13:54] -processing simnorm_cell_nk,simnorm_cell_nk_s1\n",
      "INFO [2023-07-24 17:13:54] define_signif_tumor_subclusters(), tumor: simnorm_cell_mastocyte\n",
      "INFO [2023-07-24 17:13:55] cut tree into: 1 groups\n",
      "INFO [2023-07-24 17:13:55] -processing simnorm_cell_mastocyte,simnorm_cell_mastocyte_s1\n",
      "INFO [2023-07-24 17:13:55] define_signif_tumor_subclusters(), tumor: simnorm_cell_b.cell\n",
      "INFO [2023-07-24 17:13:55] cut tree into: 1 groups\n",
      "INFO [2023-07-24 17:13:55] -processing simnorm_cell_b.cell,simnorm_cell_b.cell_s1\n",
      "INFO [2023-07-24 17:15:00] ::plot_cnv:Start\n",
      "INFO [2023-07-24 17:15:00] ::plot_cnv:Current data dimensions (r,c)=12073,6951 Total=84238650.0582476 Min=0.49605252441373 Max=2.22182072260546.\n",
      "INFO [2023-07-24 17:15:01] ::plot_cnv:Depending on the size of the matrix this may take a moment.\n",
      "INFO [2023-07-24 17:15:05] plot_cnv(): auto thresholding at: (0.733025 , 1.274583)\n",
      "INFO [2023-07-24 17:15:10] plot_cnv_observation:Start\n",
      "INFO [2023-07-24 17:15:10] Observation data size: Cells= 4371 Genes= 12073\n",
      "INFO [2023-07-24 17:15:10] clustering observations via method: ward.D\n",
      "INFO [2023-07-24 17:15:10] Number of cells in group(1) is 187\n",
      "INFO [2023-07-24 17:15:10] group size being clustered:  187,12073\n",
      "INFO [2023-07-24 17:15:11] Number of cells in group(2) is 162\n",
      "INFO [2023-07-24 17:15:11] group size being clustered:  162,12073\n",
      "INFO [2023-07-24 17:15:11] Number of cells in group(3) is 33\n",
      "INFO [2023-07-24 17:15:11] group size being clustered:  33,12073\n",
      "INFO [2023-07-24 17:15:11] Number of cells in group(4) is 268\n",
      "INFO [2023-07-24 17:15:11] group size being clustered:  268,12073\n",
      "INFO [2023-07-24 17:15:13] Number of cells in group(5) is 190\n",
      "INFO [2023-07-24 17:15:13] group size being clustered:  190,12073\n",
      "INFO [2023-07-24 17:15:14] Number of cells in group(6) is 282\n",
      "INFO [2023-07-24 17:15:14] group size being clustered:  282,12073\n",
      "INFO [2023-07-24 17:15:16] Number of cells in group(7) is 191\n",
      "INFO [2023-07-24 17:15:16] group size being clustered:  191,12073\n",
      "INFO [2023-07-24 17:15:16] Number of cells in group(8) is 362\n",
      "INFO [2023-07-24 17:15:16] group size being clustered:  362,12073\n",
      "INFO [2023-07-24 17:15:20] Number of cells in group(9) is 100\n",
      "INFO [2023-07-24 17:15:20] group size being clustered:  100,12073\n",
      "INFO [2023-07-24 17:15:20] Number of cells in group(10) is 88\n",
      "INFO [2023-07-24 17:15:20] group size being clustered:  88,12073\n",
      "INFO [2023-07-24 17:15:20] Number of cells in group(11) is 220\n",
      "INFO [2023-07-24 17:15:20] group size being clustered:  220,12073\n",
      "INFO [2023-07-24 17:15:21] Number of cells in group(12) is 114\n",
      "INFO [2023-07-24 17:15:21] group size being clustered:  114,12073\n",
      "INFO [2023-07-24 17:15:22] Number of cells in group(13) is 349\n",
      "INFO [2023-07-24 17:15:22] group size being clustered:  349,12073\n",
      "INFO [2023-07-24 17:15:25] Number of cells in group(14) is 373\n",
      "INFO [2023-07-24 17:15:25] group size being clustered:  373,12073\n",
      "INFO [2023-07-24 17:15:29] Number of cells in group(15) is 280\n",
      "INFO [2023-07-24 17:15:29] group size being clustered:  280,12073\n",
      "INFO [2023-07-24 17:15:31] Number of cells in group(16) is 16\n",
      "INFO [2023-07-24 17:15:31] group size being clustered:  16,12073\n",
      "INFO [2023-07-24 17:15:31] Number of cells in group(17) is 321\n",
      "INFO [2023-07-24 17:15:31] group size being clustered:  321,12073\n",
      "INFO [2023-07-24 17:15:34] Number of cells in group(18) is 422\n",
      "INFO [2023-07-24 17:15:34] group size being clustered:  422,12073\n",
      "INFO [2023-07-24 17:15:39] Number of cells in group(19) is 10\n",
      "INFO [2023-07-24 17:15:39] group size being clustered:  10,12073\n",
      "INFO [2023-07-24 17:15:39] Number of cells in group(20) is 1\n",
      "INFO [2023-07-24 17:15:39] Skipping group: 20, since less than 2 entries\n",
      "INFO [2023-07-24 17:15:39] Number of cells in group(21) is 391\n",
      "INFO [2023-07-24 17:15:39] group size being clustered:  391,12073\n",
      "INFO [2023-07-24 17:15:44] Number of cells in group(22) is 11\n",
      "INFO [2023-07-24 17:15:44] group size being clustered:  11,12073\n",
      "INFO [2023-07-24 17:15:44] plot_cnv_observation:Writing observation groupings/color.\n",
      "INFO [2023-07-24 17:15:44] plot_cnv_observation:Done writing observation groupings/color.\n",
      "INFO [2023-07-24 17:15:44] plot_cnv_observation:Writing observation heatmap thresholds.\n",
      "INFO [2023-07-24 17:15:44] plot_cnv_observation:Done writing observation heatmap thresholds.\n",
      "DEBUG [2023-07-24 17:15:45] inferCNV::heatmap.cnv, breaks parameter set to: [0.733025024855591,0.769128884352346,0.805232743849101,0.841336603345856,0.877440462842612,0.913544322339367,0.949648181836122,0.985752041332878,1.02185590082963,1.05795976032639,1.09406361982314,1.1301674793199,1.16627133881665,1.20237519831341,1.23847905781016,1.27458291730692]\n",
      "DEBUG [2023-07-24 17:15:47] inferCNV::heatmap.cnv, resetting breaks to adjusted x.range: [0.733025024855591,0.769128884352346,0.805232743849101,0.841336603345856,0.877440462842612,0.913544322339367,0.949648181836122,0.985752041332878,1.02185590082963,1.05795976032639,1.09406361982314,1.1301674793199,1.16627133881665,1.20237519831341,1.23847905781016,1.27458291730692]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in (function (fmt, ...) :\n",
      "“one argument not used by format 'inferCNV::heatmap.cnv x range set to:  0.733025024855591,1.27458291730692'”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DEBUG [2023-07-24 17:15:47] inferCNV::heatmap.cnv x range set to:  0.733025024855591,1.27458291730692\n",
      "INFO [2023-07-24 17:16:01] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
      "INFO [2023-07-24 17:16:01] Quantiles of plotted data range: 0.733025024855591,0.956957972789801,1,1.04043172889142,1.27458291730692\n",
      "DEBUG [2023-07-24 17:16:08] ::inferCNV::heatmap.cnv colorkey z range: 0.733025024855591,0.771707731459257,0.810390438062923,0.84907314466659,0.887755851270256,0.926438557873923,0.965121264477589,1.00380397108126,1.04248667768492,1.08116938428859,1.11985209089225,1.15853479749592,1.19721750409959,1.23590021070325,1.27458291730692\n",
      "DEBUG [2023-07-24 17:16:08] ::inferCNV::heatmap.cnv colorkey breaks range: 0.733025024855591,0.769128884352346,0.805232743849101,0.841336603345856,0.877440462842612,0.913544322339367,0.949648181836122,0.985752041332878,1.02185590082963,1.05795976032639,1.09406361982314,1.1301674793199,1.16627133881665,1.20237519831341,1.23847905781016,1.27458291730692\n",
      "DEBUG [2023-07-24 17:16:08] ::inferCNV::heatmap.cnv colorkey colors range: #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
      "INFO [2023-07-24 17:16:10] plot_cnv_references:Start\n",
      "INFO [2023-07-24 17:16:10] Reference data size: Cells= 2580 Genes= 12073\n",
      "INFO [2023-07-24 17:16:19] plot_cnv_references:Number reference groups= 24\n",
      "INFO [2023-07-24 17:16:20] plot_cnv_references:Plotting heatmap.\n",
      "DEBUG [2023-07-24 17:16:20] inferCNV::heatmap.cnv, breaks parameter set to: [0.733025024855591,0.769128884352346,0.805232743849101,0.841336603345856,0.877440462842612,0.913544322339367,0.949648181836122,0.985752041332878,1.02185590082963,1.05795976032639,1.09406361982314,1.1301674793199,1.16627133881665,1.20237519831341,1.23847905781016,1.27458291730692]\n",
      "DEBUG [2023-07-24 17:16:21] inferCNV::heatmap.cnv, resetting breaks to adjusted x.range: [0.733025024855591,0.769128884352346,0.805232743849101,0.841336603345856,0.877440462842612,0.913544322339367,0.949648181836122,0.985752041332878,1.02185590082963,1.05795976032639,1.09406361982314,1.1301674793199,1.16627133881665,1.20237519831341,1.23847905781016,1.27458291730692]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in (function (fmt, ...) :\n",
      "“one argument not used by format 'inferCNV::heatmap.cnv x range set to:  0.733025024855591,1.27458291730692'”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DEBUG [2023-07-24 17:16:21] inferCNV::heatmap.cnv x range set to:  0.733025024855591,1.27458291730692\n",
      "INFO [2023-07-24 17:16:29] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
      "INFO [2023-07-24 17:16:29] Quantiles of plotted data range: 0.733025024855591,0.975174885933373,1,1.01968617325788,1.27458291730692\n",
      "INFO [2023-07-24 17:17:39] ::plot_cnv:Start\n",
      "INFO [2023-07-24 17:17:39] ::plot_cnv:Current data dimensions (r,c)=12073,6951 Total=84238650.0582476 Min=0.49605252441373 Max=2.22182072260546.\n",
      "INFO [2023-07-24 17:17:40] ::plot_cnv:Depending on the size of the matrix this may take a moment.\n",
      "INFO [2023-07-24 17:17:45] plot_cnv(): auto thresholding at: (0.733025 , 1.274583)\n",
      "INFO [2023-07-24 17:17:50] plot_cnv_observation:Start\n",
      "INFO [2023-07-24 17:17:50] Observation data size: Cells= 4371 Genes= 12073\n",
      "INFO [2023-07-24 17:17:50] plot_cnv_observation:Writing observation groupings/color.\n",
      "INFO [2023-07-24 17:17:50] plot_cnv_observation:Done writing observation groupings/color.\n",
      "INFO [2023-07-24 17:17:51] plot_cnv_observation:Writing observation heatmap thresholds.\n",
      "INFO [2023-07-24 17:17:51] plot_cnv_observation:Done writing observation heatmap thresholds.\n",
      "DEBUG [2023-07-24 17:17:52] inferCNV::heatmap.cnv, breaks parameter set to: [0.733025024855591,0.769128884352346,0.805232743849101,0.841336603345856,0.877440462842612,0.913544322339367,0.949648181836122,0.985752041332878,1.02185590082963,1.05795976032639,1.09406361982314,1.1301674793199,1.16627133881665,1.20237519831341,1.23847905781016,1.27458291730692]\n",
      "DEBUG [2023-07-24 17:17:54] inferCNV::heatmap.cnv, resetting breaks to adjusted x.range: [0.733025024855591,0.769128884352346,0.805232743849101,0.841336603345856,0.877440462842612,0.913544322339367,0.949648181836122,0.985752041332878,1.02185590082963,1.05795976032639,1.09406361982314,1.1301674793199,1.16627133881665,1.20237519831341,1.23847905781016,1.27458291730692]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in (function (fmt, ...) :\n",
      "“one argument not used by format 'inferCNV::heatmap.cnv x range set to:  0.733025024855591,1.27458291730692'”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DEBUG [2023-07-24 17:17:54] inferCNV::heatmap.cnv x range set to:  0.733025024855591,1.27458291730692\n",
      "INFO [2023-07-24 17:18:08] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
      "INFO [2023-07-24 17:18:08] Quantiles of plotted data range: 0.733025024855591,0.956957972789801,1,1.04043172889142,1.27458291730692\n",
      "DEBUG [2023-07-24 17:18:16] ::inferCNV::heatmap.cnv colorkey z range: 0.733025024855591,0.771707731459257,0.810390438062923,0.84907314466659,0.887755851270256,0.926438557873923,0.965121264477589,1.00380397108126,1.04248667768492,1.08116938428859,1.11985209089225,1.15853479749592,1.19721750409959,1.23590021070325,1.27458291730692\n",
      "DEBUG [2023-07-24 17:18:16] ::inferCNV::heatmap.cnv colorkey breaks range: 0.733025024855591,0.769128884352346,0.805232743849101,0.841336603345856,0.877440462842612,0.913544322339367,0.949648181836122,0.985752041332878,1.02185590082963,1.05795976032639,1.09406361982314,1.1301674793199,1.16627133881665,1.20237519831341,1.23847905781016,1.27458291730692\n",
      "DEBUG [2023-07-24 17:18:16] ::inferCNV::heatmap.cnv colorkey colors range: #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
      "INFO [2023-07-24 17:18:18] plot_cnv_references:Start\n",
      "INFO [2023-07-24 17:18:18] Reference data size: Cells= 2580 Genes= 12073\n",
      "INFO [2023-07-24 17:19:05] plot_cnv_references:Number reference groups= 9\n",
      "INFO [2023-07-24 17:19:05] plot_cnv_references:Plotting heatmap.\n",
      "DEBUG [2023-07-24 17:19:06] inferCNV::heatmap.cnv, breaks parameter set to: [0.733025024855591,0.769128884352346,0.805232743849101,0.841336603345856,0.877440462842612,0.913544322339367,0.949648181836122,0.985752041332878,1.02185590082963,1.05795976032639,1.09406361982314,1.1301674793199,1.16627133881665,1.20237519831341,1.23847905781016,1.27458291730692]\n",
      "DEBUG [2023-07-24 17:19:07] inferCNV::heatmap.cnv, resetting breaks to adjusted x.range: [0.733025024855591,0.769128884352346,0.805232743849101,0.841336603345856,0.877440462842612,0.913544322339367,0.949648181836122,0.985752041332878,1.02185590082963,1.05795976032639,1.09406361982314,1.1301674793199,1.16627133881665,1.20237519831341,1.23847905781016,1.27458291730692]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in (function (fmt, ...) :\n",
      "“one argument not used by format 'inferCNV::heatmap.cnv x range set to:  0.733025024855591,1.27458291730692'”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DEBUG [2023-07-24 17:19:07] inferCNV::heatmap.cnv x range set to:  0.733025024855591,1.27458291730692\n",
      "INFO [2023-07-24 17:19:16] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
      "INFO [2023-07-24 17:19:16] Quantiles of plotted data range: 0.733025024855591,0.975174885933373,1,1.01968617325788,1.27458291730692\n",
      "INFO [2023-07-24 17:19:20] \n",
      "\n",
      "\tSTEP 17: HMM-based CNV prediction\n",
      "\n",
      "INFO [2023-07-24 17:19:20] predict_CNV_via_HMM_on_tumor_subclusters\n",
      "INFO [2023-07-24 17:19:48] -done predicting CNV based on initial tumor subclusters\n",
      "INFO [2023-07-24 17:19:48] get_predicted_CNV_regions(subcluster)\n",
      "INFO [2023-07-24 17:19:48] -processing cell_group_name: malignant_SyS1.malignant_SyS1_s2, size: 187\n",
      "INFO [2023-07-24 17:20:02] -processing cell_group_name: malignant_SyS1.malignant_SyS1_s1, size: 162\n",
      "INFO [2023-07-24 17:20:15] -processing cell_group_name: malignant_SyS10.malignant_SyS10_s1, size: 33\n",
      "INFO [2023-07-24 17:20:26] -processing cell_group_name: malignant_SyS11.malignant_SyS11_s1, size: 268\n",
      "INFO [2023-07-24 17:20:40] -processing cell_group_name: malignant_SyS11.malignant_SyS11_s2, size: 190\n",
      "INFO [2023-07-24 17:20:53] -processing cell_group_name: malignant_SyS11.met.malignant_SyS11.met_s1, size: 282\n",
      "INFO [2023-07-24 17:21:07] -processing cell_group_name: malignant_SyS11.met.malignant_SyS11.met_s2, size: 191\n",
      "INFO [2023-07-24 17:21:21] -processing cell_group_name: malignant_SyS12.malignant_SyS12_s2, size: 362\n",
      "INFO [2023-07-24 17:21:38] -processing cell_group_name: malignant_SyS12.malignant_SyS12_s3, size: 100\n",
      "INFO [2023-07-24 17:21:50] -processing cell_group_name: malignant_SyS12.malignant_SyS12_s1, size: 88\n",
      "INFO [2023-07-24 17:22:03] -processing cell_group_name: malignant_SyS12pt.malignant_SyS12pt_s1, size: 220\n",
      "INFO [2023-07-24 17:22:17] -processing cell_group_name: malignant_SyS12pt.malignant_SyS12pt_s2, size: 114\n",
      "INFO [2023-07-24 17:22:30] -processing cell_group_name: malignant_SyS13.malignant_SyS13_s1, size: 349\n",
      "INFO [2023-07-24 17:22:45] -processing cell_group_name: malignant_SyS14.malignant_SyS14_s1, size: 373\n",
      "INFO [2023-07-24 17:22:59] -processing cell_group_name: malignant_SyS16.malignant_SyS16_s1, size: 280\n",
      "INFO [2023-07-24 17:23:14] -processing cell_group_name: malignant_SyS16.malignant_SyS16_s2, size: 16\n",
      "INFO [2023-07-24 17:23:25] -processing cell_group_name: malignant_SyS2.malignant_SyS2_s1, size: 321\n",
      "INFO [2023-07-24 17:23:39] -processing cell_group_name: malignant_SyS5.malignant_SyS5_s1, size: 422\n",
      "INFO [2023-07-24 17:23:55] -processing cell_group_name: malignant_SyS5.malignant_SyS5_s2, size: 10\n",
      "INFO [2023-07-24 17:24:07] -processing cell_group_name: malignant_SyS5.malignant_SyS5_s3, size: 1\n",
      "INFO [2023-07-24 17:24:18] -processing cell_group_name: malignant_SyS7.malignant_SyS7_s1, size: 391\n",
      "INFO [2023-07-24 17:24:32] -processing cell_group_name: malignant_SyS7.malignant_SyS7_s2, size: 11\n",
      "INFO [2023-07-24 17:24:43] -processing cell_group_name: endothelial.endothelial_s1, size: 79\n",
      "INFO [2023-07-24 17:24:55] -processing cell_group_name: macrophage.macrophage_s3, size: 214\n",
      "INFO [2023-07-24 17:25:09] -processing cell_group_name: macrophage.macrophage_s5, size: 181\n",
      "INFO [2023-07-24 17:25:22] -processing cell_group_name: macrophage.macrophage_s1, size: 140\n",
      "INFO [2023-07-24 17:25:35] -processing cell_group_name: macrophage.macrophage_s6, size: 138\n",
      "INFO [2023-07-24 17:25:47] -processing cell_group_name: macrophage.macrophage_s4, size: 124\n",
      "INFO [2023-07-24 17:26:00] -processing cell_group_name: macrophage.macrophage_s2, size: 119\n",
      "INFO [2023-07-24 17:26:13] -processing cell_group_name: macrophage.macrophage_s7, size: 22\n",
      "INFO [2023-07-24 17:26:25] -processing cell_group_name: macrophage.macrophage_s8, size: 2\n",
      "INFO [2023-07-24 17:26:37] -processing cell_group_name: macrophage.macrophage_s10, size: 2\n",
      "INFO [2023-07-24 17:26:48] -processing cell_group_name: macrophage.macrophage_s9, size: 1\n",
      "INFO [2023-07-24 17:26:59] -processing cell_group_name: fibroblast.fibroblast_s1, size: 81\n",
      "INFO [2023-07-24 17:27:12] -processing cell_group_name: t.cell.t.cell_s1, size: 206\n",
      "INFO [2023-07-24 17:27:25] -processing cell_group_name: t.cd4.t.cd4_s1, size: 235\n",
      "INFO [2023-07-24 17:27:39] -processing cell_group_name: t.cd8.t.cd8_s3, size: 321\n",
      "INFO [2023-07-24 17:27:53] -processing cell_group_name: t.cd8.t.cd8_s2, size: 228\n",
      "INFO [2023-07-24 17:28:07] -processing cell_group_name: t.cd8.t.cd8_s1, size: 68\n",
      "INFO [2023-07-24 17:28:19] -processing cell_group_name: t.cd8.t.cd8_s4, size: 21\n",
      "INFO [2023-07-24 17:28:31] -processing cell_group_name: t.cd8.t.cd8_s5, size: 17\n",
      "INFO [2023-07-24 17:28:43] -processing cell_group_name: t.cd8.t.cd8_s6, size: 2\n",
      "INFO [2023-07-24 17:28:54] -processing cell_group_name: t.cd8.t.cd8_s7, size: 2\n",
      "INFO [2023-07-24 17:29:05] -processing cell_group_name: nk.nk_s1, size: 102\n",
      "INFO [2023-07-24 17:29:18] -processing cell_group_name: mastocyte.mastocyte_s1, size: 185\n",
      "INFO [2023-07-24 17:29:31] -processing cell_group_name: b.cell.b.cell_s1, size: 90\n",
      "INFO [2023-07-24 17:29:44] -writing cell clusters file: outputs/infercnv/sys_take2//17_HMM_predHMMi6.leiden.hmm_mode-subclusters.cell_groupings\n",
      "INFO [2023-07-24 17:29:44] -writing cnv regions file: outputs/infercnv/sys_take2//17_HMM_predHMMi6.leiden.hmm_mode-subclusters.pred_cnv_regions.dat\n",
      "INFO [2023-07-24 17:29:45] -writing per-gene cnv report: outputs/infercnv/sys_take2//17_HMM_predHMMi6.leiden.hmm_mode-subclusters.pred_cnv_genes.dat\n",
      "INFO [2023-07-24 17:29:46] -writing gene ordering info: outputs/infercnv/sys_take2//17_HMM_predHMMi6.leiden.hmm_mode-subclusters.genes_used.dat\n",
      "INFO [2023-07-24 17:30:27] ::plot_cnv:Start\n",
      "INFO [2023-07-24 17:30:27] ::plot_cnv:Current data dimensions (r,c)=12073,6951 Total=255166202 Min=1 Max=6.\n",
      "INFO [2023-07-24 17:30:28] ::plot_cnv:Depending on the size of the matrix this may take a moment.\n",
      "INFO [2023-07-24 17:30:33] plot_cnv_observation:Start\n",
      "INFO [2023-07-24 17:30:33] Observation data size: Cells= 4371 Genes= 12073\n",
      "INFO [2023-07-24 17:30:34] plot_cnv_observation:Writing observation groupings/color.\n",
      "INFO [2023-07-24 17:30:34] plot_cnv_observation:Done writing observation groupings/color.\n",
      "INFO [2023-07-24 17:30:34] plot_cnv_observation:Writing observation heatmap thresholds.\n",
      "INFO [2023-07-24 17:30:34] plot_cnv_observation:Done writing observation heatmap thresholds.\n",
      "DEBUG [2023-07-24 17:30:35] inferCNV::heatmap.cnv, breaks parameter set to: [0,0.4,0.8,1.2,1.6,2,2.4,2.8,3.2,3.6,4,4.4,4.8,5.2,5.6,6]\n",
      "DEBUG [2023-07-24 17:30:37] inferCNV::heatmap.cnv, resetting breaks to adjusted x.range: [0,0.4,0.8,1.2,1.6,2,2.4,2.8,3.2,3.6,4,4.4,4.8,5.2,5.6,6]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in (function (fmt, ...) :\n",
      "“one argument not used by format 'inferCNV::heatmap.cnv x range set to:  0,6'”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DEBUG [2023-07-24 17:30:37] inferCNV::heatmap.cnv x range set to:  0,6\n",
      "INFO [2023-07-24 17:30:50] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
      "INFO [2023-07-24 17:30:51] Quantiles of plotted data range: 1,3,3,3,6\n",
      "DEBUG [2023-07-24 17:30:57] ::inferCNV::heatmap.cnv colorkey z range: 0,0.428571428571429,0.857142857142857,1.28571428571429,1.71428571428571,2.14285714285714,2.57142857142857,3,3.42857142857143,3.85714285714286,4.28571428571429,4.71428571428571,5.14285714285714,5.57142857142857,6\n",
      "DEBUG [2023-07-24 17:30:57] ::inferCNV::heatmap.cnv colorkey breaks range: 0,0.4,0.8,1.2,1.6,2,2.4,2.8,3.2,3.6,4,4.4,4.8,5.2,5.6,6\n",
      "DEBUG [2023-07-24 17:30:57] ::inferCNV::heatmap.cnv colorkey colors range: #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
      "INFO [2023-07-24 17:30:59] plot_cnv_references:Start\n",
      "INFO [2023-07-24 17:30:59] Reference data size: Cells= 2580 Genes= 12073\n",
      "INFO [2023-07-24 17:31:45] plot_cnv_references:Number reference groups= 9\n",
      "INFO [2023-07-24 17:31:46] plot_cnv_references:Plotting heatmap.\n",
      "DEBUG [2023-07-24 17:31:46] inferCNV::heatmap.cnv, breaks parameter set to: [0,0.4,0.8,1.2,1.6,2,2.4,2.8,3.2,3.6,4,4.4,4.8,5.2,5.6,6]\n",
      "DEBUG [2023-07-24 17:31:48] inferCNV::heatmap.cnv, resetting breaks to adjusted x.range: [0,0.4,0.8,1.2,1.6,2,2.4,2.8,3.2,3.6,4,4.4,4.8,5.2,5.6,6]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in (function (fmt, ...) :\n",
      "“one argument not used by format 'inferCNV::heatmap.cnv x range set to:  0,6'”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DEBUG [2023-07-24 17:31:48] inferCNV::heatmap.cnv x range set to:  0,6\n",
      "INFO [2023-07-24 17:31:56] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
      "INFO [2023-07-24 17:31:56] Quantiles of plotted data range: 2,3,3,3,5\n",
      "INFO [2023-07-24 17:32:04] \n",
      "\n",
      "\tSTEP 18: Run Bayesian Network Model on HMM predicted CNVs\n",
      "\n",
      "INFO [2023-07-24 17:32:04] Creating the following Directory:  outputs/infercnv/sys_take2//BayesNetOutput.HMMi6.leiden.hmm_mode-subclusters\n",
      "INFO [2023-07-24 17:32:04] Initializing new MCM InferCNV Object.\n",
      "INFO [2023-07-24 17:32:04] validating infercnv_obj\n",
      "INFO [2023-07-24 17:32:04] Total CNV's:  805\n",
      "INFO [2023-07-24 17:32:04] Loading BUGS Model.\n",
      "INFO [2023-07-24 17:32:05] Running Sampling Using Parallel with  4 Cores\n",
      "INFO [2023-07-24 18:54:08] Obtaining probabilities post-sampling\n",
      "INFO [2023-07-24 19:02:31] Gibbs sampling time:  90.4234283089638  Minutes\n",
      "INFO [2023-07-24 19:08:41] ::plot_cnv:Start\n",
      "INFO [2023-07-24 19:08:41] ::plot_cnv:Current data dimensions (r,c)=12073,6951 Total=5564387.75386144 Min=0 Max=0.989471219982526.\n",
      "INFO [2023-07-24 19:08:42] ::plot_cnv:Depending on the size of the matrix this may take a moment.\n",
      "INFO [2023-07-24 19:08:45] plot_cnv_observation:Start\n",
      "INFO [2023-07-24 19:08:45] Observation data size: Cells= 4371 Genes= 12073\n",
      "INFO [2023-07-24 19:08:45] plot_cnv_observation:Writing observation groupings/color.\n",
      "INFO [2023-07-24 19:08:45] plot_cnv_observation:Done writing observation groupings/color.\n",
      "INFO [2023-07-24 19:08:46] plot_cnv_observation:Writing observation heatmap thresholds.\n",
      "INFO [2023-07-24 19:08:46] plot_cnv_observation:Done writing observation heatmap thresholds.\n",
      "DEBUG [2023-07-24 19:08:46] inferCNV::heatmap.cnv, breaks parameter set to: [0,0.0666666666666667,0.133333333333333,0.2,0.266666666666667,0.333333333333333,0.4,0.466666666666667,0.533333333333333,0.6,0.666666666666667,0.733333333333333,0.8,0.866666666666667,0.933333333333333,1]\n",
      "DEBUG [2023-07-24 19:08:47] inferCNV::heatmap.cnv, resetting breaks to adjusted x.range: [-1,-0.866666666666667,-0.733333333333333,-0.6,-0.466666666666667,-0.333333333333333,-0.2,-0.0666666666666667,0.0666666666666667,0.2,0.333333333333333,0.466666666666667,0.6,0.733333333333333,0.866666666666667,1]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in (function (fmt, ...) :\n",
      "“one argument not used by format 'inferCNV::heatmap.cnv x range set to:  -1,1'”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DEBUG [2023-07-24 19:08:47] inferCNV::heatmap.cnv x range set to:  -1,1\n",
      "INFO [2023-07-24 19:08:56] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
      "INFO [2023-07-24 19:08:56] Quantiles of plotted data range: 0,0,0,0,0.989471219982526\n",
      "DEBUG [2023-07-24 19:09:02] ::inferCNV::heatmap.cnv colorkey z range: -1,-0.857142857142857,-0.714285714285714,-0.571428571428571,-0.428571428571429,-0.285714285714286,-0.142857142857143,0,0.142857142857143,0.285714285714286,0.428571428571428,0.571428571428571,0.714285714285714,0.857142857142857,1\n",
      "DEBUG [2023-07-24 19:09:02] ::inferCNV::heatmap.cnv colorkey breaks range: -1,-0.866666666666667,-0.733333333333333,-0.6,-0.466666666666667,-0.333333333333333,-0.2,-0.0666666666666667,0.0666666666666667,0.2,0.333333333333333,0.466666666666667,0.6,0.733333333333333,0.866666666666667,1\n",
      "DEBUG [2023-07-24 19:09:02] ::inferCNV::heatmap.cnv colorkey colors range: #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
      "INFO [2023-07-24 19:09:03] plot_cnv_references:Start\n",
      "INFO [2023-07-24 19:09:03] Reference data size: Cells= 2580 Genes= 12073\n",
      "INFO [2023-07-24 19:09:58] plot_cnv_references:Number reference groups= 9\n",
      "INFO [2023-07-24 19:09:59] plot_cnv_references:Plotting heatmap.\n",
      "DEBUG [2023-07-24 19:09:59] inferCNV::heatmap.cnv, breaks parameter set to: [0,0.0666666666666667,0.133333333333333,0.2,0.266666666666667,0.333333333333333,0.4,0.466666666666667,0.533333333333333,0.6,0.666666666666667,0.733333333333333,0.8,0.866666666666667,0.933333333333333,1]\n",
      "DEBUG [2023-07-24 19:09:59] inferCNV::heatmap.cnv, resetting breaks to adjusted x.range: [-1,-0.866666666666667,-0.733333333333333,-0.6,-0.466666666666667,-0.333333333333333,-0.2,-0.0666666666666667,0.0666666666666667,0.2,0.333333333333333,0.466666666666667,0.6,0.733333333333333,0.866666666666667,1]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in (function (fmt, ...) :\n",
      "“one argument not used by format 'inferCNV::heatmap.cnv x range set to:  -1,1'”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DEBUG [2023-07-24 19:09:59] inferCNV::heatmap.cnv x range set to:  -1,1\n",
      "INFO [2023-07-24 19:10:05] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
      "INFO [2023-07-24 19:10:05] Quantiles of plotted data range: 0,0,0,0,0.877166442491293\n",
      "INFO [2023-07-24 19:11:26] \n",
      "\n",
      "\tSTEP 19: Filter HMM predicted CNVs based on the Bayesian Network Model results and BayesMaxPNormal\n",
      "\n",
      "INFO [2023-07-24 19:11:26] Attempting to removing CNV(s) with a probability of being normal above  0.5\n",
      "INFO [2023-07-24 19:11:26] Removing  52  CNV(s) identified by the HMM.\n",
      "INFO [2023-07-24 19:11:26] Total CNV's after removing:  753\n",
      "INFO [2023-07-24 19:11:26] Reassigning CNVs based on state probabilities.\n",
      "INFO [2023-07-24 19:11:27] Changing the following CNV's states assigned by the HMM to the following based on the CNV's state probabilities.\n",
      " chr1-region_3 : 4  (P= 0.40137920113156 ) ->  3 (P= 0.494777977852614 )\n",
      "chr8-region_17 : 4  (P= 0.378146681283209 ) ->  3 (P= 0.430251543796532 )\n",
      "chr11-region_19 : 4  (P= 0.329062532090892 ) ->  3 (P= 0.453571990715805 )\n",
      "chr14-region_28 : 4  (P= 0.321308000796993 ) ->  3 (P= 0.472444942171673 )\n",
      "chr17-region_35 : 4  (P= 0.341745961009205 ) ->  3 (P= 0.439740905753577 )\n",
      "chr15-region_44 : 4  (P= 0.2968075888007 ) ->  3 (P= 0.489457316414182 )\n",
      "chr2-region_52 : 4  (P= 0.332669393944229 ) ->  3 (P= 0.498882556267384 )\n",
      "chr5-region_56 : 4  (P= 0.19623585470884 ) ->  3 (P= 0.487010553069892 )\n",
      "chr5-region_58 : 4  (P= 0.406293086905971 ) ->  3 (P= 0.420151315892335 )\n",
      "chr10-region_72 : 4  (P= 0.425715101399966 ) ->  3 (P= 0.47409664911745 )\n",
      "chr14-region_82 : 4  (P= 0.368722019004534 ) ->  3 (P= 0.389722301044578 )\n",
      "chr12-region_91 : 4  (P= 0.39543338320629 ) ->  3 (P= 0.48391903494001 )\n",
      "chr8-region_116 : 5  (P= 0.359974911634351 ) ->  4 (P= 0.360318515307218 )\n",
      "chr8-region_121 : 4  (P= 0.28000219576144 ) ->  3 (P= 0.412138798834857 )\n",
      "chr3-region_127 : 4  (P= 0.370202769722638 ) ->  3 (P= 0.424829684713862 )\n",
      "chr14-region_129 : 4  (P= 0.298183539543492 ) ->  3 (P= 0.469157161459319 )\n",
      "chr12-region_135 : 4  (P= 0.308774735357359 ) ->  3 (P= 0.382949327678705 )\n",
      "chr12-region_140 : 5  (P= 0.16930313065029 ) ->  4 (P= 0.318682248668357 )\n",
      "chr5-region_154 : 4  (P= 0.381833235598556 ) ->  3 (P= 0.412430320067347 )\n",
      "chr1-region_160 : 4  (P= 0.431722713765126 ) ->  3 (P= 0.465304437985531 )\n",
      "chr22-region_167 : 4  (P= 0.360159160674365 ) ->  3 (P= 0.495446199790469 )\n",
      "chr8-region_179 : 4  (P= 0.448005011588193 ) ->  3 (P= 0.479024404468 )\n",
      "chr11-region_189 : 4  (P= 0.391196007087313 ) ->  3 (P= 0.495802525869552 )\n",
      "chr3-region_194 : 4  (P= 0.402562402863541 ) ->  3 (P= 0.456168031007787 )\n",
      "chr17-region_206 : 4  (P= 0.44120751808283 ) ->  3 (P= 0.463943277685139 )\n",
      "chr5-region_225 : 4  (P= 0.422478325056918 ) ->  3 (P= 0.473243496376732 )\n",
      "chr16-region_250 : 4  (P= 0.394123247044917 ) ->  3 (P= 0.46506336273632 )\n",
      "chr10-region_253 : 4  (P= 0.400217046617817 ) ->  3 (P= 0.470910006528408 )\n",
      "chr7-region_278 : 4  (P= 0.421692592082358 ) ->  3 (P= 0.425691202917613 )\n",
      "chr17-region_280 : 4  (P= 0.356197852479429 ) ->  3 (P= 0.437596408150704 )\n",
      "chr12-region_287 : 4  (P= 0.451025595893757 ) ->  3 (P= 0.460316640992071 )\n",
      "chr20-region_301 : 4  (P= 0.393734183108239 ) ->  3 (P= 0.432821433012968 )\n",
      "chr19-region_306 : 4  (P= 0.416261767554183 ) ->  3 (P= 0.476955820386132 )\n",
      "chr5-region_312 : 5  (P= 0.363687808094734 ) ->  4 (P= 0.433852123649485 )\n",
      "chr13-region_317 : 4  (P= 0.336681988295549 ) ->  3 (P= 0.345564565743464 )\n",
      "chr1-region_322 : 4  (P= 0.412075487895143 ) ->  3 (P= 0.438344226457602 )\n",
      "chr1-region_324 : 4  (P= 0.437283969739293 ) ->  3 (P= 0.477203133330127 )\n",
      "chr10-region_331 : 4  (P= 0.39700020623965 ) ->  3 (P= 0.460116734800119 )\n",
      "chr3-region_348 : 4  (P= 0.38691963164804 ) ->  3 (P= 0.489091985664379 )\n",
      "chr7-region_356 : 4  (P= 0.399376374404205 ) ->  3 (P= 0.460682761991428 )\n",
      "chr4-region_363 : 4  (P= 0.381933983770938 ) ->  3 (P= 0.461862511742486 )\n",
      "chr5-region_371 : 4  (P= 0.438531458374745 ) ->  3 (P= 0.45228317058291 )\n",
      "chr6-region_399 : 4  (P= 0.406335771831663 ) ->  3 (P= 0.454217460741901 )\n",
      "chr5-region_424 : 4  (P= 0.399539318517609 ) ->  3 (P= 0.450584437678991 )\n",
      "chr1-region_433 : 4  (P= 0.46545245625897 ) ->  3 (P= 0.479705158245074 )\n",
      "chr16-region_438 : 4  (P= 0.33610304273984 ) ->  3 (P= 0.489875832742189 )\n",
      "chr16-region_440 : 4  (P= 0.387474456738644 ) ->  3 (P= 0.489176977416867 )\n",
      "chr11-region_452 : 4  (P= 0.372199702512008 ) ->  3 (P= 0.450512675998695 )\n",
      "chr17-region_459 : 4  (P= 0.446707525614569 ) ->  3 (P= 0.460405499420975 )\n",
      "chr12-region_466 : 4  (P= 0.361656225428132 ) ->  3 (P= 0.458635130566884 )\n",
      "chr5-region_481 : 4  (P= 0.41371367351652 ) ->  3 (P= 0.475011357100597 )\n",
      "chr1-region_486 : 4  (P= 0.356032947966216 ) ->  3 (P= 0.425961261901496 )\n",
      "chr1-region_490 : 4  (P= 0.389689904842906 ) ->  3 (P= 0.459834331001961 )\n",
      "chr8-region_508 : 4  (P= 0.371589041574248 ) ->  3 (P= 0.393963789123068 )\n",
      "chr8-region_509 : 5  (P= 0.307334319531718 ) ->  4 (P= 0.349410927169986 )\n",
      "chr6-region_513 : 4  (P= 0.354487420163331 ) ->  3 (P= 0.390767874971801 )\n",
      "chr11-region_518 : 4  (P= 0.356566058305605 ) ->  3 (P= 0.428786556447196 )\n",
      "chr3-region_523 : 2  (P= 0.465877801095927 ) ->  3 (P= 0.496191286624056 )\n",
      "chr14-region_525 : 4  (P= 0.344156449509014 ) ->  3 (P= 0.381572378198072 )\n",
      "chr14-region_529 : 4  (P= 0.342685789362549 ) ->  3 (P= 0.47249645528162 )\n",
      "chr17-region_532 : 4  (P= 0.414791589590957 ) ->  3 (P= 0.454249922317795 )\n",
      "chr19-region_551 : 4  (P= 0.410317684536817 ) ->  3 (P= 0.420256818551366 )\n",
      "chr5-region_554 : 4  (P= 0.392251667842122 ) ->  3 (P= 0.4753756256968 )\n",
      "chr5-region_556 : 4  (P= 0.374019459226021 ) ->  3 (P= 0.427041767396157 )\n",
      "chr10-region_567 : 4  (P= 0.421082552081072 ) ->  3 (P= 0.441493153009839 )\n",
      "chr4-region_588 : 4  (P= 0.349825303696769 ) ->  3 (P= 0.496516625931727 )\n",
      "chr15-region_593 : 4  (P= 0.274171368886586 ) ->  3 (P= 0.473990902598204 )\n",
      "chr5-region_602 : 4  (P= 0.39024950627609 ) ->  3 (P= 0.489473218133544 )\n",
      "chr13-region_606 : 4  (P= 0.316388390989405 ) ->  3 (P= 0.424682586687703 )\n",
      "chr1-region_608 : 4  (P= 0.267948228784064 ) ->  3 (P= 0.499200262034579 )\n",
      "chr16-region_611 : 4  (P= 0.367193666089508 ) ->  3 (P= 0.418430339680156 )\n",
      "chr9-region_616 : 4  (P= 0.280786364108303 ) ->  3 (P= 0.456241753998941 )\n",
      "chr11-region_619 : 4  (P= 0.232091628497048 ) ->  3 (P= 0.435728000564695 )\n",
      "chr1-region_642 : 4  (P= 0.414745281039201 ) ->  3 (P= 0.483806240104043 )\n",
      "chr6-region_659 : 4  (P= 0.406103477744854 ) ->  3 (P= 0.494124801624907 )\n",
      "chr14-region_665 : 4  (P= 0.41304902709678 ) ->  3 (P= 0.423805347745973 )\n",
      "chr15-region_677 : 4  (P= 0.33948573041557 ) ->  3 (P= 0.430008181996748 )\n",
      "chr5-region_689 : 4  (P= 0.395792837835908 ) ->  3 (P= 0.469489715531177 )\n",
      "chr13-region_693 : 4  (P= 0.243937644961569 ) ->  3 (P= 0.47386497039688 )\n",
      "chr1-region_696 : 4  (P= 0.381401716143251 ) ->  3 (P= 0.484766985907143 )\n",
      "chr16-region_699 : 4  (P= 0.388321581384454 ) ->  3 (P= 0.439459351590744 )\n",
      "chr14-region_715 : 4  (P= 0.369529684213942 ) ->  3 (P= 0.439486832990433 )\n",
      "chr5-region_737 : 4  (P= 0.399427547540539 ) ->  3 (P= 0.480705331646177 )\n",
      "chr16-region_751 : 4  (P= 0.376194076795967 ) ->  3 (P= 0.471588011924766 )\n",
      "chr10-region_755 : 4  (P= 0.37916797550609 ) ->  3 (P= 0.484352900184377 )\n",
      "chr3-region_775 : 4  (P= 0.294757712400893 ) ->  3 (P= 0.473965524636071 )\n",
      "chr14-region_778 : 4  (P= 0.376409438807667 ) ->  3 (P= 0.377329626159547 )\n",
      "chr17-region_786 : 4  (P= 0.395623484382731 ) ->  3 (P= 0.487721179157171 )\n",
      "chr17-region_788 : 4  (P= 0.446313185901012 ) ->  3 (P= 0.454366541384789 )\n",
      "chr15-region_798 : 4  (P= 0.347251532457964 ) ->  3 (P= 0.440336419367047 )\n",
      "chr14-region_838 : 4  (P= 0.369811636823138 ) ->  3 (P= 0.484384089187715 )\n",
      "chr5-region_855 : 4  (P= 0.330544385743829 ) ->  3 (P= 0.475891750933445 )\n",
      "chr1-region_863 : 4  (P= 0.32651009462943 ) ->  3 (P= 0.396192089486673 )\n",
      "chr16-region_866 : 5  (P= 0.212880724776236 ) ->  4 (P= 0.370198401739457 )\n",
      "chr21-region_872 : 4  (P= 0.231097897547125 ) ->  3 (P= 0.404583981618791 )\n",
      "chr9-region_877 : 4  (P= 0.366361159200921 ) ->  3 (P= 0.45247026097902 )\n",
      "chr8-region_880 : 4  (P= 0.272968222920242 ) ->  3 (P= 0.455291370982052 )\n",
      "chr14-region_888 : 4  (P= 0.326071964779344 ) ->  3 (P= 0.40819348381342 )\n",
      "chr7-region_891 : 4  (P= 0.317444942314618 ) ->  3 (P= 0.362958148456481 )\n",
      "chr7-region_895 : 4  (P= 0.272699405932115 ) ->  3 (P= 0.273060108283126 )\n",
      "chr17-region_898 : 2  (P= 0.365012653867634 ) ->  3 (P= 0.453764385464608 )\n",
      "chr15-region_908 : 4  (P= 0.371112560402091 ) ->  3 (P= 0.400065116781457 )\n",
      "chr20-region_914 : 4  (P= 0.181995366858728 ) ->  3 (P= 0.362762470377907 )\n",
      "chr13-region_922 : 4  (P= 0.317130168267896 ) ->  3 (P= 0.409346674618177 )\n",
      "chr1-region_924 : 4  (P= 0.252973560690649 ) ->  3 (P= 0.484043792640586 )\n",
      "chr16-region_931 : 4  (P= 0.294706052063553 ) ->  3 (P= 0.447692009152397 )\n",
      "chr9-region_939 : 4  (P= 0.374061578354573 ) ->  3 (P= 0.467113614971307 )\n",
      "chr14-region_945 : 4  (P= 0.315872604150585 ) ->  3 (P= 0.465324332794046 )\n",
      "chr17-region_953 : 4  (P= 0.429748305797076 ) ->  3 (P= 0.441879779575324 )\n",
      "chr19-region_971 : 4  (P= 0.32660299406963 ) ->  3 (P= 0.403590339264874 )\n",
      "chr1-region_978 : 4  (P= 0.343813355364614 ) ->  3 (P= 0.36126172188169 )\n",
      "chr1-region_980 : 4  (P= 0.420345832716722 ) ->  3 (P= 0.460451723282517 )\n",
      "chr5-region_1006 : 4  (P= 0.417324592288297 ) ->  3 (P= 0.473516909405289 )\n",
      "chr1-region_1015 : 4  (P= 0.189310673880335 ) ->  3 (P= 0.436680197126312 )\n",
      "chr16-region_1018 : 4  (P= 0.258885681340693 ) ->  3 (P= 0.3555791832423 )\n",
      "chr8-region_1026 : 4  (P= 0.308600286704253 ) ->  3 (P= 0.42771366853087 )\n",
      "chr6-region_1032 : 4  (P= 0.243123185514182 ) ->  3 (P= 0.319905192320606 )\n",
      "chr6-region_1034 : 4  (P= 0.189575430733043 ) ->  3 (P= 0.437429034566142 )\n",
      "chr3-region_1042 : 4  (P= 0.15059661158831 ) ->  3 (P= 0.488885375490523 )\n",
      "chr14-region_1044 : 5  (P= 0.238934192604298 ) ->  4 (P= 0.289466205319221 )\n",
      "chr14-region_1045 : 4  (P= 0.183811608553447 ) ->  3 (P= 0.373894565482471 )\n",
      "chr14-region_1047 : 4  (P= 0.186539421633835 ) ->  3 (P= 0.439220035283822 )\n",
      "chr7-region_1050 : 4  (P= 0.164380215881816 ) ->  3 (P= 0.453179399619244 )\n",
      "chr7-region_1052 : 4  (P= 0.334475340228665 ) ->  3 (P= 0.374954842933974 )\n",
      "chr17-region_1055 : 4  (P= 0.311672980043368 ) ->  3 (P= 0.375670187433515 )\n",
      "chr17-region_1057 : 2  (P= 0.271386255092368 ) ->  3 (P= 0.416251787314606 )\n",
      "chr12-region_1062 : 4  (P= 0.310253457774376 ) ->  3 (P= 0.312761186077671 )\n",
      "chr12-region_1064 : 4  (P= 0.187446994903545 ) ->  3 (P= 0.437585744254172 )\n",
      "chr2-region_1072 : 4  (P= 0.128873559453998 ) ->  3 (P= 0.37229620569136 )\n",
      "chr2-region_1074 : 4  (P= 0.250246722584238 ) ->  5 (P= 0.310593127836753 )\n",
      "chr19-region_1077 : 4  (P= 0.186911110767207 ) ->  3 (P= 0.373680338400602 )\n",
      "chr1-region_1087 : 5  (P= 0.199185665580911 ) ->  4 (P= 0.226977151313184 )\n",
      "chr3-region_1144 : 1  (P= 0.143516300988314 ) ->  2 (P= 0.285720973813898 )\n",
      "chr14-region_1148 : 4  (P= 0.186027181352736 ) ->  3 (P= 0.237063923947907 )\n",
      "chr5-region_1209 : 5  (P= 0.173995192877753 ) ->  4 (P= 0.258298829754098 )\n",
      "chr1-region_1214 : 4  (P= 0.312473068649691 ) ->  3 (P= 0.427567626783581 )\n",
      "chr9-region_1242 : 5  (P= 0.373181157240326 ) ->  4 (P= 0.403171585042007 )\n",
      "chr8-region_1247 : 4  (P= 0.300949718676519 ) ->  3 (P= 0.476790539014913 )\n",
      "chr12-region_1276 : 4  (P= 0.339542128274935 ) ->  3 (P= 0.464862477282115 )\n",
      "chr2-region_1298 : 4  (P= 0.395187945075043 ) ->  3 (P= 0.441568668496329 )\n",
      "chr2-region_1302 : 4  (P= 0.310630104353724 ) ->  3 (P= 0.495092995196745 )\n",
      "chr20-region_1307 : 4  (P= 0.304263054596522 ) ->  3 (P= 0.497198802139122 )\n",
      "chr19-region_1309 : 4  (P= 0.366647877884414 ) ->  3 (P= 0.404680743885641 )\n",
      "chr5-region_1314 : 4  (P= 0.361416044069978 ) ->  3 (P= 0.438474458885444 )\n",
      "chr5-region_1322 : 4  (P= 0.387752864219101 ) ->  3 (P= 0.456621935281281 )\n",
      "chr1-region_1324 : 5  (P= 0.176351498544804 ) ->  4 (P= 0.293512099471166 )\n",
      "chr16-region_1333 : 5  (P= 0.177488789830097 ) ->  3 (P= 0.27721703183134 )\n",
      "chr10-region_1340 : 2  (P= 0.356589697890784 ) ->  3 (P= 0.407614134654449 )\n",
      "chr9-region_1346 : 4  (P= 0.176000528705464 ) ->  3 (P= 0.411702159643735 )\n",
      "chr8-region_1351 : 4  (P= 0.353365499310361 ) ->  3 (P= 0.411837404442496 )\n",
      "chr6-region_1356 : 4  (P= 0.236696035117733 ) ->  3 (P= 0.352797712853356 )\n",
      "chr6-region_1358 : 4  (P= 0.227622814329725 ) ->  3 (P= 0.357055833285298 )\n",
      "chr12-region_1381 : 4  (P= 0.153327138281139 ) ->  3 (P= 0.47320742391968 )\n",
      "chr12-region_1387 : 4  (P= 0.293687202619489 ) ->  3 (P= 0.295590888117779 )\n",
      "chr12-region_1389 : 4  (P= 0.302303253926301 ) ->  3 (P= 0.34401584244378 )\n",
      "chr4-region_1390 : 4  (P= 0.253932914729564 ) ->  5 (P= 0.273168366356311 )\n",
      "chr4-region_1392 : 4  (P= 0.330308189411776 ) ->  3 (P= 0.350729238573204 )\n",
      "chr15-region_1400 : 5  (P= 0.236546215064974 ) ->  4 (P= 0.29386782549061 )\n",
      "chr2-region_1406 : 4  (P= 0.371440022672189 ) ->  3 (P= 0.396474172065082 )\n",
      "chr2-region_1408 : 4  (P= 0.306770185454437 ) ->  3 (P= 0.376742857566941 )\n",
      "chr20-region_1409 : 5  (P= 0.269820188988067 ) ->  3 (P= 0.292605137287787 )\n",
      "chr20-region_1410 : 4  (P= 0.295335746889897 ) ->  3 (P= 0.295381903409532 )\n",
      "chr20-region_1413 : 5  (P= 0.293368782894498 ) ->  4 (P= 0.353071157373749 )\n",
      "chr20-region_1416 : 4  (P= 0.191731822667396 ) ->  3 (P= 0.337745298192756 )\n",
      "chr19-region_1421 : 4  (P= 0.294756952233267 ) ->  3 (P= 0.411907994777738 )\n",
      "chr6-region_1597 : 4  (P= 0.213098776525241 ) ->  3 (P= 0.464688234016736 )\n",
      "chr1-region_1617 : 4  (P= 0.249373915666667 ) ->  3 (P= 0.251713494242599 )\n",
      "chr16-region_1621 : 4  (P= 0.247494274498868 ) ->  3 (P= 0.252069057495198 )\n",
      "chr16-region_1623 : 4  (P= 0.247494274498868 ) ->  3 (P= 0.252069057495198 )\n",
      "chr16-region_1625 : 2  (P= 0.249293401311196 ) ->  3 (P= 0.252017054066811 )\n",
      "chr10-region_1628 : 2  (P= 0.249293401311196 ) ->  3 (P= 0.252017054066811 )\n",
      "chr9-region_1638 : 2  (P= 0.249229086477749 ) ->  3 (P= 0.25086238880443 )\n",
      "chr8-region_1640 : 4  (P= 0.124722228610485 ) ->  3 (P= 0.249602809519366 )\n",
      "chr8-region_1644 : 4  (P= 0.248578711044093 ) ->  3 (P= 0.249041135828671 )\n",
      "chr6-region_1646 : 4  (P= 0.247645306380682 ) ->  3 (P= 0.248497697754756 )\n",
      "chr11-region_1650 : 4  (P= 0.24713650948416 ) ->  3 (P= 0.252638942213014 )\n",
      "chr3-region_1651 : 4  (P= 0.24713650948416 ) ->  3 (P= 0.252638942213014 )\n",
      "chr3-region_1653 : 4  (P= 0.124820215781368 ) ->  3 (P= 0.252561032406788 )\n",
      "chr3-region_1655 : 4  (P= 0.248651575855881 ) ->  3 (P= 0.250851556357306 )\n",
      "chr14-region_1657 : 4  (P= 0.246865748027222 ) ->  3 (P= 0.250292691265795 )\n",
      "chr17-region_1664 : 4  (P= 0.126510352499197 ) ->  5 (P= 0.250064787504001 )\n",
      "chr20-region_1682 : 5  (P= 0.247374812970788 ) ->  4 (P= 0.24847777442246 )\n",
      "chr5-region_1692 : 2  (P= 0.247269071550275 ) ->  3 (P= 0.249780013931104 )\n",
      "chr5-region_1694 : 4  (P= 0.151861818422342 ) ->  3 (P= 0.247986870396487 )\n",
      "chr22-region_1704 : 4  (P= 0.246055898497628 ) ->  3 (P= 0.24869043482913 )\n",
      "chr3-region_1725 : 4  (P= 0.246092553780366 ) ->  3 (P= 0.252593027005566 )\n",
      "chr3-region_1727 : 4  (P= 0.124401881340592 ) ->  5 (P= 0.25101816502254 )\n",
      "chr17-region_1745 : 4  (P= 0.24978849092078 ) ->  3 (P= 0.251229782514057 )\n",
      "chr5-region_1770 : 4  (P= 0.127369974914621 ) ->  5 (P= 0.252500895501832 )\n",
      "chr5-region_1772 : 4  (P= 0.127369974914621 ) ->  5 (P= 0.252500895501832 )\n",
      "chr13-region_1775 : 4  (P= 0.134796504796683 ) ->  3 (P= 0.249061495797505 )\n",
      "chr1-region_2027 : 4  (P= 0.406485830254143 ) ->  3 (P= 0.444383454277058 )\n",
      "chr22-region_2056 : 4  (P= 0.342793116411876 ) ->  3 (P= 0.438641941501435 )\n",
      "chr1-region_2081 : 4  (P= 0.124291639135079 ) ->  2 (P= 0.252465827983909 )\n",
      "chr22-region_2084 : 4  (P= 0.24707851158769 ) ->  3 (P= 0.252605377858253 )\n",
      "chr16-region_2087 : 4  (P= 0.12570825336386 ) ->  5 (P= 0.253426392152586 )\n",
      "chr3-region_2098 : 4  (P= 0.250102935108257 ) ->  3 (P= 0.2524266316703 )\n",
      "chr3-region_2100 : 4  (P= 0.249951404632687 ) ->  3 (P= 0.25288551041847 )\n",
      "chr3-region_2102 : 4  (P= 0.248402005153062 ) ->  3 (P= 0.253265466803975 )\n",
      "chr14-region_2106 : 4  (P= 0.248239577074504 ) ->  3 (P= 0.251611030570333 )\n",
      "chr17-region_2111 : 4  (P= 0.12561530632885 ) ->  3 (P= 0.254175115460203 )\n",
      "chr12-region_2118 : 2  (P= 0.249035913235502 ) ->  3 (P= 0.254157007986485 )\n",
      "chr4-region_2120 : 4  (P= 0.250183057513243 ) ->  3 (P= 0.250368784075159 )\n",
      "chr2-region_2125 : 2  (P= 0.249035913235502 ) ->  3 (P= 0.254157007986485 )\n",
      "chr13-region_2140 : 4  (P= 0.123296966815804 ) ->  3 (P= 0.250589721920664 )\n",
      "chr22-region_2144 : 4  (P= 0.143804678433896 ) ->  3 (P= 0.250595043500635 )\n",
      "chr16-region_2146 : 2  (P= 0.249908502863324 ) ->  3 (P= 0.250108559671328 )\n",
      "chr10-region_2149 : 4  (P= 0.244817237898802 ) ->  3 (P= 0.252280501856967 )\n",
      "chr3-region_2161 : 4  (P= 0.246792767148531 ) ->  3 (P= 0.255574525729671 )\n",
      "chr3-region_2163 : 2  (P= 0.249815284905761 ) ->  3 (P= 0.250976784750991 )\n",
      "chr7-region_2166 : 4  (P= 0.156079807529111 ) ->  3 (P= 0.253182992340061 )\n",
      "chr7-region_2168 : 4  (P= 0.246767635954624 ) ->  3 (P= 0.255605262094333 )\n",
      "chr7-region_2170 : 4  (P= 0.244735308016145 ) ->  3 (P= 0.24976209620996 )\n",
      "chr17-region_2174 : 4  (P= 0.247004912938751 ) ->  3 (P= 0.249081074862415 )\n",
      "chr12-region_2177 : 4  (P= 0.173535586656407 ) ->  3 (P= 0.250914981641188 )\n",
      "chr12-region_2181 : 2  (P= 0.25060444315982 ) ->  3 (P= 0.252055490912667 )\n",
      "chr20-region_2198 : 4  (P= 0.250514895948047 ) ->  3 (P= 0.252502734106984 )\n",
      "chr5-region_2203 : 4  (P= 0.250514895948047 ) ->  3 (P= 0.252502734106984 )\n",
      "chr13-region_2205 : 4  (P= 0.125402656228628 ) ->  3 (P= 0.250224166403555 )\n",
      "INFO [2023-07-24 19:11:28] Creating Plots for CNV and cell Probabilities.\n",
      "INFO [2023-07-24 19:24:09] ::plot_cnv:Start\n",
      "INFO [2023-07-24 19:24:09] ::plot_cnv:Current data dimensions (r,c)=12073,6951 Total=5280391.24026929 Min=0 Max=0.989471219982526.\n",
      "INFO [2023-07-24 19:24:10] ::plot_cnv:Depending on the size of the matrix this may take a moment.\n",
      "INFO [2023-07-24 19:24:13] plot_cnv_observation:Start\n",
      "INFO [2023-07-24 19:24:13] Observation data size: Cells= 4371 Genes= 12073\n",
      "INFO [2023-07-24 19:24:13] plot_cnv_observation:Writing observation groupings/color.\n",
      "INFO [2023-07-24 19:24:13] plot_cnv_observation:Done writing observation groupings/color.\n",
      "INFO [2023-07-24 19:24:14] plot_cnv_observation:Writing observation heatmap thresholds.\n",
      "INFO [2023-07-24 19:24:14] plot_cnv_observation:Done writing observation heatmap thresholds.\n",
      "DEBUG [2023-07-24 19:24:14] inferCNV::heatmap.cnv, breaks parameter set to: [0,0.0666666666666667,0.133333333333333,0.2,0.266666666666667,0.333333333333333,0.4,0.466666666666667,0.533333333333333,0.6,0.666666666666667,0.733333333333333,0.8,0.866666666666667,0.933333333333333,1]\n",
      "DEBUG [2023-07-24 19:24:15] inferCNV::heatmap.cnv, resetting breaks to adjusted x.range: [-1,-0.866666666666667,-0.733333333333333,-0.6,-0.466666666666667,-0.333333333333333,-0.2,-0.0666666666666667,0.0666666666666667,0.2,0.333333333333333,0.466666666666667,0.6,0.733333333333333,0.866666666666667,1]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in (function (fmt, ...) :\n",
      "“one argument not used by format 'inferCNV::heatmap.cnv x range set to:  -1,1'”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DEBUG [2023-07-24 19:24:15] inferCNV::heatmap.cnv x range set to:  -1,1\n",
      "INFO [2023-07-24 19:24:26] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
      "INFO [2023-07-24 19:24:26] Quantiles of plotted data range: 0,0,0,0,0.989471219982526\n",
      "DEBUG [2023-07-24 19:24:32] ::inferCNV::heatmap.cnv colorkey z range: -1,-0.857142857142857,-0.714285714285714,-0.571428571428571,-0.428571428571429,-0.285714285714286,-0.142857142857143,0,0.142857142857143,0.285714285714286,0.428571428571428,0.571428571428571,0.714285714285714,0.857142857142857,1\n",
      "DEBUG [2023-07-24 19:24:32] ::inferCNV::heatmap.cnv colorkey breaks range: -1,-0.866666666666667,-0.733333333333333,-0.6,-0.466666666666667,-0.333333333333333,-0.2,-0.0666666666666667,0.0666666666666667,0.2,0.333333333333333,0.466666666666667,0.6,0.733333333333333,0.866666666666667,1\n",
      "DEBUG [2023-07-24 19:24:32] ::inferCNV::heatmap.cnv colorkey colors range: #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
      "INFO [2023-07-24 19:24:32] plot_cnv_references:Start\n",
      "INFO [2023-07-24 19:24:32] Reference data size: Cells= 2580 Genes= 12073\n",
      "INFO [2023-07-24 19:25:29] plot_cnv_references:Number reference groups= 9\n",
      "INFO [2023-07-24 19:25:30] plot_cnv_references:Plotting heatmap.\n",
      "DEBUG [2023-07-24 19:25:30] inferCNV::heatmap.cnv, breaks parameter set to: [0,0.0666666666666667,0.133333333333333,0.2,0.266666666666667,0.333333333333333,0.4,0.466666666666667,0.533333333333333,0.6,0.666666666666667,0.733333333333333,0.8,0.866666666666667,0.933333333333333,1]\n",
      "DEBUG [2023-07-24 19:25:30] inferCNV::heatmap.cnv, resetting breaks to adjusted x.range: [-1,-0.866666666666667,-0.733333333333333,-0.6,-0.466666666666667,-0.333333333333333,-0.2,-0.0666666666666667,0.0666666666666667,0.2,0.333333333333333,0.466666666666667,0.6,0.733333333333333,0.866666666666667,1]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in (function (fmt, ...) :\n",
      "“one argument not used by format 'inferCNV::heatmap.cnv x range set to:  -1,1'”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DEBUG [2023-07-24 19:25:30] inferCNV::heatmap.cnv x range set to:  -1,1\n",
      "INFO [2023-07-24 19:25:36] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
      "INFO [2023-07-24 19:25:36] Quantiles of plotted data range: 0,0,0,0,0.877166442491293\n",
      "INFO [2023-07-24 19:26:23] ::plot_cnv:Start\n",
      "INFO [2023-07-24 19:26:23] ::plot_cnv:Current data dimensions (r,c)=12073,6951 Total=253316067 Min=2 Max=6.\n",
      "INFO [2023-07-24 19:26:24] ::plot_cnv:Depending on the size of the matrix this may take a moment.\n",
      "INFO [2023-07-24 19:26:26] plot_cnv_observation:Start\n",
      "INFO [2023-07-24 19:26:26] Observation data size: Cells= 4371 Genes= 12073\n",
      "INFO [2023-07-24 19:26:27] plot_cnv_observation:Writing observation groupings/color.\n",
      "INFO [2023-07-24 19:26:27] plot_cnv_observation:Done writing observation groupings/color.\n",
      "INFO [2023-07-24 19:26:27] plot_cnv_observation:Writing observation heatmap thresholds.\n",
      "INFO [2023-07-24 19:26:27] plot_cnv_observation:Done writing observation heatmap thresholds.\n",
      "DEBUG [2023-07-24 19:26:27] inferCNV::heatmap.cnv, breaks parameter set to: [0,0.4,0.8,1.2,1.6,2,2.4,2.8,3.2,3.6,4,4.4,4.8,5.2,5.6,6]\n",
      "DEBUG [2023-07-24 19:26:28] inferCNV::heatmap.cnv, resetting breaks to adjusted x.range: [0,0.4,0.8,1.2,1.6,2,2.4,2.8,3.2,3.6,4,4.4,4.8,5.2,5.6,6]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in (function (fmt, ...) :\n",
      "“one argument not used by format 'inferCNV::heatmap.cnv x range set to:  0,6'”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DEBUG [2023-07-24 19:26:28] inferCNV::heatmap.cnv x range set to:  0,6\n",
      "INFO [2023-07-24 19:26:38] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
      "INFO [2023-07-24 19:26:38] Quantiles of plotted data range: 2,3,3,3,6\n",
      "DEBUG [2023-07-24 19:26:44] ::inferCNV::heatmap.cnv colorkey z range: 0,0.428571428571429,0.857142857142857,1.28571428571429,1.71428571428571,2.14285714285714,2.57142857142857,3,3.42857142857143,3.85714285714286,4.28571428571429,4.71428571428571,5.14285714285714,5.57142857142857,6\n",
      "DEBUG [2023-07-24 19:26:44] ::inferCNV::heatmap.cnv colorkey breaks range: 0,0.4,0.8,1.2,1.6,2,2.4,2.8,3.2,3.6,4,4.4,4.8,5.2,5.6,6\n",
      "DEBUG [2023-07-24 19:26:44] ::inferCNV::heatmap.cnv colorkey colors range: #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
      "INFO [2023-07-24 19:26:45] plot_cnv_references:Start\n",
      "INFO [2023-07-24 19:26:45] Reference data size: Cells= 2580 Genes= 12073\n",
      "INFO [2023-07-24 19:27:42] plot_cnv_references:Number reference groups= 9\n",
      "INFO [2023-07-24 19:27:43] plot_cnv_references:Plotting heatmap.\n",
      "DEBUG [2023-07-24 19:27:43] inferCNV::heatmap.cnv, breaks parameter set to: [0,0.4,0.8,1.2,1.6,2,2.4,2.8,3.2,3.6,4,4.4,4.8,5.2,5.6,6]\n",
      "DEBUG [2023-07-24 19:27:44] inferCNV::heatmap.cnv, resetting breaks to adjusted x.range: [0,0.4,0.8,1.2,1.6,2,2.4,2.8,3.2,3.6,4,4.4,4.8,5.2,5.6,6]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in (function (fmt, ...) :\n",
      "“one argument not used by format 'inferCNV::heatmap.cnv x range set to:  0,6'”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DEBUG [2023-07-24 19:27:44] inferCNV::heatmap.cnv x range set to:  0,6\n",
      "INFO [2023-07-24 19:27:50] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
      "INFO [2023-07-24 19:27:50] Quantiles of plotted data range: 2,3,3,3,5\n",
      "INFO [2023-07-24 19:27:55] \n",
      "\n",
      "\tSTEP 20: Converting HMM-based CNV states to repr expr vals\n",
      "\n",
      "INFO [2023-07-24 19:28:42] ::plot_cnv:Start\n",
      "INFO [2023-07-24 19:28:42] ::plot_cnv:Current data dimensions (r,c)=12073,6951 Total=84698368 Min=0.5 Max=3.\n",
      "INFO [2023-07-24 19:28:43] ::plot_cnv:Depending on the size of the matrix this may take a moment.\n",
      "INFO [2023-07-24 19:28:45] plot_cnv_observation:Start\n",
      "INFO [2023-07-24 19:28:45] Observation data size: Cells= 4371 Genes= 12073\n",
      "INFO [2023-07-24 19:28:46] plot_cnv_observation:Writing observation groupings/color.\n",
      "INFO [2023-07-24 19:28:46] plot_cnv_observation:Done writing observation groupings/color.\n",
      "INFO [2023-07-24 19:28:46] plot_cnv_observation:Writing observation heatmap thresholds.\n",
      "INFO [2023-07-24 19:28:46] plot_cnv_observation:Done writing observation heatmap thresholds.\n",
      "DEBUG [2023-07-24 19:28:47] inferCNV::heatmap.cnv, breaks parameter set to: [-1,-0.733333333333333,-0.466666666666667,-0.2,0.0666666666666667,0.333333333333333,0.6,0.866666666666667,1.13333333333333,1.4,1.66666666666667,1.93333333333333,2.2,2.46666666666667,2.73333333333333,3]\n",
      "DEBUG [2023-07-24 19:28:47] inferCNV::heatmap.cnv, resetting breaks to adjusted x.range: [-1,-0.733333333333333,-0.466666666666667,-0.2,0.0666666666666667,0.333333333333333,0.6,0.866666666666667,1.13333333333333,1.4,1.66666666666667,1.93333333333333,2.2,2.46666666666667,2.73333333333333,3]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in (function (fmt, ...) :\n",
      "“one argument not used by format 'inferCNV::heatmap.cnv x range set to:  -1,3'”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DEBUG [2023-07-24 19:28:47] inferCNV::heatmap.cnv x range set to:  -1,3\n",
      "INFO [2023-07-24 19:28:57] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
      "INFO [2023-07-24 19:28:57] Quantiles of plotted data range: 0.5,1,1,1,3\n",
      "DEBUG [2023-07-24 19:29:03] ::inferCNV::heatmap.cnv colorkey z range: -1,-0.714285714285714,-0.428571428571429,-0.142857142857143,0.142857142857143,0.428571428571428,0.714285714285714,1,1.28571428571429,1.57142857142857,1.85714285714286,2.14285714285714,2.42857142857143,2.71428571428571,3\n",
      "DEBUG [2023-07-24 19:29:03] ::inferCNV::heatmap.cnv colorkey breaks range: -1,-0.733333333333333,-0.466666666666667,-0.2,0.0666666666666667,0.333333333333333,0.6,0.866666666666667,1.13333333333333,1.4,1.66666666666667,1.93333333333333,2.2,2.46666666666667,2.73333333333333,3\n",
      "DEBUG [2023-07-24 19:29:03] ::inferCNV::heatmap.cnv colorkey colors range: #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
      "INFO [2023-07-24 19:29:04] plot_cnv_references:Start\n",
      "INFO [2023-07-24 19:29:04] Reference data size: Cells= 2580 Genes= 12073\n",
      "INFO [2023-07-24 19:30:01] plot_cnv_references:Number reference groups= 9\n",
      "INFO [2023-07-24 19:30:02] plot_cnv_references:Plotting heatmap.\n",
      "DEBUG [2023-07-24 19:30:02] inferCNV::heatmap.cnv, breaks parameter set to: [-1,-0.733333333333333,-0.466666666666667,-0.2,0.0666666666666667,0.333333333333333,0.6,0.866666666666667,1.13333333333333,1.4,1.66666666666667,1.93333333333333,2.2,2.46666666666667,2.73333333333333,3]\n",
      "DEBUG [2023-07-24 19:30:03] inferCNV::heatmap.cnv, resetting breaks to adjusted x.range: [-1,-0.733333333333333,-0.466666666666667,-0.2,0.0666666666666667,0.333333333333333,0.6,0.866666666666667,1.13333333333333,1.4,1.66666666666667,1.93333333333333,2.2,2.46666666666667,2.73333333333333,3]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in (function (fmt, ...) :\n",
      "“one argument not used by format 'inferCNV::heatmap.cnv x range set to:  -1,3'”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DEBUG [2023-07-24 19:30:03] inferCNV::heatmap.cnv x range set to:  -1,3\n",
      "INFO [2023-07-24 19:30:09] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
      "INFO [2023-07-24 19:30:09] Quantiles of plotted data range: 0.5,1,1,1,2\n",
      "INFO [2023-07-24 19:30:12] \n",
      "\n",
      "\tSTEP 22: Denoising\n",
      "\n",
      "INFO [2023-07-24 19:30:12] ::process_data:Remove noise, noise threshold defined via ref mean sd_amplifier:  1.5\n",
      "INFO [2023-07-24 19:30:12] denoising using mean(normal) +- sd_amplifier * sd(normal) per gene per cell across all data\n",
      "INFO [2023-07-24 19:30:13] :: **** clear_noise_via_ref_quantiles **** : removing noise between bounds:  0.906554811719053 - 1.09819581708522\n",
      "INFO [2023-07-24 19:31:49] \n",
      "\n",
      "## Making the final infercnv heatmap ##\n",
      "INFO [2023-07-24 19:31:50] ::plot_cnv:Start\n",
      "INFO [2023-07-24 19:31:50] ::plot_cnv:Current data dimensions (r,c)=12073,6951 Total=84513229.2665191 Min=0.49605252441373 Max=2.22182072260546.\n",
      "INFO [2023-07-24 19:31:51] ::plot_cnv:Depending on the size of the matrix this may take a moment.\n",
      "INFO [2023-07-24 19:31:54] plot_cnv(): auto thresholding at: (0.725417 , 1.274583)\n",
      "INFO [2023-07-24 19:31:58] plot_cnv_observation:Start\n",
      "INFO [2023-07-24 19:31:58] Observation data size: Cells= 4371 Genes= 12073\n",
      "INFO [2023-07-24 19:31:58] plot_cnv_observation:Writing observation groupings/color.\n",
      "INFO [2023-07-24 19:31:58] plot_cnv_observation:Done writing observation groupings/color.\n",
      "INFO [2023-07-24 19:31:59] plot_cnv_observation:Writing observation heatmap thresholds.\n",
      "INFO [2023-07-24 19:31:59] plot_cnv_observation:Done writing observation heatmap thresholds.\n",
      "DEBUG [2023-07-24 19:31:59] inferCNV::heatmap.cnv, breaks parameter set to: [0.72541708269308,0.762028138334002,0.798639193974925,0.835250249615848,0.871861305256771,0.908472360897693,0.945083416538616,0.981694472179539,1.01830552782046,1.05491658346138,1.09152763910231,1.12813869474323,1.16474975038415,1.20136080602507,1.237971861666,1.27458291730692]\n",
      "DEBUG [2023-07-24 19:32:00] inferCNV::heatmap.cnv, resetting breaks to adjusted x.range: [0.72541708269308,0.762028138334002,0.798639193974925,0.835250249615848,0.871861305256771,0.908472360897693,0.945083416538616,0.981694472179539,1.01830552782046,1.05491658346138,1.09152763910231,1.12813869474323,1.16474975038415,1.20136080602507,1.237971861666,1.27458291730692]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in (function (fmt, ...) :\n",
      "“one argument not used by format 'inferCNV::heatmap.cnv x range set to:  0.72541708269308,1.27458291730692'”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DEBUG [2023-07-24 19:32:00] inferCNV::heatmap.cnv x range set to:  0.72541708269308,1.27458291730692\n",
      "INFO [2023-07-24 19:32:09] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
      "INFO [2023-07-24 19:32:09] Quantiles of plotted data range: 0.72541708269308,1.00237531440214,1.00237531440214,1.00237531440214,1.27458291730692\n",
      "DEBUG [2023-07-24 19:32:15] ::inferCNV::heatmap.cnv colorkey z range: 0.72541708269308,0.764643213736925,0.803869344780771,0.843095475824617,0.882321606868463,0.921547737912308,0.960773868956154,1,1.03922613104385,1.07845226208769,1.11767839313154,1.15690452417538,1.19613065521923,1.23535678626307,1.27458291730692\n",
      "DEBUG [2023-07-24 19:32:15] ::inferCNV::heatmap.cnv colorkey breaks range: 0.72541708269308,0.762028138334002,0.798639193974925,0.835250249615848,0.871861305256771,0.908472360897693,0.945083416538616,0.981694472179539,1.01830552782046,1.05491658346138,1.09152763910231,1.12813869474323,1.16474975038415,1.20136080602507,1.237971861666,1.27458291730692\n",
      "DEBUG [2023-07-24 19:32:15] ::inferCNV::heatmap.cnv colorkey colors range: #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
      "INFO [2023-07-24 19:32:16] plot_cnv_references:Start\n",
      "INFO [2023-07-24 19:32:16] Reference data size: Cells= 2580 Genes= 12073\n",
      "INFO [2023-07-24 19:33:13] plot_cnv_references:Number reference groups= 9\n",
      "INFO [2023-07-24 19:33:13] plot_cnv_references:Plotting heatmap.\n",
      "DEBUG [2023-07-24 19:33:14] inferCNV::heatmap.cnv, breaks parameter set to: [0.72541708269308,0.762028138334002,0.798639193974925,0.835250249615848,0.871861305256771,0.908472360897693,0.945083416538616,0.981694472179539,1.01830552782046,1.05491658346138,1.09152763910231,1.12813869474323,1.16474975038415,1.20136080602507,1.237971861666,1.27458291730692]\n",
      "DEBUG [2023-07-24 19:33:14] inferCNV::heatmap.cnv, resetting breaks to adjusted x.range: [0.72541708269308,0.762028138334002,0.798639193974925,0.835250249615848,0.871861305256771,0.908472360897693,0.945083416538616,0.981694472179539,1.01830552782046,1.05491658346138,1.09152763910231,1.12813869474323,1.16474975038415,1.20136080602507,1.237971861666,1.27458291730692]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in (function (fmt, ...) :\n",
      "“one argument not used by format 'inferCNV::heatmap.cnv x range set to:  0.72541708269308,1.27458291730692'”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DEBUG [2023-07-24 19:33:14] inferCNV::heatmap.cnv x range set to:  0.72541708269308,1.27458291730692\n",
      "INFO [2023-07-24 19:33:20] Colors for breaks:  #00008B,#24249B,#4848AB,#6D6DBC,#9191CC,#B6B6DD,#DADAEE,#FFFFFF,#EEDADA,#DDB6B6,#CC9191,#BC6D6D,#AB4848,#9B2424,#8B0000\n",
      "INFO [2023-07-24 19:33:20] Quantiles of plotted data range: 0.72541708269308,1.00237531440214,1.00237531440214,1.00237531440214,1.27458291730692\n"
     ]
    }
   ],
   "source": [
    "# perform infercnv operations to reveal cnv signal\n",
    "# perform infercnv operations to reveal cnv signal\n",
    "infercnv_obj = infercnv::run(infercnv_obj,\n",
    "                             cutoff=1,  # use 1 for smart-seq, 0.1 for 10x-genomics\n",
    "                             out_dir=\"outputs/infercnv/sys_take2/\",  # dir is auto-created for storing outputs\n",
    "                             cluster_by_groups=T,   # cluster\n",
    "                             denoise=T,\n",
    "                             HMM=T,\n",
    "                             analysis_mode='subclusters',\n",
    "                             leiden_resolution=0.01, #initial run returned too many subclusters\n",
    "                             debug=TRUE\n",
    "                             )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "947f4f8a-ef0f-4141-9c03-a291cf14257c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "infercnv_obj = readRDS('outputs/infercnv/sys_take2/run.final.infercnv_obj')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f07ba300-e8f3-4c72-8222-588f04125552",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R (infercnv-env)",
   "language": "R",
   "name": "ir_infercnv"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.2.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
